{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests2\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 35 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['ecoli'] = '{}/tRNA_database/ecoli/ecoli-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked['ecoli'] = '{}/tRNA_database_masked/ecoli/ecoli-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 15\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "# index_list_fnam = 'index_list.xlsx'\n",
    "index_list_fnam = 'index_list_with-old.xlsx'\n",
    "\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Input settings ###\n",
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests2/data/AdapterRemoval\n"
     ]
    }
   ],
   "source": [
    "### Run AdapterRemoval ###\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False, \\\n",
    "                  check_input=False, AR_threads=6)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=3, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests2/data/BC_split\n",
      "Loaded results from previous run... Not running barcode split.\n"
     ]
    }
   ],
   "source": [
    "### Split files based on barcodes ###\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=12, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downsampling UMI trimmed sequences to maximum 2000000 reads.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U1</td>\n",
       "      <td>U1</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l1N</td>\n",
       "      <td>human</td>\n",
       "      <td>CC_lig</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l1N</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>TNNNNNNGACTA</td>\n",
       "      <td>7589045</td>\n",
       "      <td>1371457</td>\n",
       "      <td>6481</td>\n",
       "      <td>1377938</td>\n",
       "      <td>18.156935</td>\n",
       "      <td>0.470340</td>\n",
       "      <td>7410594</td>\n",
       "      <td>518800</td>\n",
       "      <td>524287.618942</td>\n",
       "      <td>97.648571</td>\n",
       "      <td>98.953319</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U6</td>\n",
       "      <td>U6</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l6Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>CC_lig</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l6Sp</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTACAG</td>\n",
       "      <td>506580</td>\n",
       "      <td>442970</td>\n",
       "      <td>51232</td>\n",
       "      <td>494202</td>\n",
       "      <td>97.556556</td>\n",
       "      <td>10.366611</td>\n",
       "      <td>496014</td>\n",
       "      <td>204548</td>\n",
       "      <td>320726.407557</td>\n",
       "      <td>97.914248</td>\n",
       "      <td>63.776476</td>\n",
       "      <td>496014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T100_1</td>\n",
       "      <td>T100</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P13_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P13_R2.fastq.bz2</td>\n",
       "      <td>D504</td>\n",
       "      <td>D702</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>ecoli</td>\n",
       "      <td>ecoli_mix_ctr</td>\n",
       "      <td>Percentage</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>TCAGAGCC</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>4759577</td>\n",
       "      <td>291879</td>\n",
       "      <td>4417087</td>\n",
       "      <td>4708966</td>\n",
       "      <td>98.936649</td>\n",
       "      <td>93.801633</td>\n",
       "      <td>4664217</td>\n",
       "      <td>517392</td>\n",
       "      <td>524216.227598</td>\n",
       "      <td>97.996461</td>\n",
       "      <td>98.698204</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                 U1          U1          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "1                 U6          U6          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "2             T100_1        T100          1  2023-06-06/P13_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2023-06-06/P03_R2.fastq.bz2     D501     D703     l1N   human   \n",
       "1  2023-06-06/P03_R2.fastq.bz2     D501     D703    l6Sp   human   \n",
       "2  2023-06-06/P13_R2.fastq.bz2     D504     D702    l1Sp   ecoli   \n",
       "\n",
       "      plot_group    hue_name hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0         CC_lig     Barcode       l1N          1     AGGCTATA     CGCTCATT   \n",
       "1         CC_lig     Barcode      l6Sp          2     AGGCTATA     CGCTCATT   \n",
       "2  ecoli_mix_ctr  Percentage       100          1     TCAGAGCC     TCCGGAGA   \n",
       "\n",
       "        barcode_seq  N_total     N_CC    N_CCA  N_CCA+CC  \\\n",
       "0      TNNNNNNGACTA  7589045  1371457     6481   1377938   \n",
       "1  GGCTGCCATGCTACAG   506580   442970    51232    494202   \n",
       "2  GGCTGCCATGCGACTA  4759577   291879  4417087   4708966   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             18.156935     0.470340       7410594          518800   \n",
       "1             97.556556    10.366611        496014          204548   \n",
       "2             98.936649    93.801633       4664217          517392   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524287.618942                    97.648571               98.953319   \n",
       "1   320726.407557                    97.914248               63.776476   \n",
       "2   524216.227598                    97.996461               98.698204   \n",
       "\n",
       "   N_after_downsample  \n",
       "0             2000000  \n",
       "1              496014  \n",
       "2             2000000  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=True)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=12, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Swipe on:  U6  T0  U1  T88  T100_1  T13  K05m  T50  T75  L01  T100_2  T25  K2m  K32m  K4m  K8m  K64m  K16m  L02  K1m\n",
      "Collecting alignment statistics, from sample:  U1  U6  T100_1  T50  T25  T13  T0  T100_2  L01  K05m  T88  T75  K2m  K1m  L02  K8m  K32m  K64m  K16m  K4m"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U1</td>\n",
       "      <td>U1</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l1N</td>\n",
       "      <td>human</td>\n",
       "      <td>CC_lig</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l1N</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>TNNNNNNGACTA</td>\n",
       "      <td>7589045</td>\n",
       "      <td>1371457</td>\n",
       "      <td>6481</td>\n",
       "      <td>1377938</td>\n",
       "      <td>18.156935</td>\n",
       "      <td>0.470340</td>\n",
       "      <td>7410594</td>\n",
       "      <td>518800</td>\n",
       "      <td>524287.618942</td>\n",
       "      <td>97.648571</td>\n",
       "      <td>98.953319</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1787738</td>\n",
       "      <td>62.973713</td>\n",
       "      <td>37.026287</td>\n",
       "      <td>6.138819</td>\n",
       "      <td>89.38690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U6</td>\n",
       "      <td>U6</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l6Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>CC_lig</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l6Sp</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTACAG</td>\n",
       "      <td>506580</td>\n",
       "      <td>442970</td>\n",
       "      <td>51232</td>\n",
       "      <td>494202</td>\n",
       "      <td>97.556556</td>\n",
       "      <td>10.366611</td>\n",
       "      <td>496014</td>\n",
       "      <td>204548</td>\n",
       "      <td>320726.407557</td>\n",
       "      <td>97.914248</td>\n",
       "      <td>63.776476</td>\n",
       "      <td>496014</td>\n",
       "      <td>436339</td>\n",
       "      <td>61.380257</td>\n",
       "      <td>38.619743</td>\n",
       "      <td>5.142332</td>\n",
       "      <td>87.96909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T100_1</td>\n",
       "      <td>T100</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P13_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P13_R2.fastq.bz2</td>\n",
       "      <td>D504</td>\n",
       "      <td>D702</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>ecoli</td>\n",
       "      <td>ecoli_mix_ctr</td>\n",
       "      <td>Percentage</td>\n",
       "      <td>100</td>\n",
       "      <td>1</td>\n",
       "      <td>TCAGAGCC</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>4759577</td>\n",
       "      <td>291879</td>\n",
       "      <td>4417087</td>\n",
       "      <td>4708966</td>\n",
       "      <td>98.936649</td>\n",
       "      <td>93.801633</td>\n",
       "      <td>4664217</td>\n",
       "      <td>517392</td>\n",
       "      <td>524216.227598</td>\n",
       "      <td>97.996461</td>\n",
       "      <td>98.698204</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1818546</td>\n",
       "      <td>98.077200</td>\n",
       "      <td>1.922800</td>\n",
       "      <td>0.025680</td>\n",
       "      <td>90.92730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                 U1          U1          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "1                 U6          U6          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "2             T100_1        T100          1  2023-06-06/P13_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2023-06-06/P03_R2.fastq.bz2     D501     D703     l1N   human   \n",
       "1  2023-06-06/P03_R2.fastq.bz2     D501     D703    l6Sp   human   \n",
       "2  2023-06-06/P13_R2.fastq.bz2     D504     D702    l1Sp   ecoli   \n",
       "\n",
       "      plot_group    hue_name hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0         CC_lig     Barcode       l1N          1     AGGCTATA     CGCTCATT   \n",
       "1         CC_lig     Barcode      l6Sp          2     AGGCTATA     CGCTCATT   \n",
       "2  ecoli_mix_ctr  Percentage       100          1     TCAGAGCC     TCCGGAGA   \n",
       "\n",
       "        barcode_seq  N_total     N_CC    N_CCA  N_CCA+CC  \\\n",
       "0      TNNNNNNGACTA  7589045  1371457     6481   1377938   \n",
       "1  GGCTGCCATGCTACAG   506580   442970    51232    494202   \n",
       "2  GGCTGCCATGCGACTA  4759577   291879  4417087   4708966   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             18.156935     0.470340       7410594          518800   \n",
       "1             97.556556    10.366611        496014          204548   \n",
       "2             98.936649    93.801633       4664217          517392   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524287.618942                    97.648571               98.953319   \n",
       "1   320726.407557                    97.914248               63.776476   \n",
       "2   524216.227598                    97.996461               98.698204   \n",
       "\n",
       "   N_after_downsample  N_mapped  percent_single_annotation  \\\n",
       "0             2000000   1787738                  62.973713   \n",
       "1              496014    436339                  61.380257   \n",
       "2             2000000   1818546                  98.077200   \n",
       "\n",
       "   percent_multiple_annotation  percent_multiple_codons  Mapping_percent  \n",
       "0                    37.026287                 6.138819         89.38690  \n",
       "1                    38.619743                 5.142332         87.96909  \n",
       "2                     1.922800                 0.025680         90.92730  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=None, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=12, overwrite=True, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting stats from:  U1  U6  T100_1  T50  T25  T13  T0  L01  T100_2  T75  T88  K05m  K2m  K1m  K8m  K32m  K64m  L02  K16m  K4m"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>align_gap</th>\n",
       "      <th>fmax_score&gt;0.9</th>\n",
       "      <th>count</th>\n",
       "      <th>UMIcount</th>\n",
       "      <th>UMI_percent_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U1</td>\n",
       "      <td>U1</td>\n",
       "      <td>1</td>\n",
       "      <td>l1N</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U1</td>\n",
       "      <td>U1</td>\n",
       "      <td>1</td>\n",
       "      <td>l1N</td>\n",
       "      <td>human</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Ala-TGC</td>\n",
       "      <td>72</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TGC</td>\n",
       "      <td>Ala</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>100.000572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U1</td>\n",
       "      <td>U1</td>\n",
       "      <td>1</td>\n",
       "      <td>l1N</td>\n",
       "      <td>human</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Ala-TGC</td>\n",
       "      <td>72</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TGC</td>\n",
       "      <td>Ala</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>54</td>\n",
       "      <td>54</td>\n",
       "      <td>100.005055</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode species  \\\n",
       "0                 U1          U1          1     l1N   human   \n",
       "1                 U1          U1          1     l1N   human   \n",
       "2                 U1          U1          1     l1N   human   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1                     Homo_sapiens_mito_tRNA-Ala-TGC                   72   \n",
       "2                     Homo_sapiens_mito_tRNA-Ala-TGC                   72   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  \\\n",
       "0               True     False           A   AAA       TTT   eColiLys   \n",
       "1               True     False           C   GCA       TGC        Ala   \n",
       "2               True     False           C   GCA       TGC        Ala   \n",
       "\n",
       "   align_gap  fmax_score>0.9  count  UMIcount  UMI_percent_exp  \n",
       "0      False            True      1         1       100.000000  \n",
       "1      False           False      7         7       100.000572  \n",
       "2      False            True     54        54       100.005055  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=None, \\\n",
    "                             overwrite_dir=True)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=12, load_previous=False)\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=True, pull_default=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/tRNA-charge-seq/src/plotting.py:301: UserWarning: Tight layout not applied. The left and right margins cannot be made large enough to accommodate all axes decorations.\n",
      "  fig.tight_layout()\n"
     ]
    }
   ],
   "source": [
    "# Ecoli control:\n",
    "plot_obj.plot_Ecoli_ctr(min_obs=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample/group:  AARS_kinetics  CC_lig  ecoli_mix_ctr  lig_test\n",
      "Now plotting sample/group:  AARS_kinetics  CC_lig  ecoli_mix_ctr  lig_test"
     ]
    }
   ],
   "source": [
    "# Codon abundance barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_abundance_grp', \\\n",
    "                        group=True, min_obs=500)\n",
    "\n",
    "# Codon charge barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample pairs:  (L01 - L02)  (K64m - K1m)  (T88 - T0)  (T13 - T0)  (U1 - U6)  (U6 - L01)  (U6 - L02)  (U1 - L01)  (U1 - L02)"
     ]
    }
   ],
   "source": [
    "# Plot of abundance correlation:\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['L01', 'K64m', 'T88', 'T13', \\\n",
    "                                                   'U1',  'U6',   'U6',  'U1',  'U1'], \\\n",
    "                                                  ['L02', 'K1m',  'T0',  'T0', \\\n",
    "                                                   'U6',  'L01',  'L02', 'L01', 'L02']], \\\n",
    "                             plot_type='codon', plot_name='codon_abundance_corr', \\\n",
    "                             min_obs=500, charge_plot=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  U1  U6  T100_1  T88  T75  T50  T25  T13  T100_2"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Ile2'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;31mException\u001b[0m: Exception occurred in Worker-5 with the following arguments:\nArg 0: 5\nArg 1: sample_name_unique                                     T50\nsample_name                                            T50\nreplicate                                                1\nfastq_mate1_filename           2023-06-06/P13_R1.fastq.bz2\nfastq_mate2_filename           2023-06-06/P13_R2.fastq.bz2\nP5_index                                              D504\nP7_index                                              D702\nbarcode                                               l5Sp\nspecies                                              ecoli\nplot_group                                   ecoli_mix_ctr\nhue_name                                        Percentage\nhue_value                                               50\nhue_order                                                4\nP5_index_seq                                      TCAGAGCC\nP7_index_seq                                      TCCGGAGA\nbarcode_seq                            GGCTGCCATGCAACGCATC\nN_total                                            3383672\nN_CC                                               1261993\nN_CCA                                              2083893\nN_CCA+CC                                           3345886\nCCA+CC_percent_total                             98.883284\npercent_CCA                                      62.282248\nN_after_trim                                       3313070\nN_UMI_observed                                      501091\nN_UMI_expected                               523343.542329\npercent_seqs_after_UMI_trim                       97.91345\npercent_UMI_obs-vs-exp                           95.748005\nN_after_downsample                                 2000000\nN_mapped                                           1760526\npercent_single_annotation                        98.069895\npercent_multiple_annotation                       1.930105\npercent_multiple_codons                           0.028287\nMapping_percent                                    88.0263\ncompartment                                           cyto\naa_norm                                              False\ny_norm                                                True\nName: 5, dtype: object\nArg 2: ['sample_name_unique', 'tRNA_annotation_len', 'align_5p_idx', 'align_3p_idx', 'AA_letter', 'UMIcount']\nArg 3: 10\nArg 4: True\nTraceback (most recent call last):\n  File \u001b[36m\"/home/sulab/anaconda3/lib/python3.9/site-packages/mpire/worker.py\"\u001b[39;49;00m, line \u001b[34m352\u001b[39;49;00m, in _run_safely\n    results = func()\n  File \u001b[36m\"/home/sulab/anaconda3/lib/python3.9/site-packages/mpire/worker.py\"\u001b[39;49;00m, line \u001b[34m288\u001b[39;49;00m, in _func\n    _results = func(args)\n  File \u001b[36m\"/home/sulab/anaconda3/lib/python3.9/site-packages/mpire/worker.py\"\u001b[39;49;00m, line \u001b[34m455\u001b[39;49;00m, in _helper_func_with_idx\n    \u001b[34mreturn\u001b[39;49;00m args[\u001b[34m0\u001b[39;49;00m], \u001b[36mself\u001b[39;49;00m._call_func(func, args[\u001b[34m1\u001b[39;49;00m])\n  File \u001b[36m\"/home/sulab/anaconda3/lib/python3.9/site-packages/mpire/worker.py\"\u001b[39;49;00m, line \u001b[34m480\u001b[39;49;00m, in _call_func\n    \u001b[34mreturn\u001b[39;49;00m func(*args)\n  File \u001b[36m\"/home/sulab/tRNA-charge-seq/src/plotting.py\"\u001b[39;49;00m, line \u001b[34m932\u001b[39;49;00m, in _collect_coverage_data\n    sample_stats[\u001b[33m'\u001b[39;49;00m\u001b[33mAA_letter\u001b[39;49;00m\u001b[33m'\u001b[39;49;00m] = [AAA2A[AAA] \u001b[34mfor\u001b[39;49;00m AAA \u001b[35min\u001b[39;49;00m sample_stats[\u001b[33m'\u001b[39;49;00m\u001b[33mamino_acid\u001b[39;49;00m\u001b[33m'\u001b[39;49;00m].values]\n  File \u001b[36m\"/home/sulab/tRNA-charge-seq/src/plotting.py\"\u001b[39;49;00m, line \u001b[34m932\u001b[39;49;00m, in <listcomp>\n    sample_stats[\u001b[33m'\u001b[39;49;00m\u001b[33mAA_letter\u001b[39;49;00m\u001b[33m'\u001b[39;49;00m] = [AAA2A[AAA] \u001b[34mfor\u001b[39;49;00m AAA \u001b[35min\u001b[39;49;00m sample_stats[\u001b[33m'\u001b[39;49;00m\u001b[33mamino_acid\u001b[39;49;00m\u001b[33m'\u001b[39;49;00m].values]\n\u001b[91mKeyError\u001b[39;49;00m: 'Ile2'\n",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_852002/223200175.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Coverage plots for cyto/mito transcripts:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n\u001b[0m\u001b[1;32m      3\u001b[0m                        plot_name='cov_plot_cyto_behrens_norm', n_jobs=8)\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
      "\u001b[0;32m~/tRNA-charge-seq/src/plotting.py\u001b[0m in \u001b[0;36mplot_coverage\u001b[0;34m(self, compartment, plot_type, aa_norm, y_norm, plot_name, sample_list, verbose, n_jobs, max_5p_non_temp)\u001b[0m\n\u001b[1;32m    800\u001b[0m         \u001b[0;31m# Collect data for plotting for each sample:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    801\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mWorkerPool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 802\u001b[0;31m             \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_collect_coverage_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    803\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    804\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\nNow plotting sample:'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36mmap\u001b[0;34m(self, func, iterable_of_args, iterable_len, max_tasks_active, chunk_size, n_splits, worker_lifespan, progress_bar, progress_bar_position, concatenate_numpy_output, worker_init, worker_exit, task_timeout, worker_init_timeout, worker_exit_timeout, progress_bar_options)\u001b[0m\n\u001b[1;32m    368\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0miterable_len\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mhasattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterable_of_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__len__'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    369\u001b[0m             \u001b[0miterable_len\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterable_of_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 370\u001b[0;31m         results = self.map_unordered(func, ((args_idx, args) for args_idx, args in enumerate(iterable_of_args)),\n\u001b[0m\u001b[1;32m    371\u001b[0m                                      \u001b[0miterable_len\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_tasks_active\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchunk_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_splits\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_lifespan\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    372\u001b[0m                                      \u001b[0mprogress_bar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprogress_bar_position\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_init\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_exit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtask_timeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36mmap_unordered\u001b[0;34m(self, func, iterable_of_args, iterable_len, max_tasks_active, chunk_size, n_splits, worker_lifespan, progress_bar, progress_bar_position, worker_init, worker_exit, task_timeout, worker_init_timeout, worker_exit_timeout, progress_bar_options)\u001b[0m\n\u001b[1;32m    437\u001b[0m         \"\"\"\n\u001b[1;32m    438\u001b[0m         \u001b[0;31m# Simply call imap and cast it to a list. This make sure all elements are there before returning\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m         return list(self.imap_unordered(func, iterable_of_args, iterable_len, max_tasks_active, chunk_size,\n\u001b[0m\u001b[1;32m    440\u001b[0m                                         \u001b[0mn_splits\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_lifespan\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprogress_bar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprogress_bar_position\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_init\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m                                         \u001b[0mworker_exit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtask_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_init_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_exit_timeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36mimap_unordered\u001b[0;34m(self, func, iterable_of_args, iterable_len, max_tasks_active, chunk_size, n_splits, worker_lifespan, progress_bar, progress_bar_position, worker_init, worker_exit, task_timeout, worker_init_timeout, worker_exit_timeout, progress_bar_options)\u001b[0m\n\u001b[1;32m    683\u001b[0m                     \u001b[0;31m# Terminate if exception has been thrown at this point\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    684\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_worker_comms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexception_thrown\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m                         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_handle_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprogress_bar_handler\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m                     \u001b[0;31m# All results are in: it's clean up time\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36m_handle_exception\u001b[0;34m(self, progress_bar_handler)\u001b[0m\n\u001b[1;32m    754\u001b[0m         \u001b[0mtraceback_err\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhighlight_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtraceback_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    755\u001b[0m         \u001b[0mlogger\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Re-raising obtained exception\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 756\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0merr\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mtraceback_err\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    757\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    758\u001b[0m     def stop_and_join(self, progress_bar_handler: Optional[ProgressBarHandler] = None,\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Ile2'"
     ]
    }
   ],
   "source": [
    "# Coverage plots for cyto/mito transcripts:\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_behrens_norm', n_jobs=8)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_behrens_norm', n_jobs=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMI logo:\n",
    "plot_obj.plot_UMI_logo(n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Non-template nucleotides:\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     common_seqs=None, ignore_common_count=False, \\\n",
    "                     overwrite_dir=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find mutations and save them:\n",
    "if True:\n",
    "    TM_obj.find_muts(n_jobs=8, unique_anno=True)\n",
    "    TM_obj.pickle_muts_write(pickle_name='saved_muts_unique-anno.pickle')\n",
    "else:\n",
    "    TM_obj.pickle_muts_read(pickle_name='saved_muts_unique-anno.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
