{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/collab/Alicia_2022-12-28\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Input settings ###\n",
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/collab/Alicia_2022-12-28/data/AdapterRemoval\n"
     ]
    }
   ],
   "source": [
    "### Run AdapterRemoval ###\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False, \\\n",
    "                  check_input=False)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=2, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/collab/Alicia_2022-12-28/data/BC_split\n",
      "Loaded results from previous run... Not running barcode split.\n"
     ]
    }
   ],
   "source": [
    "### Split files based on barcodes ###\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=2, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/collab/Alicia_2022-12-28/data/UMI_trimmed\n",
      "Loaded results from previous run... Not running UMI trimming.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-12-28/P8_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P8_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D702</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293t-NoLeu</td>\n",
       "      <td>Sample name</td>\n",
       "      <td>L1</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>6499165</td>\n",
       "      <td>314837</td>\n",
       "      <td>6142668</td>\n",
       "      <td>6457505</td>\n",
       "      <td>99.358995</td>\n",
       "      <td>95.124479</td>\n",
       "      <td>6385292</td>\n",
       "      <td>520600</td>\n",
       "      <td>524285.306593</td>\n",
       "      <td>98.247883</td>\n",
       "      <td>99.297080</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L2</td>\n",
       "      <td>L2</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P6_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P6_R2.fastq.bz2</td>\n",
       "      <td>D502</td>\n",
       "      <td>D703</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293t-NoLeu</td>\n",
       "      <td>Sample name</td>\n",
       "      <td>L2</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCTCTAT</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>5002866</td>\n",
       "      <td>336642</td>\n",
       "      <td>4622921</td>\n",
       "      <td>4959563</td>\n",
       "      <td>99.134436</td>\n",
       "      <td>93.212265</td>\n",
       "      <td>4910419</td>\n",
       "      <td>516459</td>\n",
       "      <td>524243.123847</td>\n",
       "      <td>98.152119</td>\n",
       "      <td>98.515169</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L3</td>\n",
       "      <td>L3</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-12-28/P8_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P8_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293t-NoLeu</td>\n",
       "      <td>Sample name</td>\n",
       "      <td>L3</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>4093393</td>\n",
       "      <td>358779</td>\n",
       "      <td>3698162</td>\n",
       "      <td>4056941</td>\n",
       "      <td>99.109492</td>\n",
       "      <td>91.156416</td>\n",
       "      <td>4021035</td>\n",
       "      <td>508194</td>\n",
       "      <td>524043.241134</td>\n",
       "      <td>98.232322</td>\n",
       "      <td>96.975585</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0                 L1          L1          1  2022-12-28/P8_R1.fastq.bz2   \n",
       "1                 L2          L2          1  2023-02-28/P6_R1.fastq.bz2   \n",
       "2                 L3          L3          1  2022-12-28/P8_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species  plot_group  \\\n",
       "0  2022-12-28/P8_R2.fastq.bz2     D503     D702    l1Sp   human  293t-NoLeu   \n",
       "1  2023-02-28/P6_R2.fastq.bz2     D502     D703    l1Sp   human  293t-NoLeu   \n",
       "2  2022-12-28/P8_R2.fastq.bz2     D503     D702    l2Sp   human  293t-NoLeu   \n",
       "\n",
       "      hue_name hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0  Sample name        L1          1     AGGATAGG     TCCGGAGA   \n",
       "1  Sample name        L2          2     GCCTCTAT     CGCTCATT   \n",
       "2  Sample name        L3          3     AGGATAGG     TCCGGAGA   \n",
       "\n",
       "          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0    GGCTGCCATGCGACTA  6499165  314837  6142668   6457505   \n",
       "1    GGCTGCCATGCGACTA  5002866  336642  4622921   4959563   \n",
       "2  GGCTGCCATGCTGTCACG  4093393  358779  3698162   4056941   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.358995    95.124479       6385292          520600   \n",
       "1             99.134436    93.212265       4910419          516459   \n",
       "2             99.109492    91.156416       4021035          508194   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524285.306593                    98.247883               99.297080   \n",
       "1   524243.123847                    98.152119               98.515169   \n",
       "2   524043.241134                    98.232322               96.975585   \n",
       "\n",
       "   N_after_downsample  \n",
       "0             2000000  \n",
       "1             2000000  \n",
       "2             2000000  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False, \\\n",
    "                       downsample_absolute=2e6)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=4, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  L1  L4  L3  L2  L5  L7  L8  L6  L10  L9  L11  L12  R1  R2  R3  R4  R5  R8  R6  R7  R9  R10  R12  R11  2Q3  2Q1  2Q2  2Q4  2Q6  2Q5  GQ1  GQ2  GQ3  GQ4  GQ5  GQ6  H1  HL1  HL2  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGL2  HGR1  HGR2  HGQ1  common-seqs  HGQ2\n",
      "Collecting alignment statistics, from sample:  L4  L2  L3  L1  L5  L7  L6  L8  L9  L10  L12  L11  R1  R2  R3  R4  R8  R5  R7  R6  R9  R10  R12  R11  2Q3  2Q1  2Q2  2Q4  2Q6  2Q5  GQ1  GQ2  GQ3  GQ4  GQ5  GQ6  HL2  HL1  H1  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGL2  HGR2  HGR1  HGQ1  HGQ2  common-seqs"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-12-28/P8_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P8_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D702</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293t-NoLeu</td>\n",
       "      <td>Sample name</td>\n",
       "      <td>L1</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>6499165</td>\n",
       "      <td>314837</td>\n",
       "      <td>6142668</td>\n",
       "      <td>6457505</td>\n",
       "      <td>99.358995</td>\n",
       "      <td>95.124479</td>\n",
       "      <td>6385292</td>\n",
       "      <td>520600</td>\n",
       "      <td>524285.306593</td>\n",
       "      <td>98.247883</td>\n",
       "      <td>99.297080</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1986745.0</td>\n",
       "      <td>81.428165</td>\n",
       "      <td>18.571835</td>\n",
       "      <td>2.586291</td>\n",
       "      <td>99.33725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L2</td>\n",
       "      <td>L2</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P6_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P6_R2.fastq.bz2</td>\n",
       "      <td>D502</td>\n",
       "      <td>D703</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293t-NoLeu</td>\n",
       "      <td>Sample name</td>\n",
       "      <td>L2</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCTCTAT</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>5002866</td>\n",
       "      <td>336642</td>\n",
       "      <td>4622921</td>\n",
       "      <td>4959563</td>\n",
       "      <td>99.134436</td>\n",
       "      <td>93.212265</td>\n",
       "      <td>4910419</td>\n",
       "      <td>516459</td>\n",
       "      <td>524243.123847</td>\n",
       "      <td>98.152119</td>\n",
       "      <td>98.515169</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1986072.0</td>\n",
       "      <td>81.314978</td>\n",
       "      <td>18.685022</td>\n",
       "      <td>2.417536</td>\n",
       "      <td>99.30360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L3</td>\n",
       "      <td>L3</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-12-28/P8_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P8_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293t-NoLeu</td>\n",
       "      <td>Sample name</td>\n",
       "      <td>L3</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>4093393</td>\n",
       "      <td>358779</td>\n",
       "      <td>3698162</td>\n",
       "      <td>4056941</td>\n",
       "      <td>99.109492</td>\n",
       "      <td>91.156416</td>\n",
       "      <td>4021035</td>\n",
       "      <td>508194</td>\n",
       "      <td>524043.241134</td>\n",
       "      <td>98.232322</td>\n",
       "      <td>96.975585</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1975947.0</td>\n",
       "      <td>80.374423</td>\n",
       "      <td>19.625577</td>\n",
       "      <td>2.643796</td>\n",
       "      <td>98.79735</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0                 L1          L1          1  2022-12-28/P8_R1.fastq.bz2   \n",
       "1                 L2          L2          1  2023-02-28/P6_R1.fastq.bz2   \n",
       "2                 L3          L3          1  2022-12-28/P8_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species  plot_group  \\\n",
       "0  2022-12-28/P8_R2.fastq.bz2     D503     D702    l1Sp   human  293t-NoLeu   \n",
       "1  2023-02-28/P6_R2.fastq.bz2     D502     D703    l1Sp   human  293t-NoLeu   \n",
       "2  2022-12-28/P8_R2.fastq.bz2     D503     D702    l2Sp   human  293t-NoLeu   \n",
       "\n",
       "      hue_name hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0  Sample name        L1          1     AGGATAGG     TCCGGAGA   \n",
       "1  Sample name        L2          2     GCCTCTAT     CGCTCATT   \n",
       "2  Sample name        L3          3     AGGATAGG     TCCGGAGA   \n",
       "\n",
       "          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0    GGCTGCCATGCGACTA  6499165  314837  6142668   6457505   \n",
       "1    GGCTGCCATGCGACTA  5002866  336642  4622921   4959563   \n",
       "2  GGCTGCCATGCTGTCACG  4093393  358779  3698162   4056941   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.358995    95.124479       6385292          520600   \n",
       "1             99.134436    93.212265       4910419          516459   \n",
       "2             99.109492    91.156416       4021035          508194   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524285.306593                    98.247883               99.297080   \n",
       "1   524243.123847                    98.152119               98.515169   \n",
       "2   524043.241134                    98.232322               96.975585   \n",
       "\n",
       "   N_after_downsample   N_mapped  percent_single_annotation  \\\n",
       "0             2000000  1986745.0                  81.428165   \n",
       "1             2000000  1986072.0                  81.314978   \n",
       "2             2000000  1975947.0                  80.374423   \n",
       "\n",
       "   percent_multiple_annotation  percent_multiple_codons  Mapping_percent  \n",
       "0                    18.571835                 2.586291         99.33725  \n",
       "1                    18.685022                 2.417536         99.30360  \n",
       "2                    19.625577                 2.643796         98.79735  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=4, overwrite=True, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/Alicia_2022-12-28/data/stats_collection\n",
      "Loaded results from previous run... Not running stats collection.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>align_gap</th>\n",
       "      <th>fmax_score&gt;0.9</th>\n",
       "      <th>count</th>\n",
       "      <th>UMIcount</th>\n",
       "      <th>UMI_percent_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>100.000191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1</td>\n",
       "      <td>L1</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>62</td>\n",
       "      <td>62</td>\n",
       "      <td>100.005818</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode species  \\\n",
       "0                 L1          L1          1    l1Sp   human   \n",
       "1                 L1          L1          1    l1Sp   human   \n",
       "2                 L1          L1          1    l1Sp   human   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  \\\n",
       "0               True     False           A   AAA       TTT   eColiLys   \n",
       "1               True     False           A   AAA       TTT   eColiLys   \n",
       "2               True     False           C   AAA       TTT   eColiLys   \n",
       "\n",
       "   align_gap  fmax_score>0.9  count  UMIcount  UMI_percent_exp  \n",
       "0      False           False      1         1       100.000000  \n",
       "1      False            True      3         3       100.000191  \n",
       "2      False           False     62        62       100.005818  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=False)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=4, load_previous=True)\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Folder exists and overwrite set to false... Doing nothing.\n"
     ]
    }
   ],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=False, pull_default=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ecoli control:\n",
    "plot_obj.plot_Ecoli_ctr(min_obs=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample/group:  293t-NoArg  293t-NoGln  293t-NoLeu  HCT116-Depletion\n",
      "Now plotting sample/group:  293t-NoArg  293t-NoGln  293t-NoLeu  HCT116-Depletion"
     ]
    }
   ],
   "source": [
    "# Codon abundance barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_abundance_grp', \\\n",
    "                        group=True, min_obs=500)\n",
    "\n",
    "# Codon charge barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Abundance correlation:\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['L1', 'L1', 'R1'], \\\n",
    "                                                  ['R1', 'L6', 'R6']], \\\n",
    "                             plot_type='transcript', plot_name='tr_abundance_corr_L-R', \\\n",
    "                             min_obs=500, charge_plot=False, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  L2  L1  L3  L4  L7  L5  L6  L8  L9  L10  L12  L11  R1  R2  R4  R3  R5  R7  R6  R8  R10  R9  R11  R12  2Q4  2Q1  2Q3  2Q2  2Q6  2Q5  GQ1  GQ2  GQ3  GQ4  GQ5  GQ6  HL1  H1  HL2  HR1  HQ1  HR2  HQ2  HG1  HGL1  HGL2  HGR1  HGR2  HGQ1  HGQ2\n",
      "Now plotting sample:  L1  L2  L3  L4  L5  L6  L7  L8  L9  L10  L11  L12  R1  R2  R3  R4  R5  R6  R7  R8  R9  R10  R11  R12  2Q1  2Q2  2Q3  2Q4  2Q5  2Q6  GQ1  GQ2  GQ3  GQ4  GQ5  GQ6  H1  HL1  HL2  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGL2  HGR1  HGR2  HGQ1  HGQ2\n",
      "Now collecting data for sample:  L1  L2  L3  L4  L5  L6  L7  L8  L9  L10  L11  L12  R1  R2  R3  R4  R5  R8  R7  R6  R9  R10  R11  R12  2Q4  2Q1  2Q2  2Q3  2Q5  2Q6  GQ1  GQ2  GQ3  GQ5  GQ4  GQ6  H1  HL1  HL2  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGR1  HGL2  HGR2  HGQ1  HGQ2\n",
      "Now plotting sample:  L1  L2  L3  L4  L5  L6  L7  L8  L9  L10  L11  L12  R1  R2  R3  R4  R5  R6  R7  R8  R9  R10  R11  R12  2Q1  2Q2  2Q3  2Q4  2Q5  2Q6  GQ1  GQ2  GQ3  GQ4  GQ5  GQ6  H1  HL1  HL2  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGL2  HGR1  HGR2  HGQ1  HGQ2"
     ]
    }
   ],
   "source": [
    "# Coverage plots for cyto/mito transcripts:\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_behrens_norm', n_jobs=4)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_behrens_norm', n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  L1  L2  L3  L4  L5  L7  L6  L8  L9  L10  L12  L11  R1  R3  R2  R4  R6  R8  R5  R7  R9  R10  R11  R12  2Q1  2Q4  2Q2  2Q3  2Q5  2Q6  GQ1  GQ2  GQ3  GQ4  GQ6  GQ5  H1  HL1  HL2  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGL2  HGR1  HGR2  HGQ1  HGQ2\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# UMI logo:\n",
    "plot_obj.plot_UMI_logo(n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  L1  L2  L3  L4  L5  L7  L6  L8  L9  L11  L12  L10  R1  R3  R2  R4  R8  R5  R6  R7  R9  R10  R11  R12  2Q4  2Q1  2Q2  2Q3  2Q5  2Q6  GQ1  GQ2  GQ3  GQ4  GQ5  GQ6  H1  HL1  HL2  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGL2  HGR1  HGR2  HGQ1  HGQ2\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  L1  L2  L3  L4  L5  L6  L7  L8  L9  L10  L12  L11  R1  R2  R3  R4  R8  R5  R7  R6  R9  R10  R12  R11  2Q3  2Q1  2Q2  2Q4  2Q6  2Q5  GQ1  GQ2  GQ4  GQ3  GQ5  GQ6  H1  HL1  HL2  HR1  HR2  HQ2  HQ1  HG1  HGL1  HGL2  HGR2  HGR1  HGQ1  HGQ2\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  L1  L2  L3  L4  L5  L6  L7  L8  L9  L10  L12  L11  R2  R1  R3  R4  R8  R5  R6  R7  R9  R10  R11  R12  2Q4  2Q1  2Q2  2Q3  2Q6  2Q5  GQ1  GQ2  GQ4  GQ3  GQ5  GQ6  H1  HL1  HL2  HR1  HR2  HQ1  HQ2  HG1  HGL1  HGR1  HGL2  HGR2  HGQ1  HGQ2\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# Non-template nucleotides:\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export data:\n",
    "plot_obj.write_charge_df(df_type='aa', fnam='charge-df_aa_2022-12-28')\n",
    "plot_obj.write_charge_df(df_type='codon', fnam='charge-df_codon_2022-12-28')\n",
    "plot_obj.write_charge_df(df_type='transcript', fnam='charge-df_transcript_2022-12-28')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Folder exists and overwrite set to false... Doing nothing.\n"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     common_seqs=common_seqs, ignore_common_count=False, \\\n",
    "                     overwrite_dir=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find mutations and save them:\n",
    "if False:\n",
    "    TM_obj.find_muts(n_jobs=4, unique_anno=True)\n",
    "    TM_obj.pickle_muts_write(pickle_name='saved_muts_unique-anno.pickle')\n",
    "else:\n",
    "    TM_obj.pickle_muts_read(pickle_name='saved_muts_unique-anno.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare mutations/gaps/RT stops for Leu/Arg depleted samples:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='mut', \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  \\\n",
    "                                                         'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', \\\n",
    "                                                         'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-gap_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='gap', \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  \\\n",
    "                                                         'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', \\\n",
    "                                                         'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-RTstops_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='RTstops', \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  \\\n",
    "                                                         'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', \\\n",
    "                                                         'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   topN=10, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare mutations/gaps/RT stops for Gln depleted samples:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoGln_tr-mut_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='mut', \\\n",
    "                                   sample_unique_pairs=[['2Q1', '2Q1', '2Q1', 'GQ1',  \\\n",
    "                                                         'GQ1', 'GQ1', '2Q1', '2Q6'], \\\n",
    "                                                        ['2Q3', '2Q4', '2Q6', 'GQ3', \\\n",
    "                                                         'GQ4', 'GQ6', 'GQ1', 'GQ6']], \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoGln_tr-gap_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='gap', \\\n",
    "                                   sample_unique_pairs=[['2Q1', '2Q1', '2Q1', 'GQ1',  \\\n",
    "                                                         'GQ1', 'GQ1', '2Q1', '2Q6'], \\\n",
    "                                                        ['2Q3', '2Q4', '2Q6', 'GQ3', \\\n",
    "                                                         'GQ4', 'GQ6', 'GQ1', 'GQ6']], \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoGln_tr-RTstops_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='RTstops', \\\n",
    "                                   sample_unique_pairs=[['2Q1', '2Q1', '2Q1', 'GQ1',  \\\n",
    "                                                         'GQ1', 'GQ1', '2Q1', '2Q6'], \\\n",
    "                                                        ['2Q3', '2Q4', '2Q6', 'GQ3', \\\n",
    "                                                         'GQ4', 'GQ6', 'GQ1', 'GQ6']], \\\n",
    "                                   topN=10, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare mutations/gaps/RT stops for Leu/Arg/Gln depleted samples (HCT116):\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='HCT_tr-mut_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='mut', \\\n",
    "                                   sample_unique_pairs=[['H1',   'H1',   'H1',  \\\n",
    "                                                         'HG1',  'HG1',  'HG1'], \\\n",
    "                                                        ['HL2',  'HR1',  'HQ2', \\\n",
    "                                                         'HGL2', 'HGR2', 'HGQ2']], \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='HCT_tr-gap_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='gap', \\\n",
    "                                   sample_unique_pairs=[['H1',   'H1',   'H1',  \\\n",
    "                                                         'HG1',  'HG1',  'HG1'], \\\n",
    "                                                        ['HL2',  'HR1',  'HQ2', \\\n",
    "                                                         'HGL2', 'HGR2', 'HGQ2']], \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='HCT_tr-RTstops_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   data_type='RTstops', \\\n",
    "                                   sample_unique_pairs=[['H1',   'H1',   'H1',  \\\n",
    "                                                         'HG1',  'HG1',  'HG1'], \\\n",
    "                                                        ['HL2',  'HR1',  'HQ2', \\\n",
    "                                                         'HGL2', 'HGR2', 'HGQ2']], \\\n",
    "                                   topN=10, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export the mutation/gap/RT stops data:\n",
    "TM_obj.write_transcript_mut(data_type='mut', csv_name='tr-mut_matrix')\n",
    "TM_obj.write_transcript_mut(data_type='gap', csv_name='tr-gap_matrix')\n",
    "TM_obj.write_transcript_mut(data_type='RTstops', csv_name='tr-RTstops_matrix')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
