{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/Kugel-samples\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False, AR_threads=6)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=2, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=12, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S01</td>\n",
       "      <td>S01</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P01_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P01_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Direct Trizol</td>\n",
       "      <td>Cell line</td>\n",
       "      <td>Panc3.27 siCtrl</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>2864349</td>\n",
       "      <td>123367</td>\n",
       "      <td>2719039</td>\n",
       "      <td>2842406</td>\n",
       "      <td>99.233927</td>\n",
       "      <td>95.659769</td>\n",
       "      <td>2806474</td>\n",
       "      <td>486471</td>\n",
       "      <td>521805.883311</td>\n",
       "      <td>97.979471</td>\n",
       "      <td>93.228347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S02</td>\n",
       "      <td>S02</td>\n",
       "      <td>2</td>\n",
       "      <td>2023-06-06/P01_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P01_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Direct Trizol</td>\n",
       "      <td>Cell line</td>\n",
       "      <td>Panc3.27 siCtrl</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>2751104</td>\n",
       "      <td>117505</td>\n",
       "      <td>2610253</td>\n",
       "      <td>2727758</td>\n",
       "      <td>99.151395</td>\n",
       "      <td>95.692250</td>\n",
       "      <td>2694732</td>\n",
       "      <td>482073</td>\n",
       "      <td>521216.264062</td>\n",
       "      <td>97.950932</td>\n",
       "      <td>92.490015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S03</td>\n",
       "      <td>S03</td>\n",
       "      <td>3</td>\n",
       "      <td>2023-06-06/P01_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P01_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Direct Trizol</td>\n",
       "      <td>Cell line</td>\n",
       "      <td>Panc3.27 siCtrl</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>2640449</td>\n",
       "      <td>119777</td>\n",
       "      <td>2500386</td>\n",
       "      <td>2620163</td>\n",
       "      <td>99.231722</td>\n",
       "      <td>95.428643</td>\n",
       "      <td>2586063</td>\n",
       "      <td>478562</td>\n",
       "      <td>520508.798590</td>\n",
       "      <td>97.940275</td>\n",
       "      <td>91.941193</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                S01         S01          1  2023-06-06/P01_R1.fastq.bz2   \n",
       "1                S02         S02          2  2023-06-06/P01_R1.fastq.bz2   \n",
       "2                S03         S03          3  2023-06-06/P01_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2023-06-06/P01_R2.fastq.bz2     D501     D701    l1Sp   human   \n",
       "1  2023-06-06/P01_R2.fastq.bz2     D501     D701    l2Sp   human   \n",
       "2  2023-06-06/P01_R2.fastq.bz2     D501     D701    l3Sp   human   \n",
       "\n",
       "      plot_group   hue_name        hue_value  hue_order P5_index_seq  \\\n",
       "0  Direct Trizol  Cell line  Panc3.27 siCtrl          1     AGGCTATA   \n",
       "1  Direct Trizol  Cell line  Panc3.27 siCtrl          1     AGGCTATA   \n",
       "2  Direct Trizol  Cell line  Panc3.27 siCtrl          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  2864349  123367  2719039   2842406   \n",
       "1     ATTACTCG  GGCTGCCATGCTGTCACG  2751104  117505  2610253   2727758   \n",
       "2     ATTACTCG    GGCTGCCATGCTGCGA  2640449  119777  2500386   2620163   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.233927    95.659769       2806474          486471   \n",
       "1             99.151395    95.692250       2694732          482073   \n",
       "2             99.231722    95.428643       2586063          478562   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \n",
       "0   521805.883311                    97.979471               93.228347  \n",
       "1   521216.264062                    97.950932               92.490015  \n",
       "2   520508.798590                    97.940275               91.941193  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "# Note, the cDNA input amount is so large that it is very unlikely to sequence\n",
    "# the same PCR amplified DNA twice. Therefore, this processing step does not\n",
    "# attempt to merge possible UMI duplicates.\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=12, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  S03  S02  S04  S01  S07  S06  S05  S08  S10  S09  S11  S12  S13  S16  S15  S14  S17  common-seqs  S18\n",
      "Collecting alignment statistics, from sample:  S02  S01  S03  S04  S06  S07  S08  S05  S09  S11  S10  S12  S13  S16  S15  S14  common-seqs  S17  S18"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S01</td>\n",
       "      <td>S01</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P01_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P01_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Direct Trizol</td>\n",
       "      <td>Cell line</td>\n",
       "      <td>Panc3.27 siCtrl</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>2864349</td>\n",
       "      <td>123367</td>\n",
       "      <td>2719039</td>\n",
       "      <td>2842406</td>\n",
       "      <td>99.233927</td>\n",
       "      <td>95.659769</td>\n",
       "      <td>2806474</td>\n",
       "      <td>486471</td>\n",
       "      <td>521805.883311</td>\n",
       "      <td>97.979471</td>\n",
       "      <td>93.228347</td>\n",
       "      <td>2690827.0</td>\n",
       "      <td>69.182374</td>\n",
       "      <td>30.817626</td>\n",
       "      <td>4.060945</td>\n",
       "      <td>95.879278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S02</td>\n",
       "      <td>S02</td>\n",
       "      <td>2</td>\n",
       "      <td>2023-06-06/P01_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P01_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Direct Trizol</td>\n",
       "      <td>Cell line</td>\n",
       "      <td>Panc3.27 siCtrl</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>2751104</td>\n",
       "      <td>117505</td>\n",
       "      <td>2610253</td>\n",
       "      <td>2727758</td>\n",
       "      <td>99.151395</td>\n",
       "      <td>95.692250</td>\n",
       "      <td>2694732</td>\n",
       "      <td>482073</td>\n",
       "      <td>521216.264062</td>\n",
       "      <td>97.950932</td>\n",
       "      <td>92.490015</td>\n",
       "      <td>2498162.0</td>\n",
       "      <td>69.980410</td>\n",
       "      <td>30.019590</td>\n",
       "      <td>4.199327</td>\n",
       "      <td>92.705397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S03</td>\n",
       "      <td>S03</td>\n",
       "      <td>3</td>\n",
       "      <td>2023-06-06/P01_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P01_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Direct Trizol</td>\n",
       "      <td>Cell line</td>\n",
       "      <td>Panc3.27 siCtrl</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>2640449</td>\n",
       "      <td>119777</td>\n",
       "      <td>2500386</td>\n",
       "      <td>2620163</td>\n",
       "      <td>99.231722</td>\n",
       "      <td>95.428643</td>\n",
       "      <td>2586063</td>\n",
       "      <td>478562</td>\n",
       "      <td>520508.798590</td>\n",
       "      <td>97.940275</td>\n",
       "      <td>91.941193</td>\n",
       "      <td>2479627.0</td>\n",
       "      <td>69.329016</td>\n",
       "      <td>30.670984</td>\n",
       "      <td>4.293307</td>\n",
       "      <td>95.884246</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                S01         S01          1  2023-06-06/P01_R1.fastq.bz2   \n",
       "1                S02         S02          2  2023-06-06/P01_R1.fastq.bz2   \n",
       "2                S03         S03          3  2023-06-06/P01_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2023-06-06/P01_R2.fastq.bz2     D501     D701    l1Sp   human   \n",
       "1  2023-06-06/P01_R2.fastq.bz2     D501     D701    l2Sp   human   \n",
       "2  2023-06-06/P01_R2.fastq.bz2     D501     D701    l3Sp   human   \n",
       "\n",
       "      plot_group   hue_name        hue_value  hue_order P5_index_seq  \\\n",
       "0  Direct Trizol  Cell line  Panc3.27 siCtrl          1     AGGCTATA   \n",
       "1  Direct Trizol  Cell line  Panc3.27 siCtrl          1     AGGCTATA   \n",
       "2  Direct Trizol  Cell line  Panc3.27 siCtrl          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  2864349  123367  2719039   2842406   \n",
       "1     ATTACTCG  GGCTGCCATGCTGTCACG  2751104  117505  2610253   2727758   \n",
       "2     ATTACTCG    GGCTGCCATGCTGCGA  2640449  119777  2500386   2620163   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.233927    95.659769       2806474          486471   \n",
       "1             99.151395    95.692250       2694732          482073   \n",
       "2             99.231722    95.428643       2586063          478562   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   521805.883311                    97.979471               93.228347   \n",
       "1   521216.264062                    97.950932               92.490015   \n",
       "2   520508.798590                    97.940275               91.941193   \n",
       "\n",
       "    N_mapped  percent_single_annotation  percent_multiple_annotation  \\\n",
       "0  2690827.0                  69.182374                    30.817626   \n",
       "1  2498162.0                  69.980410                    30.019590   \n",
       "2  2479627.0                  69.329016                    30.670984   \n",
       "\n",
       "   percent_multiple_codons  Mapping_percent  \n",
       "0                 4.060945        95.879278  \n",
       "1                 4.199327        92.705397  \n",
       "2                 4.293307        95.884246  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=4, overwrite=False, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Collecting stats from:  S01  S02  S03  S04  S05  S06  S10  S12  S08  S09  S11  S07  S15  S13  S14  S17  S16  S18"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S01</td>\n",
       "      <td>Pa-Ctr</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S01</td>\n",
       "      <td>Pa-Ctr</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S01</td>\n",
       "      <td>Pa-Ctr</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode species  \\\n",
       "0                S01      Pa-Ctr          1    l1Sp   human   \n",
       "1                S01      Pa-Ctr          1    l1Sp   human   \n",
       "2                S01      Pa-Ctr          1    l1Sp   human   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  count  \n",
       "0               True     False           A   AAA       TTT   eColiLys      1  \n",
       "1               True     False           C   AAA       TTT   eColiLys    688  \n",
       "2               True      True           A   AAA       TTT   eColiLys      2  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=False)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=6, load_previous=True)\n",
    "# The dataframe returned is the \"ALL_stats_aggregate_filtered.csv\"\n",
    "# which is the aggregated data filtered to contain only the\n",
    "# most relevant columnns and requiring the 3' must be covered\n",
    "# and have no 3' non-template bases.\n",
    "# The CSV file output \"ALL_stats_aggregate.csv\" is the data\n",
    "# aggregated based on all values identical except\n",
    "# readID, 5p_UMI and 3p_BC. I.e. every information, except\n",
    "# the UMI seequence is maintained in the aggregated CSV.\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Folder exists and overwrite set to false... Doing nothing.\n"
     ]
    }
   ],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=False, pull_default=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  S01  S02  S03  S04  S05  S06  S07  S08  S10  S09  S11  S12  S14  S16  S15  S13  S18  S17\n",
      "Now plotting sample:  S01  S02  S03  S04  S05  S06  S07  S08  S09  S10  S11  S12  S13  S14  S15  S16  S17  S18\n",
      "Now collecting data for sample:  S01  S02  S03  S04  S05  S06  S07  S09  S08  S11  S10  S12  S14  S15  S18  S16  S13  S17\n",
      "Now plotting sample:  S01  S02  S03  S04  S05  S06  S07  S08  S09  S10  S11  S12  S13  S14  S15  S16  S17  S18\n",
      "Now plotting sample/group:  Direct Trizol  Snap frozen\n",
      "Now plotting sample/group:  Direct Trizol  Snap frozen\n",
      "Now plotting sample/group:  Direct Trizol  Snap frozen\n",
      "Now plotting sample/group:  Direct Trizol  Snap frozen\n",
      "Now plotting sample pairs:  (S01 - S02)  (S01 - S03)  (S04 - S05)  (S04 - S06)  (S07 - S08)  (S07 - S09)  (S10 - S11)  (S10 - S12)  (S13 - S14)  (S13 - S15)  (S16 - S17)  (S16 - S18)\n",
      "Now plotting sample pairs:  (S01 - S02)  (S01 - S03)  (S04 - S05)  (S04 - S06)  (S07 - S08)  (S07 - S09)  (S10 - S11)  (S10 - S12)  (S13 - S14)  (S13 - S15)  (S16 - S17)  (S16 - S18)\n",
      "Now plotting sample pairs:  (Pa-Ctr - Pa-HMGA)  (KP4-Ctr - KP4-HMGA)  (KP4-Ctr-SF - KP4-HMGA-SF)\n",
      "Now plotting sample pairs:  (Pa-Ctr - Pa-HMGA)  (KP4-Ctr - KP4-HMGA)  (KP4-Ctr-SF - KP4-HMGA-SF)"
     ]
    }
   ],
   "source": [
    "# Coverage plots:\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_norm', n_jobs=12)\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_norm', n_jobs=12)\n",
    "\n",
    "# Charge plots:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_charge', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)\n",
    "\n",
    "# RPM plots:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_rpm', \\\n",
    "                        group=True, min_obs=500, charge_plot=False)\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_rpm', \\\n",
    "                        group=True, min_obs=500, charge_plot=False)\n",
    "\n",
    "# Ecoli control plots:\n",
    "plot_obj.plot_Ecoli_ctr(plot_name='ecoli-ctr_charge', charge_plot=True)\n",
    "plot_obj.plot_Ecoli_ctr(plot_name='ecoli-ctr_rpm', charge_plot=False)\n",
    "\n",
    "# Sample replicability plots:\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['S01', 'S01', 'S04', 'S04', 'S07', 'S07', \\\n",
    "                                                   'S10', 'S10', 'S13', 'S13', 'S16', 'S16'], \\\n",
    "                                                  ['S02', 'S03', 'S05', 'S06', 'S08', 'S09', \\\n",
    "                                                   'S11', 'S12', 'S14', 'S15', 'S17', 'S18']], \\\n",
    "                             plot_type='codon', plot_name='codon_charge_replicability', \\\n",
    "                             min_obs=500, charge_plot=True)\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['S01', 'S01', 'S04', 'S04', 'S07', 'S07', \\\n",
    "                                                   'S10', 'S10', 'S13', 'S13', 'S16', 'S16'], \\\n",
    "                                                  ['S02', 'S03', 'S05', 'S06', 'S08', 'S09', \\\n",
    "                                                   'S11', 'S12', 'S14', 'S15', 'S17', 'S18']], \\\n",
    "                             plot_type='codon', plot_name='codon_rpm_replicability', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "\n",
    "# Charge/RPM correlation between sample group:\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['Pa-Ctr',  'KP4-Ctr',  'KP4-Ctr-SF'], \\\n",
    "                                           ['Pa-HMGA', 'KP4-HMGA', 'KP4-HMGA-SF']], \\\n",
    "                             plot_type='codon', plot_name='codon_charge_diff', \\\n",
    "                             min_obs=500, charge_plot=True)\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['Pa-Ctr',  'KP4-Ctr',  'KP4-Ctr-SF'], \\\n",
    "                                           ['Pa-HMGA', 'KP4-HMGA', 'KP4-HMGA-SF']], \\\n",
    "                             plot_type='codon', plot_name='codon_rpm_diff', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['Pa-Ctr',  'KP4-Ctr',  'KP4-Ctr-SF'], \\\n",
    "                                           ['Pa-HMGA', 'KP4-HMGA', 'KP4-HMGA-SF']], \\\n",
    "                             plot_type='transcript', plot_name='tr_charge_diff', \\\n",
    "                             min_obs=500, charge_plot=True)\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['Pa-Ctr',  'KP4-Ctr',  'KP4-Ctr-SF'], \\\n",
    "                                           ['Pa-HMGA', 'KP4-HMGA', 'KP4-HMGA-SF']], \\\n",
    "                             plot_type='transcript', plot_name='tr_rpm_diff', \\\n",
    "                             min_obs=500, charge_plot=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  S02  S01  S03  S04  S06  S07  S08  S05  S09  S10  S11  S12  S16  S14  S13  S15  S17  S18\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  S02  S03  S01  S04  S07  S06  S05  S08  S10  S09  S12  S11  S13  S15  S14  S16  S17  S18\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# Logo plots of 5/3 prime non-template sequence:\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=4)\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=4)\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write charge/rpm data to csv file:\n",
    "fnam = 'data/plotting/aa_charge.csv'\n",
    "plot_obj.charge_filt['aa'].reset_index(drop=True).to_csv(fnam, index=False)\n",
    "fnam = 'data/plotting/codon_charge.csv'\n",
    "plot_obj.charge_filt['codon'].reset_index(drop=True).to_csv(fnam, index=False)\n",
    "fnam = 'data/plotting/tr_charge.csv'\n",
    "plot_obj.charge_filt['tr'].reset_index(drop=True).to_csv(fnam, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Folder exists and overwrite set to false... Doing nothing.\n"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     common_seqs=common_seqs, ignore_common_count=False, \\\n",
    "                     overwrite_dir=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting stats from:  S01  S04  S02  S03  S05  S06  S07  S09  S10  S08  S11  S12  S15  S14  S16  S18  S13  S17"
     ]
    }
   ],
   "source": [
    "TM_obj.find_muts(n_jobs=12, unique_anno=True, fix_end=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot transcript mutations/gaps:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='tr-mut_matrix_comp_top5-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['Pa-Ctr',  'KP4-Ctr',  'KP4-Ctr-SF'], \\\n",
    "                                                 ['Pa-HMGA', 'KP4-HMGA', 'KP4-HMGA-SF']],\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   topN=5, topN_select='max_diff')\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='tr-gap_matrix_comp_top5-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=True, \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['Pa-Ctr',  'KP4-Ctr',  'KP4-Ctr-SF'], \\\n",
    "                                                 ['Pa-HMGA', 'KP4-HMGA', 'KP4-HMGA-SF']],\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   topN=5, topN_select='max_diff')\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='tr-gap_matrix_comp_Cys-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=True, \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['Pa-Ctr',  'KP4-Ctr',  'KP4-Ctr-SF'], \\\n",
    "                                                 ['Pa-HMGA', 'KP4-HMGA', 'KP4-HMGA-SF']],\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   topN=5, topN_select='max_diff', \\\n",
    "                                   anno_substring_compare='Cys')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write transcript mutations/gaps to csv file:\n",
    "TM_obj.write_transcript_mut(csv_name='mut-matrix_right-aligned', \\\n",
    "                            sample_list=list(sample_df['sample_name_unique'].values))\n",
    "TM_obj.write_transcript_mut(csv_name='gap-matrix_right-aligned', gap_only=True, \\\n",
    "                            sample_list=list(sample_df['sample_name_unique'].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
