{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy, warnings\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "# Numpy will complain about subnormals if python\n",
    "# is compiled with the -ffast-math compiler flag:\n",
    "# https://github.com/clearlinux/distribution/issues/2809\n",
    "with warnings.catch_warnings():\n",
    "    warnings.filterwarnings(\"ignore\", message=\"The value of the smallest subnormal for\")\n",
    "    import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/masked-align_analysis\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))\n",
    "\n",
    "# Read sample list:\n",
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment effect on reads mapping to transcripts/codons\n",
    "Process a set of samples covering all adapter barcodes with/without using a masked reference for alignment.\n",
    "Gather alignment statistics from each case with the purpose of showing the effect of the alignment method on the mapping to transcripts/codons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/masked-align_analysis/data/AdapterRemoval\n",
      "Downsampling UMI trimmed sequences to maximum 2000000 reads.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8h_p1</td>\n",
       "      <td>8h_p1</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>3789831</td>\n",
       "      <td>1854972</td>\n",
       "      <td>1889882</td>\n",
       "      <td>3744854</td>\n",
       "      <td>98.813219</td>\n",
       "      <td>50.466106</td>\n",
       "      <td>3701708</td>\n",
       "      <td>502544</td>\n",
       "      <td>523837.957508</td>\n",
       "      <td>97.674751</td>\n",
       "      <td>95.935011</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8h_p2</td>\n",
       "      <td>8h_p2</td>\n",
       "      <td>2</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3330318</td>\n",
       "      <td>1694269</td>\n",
       "      <td>1594075</td>\n",
       "      <td>3288344</td>\n",
       "      <td>98.739640</td>\n",
       "      <td>48.476528</td>\n",
       "      <td>3254876</td>\n",
       "      <td>491336</td>\n",
       "      <td>523232.671659</td>\n",
       "      <td>97.734691</td>\n",
       "      <td>93.903922</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8h_p3</td>\n",
       "      <td>8h_p3</td>\n",
       "      <td>3</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>3623618</td>\n",
       "      <td>1789206</td>\n",
       "      <td>1790632</td>\n",
       "      <td>3579838</td>\n",
       "      <td>98.791815</td>\n",
       "      <td>50.019917</td>\n",
       "      <td>3537475</td>\n",
       "      <td>497838</td>\n",
       "      <td>523672.403392</td>\n",
       "      <td>97.622735</td>\n",
       "      <td>95.066686</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0              8h_p1       8h_p1          1  2023-02-28/P7_R1.fastq.bz2   \n",
       "1              8h_p2       8h_p2          2  2023-02-28/P7_R1.fastq.bz2   \n",
       "2              8h_p3       8h_p3          3  2023-02-28/P7_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species    plot_group  \\\n",
       "0  2023-02-28/P7_R2.fastq.bz2     D503     D701    l1Sp   human  Barcode test   \n",
       "1  2023-02-28/P7_R2.fastq.bz2     D503     D701    l2Sp   human  Barcode test   \n",
       "2  2023-02-28/P7_R2.fastq.bz2     D503     D701    l3Sp   human  Barcode test   \n",
       "\n",
       "  hue_name hue_value  hue_order P5_index_seq P7_index_seq         barcode_seq  \\\n",
       "0  Barcode      l1Sp          1     AGGATAGG     ATTACTCG    GGCTGCCATGCGACTA   \n",
       "1  Barcode      l2Sp          2     AGGATAGG     ATTACTCG  GGCTGCCATGCTGTCACG   \n",
       "2  Barcode      l3Sp          3     AGGATAGG     ATTACTCG    GGCTGCCATGCTGCGA   \n",
       "\n",
       "   N_total     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  \\\n",
       "0  3789831  1854972  1889882   3744854             98.813219    50.466106   \n",
       "1  3330318  1694269  1594075   3288344             98.739640    48.476528   \n",
       "2  3623618  1789206  1790632   3579838             98.791815    50.019917   \n",
       "\n",
       "   N_after_trim  N_UMI_observed  N_UMI_expected  percent_seqs_after_UMI_trim  \\\n",
       "0       3701708          502544   523837.957508                    97.674751   \n",
       "1       3254876          491336   523232.671659                    97.734691   \n",
       "2       3537475          497838   523672.403392                    97.622735   \n",
       "\n",
       "   percent_UMI_obs-vs-exp  N_after_downsample  \n",
       "0               95.935011             2000000  \n",
       "1               93.903922             2000000  \n",
       "2               95.066686             2000000  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False, \\\n",
    "                  check_input=False, AR_threads=12)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=4, overwrite=False)\n",
    "\n",
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=9, load_previous=True)\n",
    "\n",
    "# Trim UMIs:\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=9, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  common-seqs  8h_p6  8h_p1  8h_p4  8h_p2  8h_p7  8h_p9  8h_p3  8h_p5  8h_p8\n",
      "Collecting alignment statistics, from sample:  common-seqs  8h_p3  8h_p7  8h_p2  8h_p1  8h_p5  8h_p6  8h_p8  8h_p9  8h_p4Using common sequences...\n",
      "Collecting stats from:  8h_p1  8h_p2  8h_p3  8h_p6  8h_p7  8h_p8  8h_p9  8h_p4  8h_p5"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8h_p1</td>\n",
       "      <td>8h_p1</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>3789831</td>\n",
       "      <td>1854972</td>\n",
       "      <td>1889882</td>\n",
       "      <td>3744854</td>\n",
       "      <td>98.813219</td>\n",
       "      <td>50.466106</td>\n",
       "      <td>3701708</td>\n",
       "      <td>502544</td>\n",
       "      <td>523837.957508</td>\n",
       "      <td>97.674751</td>\n",
       "      <td>95.935011</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1976609.0</td>\n",
       "      <td>68.947728</td>\n",
       "      <td>31.052272</td>\n",
       "      <td>7.087188</td>\n",
       "      <td>98.83045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8h_p2</td>\n",
       "      <td>8h_p2</td>\n",
       "      <td>2</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3330318</td>\n",
       "      <td>1694269</td>\n",
       "      <td>1594075</td>\n",
       "      <td>3288344</td>\n",
       "      <td>98.739640</td>\n",
       "      <td>48.476528</td>\n",
       "      <td>3254876</td>\n",
       "      <td>491336</td>\n",
       "      <td>523232.671659</td>\n",
       "      <td>97.734691</td>\n",
       "      <td>93.903922</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1969357.0</td>\n",
       "      <td>68.129648</td>\n",
       "      <td>31.870352</td>\n",
       "      <td>7.179298</td>\n",
       "      <td>98.46785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8h_p3</td>\n",
       "      <td>8h_p3</td>\n",
       "      <td>3</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>3623618</td>\n",
       "      <td>1789206</td>\n",
       "      <td>1790632</td>\n",
       "      <td>3579838</td>\n",
       "      <td>98.791815</td>\n",
       "      <td>50.019917</td>\n",
       "      <td>3537475</td>\n",
       "      <td>497838</td>\n",
       "      <td>523672.403392</td>\n",
       "      <td>97.622735</td>\n",
       "      <td>95.066686</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1976978.0</td>\n",
       "      <td>68.889942</td>\n",
       "      <td>31.110058</td>\n",
       "      <td>7.191076</td>\n",
       "      <td>98.84890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0              8h_p1       8h_p1          1  2023-02-28/P7_R1.fastq.bz2   \n",
       "1              8h_p2       8h_p2          2  2023-02-28/P7_R1.fastq.bz2   \n",
       "2              8h_p3       8h_p3          3  2023-02-28/P7_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species    plot_group  \\\n",
       "0  2023-02-28/P7_R2.fastq.bz2     D503     D701    l1Sp   human  Barcode test   \n",
       "1  2023-02-28/P7_R2.fastq.bz2     D503     D701    l2Sp   human  Barcode test   \n",
       "2  2023-02-28/P7_R2.fastq.bz2     D503     D701    l3Sp   human  Barcode test   \n",
       "\n",
       "  hue_name hue_value  hue_order P5_index_seq P7_index_seq         barcode_seq  \\\n",
       "0  Barcode      l1Sp          1     AGGATAGG     ATTACTCG    GGCTGCCATGCGACTA   \n",
       "1  Barcode      l2Sp          2     AGGATAGG     ATTACTCG  GGCTGCCATGCTGTCACG   \n",
       "2  Barcode      l3Sp          3     AGGATAGG     ATTACTCG    GGCTGCCATGCTGCGA   \n",
       "\n",
       "   N_total     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  \\\n",
       "0  3789831  1854972  1889882   3744854             98.813219    50.466106   \n",
       "1  3330318  1694269  1594075   3288344             98.739640    48.476528   \n",
       "2  3623618  1789206  1790632   3579838             98.791815    50.019917   \n",
       "\n",
       "   N_after_trim  N_UMI_observed  N_UMI_expected  percent_seqs_after_UMI_trim  \\\n",
       "0       3701708          502544   523837.957508                    97.674751   \n",
       "1       3254876          491336   523232.671659                    97.734691   \n",
       "2       3537475          497838   523672.403392                    97.622735   \n",
       "\n",
       "   percent_UMI_obs-vs-exp  N_after_downsample   N_mapped  \\\n",
       "0               95.935011             2000000  1976609.0   \n",
       "1               93.903922             2000000  1969357.0   \n",
       "2               95.066686             2000000  1976978.0   \n",
       "\n",
       "   percent_single_annotation  percent_multiple_annotation  \\\n",
       "0                  68.947728                    31.052272   \n",
       "1                  68.129648                    31.870352   \n",
       "2                  68.889942                    31.110058   \n",
       "\n",
       "   percent_multiple_codons  Mapping_percent  \n",
       "0                 7.087188         98.83045  \n",
       "1                 7.179298         98.46785  \n",
       "2                 7.191076         98.84890  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Align reads to unmasked reference tRNAs:\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=12, overwrite=False, load_previous=False)\n",
    "\n",
    "# Collect alignment statistics:\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=True)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=12, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  common-seqs  8h_p9  8h_p5  8h_p7  8h_p3  8h_p4  8h_p1  8h_p2  8h_p6  8h_p8\n",
      "Collecting alignment statistics, from sample:  common-seqs  8h_p9  8h_p7  8h_p3  8h_p1  8h_p2  8h_p6  8h_p8  8h_p4  8h_p5Using common sequences...\n",
      "Collecting stats from:  8h_p1  8h_p3  8h_p2  8h_p6  8h_p7  8h_p8  8h_p9  8h_p4  8h_p5"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8h_p1</td>\n",
       "      <td>8h_p1</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>3789831</td>\n",
       "      <td>1854972</td>\n",
       "      <td>1889882</td>\n",
       "      <td>3744854</td>\n",
       "      <td>98.813219</td>\n",
       "      <td>50.466106</td>\n",
       "      <td>3701708</td>\n",
       "      <td>502544</td>\n",
       "      <td>523837.957508</td>\n",
       "      <td>97.674751</td>\n",
       "      <td>95.935011</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1978910.0</td>\n",
       "      <td>75.396102</td>\n",
       "      <td>24.603898</td>\n",
       "      <td>2.779358</td>\n",
       "      <td>98.94550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8h_p2</td>\n",
       "      <td>8h_p2</td>\n",
       "      <td>2</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3330318</td>\n",
       "      <td>1694269</td>\n",
       "      <td>1594075</td>\n",
       "      <td>3288344</td>\n",
       "      <td>98.739640</td>\n",
       "      <td>48.476528</td>\n",
       "      <td>3254876</td>\n",
       "      <td>491336</td>\n",
       "      <td>523232.671659</td>\n",
       "      <td>97.734691</td>\n",
       "      <td>93.903922</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1972529.0</td>\n",
       "      <td>74.093917</td>\n",
       "      <td>25.906083</td>\n",
       "      <td>2.865864</td>\n",
       "      <td>98.62645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8h_p3</td>\n",
       "      <td>8h_p3</td>\n",
       "      <td>3</td>\n",
       "      <td>2023-02-28/P7_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P7_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Barcode test</td>\n",
       "      <td>Barcode</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>3623618</td>\n",
       "      <td>1789206</td>\n",
       "      <td>1790632</td>\n",
       "      <td>3579838</td>\n",
       "      <td>98.791815</td>\n",
       "      <td>50.019917</td>\n",
       "      <td>3537475</td>\n",
       "      <td>497838</td>\n",
       "      <td>523672.403392</td>\n",
       "      <td>97.622735</td>\n",
       "      <td>95.066686</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1979241.0</td>\n",
       "      <td>75.182810</td>\n",
       "      <td>24.817190</td>\n",
       "      <td>2.826336</td>\n",
       "      <td>98.96205</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0              8h_p1       8h_p1          1  2023-02-28/P7_R1.fastq.bz2   \n",
       "1              8h_p2       8h_p2          2  2023-02-28/P7_R1.fastq.bz2   \n",
       "2              8h_p3       8h_p3          3  2023-02-28/P7_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species    plot_group  \\\n",
       "0  2023-02-28/P7_R2.fastq.bz2     D503     D701    l1Sp   human  Barcode test   \n",
       "1  2023-02-28/P7_R2.fastq.bz2     D503     D701    l2Sp   human  Barcode test   \n",
       "2  2023-02-28/P7_R2.fastq.bz2     D503     D701    l3Sp   human  Barcode test   \n",
       "\n",
       "  hue_name hue_value  hue_order P5_index_seq P7_index_seq         barcode_seq  \\\n",
       "0  Barcode      l1Sp          1     AGGATAGG     ATTACTCG    GGCTGCCATGCGACTA   \n",
       "1  Barcode      l2Sp          2     AGGATAGG     ATTACTCG  GGCTGCCATGCTGTCACG   \n",
       "2  Barcode      l3Sp          3     AGGATAGG     ATTACTCG    GGCTGCCATGCTGCGA   \n",
       "\n",
       "   N_total     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  \\\n",
       "0  3789831  1854972  1889882   3744854             98.813219    50.466106   \n",
       "1  3330318  1694269  1594075   3288344             98.739640    48.476528   \n",
       "2  3623618  1789206  1790632   3579838             98.791815    50.019917   \n",
       "\n",
       "   N_after_trim  N_UMI_observed  N_UMI_expected  percent_seqs_after_UMI_trim  \\\n",
       "0       3701708          502544   523837.957508                    97.674751   \n",
       "1       3254876          491336   523232.671659                    97.734691   \n",
       "2       3537475          497838   523672.403392                    97.622735   \n",
       "\n",
       "   percent_UMI_obs-vs-exp  N_after_downsample   N_mapped  \\\n",
       "0               95.935011             2000000  1978910.0   \n",
       "1               93.903922             2000000  1972529.0   \n",
       "2               95.066686             2000000  1979241.0   \n",
       "\n",
       "   percent_single_annotation  percent_multiple_annotation  \\\n",
       "0                  75.396102                    24.603898   \n",
       "1                  74.093917                    25.906083   \n",
       "2                  75.182810                    24.817190   \n",
       "\n",
       "   percent_multiple_codons  Mapping_percent  \n",
       "0                 2.779358         98.94550  \n",
       "1                 2.865864         98.62645  \n",
       "2                 2.826336         98.96205  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make new folders for the masked alignment/stats:\n",
    "dir_dict_masked = copy.deepcopy(dir_dict)\n",
    "dir_dict_masked['align_dir'] = 'SWalign_masked'\n",
    "dir_dict_masked['stats_dir'] = 'stats_collection_masked'\n",
    "\n",
    "# Align reads to masked reference tRNAs:\n",
    "align_obj = SWIPE_align(dir_dict_masked, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=12, overwrite=False, load_previous=False)\n",
    "\n",
    "# Collect alignment statistics:\n",
    "stats_obj = STATS_collection(dir_dict_masked, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=True)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=12, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
