{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Input settings ###\n",
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests/data/AdapterRemoval\n"
     ]
    }
   ],
   "source": [
    "### Run AdapterRemoval ###\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False, \\\n",
    "                  check_input=False)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=2, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests/data/BC_split\n",
      "Loaded results from previous run... Not running barcode split.\n"
     ]
    }
   ],
   "source": [
    "### Split files based on barcodes ###\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=6, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests/data/UMI_trimmed\n",
      "Loaded results from previous run... Not running UMI trimming.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1_v1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-12-28/P1_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P1_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>4882095</td>\n",
       "      <td>223688</td>\n",
       "      <td>4625327</td>\n",
       "      <td>4849015</td>\n",
       "      <td>99.322422</td>\n",
       "      <td>95.386939</td>\n",
       "      <td>4801265</td>\n",
       "      <td>515457</td>\n",
       "      <td>524232.737118</td>\n",
       "      <td>98.344358</td>\n",
       "      <td>98.325985</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p3_v2</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>2022-12-28/P3_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P3_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3002203</td>\n",
       "      <td>149777</td>\n",
       "      <td>2828460</td>\n",
       "      <td>2978237</td>\n",
       "      <td>99.201720</td>\n",
       "      <td>94.970951</td>\n",
       "      <td>2951343</td>\n",
       "      <td>492411</td>\n",
       "      <td>522405.132752</td>\n",
       "      <td>98.305911</td>\n",
       "      <td>94.258454</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0m_p1_v2</td>\n",
       "      <td>100p_0m</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P4_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P4_R2.fastq.bz2</td>\n",
       "      <td>D502</td>\n",
       "      <td>D701</td>\n",
       "      <td>l5Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p_0m</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCTCTAT</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCAACGCATC</td>\n",
       "      <td>2989616</td>\n",
       "      <td>163303</td>\n",
       "      <td>2798525</td>\n",
       "      <td>2961828</td>\n",
       "      <td>99.070516</td>\n",
       "      <td>94.486412</td>\n",
       "      <td>2932028</td>\n",
       "      <td>494676</td>\n",
       "      <td>522334.473460</td>\n",
       "      <td>98.073733</td>\n",
       "      <td>94.704835</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0           100p1_v1        100p          1  2022-12-28/P1_R1.fastq.bz2   \n",
       "1           100p3_v2        100p          3  2022-12-28/P3_R1.fastq.bz2   \n",
       "2           0m_p1_v2     100p_0m          1  2023-02-28/P4_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2022-12-28/P1_R2.fastq.bz2     D501     D701    l1Sp   human   \n",
       "1  2022-12-28/P3_R2.fastq.bz2     D501     D703    l2Sp   human   \n",
       "2  2023-02-28/P4_R2.fastq.bz2     D502     D701    l5Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge   100p_0m          2     GCCTCTAT   \n",
       "\n",
       "  P7_index_seq          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0     ATTACTCG     GGCTGCCATGCGACTA  4882095  223688  4625327   4849015   \n",
       "1     CGCTCATT   GGCTGCCATGCTGTCACG  3002203  149777  2828460   2978237   \n",
       "2     ATTACTCG  GGCTGCCATGCAACGCATC  2989616  163303  2798525   2961828   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.322422    95.386939       4801265          515457   \n",
       "1             99.201720    94.970951       2951343          492411   \n",
       "2             99.070516    94.486412       2932028          494676   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524232.737118                    98.344358               98.325985   \n",
       "1   522405.132752                    98.305911               94.258454   \n",
       "2   522334.473460                    98.073733               94.704835   \n",
       "\n",
       "   N_after_downsample  \n",
       "0             2000000  \n",
       "1             2000000  \n",
       "2             2000000  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=6, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_mtests/data/SWalign\n",
      "Loaded results from previous run... Not running alignment.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1_v1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-12-28/P1_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P1_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>4882095</td>\n",
       "      <td>223688</td>\n",
       "      <td>4625327</td>\n",
       "      <td>4849015</td>\n",
       "      <td>99.322422</td>\n",
       "      <td>95.386939</td>\n",
       "      <td>4801265</td>\n",
       "      <td>515457</td>\n",
       "      <td>524232.737118</td>\n",
       "      <td>98.344358</td>\n",
       "      <td>98.325985</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1988871</td>\n",
       "      <td>78.933475</td>\n",
       "      <td>21.066525</td>\n",
       "      <td>2.359127</td>\n",
       "      <td>99.44355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p3_v2</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>2022-12-28/P3_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P3_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3002203</td>\n",
       "      <td>149777</td>\n",
       "      <td>2828460</td>\n",
       "      <td>2978237</td>\n",
       "      <td>99.201720</td>\n",
       "      <td>94.970951</td>\n",
       "      <td>2951343</td>\n",
       "      <td>492411</td>\n",
       "      <td>522405.132752</td>\n",
       "      <td>98.305911</td>\n",
       "      <td>94.258454</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1977342</td>\n",
       "      <td>76.058416</td>\n",
       "      <td>23.941584</td>\n",
       "      <td>2.572494</td>\n",
       "      <td>98.86710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0m_p1_v2</td>\n",
       "      <td>100p_0m</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P4_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P4_R2.fastq.bz2</td>\n",
       "      <td>D502</td>\n",
       "      <td>D701</td>\n",
       "      <td>l5Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p_0m</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCTCTAT</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCAACGCATC</td>\n",
       "      <td>2989616</td>\n",
       "      <td>163303</td>\n",
       "      <td>2798525</td>\n",
       "      <td>2961828</td>\n",
       "      <td>99.070516</td>\n",
       "      <td>94.486412</td>\n",
       "      <td>2932028</td>\n",
       "      <td>494676</td>\n",
       "      <td>522334.473460</td>\n",
       "      <td>98.073733</td>\n",
       "      <td>94.704835</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1975184</td>\n",
       "      <td>73.976349</td>\n",
       "      <td>26.023651</td>\n",
       "      <td>2.819383</td>\n",
       "      <td>98.75920</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0           100p1_v1        100p          1  2022-12-28/P1_R1.fastq.bz2   \n",
       "1           100p3_v2        100p          3  2022-12-28/P3_R1.fastq.bz2   \n",
       "2           0m_p1_v2     100p_0m          1  2023-02-28/P4_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2022-12-28/P1_R2.fastq.bz2     D501     D701    l1Sp   human   \n",
       "1  2022-12-28/P3_R2.fastq.bz2     D501     D703    l2Sp   human   \n",
       "2  2023-02-28/P4_R2.fastq.bz2     D502     D701    l5Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge   100p_0m          2     GCCTCTAT   \n",
       "\n",
       "  P7_index_seq          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0     ATTACTCG     GGCTGCCATGCGACTA  4882095  223688  4625327   4849015   \n",
       "1     CGCTCATT   GGCTGCCATGCTGTCACG  3002203  149777  2828460   2978237   \n",
       "2     ATTACTCG  GGCTGCCATGCAACGCATC  2989616  163303  2798525   2961828   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.322422    95.386939       4801265          515457   \n",
       "1             99.201720    94.970951       2951343          492411   \n",
       "2             99.070516    94.486412       2932028          494676   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524232.737118                    98.344358               98.325985   \n",
       "1   522405.132752                    98.305911               94.258454   \n",
       "2   522334.473460                    98.073733               94.704835   \n",
       "\n",
       "   N_after_downsample  N_mapped  percent_single_annotation  \\\n",
       "0             2000000   1988871                  78.933475   \n",
       "1             2000000   1977342                  76.058416   \n",
       "2             2000000   1975184                  73.976349   \n",
       "\n",
       "   percent_multiple_annotation  percent_multiple_codons  Mapping_percent  \n",
       "0                    21.066525                 2.359127         99.44355  \n",
       "1                    23.941584                 2.572494         98.86710  \n",
       "2                    26.023651                 2.819383         98.75920  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=4, overwrite=False, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Collecting stats from:  100p1_v1  100p3_v2  0m_p1_v2  0m_p2_v2  0m_p3_v2  0m_p4_v2  85p1_v1  85p2_v1  70p2_v2  85p4_v1  85p2_v2  70p1_v1  85p1_v2  85p3_v2  70p1_v2  85p4_v2  70p5_v2  55p1_v1  70p3_v2  70p4_v2  55p3_v1  55p3_v2  55p2_v2  55p1_v2  40p2_v1  40p1_v1  55p4_v2  55p5_v2  40p2_v2  40p4_v1  40p1_v2  40p3_v2  25p1_v1  40p4_v2  25p1_v2  25p3_v2  25p5_v2  25p2_v2  25p4_v2  25p6_v2  10p1_v1  10p3_v1  10p4_v2  10p1_v2  0p1_v1  10p2_v2  10p3_v2  10p5_v2  0p4_v1  0p2_v1  0p1_v2  0p3_v2  0p2_v2  0p4_v2  0p_90m1_v1  0p_90m2_v1  0p_90m3_v1  0p_90m4_v1  0m_1  8m_1  4m_1  32m_1  16m_1  2h_1  1h_1  4h_1  8h_1  40h_NoOx_1  40h_1  16h_1  4m_2  0m_2  16m_2  8m_2  32m_2  1h_2  2h_2  8h_2  4h_2  16h_2  40h_NoOx_2  40h_2  0m_3  16m_3  8m_3  4m_3  32m_3  1h_3  4h_3  2h_3  8h_3  40h_3  16h_3  8m_4  40h_NoOx_3  0m_4  32m_4  4m_4  16m_4  1h_4  2h_4  4h_4  8h_4  16h_4  8h_p1  40h_NoOx_4  40h_4  8h_p2  8h_p4  8h_p3  8h_p5  8h_p7  8h_p8  8h_p9  8h_p6"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>align_gap</th>\n",
       "      <th>fmax_score&gt;0.9</th>\n",
       "      <th>count</th>\n",
       "      <th>UMIcount</th>\n",
       "      <th>UMI_percent_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1_v1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p1_v1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>50</td>\n",
       "      <td>49</td>\n",
       "      <td>98.00458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p1_v1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>100.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode species  \\\n",
       "0           100p1_v1        100p          1    l1Sp   human   \n",
       "1           100p1_v1        100p          1    l1Sp   human   \n",
       "2           100p1_v1        100p          1    l1Sp   human   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  \\\n",
       "0               True     False           A   AAA       TTT   eColiLys   \n",
       "1               True     False           A   AAA       TTT   eColiLys   \n",
       "2               True     False           A   AAA       TTT   eColiLys   \n",
       "\n",
       "   align_gap  fmax_score>0.9  count  UMIcount  UMI_percent_exp  \n",
       "0      False           False      1         1        100.00000  \n",
       "1      False            True     50        49         98.00458  \n",
       "2       True           False      1         1        100.00000  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=False)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=8, load_previous=True)\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=False, pull_default=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ecoli control:\n",
    "plot_obj.plot_Ecoli_ctr(min_obs=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample/group:  Acylation-half-life  Barcode test  Charge-titration\n",
      "Now plotting sample/group:  Acylation-half-life  Barcode test  Charge-titration"
     ]
    }
   ],
   "source": [
    "# Codon abundance barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_abundance_grp', \\\n",
    "                        group=True, min_obs=500)\n",
    "\n",
    "# Codon charge barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  100p1_v1  100p3_v2  0m_p1_v2  0m_p4_v2  85p1_v1  85p2_v1  85p4_v1  85p1_v2  85p2_v2  85p3_v2  0m_p2_v2  0m_p3_v2  85p4_v2  70p1_v1  55p3_v1  55p1_v1  70p1_v2  70p5_v2  70p3_v2  55p1_v2  55p2_v2  55p3_v2  70p4_v2  70p2_v2  40p1_v1  55p5_v2  55p4_v2  40p2_v1  40p4_v1  40p2_v2  40p4_v2  40p3_v2  40p1_v2  25p1_v2  25p1_v1  25p2_v2  25p3_v2  25p6_v2  10p1_v1  25p4_v2  25p5_v2  10p3_v1  10p5_v2  10p2_v2  10p4_v2  10p1_v2  10p3_v2  0p1_v1  0p2_v1  0p1_v2  0p4_v1  0p3_v2  0p2_v2  0p4_v2  4m_1  0p_90m1_v1  0m_1  0p_90m2_v1  0p_90m4_v1  0p_90m3_v1  8m_1  16m_1  32m_1  40h_1  2h_1  1h_1  8h_1  4h_1  16h_1  40h_NoOx_1  4m_2  0m_2  8m_2  32m_2  16m_2  40h_NoOx_2  2h_2  4h_2  0m_3  16h_2  1h_2  40h_2  8h_2  4m_3  8m_3  16m_3  32m_3  1h_3  2h_3  8h_3  4h_3  40h_NoOx_3  40h_3  16h_3  0m_4  4m_4  8m_4  16m_4  1h_4  32m_4  2h_4  16h_4  4h_4  8h_4  40h_4  8h_p1  40h_NoOx_4  8h_p2  8h_p3  8h_p5  8h_p4  8h_p6  8h_p7  8h_p8  8h_p9\n",
      "Now plotting sample:  100p1_v1  100p3_v2  0m_p1_v2  0m_p2_v2  0m_p3_v2  0m_p4_v2  85p1_v1  85p2_v1  85p4_v1  85p1_v2  85p2_v2  85p3_v2  85p4_v2  70p1_v1  70p1_v2  70p2_v2  70p3_v2  70p4_v2  70p5_v2  55p1_v1  55p3_v1  55p1_v2  55p2_v2  55p3_v2  55p4_v2  55p5_v2  40p1_v1  40p2_v1  40p4_v1  40p1_v2  40p2_v2  40p3_v2  40p4_v2  25p1_v1  25p1_v2  25p2_v2  25p3_v2  25p4_v2  25p5_v2  25p6_v2  10p1_v1  10p3_v1  10p1_v2  10p2_v2  10p3_v2  10p4_v2  10p5_v2  0p1_v1  0p2_v1  0p4_v1  0p1_v2  0p2_v2  0p3_v2  0p4_v2  0p_90m1_v1  0p_90m2_v1  0p_90m3_v1  0p_90m4_v1  0m_1  4m_1  8m_1  16m_1  32m_1  1h_1  2h_1  4h_1  8h_1  16h_1  40h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  32m_2  1h_2  2h_2  4h_2  8h_2  16h_2  40h_2  40h_NoOx_2  0m_3  4m_3  8m_3  16m_3  32m_3  1h_3  2h_3  4h_3  8h_3  16h_3  40h_3  40h_NoOx_3  0m_4  4m_4  8m_4  16m_4  32m_4  1h_4  2h_4  4h_4  8h_4  16h_4  40h_4  40h_NoOx_4  8h_p1  8h_p2  8h_p3  8h_p4  8h_p5  8h_p6  8h_p7  8h_p8  8h_p9\n",
      "Now collecting data for sample:  100p1_v1  100p3_v2  0m_p2_v2  0m_p1_v2  0m_p4_v2  0m_p3_v2  85p1_v1  85p2_v1  85p1_v2  85p4_v1  85p3_v2  85p2_v2  85p4_v2  70p1_v1  70p5_v2  55p3_v1  55p1_v1  55p1_v2  70p1_v2  70p2_v2  70p3_v2  55p2_v2  55p3_v2  70p4_v2  55p5_v2  55p4_v2  40p2_v1  40p1_v1  40p4_v1  40p1_v2  40p2_v2  40p3_v2  40p4_v2  25p2_v2  25p1_v1  25p1_v2  25p5_v2  25p3_v2  25p4_v2  10p1_v1  25p6_v2  10p1_v2  10p3_v1  10p5_v2  10p2_v2  10p3_v2  10p4_v2  0p1_v1  0p2_v2  0p2_v1  0p4_v1  0p1_v2  0p3_v2  0p_90m1_v1  0p4_v2  0p_90m3_v1  0p_90m2_v1  4m_1  0p_90m4_v1  0m_1  16m_1  32m_1  8m_1  1h_1  2h_1  4h_1  16h_1  40h_NoOx_1  8h_1  40h_1  0m_2  4m_2  16m_2  8m_2  32m_2  16h_2  1h_2  2h_2  4h_2  8h_2  40h_2  40h_NoOx_2  4m_3  0m_3  8m_3  32m_3  16m_3  8h_3  4h_3  2h_3  1h_3  40h_NoOx_3  16h_3  40h_3  0m_4  4m_4  8m_4  32m_4  16m_4  16h_4  8h_4  2h_4  4h_4  1h_4  40h_4  40h_NoOx_4  8h_p1  8h_p2  8h_p3  8h_p4  8h_p5  8h_p8  8h_p7  8h_p6  8h_p9\n",
      "Now plotting sample:  100p1_v1  100p3_v2  0m_p1_v2  0m_p2_v2  0m_p3_v2  0m_p4_v2  85p1_v1  85p2_v1  85p4_v1  85p1_v2  85p2_v2  85p3_v2  85p4_v2  70p1_v1  70p1_v2  70p2_v2  70p3_v2  70p4_v2  70p5_v2  55p1_v1  55p3_v1  55p1_v2  55p2_v2  55p3_v2  55p4_v2  55p5_v2  40p1_v1  40p2_v1  40p4_v1  40p1_v2  40p2_v2  40p3_v2  40p4_v2  25p1_v1  25p1_v2  25p2_v2  25p3_v2  25p4_v2  25p5_v2  25p6_v2  10p1_v1  10p3_v1  10p1_v2  10p2_v2  10p3_v2  10p4_v2  10p5_v2  0p1_v1  0p2_v1  0p4_v1  0p1_v2  0p2_v2  0p3_v2  0p4_v2  0p_90m1_v1  0p_90m2_v1  0p_90m3_v1  0p_90m4_v1  0m_1  4m_1  8m_1  16m_1  32m_1  1h_1  2h_1  4h_1  8h_1  16h_1  40h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  32m_2  1h_2  2h_2  4h_2  8h_2  16h_2  40h_2  40h_NoOx_2  0m_3  4m_3  8m_3  16m_3  32m_3  1h_3  2h_3  4h_3  8h_3  16h_3  40h_3  40h_NoOx_3  0m_4  4m_4  8m_4  16m_4  32m_4  1h_4  2h_4  4h_4  8h_4  16h_4  40h_4  40h_NoOx_4  8h_p1  8h_p2  8h_p3  8h_p4  8h_p5  8h_p6  8h_p7  8h_p8  8h_p9"
     ]
    }
   ],
   "source": [
    "# Coverage plots for cyto/mito transcripts:\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_behrens_norm', n_jobs=12)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_behrens_norm', n_jobs=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  100p1_v1  100p3_v2  0m_p1_v2  0m_p2_v2  0m_p3_v2  0m_p4_v2  85p1_v1  85p2_v1  85p4_v1  85p1_v2  85p2_v2  85p3_v2  85p4_v2  70p1_v2  70p1_v1  70p2_v2  70p3_v2  70p5_v2  70p4_v2  55p1_v1  55p3_v1  55p1_v2  55p2_v2  55p3_v2  55p4_v2  40p1_v1  55p5_v2  40p2_v1  40p4_v1  40p1_v2  40p2_v2  40p3_v2  40p4_v2  25p1_v1  25p1_v2  25p2_v2  25p3_v2  25p4_v2  25p5_v2  25p6_v2  10p1_v1  10p3_v1  10p1_v2  10p2_v2  10p3_v2  10p4_v2  10p5_v2  0p1_v1  0p4_v1  0p2_v1  0p1_v2  0p2_v2  0p3_v2  0p4_v2  0p_90m1_v1  0p_90m2_v1  0p_90m3_v1  0p_90m4_v1  0m_1  4m_1  8m_1  32m_1  16m_1  1h_1  2h_1  4h_1  8h_1  16h_1  40h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  32m_2  1h_2  2h_2  4h_2  8h_2  16h_2  40h_2  40h_NoOx_2  0m_3  4m_3  8m_3  16m_3  32m_3  2h_3  1h_3  4h_3  8h_3  16h_3  40h_3  40h_NoOx_3  0m_4  4m_4  16m_4  8m_4  32m_4  1h_4  4h_4  2h_4  8h_4  16h_4  40h_NoOx_4  40h_4  8h_p2  8h_p1  8h_p3  8h_p4  8h_p5  8h_p6  8h_p7  8h_p8  8h_p9\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# UMI logo:\n",
    "plot_obj.plot_UMI_logo(n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  100p1_v1  100p3_v2  0m_p1_v2  0m_p2_v2  0m_p3_v2  0m_p4_v2  85p1_v1  85p2_v1  85p4_v1  85p1_v2  85p3_v2  85p2_v2  85p4_v2  70p1_v1  70p1_v2  70p2_v2  70p3_v2  70p4_v2  70p5_v2  55p1_v1  55p3_v1  55p1_v2  55p3_v2  55p2_v2  55p4_v2  55p5_v2  40p1_v1  40p2_v1  40p4_v1  40p3_v2  40p2_v2  40p1_v2  40p4_v2  25p1_v1  25p1_v2  25p2_v2  25p3_v2  25p4_v2  25p5_v2  25p6_v2  10p1_v1  10p3_v1  10p1_v2  10p2_v2  10p3_v2  10p4_v2  10p5_v2  0p1_v1  0p2_v1  0p4_v1  0p2_v2  0p1_v2  0p3_v2  0p4_v2  0p_90m1_v1  0p_90m2_v1  0p_90m3_v1  0p_90m4_v1  0m_1  4m_1  8m_1  16m_1  32m_1  2h_1  1h_1  4h_1  8h_1  40h_1  16h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  32m_2  1h_2  2h_2  4h_2  8h_2  16h_2  40h_2  40h_NoOx_2  0m_3  16m_3  4m_3  8m_3  32m_3  1h_3  2h_3  4h_3  8h_3  16h_3  40h_3  40h_NoOx_3  0m_4  4m_4  16m_4  8m_4  32m_4  1h_4  4h_4  2h_4  8h_4  16h_4  40h_4  40h_NoOx_4  8h_p1  8h_p3  8h_p2  8h_p4  8h_p5  8h_p6  8h_p7  8h_p8  8h_p9\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  100p1_v1  100p3_v2  0m_p1_v2  0m_p2_v2  0m_p3_v2  0m_p4_v2  85p1_v1  85p2_v1  85p4_v1  85p1_v2  85p2_v2  85p3_v2  85p4_v2  70p2_v2  70p1_v1  70p1_v2  70p3_v2  70p5_v2  70p4_v2  55p1_v1  55p3_v1  55p3_v2  55p1_v2  55p2_v2  55p4_v2  55p5_v2  40p1_v1  40p2_v1  40p4_v1  40p3_v2  40p2_v2  40p1_v2  40p4_v2  25p1_v1  25p1_v2  25p2_v2  25p3_v2  25p4_v2  25p6_v2  25p5_v2  10p1_v1  10p3_v1  10p1_v2  10p2_v2  10p3_v2  10p4_v2  10p5_v2  0p1_v1  0p2_v1  0p4_v1  0p2_v2  0p1_v2  0p3_v2  0p4_v2  0p_90m1_v1  0p_90m2_v1  0p_90m3_v1  0p_90m4_v1  0m_1  4m_1  8m_1  16m_1  32m_1  1h_1  2h_1  4h_1  8h_1  16h_1  40h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  1h_2  32m_2  4h_2  2h_2  16h_2  8h_2  40h_2  40h_NoOx_2  0m_3  16m_3  4m_3  8m_3  32m_3  1h_3  2h_3  4h_3  8h_3  16h_3  40h_NoOx_3  40h_3  0m_4  8m_4  4m_4  16m_4  1h_4  32m_4  2h_4  4h_4  8h_4  16h_4  40h_4  40h_NoOx_4  8h_p3  8h_p1  8h_p2  8h_p4  8h_p6  8h_p5  8h_p7  8h_p8  8h_p9\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  100p1_v1  100p3_v2  0m_p1_v2  0m_p2_v2  0m_p3_v2  0m_p4_v2  85p1_v1  85p2_v1  85p4_v1  85p1_v2  85p3_v2  85p2_v2  85p4_v2  70p1_v1  70p1_v2  70p2_v2  70p3_v2  70p4_v2  70p5_v2  55p1_v1  55p3_v1  55p1_v2  55p3_v2  55p2_v2  55p4_v2  55p5_v2  40p2_v1  40p1_v1  40p4_v1  40p2_v2  40p1_v2  40p3_v2  40p4_v2  25p1_v1  25p1_v2  25p2_v2  25p3_v2  25p4_v2  25p6_v2  25p5_v2  10p1_v1  10p3_v1  10p2_v2  10p1_v2  10p3_v2  10p4_v2  0p1_v1  10p5_v2  0p2_v1  0p1_v2  0p4_v1  0p2_v2  0p3_v2  0p_90m1_v1  0p4_v2  0p_90m2_v1  0p_90m3_v1  0p_90m4_v1  0m_1  4m_1  8m_1  16m_1  32m_1  1h_1  2h_1  4h_1  8h_1  16h_1  40h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  1h_2  32m_2  2h_2  4h_2  8h_2  16h_2  40h_2  40h_NoOx_2  0m_3  16m_3  8m_3  4m_3  32m_3  2h_3  1h_3  4h_3  16h_3  8h_3  40h_3  40h_NoOx_3  8m_4  4m_4  0m_4  16m_4  32m_4  1h_4  2h_4  4h_4  8h_4  16h_4  40h_4  40h_NoOx_4  8h_p1  8h_p2  8h_p3  8h_p4  8h_p5  8h_p6  8h_p7  8h_p8  8h_p9\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# Non-template nucleotides:\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample pairs:  (100p - 85p)  (100p - 0p)  (85p - 40p)  (85p - 0p)  (40p - 0p)  (0m - 40h)  (40h - 40h_NoOx)\n",
      "Now plotting sample pairs:  (100p - 85p)  (100p - 0p)  (85p - 40p)  (85p - 0p)  (40p - 0p)  (0m - 40h)  (40h - 40h_NoOx)\n",
      "Now plotting sample pairs:  (100p1_v1 - 100p3_v2)  (85p1_v1 - 85p4_v1)  (55p1_v1 - 55p3_v2)  (40p1_v1 - 40p2_v2)  (0p1_v1 - 0p2_v2)  (40h_1 - 40h_2)  (40h_NoOx_1 - 40h_NoOx_2)"
     ]
    }
   ],
   "source": [
    "# Plot of abundance and charge correlation:\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['100p', '100p', '85p', '85p', \\\n",
    "                                            '40p', '0m',  '40h'], \\\n",
    "                                           ['85p',  '0p',   '40p', '0p', \\\n",
    "                                            '0p',  '40h', '40h_NoOx']], \\\n",
    "                             sample_pairs_col='sample_name', \\\n",
    "                             plot_type='transcript', plot_name='tr_abundance_corr', \\\n",
    "                             min_obs=500, charge_plot=False, log=True)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['100p', '100p', '85p', '85p', \\\n",
    "                                            '40p', '0m',  '40h'], \\\n",
    "                                           ['85p',  '0p',   '40p', '0p', \\\n",
    "                                            '0p',  '40h', '40h_NoOx']], \\\n",
    "                             sample_pairs_col='sample_name', \\\n",
    "                             plot_type='codon', plot_name='codon_abundance_corr', \\\n",
    "                             min_obs=500, charge_plot=False, log=True)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['100p1_v1', '85p1_v1', '55p1_v1', \\\n",
    "                                                   '40p1_v1', '0p1_v1', '40h_1', '40h_NoOx_1'], \\\n",
    "                                                  ['100p3_v2', '85p4_v1', '55p3_v2', \\\n",
    "                                                   '40p2_v2', '0p2_v2', '40h_2', '40h_NoOx_2']], \\\n",
    "                             plot_type='codon', plot_name='codon_charge_corr', \\\n",
    "                             min_obs=500, charge_plot=True, one2one_corr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     common_seqs=common_seqs, ignore_common_count=False, \\\n",
    "                     overwrite_dir=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting stats from:  100p1_v1  0m_p2_v2  100p3_v2  0m_p1_v2  85p1_v1  0m_p4_v2  0m_p3_v2  85p2_v1  70p1_v1  85p2_v2  70p1_v2  85p1_v2  85p4_v1  70p2_v2  85p4_v2  85p3_v2  70p5_v2  70p4_v2  70p3_v2  55p3_v1  55p1_v2  55p1_v1  55p2_v2  55p3_v2  55p4_v2  40p2_v1  55p5_v2  40p1_v1  40p4_v1  40p1_v2  40p4_v2  40p2_v2  25p1_v1  40p3_v2  25p1_v2  25p2_v2  25p5_v2  25p3_v2  25p6_v2  10p1_v1  25p4_v2  10p1_v2  10p3_v1  10p4_v2  0p1_v1  10p2_v2  10p3_v2  10p5_v2  0p2_v1  0p4_v1  0p1_v2  0p3_v2  0p2_v2  0p4_v2  0p_90m3_v1  0p_90m1_v1  0m_1  0p_90m4_v1  0p_90m2_v1  4m_1  8m_1  32m_1  16m_1  2h_1  8h_1  4h_1  1h_1  16h_1  40h_1  40h_NoOx_1  4m_2  0m_2  8m_2  32m_2  16m_2  1h_2  2h_2  4h_2  8h_2  16h_2  0m_3  8m_3  40h_2  40h_NoOx_2  4m_3  16m_3  32m_3  1h_3  2h_3  4h_3  8h_3  16h_3  40h_3  40h_NoOx_3  0m_4  4m_4  32m_4  8m_4  16m_4  1h_4  2h_4  16h_4  4h_4  8h_4  40h_4  40h_NoOx_4  8h_p1  8h_p2  8h_p4  8h_p3  8h_p6  8h_p5  8h_p7  8h_p8  8h_p9"
     ]
    }
   ],
   "source": [
    "# Find mutations and save them:\n",
    "if False:\n",
    "    TM_obj.find_muts(n_jobs=8, unique_anno=True)\n",
    "    TM_obj.pickle_muts_write(pickle_name='saved_muts_unique-anno.pickle')\n",
    "else:\n",
    "    TM_obj.pickle_muts_read(pickle_name='saved_muts_unique-anno.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap/RT stops for charge titration samples:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='charge-tit_mut_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='mut', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['100p', '100p', '85p', '85p', '40p'], \\\n",
    "                                                 ['85p',  '0p',   '40p', '0p',  '0p']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='charge-tit_gap_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='gap', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['100p', '100p', '85p', '85p', '40p'], \\\n",
    "                                                 ['85p',  '0p',   '40p', '0p',  '0p']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='charge-tit_RTstops_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='RTstops', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['100p', '100p', '85p', '85p', '40p'], \\\n",
    "                                                 ['85p',  '0p',   '40p', '0p',  '0p']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap/RT stops for charge titration samples:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='hl_mut_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='mut', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['0m', '0m', '0m',  '4h',  '40h',], \\\n",
    "                                                 ['1h', '4h', '40h', '16h', '40h_NoOx']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='hl_gap_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='gap', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['0m', '0m', '0m',  '4h',  '40h',], \\\n",
    "                                                 ['1h', '4h', '40h', '16h', '40h_NoOx']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='hl_RTstops_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='RTstops', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['0m', '0m', '0m',  '4h',  '40h',], \\\n",
    "                                                 ['1h', '4h', '40h', '16h', '40h_NoOx']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
