{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:549: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:549: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/RT-comp\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Input settings ###\n",
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Run AdapterRemoval ###\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=True, \\\n",
    "                  AR_threads=6)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=2, overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Split files based on barcodes ###\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=True)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=12, load_previous=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downsampling UMI trimmed sequences to maximum 2000000 reads.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A70</td>\n",
       "      <td>A70</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P12_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P12_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D704</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>RT-comp</td>\n",
       "      <td>Polymerase-sample</td>\n",
       "      <td>Maxima_S1</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>GAGATTCC</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>3641249</td>\n",
       "      <td>363080</td>\n",
       "      <td>3247507</td>\n",
       "      <td>3610587</td>\n",
       "      <td>99.157926</td>\n",
       "      <td>89.944017</td>\n",
       "      <td>3565300</td>\n",
       "      <td>503559</td>\n",
       "      <td>523704.222531</td>\n",
       "      <td>97.914205</td>\n",
       "      <td>96.153321</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A71</td>\n",
       "      <td>A71</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P12_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P12_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D704</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>RT-comp</td>\n",
       "      <td>Polymerase-sample</td>\n",
       "      <td>Maxima_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>GAGATTCC</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3438157</td>\n",
       "      <td>229039</td>\n",
       "      <td>3179334</td>\n",
       "      <td>3408373</td>\n",
       "      <td>99.133722</td>\n",
       "      <td>93.280108</td>\n",
       "      <td>3365403</td>\n",
       "      <td>498815</td>\n",
       "      <td>523433.263275</td>\n",
       "      <td>97.883924</td>\n",
       "      <td>95.296771</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A72</td>\n",
       "      <td>A72</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P12_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P12_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D704</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>RT-comp</td>\n",
       "      <td>Polymerase-sample</td>\n",
       "      <td>Maxima_S3</td>\n",
       "      <td>5</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>GAGATTCC</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>4065779</td>\n",
       "      <td>297561</td>\n",
       "      <td>3734550</td>\n",
       "      <td>4032111</td>\n",
       "      <td>99.171918</td>\n",
       "      <td>92.620218</td>\n",
       "      <td>3979477</td>\n",
       "      <td>507719</td>\n",
       "      <td>524023.050517</td>\n",
       "      <td>97.877356</td>\n",
       "      <td>96.888677</td>\n",
       "      <td>2000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                A70         A70          1  2023-06-06/P12_R1.fastq.bz2   \n",
       "1                A71         A71          1  2023-06-06/P12_R1.fastq.bz2   \n",
       "2                A72         A72          1  2023-06-06/P12_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species plot_group  \\\n",
       "0  2023-06-06/P12_R2.fastq.bz2     D503     D704    l1Sp   human    RT-comp   \n",
       "1  2023-06-06/P12_R2.fastq.bz2     D503     D704    l2Sp   human    RT-comp   \n",
       "2  2023-06-06/P12_R2.fastq.bz2     D503     D704    l3Sp   human    RT-comp   \n",
       "\n",
       "            hue_name  hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0  Polymerase-sample  Maxima_S1          1     AGGATAGG     GAGATTCC   \n",
       "1  Polymerase-sample  Maxima_S2          3     AGGATAGG     GAGATTCC   \n",
       "2  Polymerase-sample  Maxima_S3          5     AGGATAGG     GAGATTCC   \n",
       "\n",
       "          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0    GGCTGCCATGCGACTA  3641249  363080  3247507   3610587   \n",
       "1  GGCTGCCATGCTGTCACG  3438157  229039  3179334   3408373   \n",
       "2    GGCTGCCATGCTGCGA  4065779  297561  3734550   4032111   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.157926    89.944017       3565300          503559   \n",
       "1             99.133722    93.280108       3365403          498815   \n",
       "2             99.171918    92.620218       3979477          507719   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   523704.222531                    97.914205               96.153321   \n",
       "1   523433.263275                    97.883924               95.296771   \n",
       "2   524023.050517                    97.877356               96.888677   \n",
       "\n",
       "   N_after_downsample  \n",
       "0             2000000  \n",
       "1             2000000  \n",
       "2             2000000  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=True)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=6, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  A76  A73  A74  A71  A72  A75  A70  A70_T  A71_T  A75_T  A72_T  A76_T  A74_T  common-seqs  A73_T\n",
      "Collecting alignment statistics, from sample:  A70  A71  A75  A76  A73  A74  A72  A70_T  common-seqs  A71_T  A74_T  A76_T  A72_T  A73_T  A75_T"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A70</td>\n",
       "      <td>A70</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P12_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P12_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D704</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>RT-comp</td>\n",
       "      <td>Polymerase-sample</td>\n",
       "      <td>Maxima_S1</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>GAGATTCC</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>3641249</td>\n",
       "      <td>363080</td>\n",
       "      <td>3247507</td>\n",
       "      <td>3610587</td>\n",
       "      <td>99.157926</td>\n",
       "      <td>89.944017</td>\n",
       "      <td>3565300</td>\n",
       "      <td>503559</td>\n",
       "      <td>523704.222531</td>\n",
       "      <td>97.914205</td>\n",
       "      <td>96.153321</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1904893.0</td>\n",
       "      <td>64.536276</td>\n",
       "      <td>35.463724</td>\n",
       "      <td>6.109215</td>\n",
       "      <td>95.24465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A71</td>\n",
       "      <td>A71</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P12_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P12_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D704</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>RT-comp</td>\n",
       "      <td>Polymerase-sample</td>\n",
       "      <td>Maxima_S2</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>GAGATTCC</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3438157</td>\n",
       "      <td>229039</td>\n",
       "      <td>3179334</td>\n",
       "      <td>3408373</td>\n",
       "      <td>99.133722</td>\n",
       "      <td>93.280108</td>\n",
       "      <td>3365403</td>\n",
       "      <td>498815</td>\n",
       "      <td>523433.263275</td>\n",
       "      <td>97.883924</td>\n",
       "      <td>95.296771</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1839343.0</td>\n",
       "      <td>64.155788</td>\n",
       "      <td>35.844212</td>\n",
       "      <td>6.244404</td>\n",
       "      <td>91.96715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A72</td>\n",
       "      <td>A72</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P12_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P12_R2.fastq.bz2</td>\n",
       "      <td>D503</td>\n",
       "      <td>D704</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>RT-comp</td>\n",
       "      <td>Polymerase-sample</td>\n",
       "      <td>Maxima_S3</td>\n",
       "      <td>5</td>\n",
       "      <td>AGGATAGG</td>\n",
       "      <td>GAGATTCC</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>4065779</td>\n",
       "      <td>297561</td>\n",
       "      <td>3734550</td>\n",
       "      <td>4032111</td>\n",
       "      <td>99.171918</td>\n",
       "      <td>92.620218</td>\n",
       "      <td>3979477</td>\n",
       "      <td>507719</td>\n",
       "      <td>524023.050517</td>\n",
       "      <td>97.877356</td>\n",
       "      <td>96.888677</td>\n",
       "      <td>2000000</td>\n",
       "      <td>1904997.0</td>\n",
       "      <td>64.601362</td>\n",
       "      <td>35.398638</td>\n",
       "      <td>6.004366</td>\n",
       "      <td>95.24985</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                A70         A70          1  2023-06-06/P12_R1.fastq.bz2   \n",
       "1                A71         A71          1  2023-06-06/P12_R1.fastq.bz2   \n",
       "2                A72         A72          1  2023-06-06/P12_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species plot_group  \\\n",
       "0  2023-06-06/P12_R2.fastq.bz2     D503     D704    l1Sp   human    RT-comp   \n",
       "1  2023-06-06/P12_R2.fastq.bz2     D503     D704    l2Sp   human    RT-comp   \n",
       "2  2023-06-06/P12_R2.fastq.bz2     D503     D704    l3Sp   human    RT-comp   \n",
       "\n",
       "            hue_name  hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0  Polymerase-sample  Maxima_S1          1     AGGATAGG     GAGATTCC   \n",
       "1  Polymerase-sample  Maxima_S2          3     AGGATAGG     GAGATTCC   \n",
       "2  Polymerase-sample  Maxima_S3          5     AGGATAGG     GAGATTCC   \n",
       "\n",
       "          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0    GGCTGCCATGCGACTA  3641249  363080  3247507   3610587   \n",
       "1  GGCTGCCATGCTGTCACG  3438157  229039  3179334   3408373   \n",
       "2    GGCTGCCATGCTGCGA  4065779  297561  3734550   4032111   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.157926    89.944017       3565300          503559   \n",
       "1             99.133722    93.280108       3365403          498815   \n",
       "2             99.171918    92.620218       3979477          507719   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   523704.222531                    97.914205               96.153321   \n",
       "1   523433.263275                    97.883924               95.296771   \n",
       "2   524023.050517                    97.877356               96.888677   \n",
       "\n",
       "   N_after_downsample   N_mapped  percent_single_annotation  \\\n",
       "0             2000000  1904893.0                  64.536276   \n",
       "1             2000000  1839343.0                  64.155788   \n",
       "2             2000000  1904997.0                  64.601362   \n",
       "\n",
       "   percent_multiple_annotation  percent_multiple_codons  Mapping_percent  \n",
       "0                    35.463724                 6.109215         95.24465  \n",
       "1                    35.844212                 6.244404         91.96715  \n",
       "2                    35.398638                 6.004366         95.24985  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=8, overwrite=True, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Collecting stats from:  A70  A71  A72  A75  A76  A70_T  A71_T  A72_T  A74_T  A73_T  A74  A73  A76_T  A75_T"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>align_gap</th>\n",
       "      <th>fmax_score&gt;0.9</th>\n",
       "      <th>count</th>\n",
       "      <th>UMIcount</th>\n",
       "      <th>UMI_percent_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A70</td>\n",
       "      <td>A70</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>100.000095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A70</td>\n",
       "      <td>A70</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>100.000286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A70</td>\n",
       "      <td>A70</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>100.008583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode species  \\\n",
       "0                A70         A70          1    l1Sp   human   \n",
       "1                A70         A70          1    l1Sp   human   \n",
       "2                A70         A70          1    l1Sp   human   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  \\\n",
       "0               True     False           A   AAA       TTT   eColiLys   \n",
       "1               True     False           A   AAA       TTT   eColiLys   \n",
       "2               True     False           C   AAA       TTT   eColiLys   \n",
       "\n",
       "   align_gap  fmax_score>0.9  count  UMIcount  UMI_percent_exp  \n",
       "0      False           False      2         2       100.000095  \n",
       "1      False            True      4         4       100.000286  \n",
       "2      False           False     91        91       100.008583  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=True)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=12, load_previous=False)\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=True, pull_default=False)\n",
    "\n",
    "# Add polymerase column to sample_df\n",
    "# to group plots according to this:\n",
    "sample_df['polymerase'] = [v.split('_')[0] for v in sample_df['hue_value']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ecoli control:\n",
    "plot_obj.plot_Ecoli_ctr()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample/group:  RT-comp\n",
      "Now plotting sample/group:  RT-comp"
     ]
    }
   ],
   "source": [
    "# Codon abundance barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_abundance_grp', \\\n",
    "                        group=True, min_obs=500)\n",
    "\n",
    "# Codon charge barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample pairs:  (A70 - A70_T)  (A71 - A71_T)  (A72 - A72_T)  (A73 - A73_T)  (A74 - A74_T)  (A75 - A75_T)  (A76 - A76_T)\n",
      "Now plotting sample pairs:  (A70 - A70_T)  (A71 - A71_T)  (A72 - A72_T)  (A73 - A73_T)  (A74 - A74_T)  (A75 - A75_T)  (A76 - A76_T)"
     ]
    }
   ],
   "source": [
    "# Transcript abundance comparison Maxima vs. TGIRT:\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['A70', 'A71', 'A72', 'A73', \\\n",
    "                                                   'A74', 'A75', 'A76'], \\\n",
    "                                                  ['A70_T', 'A71_T', 'A72_T', 'A73_T', \\\n",
    "                                                   'A74_T', 'A75_T', 'A76_T']], \\\n",
    "                             plot_type='transcript', plot_name='tr_abundance_corr_samples', \\\n",
    "                             min_obs=500, charge_plot=False, log=True)\n",
    "\n",
    "# Transcript abundance comparison Maxima vs. TGIRT:\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['A70', 'A71', 'A72', 'A73', \\\n",
    "                                                   'A74', 'A75', 'A76'], \\\n",
    "                                                  ['A70_T', 'A71_T', 'A72_T', 'A73_T', \\\n",
    "                                                   'A74_T', 'A75_T', 'A76_T']], \\\n",
    "                             plot_type='transcript', plot_name='tr_charge_corr_samples', \\\n",
    "                             min_obs=500, charge_plot=True, one2one_corr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample pairs:  (Maxima - TGIRT)\n",
      "Now plotting sample pairs:  (Maxima - TGIRT)"
     ]
    }
   ],
   "source": [
    "# Transcript abundance comparison Maxima vs. TGIRT.\n",
    "# Average over all the samples in each RT group:\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['Maxima'], \\\n",
    "                                           ['TGIRT']], \\\n",
    "                             sample_pairs_col='polymerase', \\\n",
    "                             plot_type='codon', plot_name='codon_abundance_corr_RT', \\\n",
    "                             min_obs=1000, charge_plot=False, log=True)\n",
    "\n",
    "# Transcript abundance comparison Maxima vs. TGIRT:\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['Maxima'], \\\n",
    "                                           ['TGIRT']], \\\n",
    "                             sample_pairs_col='polymerase', \\\n",
    "                             plot_type='codon', plot_name='codon_charge_corr_RT', \\\n",
    "                             min_obs=1000, charge_plot=True, one2one_corr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  A70  A71  A72  A73  A74  A75  A76  A70_T  A72_T  A73_T  A74_T  A71_T  A76_T  A75_T\n",
      "Now plotting sample:  A70  A71  A72  A73  A74  A75  A76  A70_T  A71_T  A72_T  A73_T  A74_T  A75_T  A76_T\n",
      "Now collecting data for sample:  A70  A72  A71  A75  A76  A70_T  A71_T  A72_T  A73  A73_T  A74_T  A74  A75_T  A76_T\n",
      "Now plotting sample:  A70  A71  A72  A73  A74  A75  A76  A70_T  A71_T  A72_T  A73_T  A74_T  A75_T  A76_T"
     ]
    }
   ],
   "source": [
    "### Coverage plots for cyto/mito transcripts:\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_behrens_norm', n_jobs=12)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_behrens_norm', n_jobs=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  A70  A76  A71  A72  A73  A74  A75  A70_T  A72_T  A71_T  A75_T  A73_T  A74_T  A76_T\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# UMI logo:\n",
    "plot_obj.plot_UMI_logo(n_jobs=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  A70  A72  A71  A73  A74  A75  A76  A70_T  A74_T  A71_T  A73_T  A76_T  A72_T  A75_T\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  A70  A76  A71  A72  A73  A74  A75  A70_T  A75_T  A72_T  A73_T  A74_T  A76_T  A71_T\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  A70  A71  A72  A73  A74  A75  A76  A70_T  A74_T  A76_T  A72_T  A71_T  A73_T  A75_T\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# Non-template nucleotides:\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     common_seqs=common_seqs, ignore_common_count=False, \\\n",
    "                     overwrite_dir=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting stats from:  A70  A71  A72  A75  A76  A70_T  A71_T  A72_T  A73  A73_T  A74_T  A74  A76_T  A75_T"
     ]
    }
   ],
   "source": [
    "# Find mutations and save them:\n",
    "TM_obj.find_muts(n_jobs=12, unique_anno=True)\n",
    "TM_obj.pickle_muts_write(pickle_name='saved_muts_unique-anno.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare mutations/gaps/RT stops by polymerase used:\n",
    "TM_obj.plot_transcript_mut_compare(plot_name='mut_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   data_type='mut', \\\n",
    "                                   sample_pairs=[['Maxima'], \\\n",
    "                                                 ['TGIRT']], \\\n",
    "                                   sample_pairs_col='polymerase', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(plot_name='gap_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   data_type='gap', \\\n",
    "                                   sample_pairs=[['Maxima'], \\\n",
    "                                                 ['TGIRT']], \\\n",
    "                                   sample_pairs_col='polymerase', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(plot_name='RTstops_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   data_type='RTstops', \\\n",
    "                                   sample_pairs=[['Maxima'], \\\n",
    "                                                 ['TGIRT']], \\\n",
    "                                   sample_pairs_col='polymerase', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=10, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare mutations/gaps/RT stops by polymerase used\n",
    "# but only for mitochondrial tRNA transcripts:\n",
    "TM_obj.plot_transcript_mut_compare(plot_name='mut_matrix_comp_top10-max-diff_mito', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   data_type='mut', \\\n",
    "                                   mito=True, \\\n",
    "                                   sample_pairs=[['Maxima'], \\\n",
    "                                                 ['TGIRT']], \\\n",
    "                                   sample_pairs_col='polymerase', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(plot_name='gap_matrix_comp_top10-max-diff_mito', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   data_type='gap', \\\n",
    "                                   mito=True, \\\n",
    "                                   sample_pairs=[['Maxima'], \\\n",
    "                                                 ['TGIRT']], \\\n",
    "                                   sample_pairs_col='polymerase', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(plot_name='RTstops_matrix_comp_top10-max-diff_mito', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   data_type='RTstops', \\\n",
    "                                   mito=True, \\\n",
    "                                   sample_pairs=[['Maxima'], \\\n",
    "                                                 ['TGIRT']], \\\n",
    "                                   sample_pairs_col='polymerase', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=10, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
