{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/tRNAseq_third-gen\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/AdapterRemoval\n"
     ]
    }
   ],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=4, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/BC_split\n",
      "Loaded results from previous run... Not running barcode split.\n"
     ]
    }
   ],
   "source": [
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=12, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Perform Kmer analysis on unmapped reads ###\n",
    "if False: # barcode analysis is typically suficient \n",
    "    kmer_obj = Kmer_analysis(dir_dict, inp_file_df, index_df)\n",
    "    # Add a filter to avoid Kmers from the end of tRNA sequences:\n",
    "    kmer_obj.filter_3p_fasta(tRNA_database['human'])\n",
    "    # Add to this filter the constant region of the adapters:\n",
    "    kmer_obj.filter_window_BC(filter_window=(0, 11))\n",
    "    # Search for Kmers:\n",
    "    all_kmer = kmer_obj.search_unmapped(search_size=13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "if False:\n",
    "    ### Perform barcode analysis on unmapped reads ###\n",
    "    # Recall that adapters look like this:\n",
    "    # GGCTGCCATGC    GACTA\n",
    "    # GGCTGCCATGCA   AGTGC\n",
    "    # GGCTGCCATGCTG  TCACG\n",
    "    # GGCTGCCATGCAAC CTGAT\n",
    "    # With the barcode as the rightmost 5 nt.\n",
    "    # Search these 5 nt. barcodes by specifying BC_size_3p=5\n",
    "    bc_analysis_obj = BC_analysis(dir_dict, inp_file_df, index_df, BC_size_3p=5, overwrite=True)\n",
    "    # Search for barcodes in the unmapped reads.\n",
    "    # For the summary output, filter by a max distance\n",
    "    # to any barcode of 1 (group_dist=1), then group by barcode name:\n",
    "    bc_analysis_df = bc_analysis_obj.search_unmapped(group_dist=1)\n",
    "    bc_analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/UMI_trimmed\n",
      "Loaded results from previous run... Not running UMI trimming.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>P1_R1.fastq.bz2</td>\n",
       "      <td>P1_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>4882155</td>\n",
       "      <td>223687</td>\n",
       "      <td>4625383</td>\n",
       "      <td>4849070</td>\n",
       "      <td>99.322328</td>\n",
       "      <td>95.387012</td>\n",
       "      <td>4801289</td>\n",
       "      <td>515449</td>\n",
       "      <td>524232.739648</td>\n",
       "      <td>98.343641</td>\n",
       "      <td>98.324458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>P2_R1.fastq.bz2</td>\n",
       "      <td>P2_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>4927612</td>\n",
       "      <td>175921</td>\n",
       "      <td>4722585</td>\n",
       "      <td>4898506</td>\n",
       "      <td>99.409328</td>\n",
       "      <td>96.408681</td>\n",
       "      <td>4850600</td>\n",
       "      <td>513467</td>\n",
       "      <td>524237.700138</td>\n",
       "      <td>98.437133</td>\n",
       "      <td>97.945455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>P3_R1.fastq.bz2</td>\n",
       "      <td>P3_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3002237</td>\n",
       "      <td>149809</td>\n",
       "      <td>2828476</td>\n",
       "      <td>2978285</td>\n",
       "      <td>99.202195</td>\n",
       "      <td>94.969958</td>\n",
       "      <td>2951349</td>\n",
       "      <td>492424</td>\n",
       "      <td>522405.154299</td>\n",
       "      <td>98.304997</td>\n",
       "      <td>94.260938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate fastq_mate1_filename  \\\n",
       "0              100p1        100p          1      P1_R1.fastq.bz2   \n",
       "1              100p2        100p          2      P2_R1.fastq.bz2   \n",
       "2              100p3        100p          3      P3_R1.fastq.bz2   \n",
       "\n",
       "  fastq_mate2_filename P5_index P7_index barcode species        plot_group  \\\n",
       "0      P1_R2.fastq.bz2     D501     D701    l1Sp   human  Charge-titration   \n",
       "1      P2_R2.fastq.bz2     D501     D702    l2Sp   human  Charge-titration   \n",
       "2      P3_R2.fastq.bz2     D501     D703    l3Sp   human  Charge-titration   \n",
       "\n",
       "         hue_name hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0  Percent charge      100p          1     AGGCTATA     ATTACTCG   \n",
       "1  Percent charge      100p          1     AGGCTATA     TCCGGAGA   \n",
       "2  Percent charge      100p          1     AGGCTATA     CGCTCATT   \n",
       "\n",
       "          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0    GGCTGCCATGCGACTA  4882155  223687  4625383   4849070   \n",
       "1   GGCTGCCATGCAAGTGC  4927612  175921  4722585   4898506   \n",
       "2  GGCTGCCATGCTGTCACG  3002237  149809  2828476   2978285   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.322328    95.387012       4801289          515449   \n",
       "1             99.409328    96.408681       4850600          513467   \n",
       "2             99.202195    94.969958       2951349          492424   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \n",
       "0   524232.739648                    98.343641               98.324458  \n",
       "1   524237.700138                    98.437133               97.945455  \n",
       "2   522405.154299                    98.304997               94.260938  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "# Note, the cDNA input amount is so large that it is very unlikely to sequence\n",
    "# the same PCR amplified DNA twice. Therefore, this processing step does not\n",
    "# attempt to merge possible UMI duplicates.\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=12, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  100p3  100p4  85p2  100p1  100p2  85p1  70p1  70p4  70p2  85p3  85p4  55p4  70p3  55p1  40p1  55p2  40p3  40p4  25p3  25p2  10p1  10p3  25p1  55p3  25p4  0p3  0p4  10p2  0p2  0p1  10p4  40p2  0p_90m2  0p_90m1  0p_90m3  4m_1  8m_1  0p_90m4  0m_1  1h_1  16m_1  4h_1  32m_1  8h_1  2h_1  16h_1  40h_1  40h_NoOx_1  16m_2  0m_2  8m_2  4m_2  2h_2  32m_2  1h_2  4h_2  0m_3  8h_2  16h_2  40h_NoOx_2  40h_2  8m_3  4m_3  1h_3  16m_3  32m_3  2h_3  8h_3  40h_NoOx_3  4h_3  40h_3  16h_3  8m_4  0m_4  4m_4  16m_4  16h_4  32m_4  2h_4  4h_4  1h_4  8h_4  40h_NoOx_4  40h_4  L1  L3  L5  L2  L4  L6  L8  L10  L7  L12  L11  R3  L9  R1  R4  R2  R6  R9  R7  R5  common-seqs  R10  R11  R8  R12\n",
      "Collecting alignment statistics, from sample:  100p2  100p4  100p3  85p2  100p1  85p1  70p1  70p2  85p3  70p3  70p4  85p4  55p3  55p1  40p3  55p2  40p4  40p1  40p2  25p1  25p2  25p3  25p4  10p3  10p4  0p1  10p2  55p4  10p1  0p2  0p4  0p_90m1  0p3  0p_90m3  0p_90m2  0m_1  8m_1  0p_90m4  4m_1  32m_1  2h_1  1h_1  16m_1  8h_1  4h_1  40h_1  40h_NoOx_1  16h_1  8m_2  4m_2  16m_2  0m_2  32m_2  2h_2  4h_2  8h_2  1h_2  16h_2  40h_2  8m_3  0m_3  4m_3  40h_NoOx_2  16m_3  32m_3  1h_3  4h_3  2h_3  40h_3  16h_3  8h_3  40h_NoOx_3  0m_4  4m_4  8m_4  1h_4  16m_4  2h_4  40h_4  4h_4  8h_4  32m_4  16h_4  L1  40h_NoOx_4  L4  L5  L7  L2  L6  L3  L10  L8  L9  R1  L11  R2  L12  R3  R4  R7  R6  R5  R9  R10  R11  R8  R12  common-seqs"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>P1_R1.fastq.bz2</td>\n",
       "      <td>P1_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>4882155</td>\n",
       "      <td>223687</td>\n",
       "      <td>4625383</td>\n",
       "      <td>4849070</td>\n",
       "      <td>99.322328</td>\n",
       "      <td>95.387012</td>\n",
       "      <td>4801289</td>\n",
       "      <td>515449</td>\n",
       "      <td>524232.739648</td>\n",
       "      <td>98.343641</td>\n",
       "      <td>98.324458</td>\n",
       "      <td>4774572.0</td>\n",
       "      <td>79.222577</td>\n",
       "      <td>20.777423</td>\n",
       "      <td>1.408608</td>\n",
       "      <td>99.443545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>P2_R1.fastq.bz2</td>\n",
       "      <td>P2_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>4927612</td>\n",
       "      <td>175921</td>\n",
       "      <td>4722585</td>\n",
       "      <td>4898506</td>\n",
       "      <td>99.409328</td>\n",
       "      <td>96.408681</td>\n",
       "      <td>4850600</td>\n",
       "      <td>513467</td>\n",
       "      <td>524237.700138</td>\n",
       "      <td>98.437133</td>\n",
       "      <td>97.945455</td>\n",
       "      <td>4821834.0</td>\n",
       "      <td>80.276115</td>\n",
       "      <td>19.723885</td>\n",
       "      <td>1.333559</td>\n",
       "      <td>99.406960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>P3_R1.fastq.bz2</td>\n",
       "      <td>P3_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3002237</td>\n",
       "      <td>149809</td>\n",
       "      <td>2828476</td>\n",
       "      <td>2978285</td>\n",
       "      <td>99.202195</td>\n",
       "      <td>94.969958</td>\n",
       "      <td>2951349</td>\n",
       "      <td>492424</td>\n",
       "      <td>522405.154299</td>\n",
       "      <td>98.304997</td>\n",
       "      <td>94.260938</td>\n",
       "      <td>2918119.0</td>\n",
       "      <td>76.456443</td>\n",
       "      <td>23.543557</td>\n",
       "      <td>2.047518</td>\n",
       "      <td>98.874074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate fastq_mate1_filename  \\\n",
       "0              100p1        100p          1      P1_R1.fastq.bz2   \n",
       "1              100p2        100p          2      P2_R1.fastq.bz2   \n",
       "2              100p3        100p          3      P3_R1.fastq.bz2   \n",
       "\n",
       "  fastq_mate2_filename P5_index P7_index barcode species        plot_group  \\\n",
       "0      P1_R2.fastq.bz2     D501     D701    l1Sp   human  Charge-titration   \n",
       "1      P2_R2.fastq.bz2     D501     D702    l2Sp   human  Charge-titration   \n",
       "2      P3_R2.fastq.bz2     D501     D703    l3Sp   human  Charge-titration   \n",
       "\n",
       "         hue_name hue_value  hue_order P5_index_seq P7_index_seq  \\\n",
       "0  Percent charge      100p          1     AGGCTATA     ATTACTCG   \n",
       "1  Percent charge      100p          1     AGGCTATA     TCCGGAGA   \n",
       "2  Percent charge      100p          1     AGGCTATA     CGCTCATT   \n",
       "\n",
       "          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0    GGCTGCCATGCGACTA  4882155  223687  4625383   4849070   \n",
       "1   GGCTGCCATGCAAGTGC  4927612  175921  4722585   4898506   \n",
       "2  GGCTGCCATGCTGTCACG  3002237  149809  2828476   2978285   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.322328    95.387012       4801289          515449   \n",
       "1             99.409328    96.408681       4850600          513467   \n",
       "2             99.202195    94.969958       2951349          492424   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524232.739648                    98.343641               98.324458   \n",
       "1   524237.700138                    98.437133               97.945455   \n",
       "2   522405.154299                    98.304997               94.260938   \n",
       "\n",
       "    N_mapped  percent_single_annotation  percent_multiple_annotation  \\\n",
       "0  4774572.0                  79.222577                    20.777423   \n",
       "1  4821834.0                  80.276115                    19.723885   \n",
       "2  2918119.0                  76.456443                    23.543557   \n",
       "\n",
       "   percent_multiple_codons  Mapping_percent  \n",
       "0                 1.408608        99.443545  \n",
       "1                 1.333559        99.406960  \n",
       "2                 2.047518        98.874074  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=6, overwrite=False, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Collecting stats from:  100p1  100p2  100p3  100p4  85p1  85p2  85p3  85p4  55p3  55p2  70p1  70p3  70p4  55p1  40p1  55p4  70p2  40p2  40p3  10p3  25p3  25p2  0p2  10p2  0p1  25p4  10p4  40p4  25p1  0p_90m2  0p4  10p1  0p3  0p_90m1  0p_90m3  0p_90m4  32m_1  4m_1  0m_1  16m_1  2h_1  8m_1  4h_1  16h_1  0m_2  8h_1  1h_1  4m_2  40h_NoOx_1  32m_2  40h_1  16m_2  8m_2  4h_2  2h_2  40h_NoOx_2  16h_2  8h_2  0m_3  1h_2  40h_2  16m_3  4h_3  2h_3  8m_3  4m_3  1h_3  8h_3  32m_3  16h_3  40h_NoOx_3  32m_4  4m_4  8m_4  0m_4  40h_3  1h_4  2h_4  16m_4  16h_4  L3  40h_4  8h_4  4h_4  40h_NoOx_4  L1  L2  L5  L4  L6  L9  L10  L11  L8  L7  R7  R1  R8  L12  R2  R6  R5  R4  R3  R10  R11  R12  R9"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>1295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode  \\\n",
       "0              100p1        100p          1    l1Sp   \n",
       "1              100p1        100p          1    l1Sp   \n",
       "2              100p1        100p          1    l1Sp   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  count  \n",
       "0               True     False           A   AAA       TTT   eColiLys    146  \n",
       "1               True     False           C   AAA       TTT   eColiLys   1295  \n",
       "2               True      True           A   AAA       TTT   eColiLys     67  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=True)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=8, load_previous=False)\n",
    "# The dataframe returned is the \"ALL_stats_aggregate_filtered.csv\"\n",
    "# which is the aggregated data filtered to contain only the\n",
    "# most relevant columnns and requiring the 3' must be covered\n",
    "# and have no 3' non-template bases.\n",
    "# The CSV file output \"ALL_stats_aggregate.csv\" is the data\n",
    "# aggregated based on all values identical except\n",
    "# readID, 5p_UMI and 3p_BC. I.e. every information, except\n",
    "# the UMI seequence is maintained in the aggregated CSV.\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=True, pull_default=False)\n",
    "\n",
    "# Plot UMI nucleotide content logo per sample:\n",
    "#plot_obj.plot_UMI_logo()\n",
    "\n",
    "# Plot the 5p non-template nucleotide logo per sample,\n",
    "# using only those at, or below, the legth determined as a percentile:\n",
    "#plot_obj.plot_non_temp('5p', '_5p-non-template_logo', seq_len_percentile=99)\n",
    "\n",
    "# 3p non-template nucleotides are more rare, therefore 99.9 percentile:\n",
    "#plot_obj.plot_non_temp('3p', '_3p-non-template_logo', seq_len_percentile=99.9)\n",
    "\n",
    "# 3p non-template nucleotides, but require 3p coverage.\n",
    "# This is more likely to reflect either:\n",
    "# 1) Erronous CCA addition, or 2) additional bases in the adapter.\n",
    "#plot_obj.plot_non_temp('3p', '_3p-non-template_3p-cover_logo', seq_len_percentile=99.9, _3p_cover=True)\n",
    "\n",
    "# Plot the 5p to 3p coverage for each amino acid:\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='needle', aa_norm=False, plot_name='cov_plot_cyto_needle')\n",
    "#plot_obj.plot_coverage(compartment='mito', plot_type='needle', aa_norm=False, plot_name='cov_plot_mito_needle')\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', aa_norm=False, plot_name='cov_plot_cyto_behrens')\n",
    "#plot_obj.plot_coverage(compartment='mito', plot_type='behrens', aa_norm=False, plot_name='cov_plot_mito_behrens')\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='needle', aa_norm=False, #plot_name='cov_plot_cyto_needle_100p', sample_list=['100p1', '100p2', '100p3', '100p4'])\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, plot_name='cov_plot_cyto_behrens_norm')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  100p2  100p1  100p3  85p1  100p4  85p2  85p3  85p4  55p3  70p4  70p3  55p2  70p2  55p1  70p1  40p4  40p1  55p4  25p4  10p3  25p3  0p1  10p1  25p2  0p4  40p2  25p1  10p2  10p4  0p_90m1  0p3  0p2  8m_1  0p_90m2  0p_90m4  32m_1  16m_1  0m_1  40p3  4m_1  0p_90m3  1h_1  40h_1  8h_1  4h_1  40h_NoOx_1  2h_1  0m_2  4m_2  16h_1  2h_2  8m_2  32m_2  8h_2  4h_2  16h_2  1h_2  40h_2  16m_2  40h_NoOx_2  32m_3  8h_3  0m_3  4m_3  2h_3  4h_3  1h_3  16m_3  8m_3  40h_NoOx_3  16h_3  40h_3  0m_4  1h_4  16m_4  4m_4  8m_4  32m_4  40h_4  4h_4  2h_4  8h_4  40h_NoOx_4  L1  16h_4  L3  L2  L5  L7  L4  R1  L11  L10  L6  L9  L8  L12  R2  R6  R5  R4  R8  R9  R3  R10  R7  R11  R12\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  100p2  100p1  100p3  100p4  85p1  85p2  85p3  85p4  55p3  55p2  70p3  70p4  70p2  70p1  40p2  40p1  55p1  55p4  25p2  10p3  25p3  0p1  25p4  0p3  0p4  40p4  25p1  0p2  10p1  10p2  0p_90m2  10p4  0p_90m1  0p_90m4  0p_90m3  4m_1  0m_1  8m_1  40p3  16m_1  40h_1  16h_1  4h_1  32m_1  1h_1  2h_1  8h_1  40h_NoOx_1  1h_2  4m_2  8h_2  32m_2  8m_2  0m_2  4h_2  16m_2  16h_2  40h_NoOx_2  2h_2  0m_3  40h_2  4m_3  8m_3  1h_3  16m_3  4h_3  2h_3  16h_3  32m_3  8h_3  32m_4  40h_3  8m_4  0m_4  4m_4  40h_NoOx_3  16m_4  2h_4  L1  1h_4  8h_4  40h_NoOx_4  16h_4  40h_4  4h_4  L3  L2  L6  L8  L10  L7  L4  L11  L5  R3  L9  L12  R1  R2  R7  R12  R6  R4  R5  R9  R11  R8  R10\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  100p3  100p1  100p2  100p4  85p1  85p2  85p3  85p4  55p3  70p4  55p2  70p3  55p4  70p2  70p1  40p1  55p1  40p3  10p2  40p2  25p4  0p1  10p3  25p3  0p4  25p1  10p1  10p4  0p_90m3  0p2  0p_90m1  0p3  0p_90m2  0p_90m4  16m_1  25p2  4h_1  40p4  0m_1  16h_1  4m_1  32m_1  2h_1  8m_2  8h_1  8m_1  40h_1  1h_1  0m_2  40h_NoOx_1  16m_2  4m_2  1h_2  8h_2  2h_2  0m_3  4h_2  16h_2  40h_NoOx_2  40h_2  32m_2  8m_3  16m_3  2h_3  4m_3  16h_3  40h_3  4h_3  32m_3  1h_3  8h_3  40h_NoOx_3  0m_4  32m_4  8m_4  4h_4  4m_4  8h_4  16h_4  1h_4  16m_4  2h_4  40h_4  40h_NoOx_4  L5  L1  L8  L2  L7  L6  L3  R1  L9  L4  L10  L11  R2  L12  R3  R5  R7  R9  R6  R11  R10  R4  R8  R12\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  100p2  100p1  100p3  100p4  85p1  85p2  85p3  85p4  55p3  70p3  70p4  55p2  70p2  55p4  70p1  40p4  40p1  55p1  25p3  10p3  10p1  25p4  0p2  25p2  0p4  40p3  10p2  0p_90m2  0p1  40p2  0p_90m1  25p1  0p3  10p4  0p_90m3  32m_1  4m_1  0p_90m4  16m_1  0m_1  1h_1  8h_1  8m_1  2h_1  4h_1  40h_1  16m_2  0m_2  8m_2  40h_NoOx_1  16h_1  32m_2  2h_2  4h_2  4m_2  8h_2  1h_2  16h_2  40h_2  32m_3  40h_NoOx_2  0m_3  8m_3  4h_3  4m_3  16m_3  2h_3  16h_3  8h_3  0m_4  1h_3  40h_3  40h_NoOx_3  4m_4  1h_4  32m_4  2h_4  4h_4  8m_4  8h_4  16m_4  L5  16h_4  40h_NoOx_4  L3  L2  L1  L6  L4  40h_4  L8  L9  L7  R5  L12  L10  R1  L11  R2  R3  R7  R6  R11  R4  R12  R8  R9  R10\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  100p2  100p1  85p2  100p3  100p4  85p3  85p1  85p4  70p1  70p2  70p3  70p4  40p3  10p1  55p3  40p2  55p4  25p1  55p1  55p2  40p1  40p4  10p2  10p4  0p4  10p3  25p4  8m_1  0p_90m1  0p_90m3  0p2  0p_90m2  0p1  25p2  16h_1  0p3  0p_90m4  40h_NoOx_1  0m_1  16m_1  4m_1  25p3  1h_2  4h_1  2h_1  8h_1  1h_1  2h_2  40h_1  8h_2  32m_1  32m_2  16m_2  4m_2  8m_2  4h_2  0m_2  40h_2  40h_NoOx_2  8m_3  16h_2  0m_3  4m_3  2h_3  1h_3  32m_3  16m_3  8h_3  1h_4  40h_3  16h_3  4h_3  8m_4  4m_4  40h_NoOx_3  2h_4  32m_4  16m_4  0m_4  16h_4  L6  4h_4  L1  8h_4  L3  40h_4  L2  40h_NoOx_4  L7  L5  R5  L4  L9  L8  R1  R2  R4  R3  L10  L11  L12  R8  R7  R6  R10  R12  R9  R11\n",
      "Now plotting sample:  100p1  100p2  100p3  100p4  85p1  85p2  85p3  85p4  70p1  70p2  70p3  70p4  55p1  55p2  55p3  55p4  40p1  40p2  40p3  40p4  25p1  25p2  25p3  25p4  10p1  10p2  10p3  10p4  0p1  0p2  0p3  0p4  0p_90m1  0p_90m2  0p_90m3  0p_90m4  0m_1  4m_1  8m_1  16m_1  32m_1  1h_1  2h_1  4h_1  8h_1  16h_1  40h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  32m_2  1h_2  2h_2  4h_2  8h_2  16h_2  40h_2  40h_NoOx_2  0m_3  4m_3  8m_3  16m_3  32m_3  1h_3  2h_3  4h_3  8h_3  16h_3  40h_3  40h_NoOx_3  0m_4  4m_4  8m_4  16m_4  32m_4  1h_4  2h_4  4h_4  8h_4  16h_4  40h_4  40h_NoOx_4  L1  L2  L3  L4  L5  L6  L7  L8  L9  L10  L11  L12  R1  R2  R3  R4  R5  R6  R7  R8  R9  R10  R11  R12\n",
      "Now collecting data for sample:  100p1  100p2  100p3  85p1  100p4  85p2  85p3  85p4  70p1  70p2  70p3  70p4  40p3  10p1  55p3  40p2  55p4  25p1  55p1  55p2  40p1  40p4  10p2  10p4  0p4  10p3  25p4  8m_1  0p_90m1  0p2  0p_90m3  0p_90m2  0p1  25p2  16h_1  0p3  40h_NoOx_1  0p_90m4  0m_1  16m_1  25p3  32m_2  4m_1  4h_1  8h_1  1h_1  2h_1  40h_1  32m_1  16m_2  40h_2  0m_2  2h_2  8m_2  4h_2  1h_2  16h_2  4m_2  4h_3  8h_2  4m_3  16m_3  40h_NoOx_2  8m_3  0m_3  32m_3  1h_3  1h_4  16h_3  2h_3  0m_4  8h_3  32m_4  4m_4  8m_4  40h_3  40h_NoOx_3  16m_4  2h_4  L6  40h_4  4h_4  8h_4  16h_4  L2  L5  L3  40h_NoOx_4  L4  L1  L9  L8  L7  R1  R9  R4  R3  L12  R5  R12  L10  L11  R2  R6  R8  R7  R11  R10\n",
      "Now plotting sample:  100p1  100p2  100p3  100p4  85p1  85p2  85p3  85p4  70p1  70p2  70p3  70p4  55p1  55p2  55p3  55p4  40p1  40p2  40p3  40p4  25p1  25p2  25p3  25p4  10p1  10p2  10p3  10p4  0p1  0p2  0p3  0p4  0p_90m1  0p_90m2  0p_90m3  0p_90m4  0m_1  4m_1  8m_1  16m_1  32m_1  1h_1  2h_1  4h_1  8h_1  16h_1  40h_1  40h_NoOx_1  0m_2  4m_2  8m_2  16m_2  32m_2  1h_2  2h_2  4h_2  8h_2  16h_2  40h_2  40h_NoOx_2  0m_3  4m_3  8m_3  16m_3  32m_3  1h_3  2h_3  4h_3  8h_3  16h_3  40h_3  40h_NoOx_3  0m_4  4m_4  8m_4  16m_4  32m_4  1h_4  2h_4  4h_4  8h_4  16h_4  40h_4  40h_NoOx_4  L1  L2  L3  L4  L5  L6  L7  L8  L9  L10  L11  L12  R1  R2  R3  R4  R5  R6  R7  R8  R9  R10  R11  R12\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  NoLeu  NoArg\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  NoLeu  NoArg\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  NoLeu  NoArg\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  NoLeu  NoArg\n",
      "Now plotting sample pairs:  (100p - 85p)  (100p - 0p)  (85p - 40p)  (85p - 0p)  (40p - 0p)  (40h - 40h_NoOx)\n",
      "Now plotting sample pairs:  (100p - 85p)  (100p - 0p)  (85p - 40p)  (85p - 0p)  (40p - 0p)  (40h - 40h_NoOx)\n",
      "Now plotting sample pairs:  (L1 - R1)  (L1 - L6)  (R1 - R6)\n",
      "Now plotting sample pairs:  (L1 - R1)  (L1 - L6)  (R1 - R6)\n",
      "Now plotting sample pairs:  (100p1 - 100p3)  (85p1 - 85p4)  (55p1 - 55p3)  (40p1 - 40p2)  (0p1 - 0p2)  (40h_1 - 40h_2)  (40h_NoOx_1 - 40h_NoOx_2)"
     ]
    }
   ],
   "source": [
    "plot_obj.plot_UMI_logo(n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=8)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_behrens_norm', n_jobs=12)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_behrens_norm', n_jobs=12)\n",
    "\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_abundance_grp', \\\n",
    "                        group=True, min_obs=500, \\\n",
    "                        sample_list_exl=['70p2', '70p3', '70p4', '25p2', '25p3', '25p4'], \\\n",
    "                        bc_list_exl=['l2Sp', 'l4Sp', 'l7Sp'])\n",
    "\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_abundance_grp', \\\n",
    "                        group=True, min_obs=500, \\\n",
    "                        sample_list_exl=['70p2', '70p3', '70p4', '25p2', '25p3', '25p4'], \\\n",
    "                        bc_list_exl=['l2Sp', 'l4Sp', 'l7Sp'])\n",
    "\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True, \\\n",
    "                        sample_list_exl=['70p2', '70p3', '70p4', '25p2', '25p3', '25p4'], \\\n",
    "                        bc_list_exl=['l2Sp', 'l4Sp', 'l7Sp'])\n",
    "\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True, \\\n",
    "                        sample_list_exl=['70p2', '70p3', '70p4', '25p2', '25p3', '25p4'], \\\n",
    "                        bc_list_exl=['l2Sp', 'l4Sp', 'l7Sp'])\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['100p', '100p', '85p', '85p', '40p', '40h'], \\\n",
    "                                           ['85p', '0p', '40p', '0p', '0p', '40h_NoOx']], \\\n",
    "                             plot_type='codon', plot_name='codon_abundance_corr', \\\n",
    "                             min_obs=500, charge_plot=False,\n",
    "                             sample_list_exl=['100p2', '100p4', '85p3', \\\n",
    "                                              '55p2', '55p4', '40p3', '0p3'])\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['100p', '100p', '85p', '85p', '40p', '40h'], \\\n",
    "                                           ['85p', '0p', '40p', '0p', '0p', '40h_NoOx']], \\\n",
    "                             plot_type='transcript', plot_name='tr_abundance_corr', \\\n",
    "                             min_obs=500, charge_plot=False,\n",
    "                             sample_list_exl=['100p2', '100p4', '85p3', \\\n",
    "                                              '55p2', '55p4', '40p3', '0p3'])\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['L1', 'L1', 'R1'], ['R1', 'L6', 'R6']], \\\n",
    "                             plot_type='codon', plot_name='codon_abundance_corr_L-R', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['L1', 'L1', 'R1'], ['R1', 'L6', 'R6']], \\\n",
    "                             plot_type='transcript', plot_name='tr_abundance_corr_L-R', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['100p1', '85p1', '55p1', '40p1', '0p1', '40h_1', '40h_NoOx_1'], \\\n",
    "                                           ['100p3', '85p4', '55p3', '40p2', '0p2', '40h_2', '40h_NoOx_2']], \\\n",
    "                             plot_type='codon', plot_name='codon_charge_corr', \\\n",
    "                             min_obs=500, charge_plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the TM_analysis to find the most mutated positions in the tRNA transcripts.\n",
    "# Then generate a new version of the tRNA transcripts database with these positions masked\n",
    "# Then, using the masked sequences, re-run the alignment, stats collected and plotting\n",
    "\n",
    "# Re-run using gap_open = -3 and gap_extension = -2 to reflect how gaps are more tolerated\n",
    "\n",
    "### The masked tRNA database will only be used for alignment\n",
    "# All other steps will use the old unmasked database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict_masked, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     common_seqs=common_seqs, ignore_common_count=False, \\\n",
    "                     overwrite_dir=True)\n",
    "\n",
    "\n",
    "#TM_obj.find_muts(n_jobs=12, unique_anno=True)\n",
    "#TM_obj.find_muts(n_jobs=4)\n",
    "\n",
    "#TM_obj.plot_transcript_logo()\n",
    "\n",
    "\n",
    "\n",
    "# tr_mut_out = TM_obj.plot_transcript_mut(topN=30, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat', min_count_show=10)\n",
    "#tr_mut_out = TM_obj.plot_transcript_mut(topN=30, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat2')\n",
    "\n",
    "\n",
    "#tr_cov_out = TM_obj.plot_transcript_cov(topN=40, no_plot_return=False, mito=False, plot_name='test_heat_cov', sort_rows=True)\n",
    "#tr_mut_out = TM_obj.plot_transcript_mut(topN=40, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat2', sort_rows=tr_cov_out[2])\n",
    "\n",
    "\n",
    "#TM_obj.mask_tRNA_database(min_mut_freq=0.5, min_pos_count=50, min_tr_count=100, frac_max_score=0.90)\n",
    "\n",
    "#tRNA_database_masked = TM_obj.write_masked_tRNA_database(out_dir='tRNA_database_masked')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TM_obj.find_muts(n_jobs=10, unique_anno=True, fix_end=True, \\\n",
    "#                 sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "#                              '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3',\n",
    "#                              'L1', 'L3', 'L6', 'L8', 'L12', \\\n",
    "#                              'R1', 'R3', 'R6', 'R8', 'R12'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting stats from:  100p1  40h_NoOx_1  100p3  0p1  0p2  0p4  0m_1  L1  40h_NoOx_2  0m_3  L3  L9  L6  L10  L4  L5  L2  L8  L7  L11  R8  L12  R2  R4  R6  R5  R3  R1  R7  R10  R11  R9  R12"
     ]
    }
   ],
   "source": [
    "TM_obj.find_muts(n_jobs=10, unique_anno=False, fix_end=True, \\\n",
    "                 sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "                              '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3',\n",
    "                              'L1', 'L2', 'L3', 'L4',  'L5',  'L6', \\\n",
    "                              'L7', 'L8', 'L9', 'L10', 'L11', 'L12', \\\n",
    "                              'R1', 'R2', 'R3', 'R4',  'R5',  'R6', \\\n",
    "                              'R7', 'R8', 'R9', 'R10', 'R11', 'R12'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Arg', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Arg')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Leu', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Leu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Asn', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Asn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Ala', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Ala')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_logo(sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "                                         '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3', \\\n",
    "                                         'L1', 'L3', 'L6', 'R1', 'R3', 'R6'], \\\n",
    "                            plot_name='tr-muts_logo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr_cov_out = TM_obj.plot_transcript_cov(topN=40, no_plot_return=True, mito=False, \\\n",
    "                                        plot_name='tr-cov_matrix', sort_rows=True, \\\n",
    "                                        sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "                                                     '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3', \\\n",
    "                                                     'L1', 'L3', 'L6', 'R1', 'R3', 'R6'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr_mut_out = TM_obj.plot_transcript_mut(topN=40, no_plot_return=True, mito=False, \\\n",
    "                                        gap_only=False, min_count_show=1000, \\\n",
    "                                        plot_name='tr-mut_matrix', \\\n",
    "                                        sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "                                                     '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3', \\\n",
    "                                                     'L1', 'L3', 'L6', 'R1', 'R3', 'R6'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.write_transcript_mut(['L1', 'L2', 'L3', 'L4',  'L5',  'L6', \\\n",
    "                             'L7', 'L8', 'L9', 'L10', 'L11', 'L12', \\\n",
    "                             'R1', 'R2', 'R3', 'R4',  'R5',  'R6', \\\n",
    "                             'R7', 'R8', 'R9', 'R10', 'R11', 'R12'], \\\n",
    "                            csv_name='mut-matrix_right-aligned')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.write_transcript_mut(['L1', 'L2', 'L3', 'L4',  'L5',  'L6', \\\n",
    "                             'L7', 'L8', 'L9', 'L10', 'L11', 'L12', \\\n",
    "                             'R1', 'R2', 'R3', 'R4',  'R5',  'R6', \\\n",
    "                             'R7', 'R8', 'R9', 'R10', 'R11', 'R12'], \\\n",
    "                            csv_name='mut-matrix_left-aligned', \\\n",
    "                            right_align=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
