{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy, warnings\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "# Numpy will complain about subnormals if python\n",
    "# is compiled with the -ffast-math compiler flag:\n",
    "# https://github.com/clearlinux/distribution/issues/2809\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter(\"ignore\", message=\"The value of the smallest subnormal for\")\n",
    "    import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/Ser-Pro_anno\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))\n",
    "\n",
    "# Read sample list:\n",
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment effect on reads mapping to Ser/Pro codons\n",
    "Process a set of samples covering all adapter barcodes with/without using a masked reference for alignment.\n",
    "Gather alignment statistics from each case with the purpose of showing the effect of the alignment method on the mapping to Ser/Pro codons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=True, AR_threads=12)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=4, overwrite=True)\n",
    "\n",
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=True)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=9, load_previous=False)\n",
    "\n",
    "# Trim UMIs:\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=True)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=9, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  100p3_v2  0m_p1_v2  0m_p3_v2  0m_p2_v2  100p1_v1  0m_p4_v2  85p2_v1  85p4_v1  85p1_v1  85p2_v2  85p1_v2  85p3_v2  85p4_v2  70p1_v1  70p3_v2  70p2_v2  70p1_v2  70p4_v2  70p5_v2  55p1_v1  55p3_v2  55p1_v2  55p2_v2  55p3_v1  55p4_v2  55p5_v2  40p4_v1  40p2_v1  40p1_v1  40p1_v2  40p2_v2  40p3_v2  40p4_v2  25p2_v2  25p1_v2  25p1_v1  25p6_v2  10p1_v1  25p3_v2  25p4_v2  25p5_v2  10p3_v1  10p2_v2  10p1_v2  10p3_v2  10p4_v2  10p5_v2  0p1_v1  0p2_v1  0p4_v1  0p4_v2  0p1_v2  0p_90m2_v1  0p_90m1_v1  0p2_v2  0p3_v2  4m_1  0p_90m3_v1  0p_90m4_v1  8m_1  0m_1  32m_1  16m_1  1h_1  8h_1  40h_1  4h_1  16h_1  2h_1  40h_NoOx_1  8m_2  4m_2  32m_2  0m_2  16m_2  1h_2  8h_2  2h_2  4h_2  40h_2  16h_2  40h_NoOx_2  4m_3  0m_3  16m_3  1h_3  8m_3  32m_3  2h_3  4h_3  8h_3  40h_NoOx_3  40h_3  16h_3  0m_4  4m_4  16m_4  8m_4  1h_4  32m_4  2h_4  4h_4  8h_4  16h_4  40h_NoOx_4  40h_4  8h_p1  8h_p2  8h_p4  8h_p5  8h_p6  8h_p3  8h_p7  8h_p8  8h_p9  L1  L3  L2  L4  L5  L6  L8  L9  L7  L10  L12  R1  L11  R3  R2  R5  R4  R11  R8  R7  R6  R9  R10  R12  2Q2  2Q3  2Q6  GQ1  2Q1  2Q4  GQ5  2Q5  GQ3  GQ6  GQ2  HL1  GQ4  H1  HR1  HL2  HG1  HQ2  HQ1  HR2  HGL2  HGL1  HGR2  HGR1  HGQ1  common-seqs  HGQ2\n",
      "Collecting alignment statistics, from sample:  100p1_v1  100p3_v2  0m_p3_v2  0m_p2_v2  0m_p1_v2  0m_p4_v2  85p4_v1  85p1_v2  85p2_v2  85p3_v2  85p2_v1  85p1_v1  70p1_v2  70p1_v1  70p2_v2  85p4_v2  70p3_v2  70p4_v2  55p1_v1  70p5_v2  55p3_v1  55p3_v2  55p4_v2  55p1_v2  40p1_v1  55p5_v2  55p2_v2  40p2_v1  40p2_v2  40p4_v1  40p1_v2  40p4_v2  40p3_v2  25p2_v2  25p1_v2  25p1_v1  25p3_v2  25p4_v2  25p5_v2  25p6_v2  10p1_v1  10p1_v2  10p2_v2  10p3_v2  10p3_v1  10p4_v2  10p5_v2  0p1_v1  0p2_v1  0p4_v1  0p1_v2  0p2_v2  0p4_v2  0p3_v2  0p_90m2_v1  0p_90m1_v1  0p_90m4_v1  0m_1  0p_90m3_v1  8m_1  4m_1  4h_1  32m_1  8h_1  2h_1  1h_1  16m_1  0m_2  4m_2  16h_1  40h_NoOx_1  8m_2  40h_1  16m_2  32m_2  1h_2  2h_2  16h_2  40h_2  40h_NoOx_2  0m_3  4h_2  8h_2  4m_3  16m_3  8m_3  1h_3  2h_3  4h_3  32m_3  8h_3  16h_3  40h_NoOx_3  0m_4  40h_3  4m_4  8m_4  16m_4  2h_4  32m_4  1h_4  8h_4  4h_4  40h_4  40h_NoOx_4  8h_p1  8h_p3  16h_4  8h_p2  8h_p4  8h_p5  8h_p6  8h_p9  L1  L2  8h_p7  8h_p8  L3  L4  L5  L6  L9  L7  L8  L10  L12  L11  R3  R1  R5  R2  R4  R7  R6  R11  R8  R9  R10  2Q1  R12  2Q2  2Q3  GQ1  2Q6  2Q4  2Q5  GQ3  GQ2  GQ6  H1  GQ5  GQ4  HL1  HR2  HL2  HR1  HG1  HQ2  HQ1  HGL1  HGL2  HGR1  HGR2  HGQ1  HGQ2  common-seqs"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1_v1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>2022-12-28/P1_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P1_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>4882120</td>\n",
       "      <td>223691</td>\n",
       "      <td>4625330</td>\n",
       "      <td>4849021</td>\n",
       "      <td>99.322036</td>\n",
       "      <td>95.386883</td>\n",
       "      <td>4801306</td>\n",
       "      <td>515441</td>\n",
       "      <td>524232.741440</td>\n",
       "      <td>98.344695</td>\n",
       "      <td>98.322932</td>\n",
       "      <td>4774581.0</td>\n",
       "      <td>79.221821</td>\n",
       "      <td>20.778179</td>\n",
       "      <td>1.408794</td>\n",
       "      <td>99.443381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p3_v2</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>2022-12-28/P3_R1.fastq.bz2</td>\n",
       "      <td>2022-12-28/P3_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>3002197</td>\n",
       "      <td>149772</td>\n",
       "      <td>2828518</td>\n",
       "      <td>2978290</td>\n",
       "      <td>99.203683</td>\n",
       "      <td>94.971208</td>\n",
       "      <td>2951336</td>\n",
       "      <td>492419</td>\n",
       "      <td>522405.107612</td>\n",
       "      <td>98.305874</td>\n",
       "      <td>94.259990</td>\n",
       "      <td>2918101.0</td>\n",
       "      <td>76.457909</td>\n",
       "      <td>23.542091</td>\n",
       "      <td>2.047804</td>\n",
       "      <td>98.873900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0m_p1_v2</td>\n",
       "      <td>100p_0m</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-02-28/P4_R1.fastq.bz2</td>\n",
       "      <td>2023-02-28/P4_R2.fastq.bz2</td>\n",
       "      <td>D502</td>\n",
       "      <td>D701</td>\n",
       "      <td>l5Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p_0m</td>\n",
       "      <td>2</td>\n",
       "      <td>GCCTCTAT</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCAACGCATC</td>\n",
       "      <td>2989860</td>\n",
       "      <td>163356</td>\n",
       "      <td>2798679</td>\n",
       "      <td>2962035</td>\n",
       "      <td>99.069354</td>\n",
       "      <td>94.485008</td>\n",
       "      <td>2931995</td>\n",
       "      <td>494668</td>\n",
       "      <td>522334.350496</td>\n",
       "      <td>98.064625</td>\n",
       "      <td>94.703325</td>\n",
       "      <td>2895784.0</td>\n",
       "      <td>74.423127</td>\n",
       "      <td>25.576873</td>\n",
       "      <td>2.288948</td>\n",
       "      <td>98.764971</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate        fastq_mate1_filename  \\\n",
       "0           100p1_v1        100p          1  2022-12-28/P1_R1.fastq.bz2   \n",
       "1           100p3_v2        100p          3  2022-12-28/P3_R1.fastq.bz2   \n",
       "2           0m_p1_v2     100p_0m          1  2023-02-28/P4_R1.fastq.bz2   \n",
       "\n",
       "         fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2022-12-28/P1_R2.fastq.bz2     D501     D701    l1Sp   human   \n",
       "1  2022-12-28/P3_R2.fastq.bz2     D501     D703    l2Sp   human   \n",
       "2  2023-02-28/P4_R2.fastq.bz2     D502     D701    l5Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge   100p_0m          2     GCCTCTAT   \n",
       "\n",
       "  P7_index_seq          barcode_seq  N_total    N_CC    N_CCA  N_CCA+CC  \\\n",
       "0     ATTACTCG     GGCTGCCATGCGACTA  4882120  223691  4625330   4849021   \n",
       "1     CGCTCATT   GGCTGCCATGCTGTCACG  3002197  149772  2828518   2978290   \n",
       "2     ATTACTCG  GGCTGCCATGCAACGCATC  2989860  163356  2798679   2962035   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.322036    95.386883       4801306          515441   \n",
       "1             99.203683    94.971208       2951336          492419   \n",
       "2             99.069354    94.485008       2931995          494668   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0   524232.741440                    98.344695               98.322932   \n",
       "1   522405.107612                    98.305874               94.259990   \n",
       "2   522334.350496                    98.064625               94.703325   \n",
       "\n",
       "    N_mapped  percent_single_annotation  percent_multiple_annotation  \\\n",
       "0  4774581.0                  79.221821                    20.778179   \n",
       "1  2918101.0                  76.457909                    23.542091   \n",
       "2  2895784.0                  74.423127                    25.576873   \n",
       "\n",
       "   percent_multiple_codons  Mapping_percent  \n",
       "0                 1.408794        99.443381  \n",
       "1                 2.047804        98.873900  \n",
       "2                 2.288948        98.764971  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Align reads to unmasked reference tRNAs:\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=5, overwrite=False, load_previous=False)\n",
    "\n",
    "# Collect alignment statistics:\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=True)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=9, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make new folders for the masked alignment/stats:\n",
    "dir_dict_masked = copy.deepcopy(dir_dict)\n",
    "dir_dict_masked['align_dir'] = 'SWalign_masked'\n",
    "dir_dict_masked['stats_dir'] = 'stats_collection_masked'\n",
    "\n",
    "# Align reads to unmasked reference tRNAs:\n",
    "align_obj = SWIPE_align(dir_dict_masked, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=True)\n",
    "sample_df = align_obj.run_parallel(n_jobs=5, overwrite=False, load_previous=False)\n",
    "\n",
    "# Collect alignment statistics:\n",
    "stats_obj = STATS_collection(dir_dict_masked, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=True)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=9, load_previous=False)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
