{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "workbookDir: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/3-stats_collection\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sys, os, subprocess, shutil, glob, bz2\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import json\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.ticker as ticker\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "%matplotlib inline\n",
    "\n",
    "\n",
    "# Navigate back to workbookDir in case of re-running a code block:\n",
    "if not 'workbookDir' in globals():\n",
    "    workbookDir = os.getcwd()\n",
    "print('workbookDir: ' + workbookDir)\n",
    "os.chdir(workbookDir)  # If you changed the current working dir, this will take you back to the workbook dir."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Requirements\n",
    "1. gfsa\n",
    "2. hgrewsa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNAcompRNA = {a: b for a, b in zip('ATGC', 'UACG')}\n",
    "def anticodon2codon(anticodon):\n",
    "    codon = ''.join([DNAcompRNA[b] for b in anticodon[::-1]])\n",
    "    return(codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Comments ######\n",
    "\n",
    "# Parse all data and for each sample generate the following dataframe:\n",
    "# readID, sample_name, replicate, barcode, tRNA_annotation, align_score, unique_annotation, tRNA_annotation_len, align_5p_idx, align_3p_idx, align_5p_nt, align_3p_nt, codon, anticodon, amino_acid, 5p_cover, 3p_cover, 5p_non-temp, 3p_non-temp\n",
    "\n",
    "# readID -> (string) fastq raw data read header/ID\n",
    "# sample_name -> (string) non unique sample name\n",
    "# replicate -> (integer) replicate number (sample_name - replicate should be unique)\n",
    "# barcode -> (string) barcode name\n",
    "# tRNA_annotation -> (string) tRNA annotation string, either unique or multiple separated by @\n",
    "# align_score -> (integer) alignment score\n",
    "# unique_annotation -> (True/False) if annotation is unique or (multiple or centroid)\n",
    "# tRNA_annotation_len -> (integer) the length of the tRNA annotated, if multiple annotations take the first\n",
    "# align_5p_idx -> (integer) first base (by index) of annotated tRNA in the alignement\n",
    "# align_3p_idx -> (integer) last base (by index) of annotated tRNA in the alignement (if CC add one to index)\n",
    "# align_5p_nt -> (string) first base of annotated tRNA in the alignement\n",
    "# align_3p_nt -> (string) last base of annotated tRNA in the alignement\n",
    "# codon -> (string) codon\n",
    "# anticodon -> (string) anticodon\n",
    "# amino_acid -> (string) amino acid\n",
    "# 5p_cover -> (True/False) if align_5p_idx == 1\n",
    "# 3p_cover -> (True/False) if align_3p_idx == tRNA_annotation_len\n",
    "# 5p_non-temp -> (string) 5' bases not part of the alignment to the annotation i.e. non-template bases\n",
    "# 3p_non-temp -> (string) 3' bases not part of the alignment to the annotation i.e. non-template bases\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Write by appending to previous sample, at the end bzip the whole thing.\n",
    "# Generate a smaller \"charging statistics\" dataframe by selecting:\n",
    "# (df['3p_cover'] == True) & (df['align_score'] >= MIN_SCORE)\n",
    "# and then aggregating on 3p_end == 'CCA' or 3p_end == 'CC'\n",
    "# Also write the file. When this is done all statistics and plots can be derived from this.\n",
    "\n",
    "\n",
    "# Stats general:\n",
    "# Percent covering 3' (read covering end[-1]), percent with 3' truncation\n",
    "# Percent of those covering 3' are full length\n",
    "# Percent sequences with 0, 1, 2, ..., n non-template extensions (defined as not aligned) and their composition e.g. columns of 0, 1, 2,.. n and rows of ATGC.\n",
    "# Break ^ down by preceeding base and plot as frequency logo plot for ATCG and all\n",
    "# Break ^^ down by 5p_cover True/False\n",
    "\n",
    "\n",
    "# TODO:\n",
    "# Plot/use the NaCl data\n",
    "# Test for bias in any of the barcodes\n",
    "# Add unspliced tRNA as template and dumb seqeunces that map to unspliced tRNA\n",
    "# Plot for specific tRNA sequences to see if any are more or less charged i.e. has regulatory potential.\n",
    "# Read count ratio cyto vs. mito"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables that changes from batch to batch:\n",
    "MIN_SCORE = 25\n",
    "data_folder = 'data/pilot_exp'\n",
    "project_folder = 'projects/pilot_exp'\n",
    "tRNA_database = '../../../2-align_reads/tRNA_database/hg19_mature-tRNA.fa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables that should not change:\n",
    "data_dir = 'stats_collection'\n",
    "align_dir = 'SWalign'\n",
    "umi_dir = 'UMI_trimmed'\n",
    "sample_list = 'sample_list.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read sample information ###\n",
    "sample_df = pd.read_excel('../' + project_folder + '/' + sample_list)\n",
    "sample_dict = {un: {'sample_name': n, 'replicate': r, 'barcode': b} for un, n, r, b in zip(sample_df['sample_name_unique'].values, sample_df['sample_name'].values, sample_df['replicate'].values, sample_df['barcode'].values)}\n",
    "\n",
    "# Create folder for data and stats:\n",
    "os.chdir('../' + data_folder)\n",
    "stats_dir = '../../' + project_folder + '/stats_collection'\n",
    "try:\n",
    "    os.mkdir(stats_dir) # For stats\n",
    "except:\n",
    "    shutil.rmtree(stats_dir)\n",
    "    os.mkdir(stats_dir)\n",
    "# For manipulations and final data:\n",
    "try:\n",
    "    os.mkdir(data_dir) # For data\n",
    "except:\n",
    "    shutil.rmtree(data_dir)\n",
    "    os.mkdir(data_dir)\n",
    "os.chdir(data_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the tRNA database to find the length of each sequence:\n",
    "tRNA_data = dict()\n",
    "for record in SeqIO.parse(tRNA_database, \"fasta\"):\n",
    "    tRNA_data[record.id] = dict()\n",
    "    tRNA_data[record.id]['len'] = len(record.seq)\n",
    "    tRNA_data[record.id]['codon'] = anticodon2codon(record.id.split('-')[2])\n",
    "    tRNA_data[record.id]['anticodon'] = record.id.split('-')[2]\n",
    "    tRNA_data[record.id]['amino_acid'] = record.id.split('-')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_csv_fnam = 'stats_collection.csv.bz2'\n",
    "agg_csv_fnam = 'stats_filtered_CC-CCA-aggregate.csv'\n",
    "try:\n",
    "    os.remove(stat_csv_fnam)\n",
    "    os.remove(agg_csv_fnam)\n",
    "except:\n",
    "    pass\n",
    "\n",
    "# Open filehandles and printer headers:\n",
    "fh_stats_out = bz2.open(stat_csv_fnam, 'ab')\n",
    "header = ['readID', 'sample_name', 'replicate', 'barcode', 'tRNA_annotation', 'align_score', 'unique_annotation', 'tRNA_annotation_len', 'align_5p_idx', 'align_3p_idx', 'align_5p_nt', 'align_3p_nt', 'codon', 'anticodon', 'amino_acid', '5p_cover', '3p_cover', '5p_non-temp', '3p_non-temp']\n",
    "header_bin = str.encode(','.join(header))\n",
    "fh_stats_out.write(header_bin)\n",
    "\n",
    "fh_agg_out = open(agg_csv_fnam, 'a')\n",
    "agg_cols = ['sample_name', 'replicate', 'barcode', 'tRNA_annotation', 'tRNA_annotation_len', 'unique_annotation', 'align_3p_nt', 'codon', 'anticodon', 'amino_acid', 'count']\n",
    "print(','.join(agg_cols), file=fh_agg_out)\n",
    "\n",
    "# Files to collect stats from:\n",
    "json_files = glob.glob('../' + align_dir + '/*.json.bz2')\n",
    "for json_file in json_files:\n",
    "    fnam_base = '_'.join(json_file.split('/')[-1].split('_')[0:-1])\n",
    "    uidx = fnam_base.index('UMI')\n",
    "    unique_sample_name = fnam_base[0:uidx-1]\n",
    "    \n",
    "    with bz2.open(json_file, 'rt', encoding=\"utf-8\") as fh_gz:\n",
    "        SWres = json.loads(fh_gz.read())\n",
    "    \n",
    "    # Extract non-template bases from UMI processed reads:\n",
    "    fastq_fnam = '../' + umi_dir + '/' + fnam_base + '.fastq.bz2'\n",
    "    with bz2.open(fastq_fnam, 'rt') as fh_gz:\n",
    "        for UMIread in SeqIO.parse(fh_gz, \"fastq\"):\n",
    "            if UMIread.id in SWres:\n",
    "                qpos = SWres[UMIread.id]['qpos'][0]\n",
    "                SWres[UMIread.id]['5p_non-temp'] = str(UMIread.seq)[0:(qpos[0]-1)]\n",
    "                SWres[UMIread.id]['3p_non-temp'] = str(UMIread.seq)[qpos[1]:]\n",
    "\n",
    "    ### Collect stats on a per sample basis and store in tmp file ###\n",
    "    with open('tmp_stat.csv', 'w') as tmp_csv:\n",
    "        print(','.join(header), file=tmp_csv)\n",
    "        for i, readID in enumerate(SWres):\n",
    "            sample_name = sample_dict[unique_sample_name]['sample_name']\n",
    "            replicate = sample_dict[unique_sample_name]['replicate']\n",
    "            barcode = sample_dict[unique_sample_name]['barcode']\n",
    "            tRNA_annotation = SWres[readID]['name']\n",
    "            tRNA_annotation_first = tRNA_annotation.split('@')[0]\n",
    "            align_score = SWres[readID]['score']\n",
    "            unique_annotation = '@' not in tRNA_annotation\n",
    "            tRNA_annotation_len = tRNA_data[tRNA_annotation_first]['len']\n",
    "            align_5p_idx, align_3p_idx = SWres[readID]['dpos'][0]\n",
    "            align_5p_nt = SWres[readID]['qseq'][0]\n",
    "            align_3p_nt = SWres[readID]['qseq'][-1]\n",
    "            # Move index for reads with beta-eliminated A:\n",
    "            if align_3p_idx == (tRNA_annotation_len - 1) and align_3p_nt == 'C':\n",
    "                align_3p_idx += 1\n",
    "            codon = tRNA_data[tRNA_annotation_first]['codon']\n",
    "            anticodon = tRNA_data[tRNA_annotation_first]['anticodon']\n",
    "            amino_acid = tRNA_data[tRNA_annotation_first]['amino_acid']\n",
    "            _5p_cover = align_5p_idx == 1\n",
    "            _3p_cover = align_3p_idx == tRNA_annotation_len\n",
    "            _5p_non_temp = SWres[readID]['5p_non-temp']\n",
    "            _3p_non_temp = SWres[readID]['3p_non-temp']\n",
    "            # Print line to tmp file:\n",
    "            csv_line = ','.join(map(str, [readID, sample_name, replicate, barcode, tRNA_annotation, align_score, unique_annotation, tRNA_annotation_len, align_5p_idx, align_3p_idx, align_5p_nt, align_3p_nt, codon, anticodon, amino_acid, _5p_cover, _3p_cover, _5p_non_temp, _3p_non_temp]))\n",
    "            print(csv_line, file=tmp_csv)\n",
    "    # Append the sample statistics to the final csv file using bz2 compression:\n",
    "    with open('tmp_stat.csv', 'rb') as tmp_csv:\n",
    "        fh_stats_out.write(tmp_csv.read())\n",
    "\n",
    "    ### Aggregate filtered data to count charged/uncharged tRNAs ###\n",
    "    # Read stats from tmp csv file (this is faster than building row by row):\n",
    "    stat_df = pd.read_csv('tmp_stat.csv', keep_default_na=False)\n",
    "    os.remove('tmp_stat.csv')\n",
    "    # 3' must be covered and no 3' non-template bases:\n",
    "    row_mask = (stat_df['3p_cover']) & (stat_df['3p_non-temp'] == '')\n",
    "    agg_df = stat_df.loc[row_mask, agg_cols[0:-1]]\n",
    "    agg_df['count'] = stat_df.loc[row_mask, ['align_3p_nt']]  # dummy for groupby count\n",
    "    agg_df = agg_df.groupby(agg_cols, as_index=False).agg({\"count\": \"count\"})\n",
    "    agg_df.to_csv(fh_agg_out, header=False, index=False, mode='a')\n",
    "\n",
    "fh_stats_out.close()\n",
    "fh_agg_out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(agg_csv_fnam)\n",
    "dfcol = df.columns.values\n",
    "col_mask = (dfcol != 'align_3p_nt') & (dfcol != 'count')\n",
    "cols_sele = dfcol[col_mask]\n",
    "\n",
    "df_charge = pd.DataFrame(columns=cols_sele)\n",
    "df_charge['charge'] = []\n",
    "\n",
    "df_dict = dict()\n",
    "for row in df.iterrows():\n",
    "    key = tuple(row[1].loc[cols_sele])\n",
    "    if key in df_dict:\n",
    "        df_dict[key][row[1]['align_3p_nt']] = row[1]['count']\n",
    "    else:\n",
    "        df_dict[key] = dict()\n",
    "        df_dict[key][row[1]['align_3p_nt']] = row[1]['count']\n",
    "\n",
    "for key in df_dict.keys():\n",
    "    if 'A' not in df_dict[key]:\n",
    "        df_dict[key]['A'] = 0\n",
    "    if 'C' not in df_dict[key]:\n",
    "        df_dict[key]['C'] = 0\n",
    "    \n",
    "    df_dict[key]['charge'] = df_dict[key]['A'] / (df_dict[key]['A'] + df_dict[key]['C']) * 100\n",
    "\n",
    "for i, key in enumerate(df_dict.keys()):\n",
    "    row = list(key)\n",
    "    row.append(df_dict[key]['charge'])\n",
    "    df_charge.loc[i] = row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes = set(df_charge['barcode'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "charge_norm = dict()\n",
    "for barcode in barcodes:\n",
    "    bar_mask = (df_charge['barcode'] == barcode)\n",
    "    eColi_mask = (df_charge['amino_acid'] == 'eColiLys')\n",
    "    eColi_charge = df_charge[bar_mask&eColi_mask]['charge'].values.mean()\n",
    "    charge_norm[barcode] = eColi_charge\n",
    "\n",
    "max_charge = max(charge_norm.values())\n",
    "for barcode in barcodes:\n",
    "    charge_norm[barcode] = max_charge / charge_norm[barcode]\n",
    "\n",
    "df_charge['charge_norm'] = [0] * len(df_charge)\n",
    "for barcode in barcodes:\n",
    "    bar_mask = (df_charge['barcode'] == barcode)\n",
    "    df_charge.loc[bar_mask, 'charge_norm'] = df_charge.loc[bar_mask, 'charge'] * charge_norm[barcode]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Rich</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Glu-TTC-1-1</td>\n",
       "      <td>72</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>GAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Rich</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Glu-TTC-1-1</td>\n",
       "      <td>72</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>GAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>Rich</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-CTC-1-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>GAG</td>\n",
       "      <td>CTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>1208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>Rich</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-CTC-1-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>GAG</td>\n",
       "      <td>CTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>Rich</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-CTC-1-1@Homo_sapiens_tRN...</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>GAG</td>\n",
       "      <td>CTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9976</th>\n",
       "      <td>A-NaCl</td>\n",
       "      <td>1</td>\n",
       "      <td>l4</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-TTC-2-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>GAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9977</th>\n",
       "      <td>A-NaCl</td>\n",
       "      <td>1</td>\n",
       "      <td>l4</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-TTC-2-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>GAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9978</th>\n",
       "      <td>A-NaCl</td>\n",
       "      <td>1</td>\n",
       "      <td>l4</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-TTC-3-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>GAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9979</th>\n",
       "      <td>A-NaCl</td>\n",
       "      <td>1</td>\n",
       "      <td>l4</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-TTC-4-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>GAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9980</th>\n",
       "      <td>A-NaCl</td>\n",
       "      <td>1</td>\n",
       "      <td>l4</td>\n",
       "      <td>Homo_sapiens_tRNA-Glu-TTC-4-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>GAA</td>\n",
       "      <td>TTC</td>\n",
       "      <td>Glu</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>237 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_name  replicate barcode  \\\n",
       "14          Rich          1      l1   \n",
       "15          Rich          1      l1   \n",
       "263         Rich          1      l1   \n",
       "264         Rich          1      l1   \n",
       "265         Rich          1      l1   \n",
       "...          ...        ...     ...   \n",
       "9976      A-NaCl          1      l4   \n",
       "9977      A-NaCl          1      l4   \n",
       "9978      A-NaCl          1      l4   \n",
       "9979      A-NaCl          1      l4   \n",
       "9980      A-NaCl          1      l4   \n",
       "\n",
       "                                        tRNA_annotation  tRNA_annotation_len  \\\n",
       "14                   Homo_sapiens_mito_tRNA-Glu-TTC-1-1                   72   \n",
       "15                   Homo_sapiens_mito_tRNA-Glu-TTC-1-1                   72   \n",
       "263                       Homo_sapiens_tRNA-Glu-CTC-1-1                   75   \n",
       "264                       Homo_sapiens_tRNA-Glu-CTC-1-1                   75   \n",
       "265   Homo_sapiens_tRNA-Glu-CTC-1-1@Homo_sapiens_tRN...                   75   \n",
       "...                                                 ...                  ...   \n",
       "9976                      Homo_sapiens_tRNA-Glu-TTC-2-1                   75   \n",
       "9977                      Homo_sapiens_tRNA-Glu-TTC-2-1                   75   \n",
       "9978                      Homo_sapiens_tRNA-Glu-TTC-3-1                   75   \n",
       "9979                      Homo_sapiens_tRNA-Glu-TTC-4-1                   75   \n",
       "9980                      Homo_sapiens_tRNA-Glu-TTC-4-1                   75   \n",
       "\n",
       "      unique_annotation align_3p_nt codon anticodon amino_acid  count  \n",
       "14                 True           A   GAA       TTC        Glu     19  \n",
       "15                 True           C   GAA       TTC        Glu     33  \n",
       "263                True           A   GAG       CTC        Glu   1208  \n",
       "264                True           C   GAG       CTC        Glu    434  \n",
       "265               False           A   GAG       CTC        Glu      1  \n",
       "...                 ...         ...   ...       ...        ...    ...  \n",
       "9976               True           A   GAA       TTC        Glu    627  \n",
       "9977               True           C   GAA       TTC        Glu      3  \n",
       "9978               True           A   GAA       TTC        Glu      2  \n",
       "9979               True           A   GAA       TTC        Glu    348  \n",
       "9980               True           C   GAA       TTC        Glu      3  \n",
       "\n",
       "[237 rows x 11 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['amino_acid'] == 'Glu']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfAAAAFgCAYAAABEyiulAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAVOElEQVR4nO3df7DldX3f8dc7rLBQRJYQb3FBlgbGrpI2U64mwpiy3dimJde1DFQdSUiGuuko/ui0tETDOA1lYofEMQ7tZKimIdG4xXUz7JU06mzEpJlq3FUr4NZyDS2w0iXaRTTC8sNP/7hn63XZH2eXe+73fnYfjxnmnO/3/Hrf+c6dJ5/vOXtutdYCAPTlh4YeAAA4cgIOAB0ScADokIADQIcEHAA6tGLoAZ6LM888s61Zs2boMQBgYnbs2PGN1tqP7L+/64CvWbMm27dvH3oMAJiYqvrfB9rvFDoAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0aGIBr6rfrqpHquqeBfvOqKpPVdV9o8tVC2775aqaq6qvVtU/mNRcAHAsmOQK/HeS/Mx++65Psq21dkGSbaPtVNVLk7w+yctGj/kPVXXCBGcDgK5NLOCttT9J8n/3270hyW2j67clee2C/Ztaa3tba/cnmUvyiknNBgC9W+o/ZjLVWns4SVprD1fVC0f7Vyf57IL7PTTa9yxVtTHJxiSZmprKXXfdNblpAWCZWi5/jawOsK8d6I6ttVuT3Jok09PT7dJLL53gWAxtZmZm6BEO6rvf/e4RP+aUU06ZwCTP3ezs7NAjAEdoqQO+u6rOGq2+z0ryyGj/Q0nOWXC/s5N8fYlnY5nau+uew99pCKv+xhE/ZDn+LCetvnDoEYCjsNQB35rk6iTvGV3esWD/71fVe5O8KMkFSf58iWdjGfvQdZcNPcKzvOEDO4/4Mcvt57jq5juHHgE4ShMLeFV9JMmlSc6sqoeSvDvz4b69qq5J8kCSK5OktXZvVd2e5CtJnk7yltbaM5OaDRbDR/7p2qFHAI5jEwt4a+0NB7lp/UHuf1OSmyY1DwAcS3wTGwB0SMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADokIADQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADokIADQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6NAgAa+qf15V91bVPVX1kapaWVVnVNWnquq+0eWqIWYDgB4secCranWStyWZbq1dmOSEJK9Pcn2Sba21C5JsG20DAAcw1Cn0FUlOrqoVSU5J8vUkG5LcNrr9tiSvHWY0AFj+Viz1C7bWdlXVryd5IMnjST7ZWvtkVU211h4e3efhqnrhgR5fVRuTbEySqamp3HXXXUs0OUPYsGFD2lN/P18+6fShRzkmXfmml6Sed7LfI+hQtdaW9gXn39v+WJLXJXk0yUeTbE5yS2vt9AX329NaO+T74NPT02379u2TG5bBzczMZO+ue/Kh6y4bepRj0lU335mTVl+Y2dnZoUcBDqKqdrTWpvffP8Qp9J9Ocn9r7S9ba08l2ZLk4iS7q+qsJBldPjLAbADQhSEC/kCSn6yqU6qqkqxPsjPJ1iRXj+5zdZI7BpgNALowxHvgn6uqzUm+kOTpJF9McmuSU5PcXlXXZD7yVy71bADQiyUPeJK01t6d5N377d6b+dU4AHAYvokNADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADokIADQIdWDD0AAIxj/fr1R/yYbdu2TWCS5cEKHAA6ZAUOQBcOtJretyo/llfaB2MFDgAdEnAA6JBT6AD8fzMzM0OPcFR6nHt2dvY5PV7AAfgBj+3cOfQIY1tx7rlJ+po5SU5bu/Y5P4eAA/Ast1xxxdAjjOUdn/98kn7mTZJrN29elOfxHjgAdMgKHIAu7FttH8lt73v5yyc1zuCswAGgQ1bgAHThWF5NHw0rcADokIADQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADo0CABr6rTq2pzVf2PqtpZVa+sqjOq6lNVdd/octUQswFAD1YM9Lq/meSPWmtXVNWJSU5J8s4k21pr76mq65Ncn+RfDzQfcIxav379ET9m27ZtE5gEnpslX4FX1WlJfirJB5OktfZka+3RJBuS3Da6221JXrvUswFAL6q1trQvWPXjSW5N8pUkfzvJjiRvT7KrtXb6gvvtaa096zR6VW1MsjFJpqamLtq0adMSTM1Q5ubm0p56POdNnT70KMek+3c/mnreyTn//POHHmVwN954Y5LkhhtuGHiSYc3NzeWZJ57Ii1d5F3NSHtizJyesXDn27926det2tNam998/RMCnk3w2ySWttc9V1W8meSzJW8cJ+ELT09Nt+/btE52XYc3MzGTvrnvyoesuG3qUY9JVN9+Zk1ZfmNnZ2aFHGdy+U+vH++nymZmZPLZzZ2654oqhRzlmXbt5c05bu3bs37uqOmDAh/gQ20NJHmqtfW60vTnJ30myu6rOSpLR5SMDzAYAXVjygLfW/k+SB6vqJaNd6zN/On1rkqtH+65OcsdSzwYAvRjqU+hvTfLh0SfQ/yLJL2b+fyZur6prkjyQ5MqBZgMWwczMzNAjHJXe5vb2x/FrkIC31r6U5Fnn8zO/GgeOETt37xx6hLGd+/xzk/Q189qptUOPwICGWoEDx4nLb7h86BHGsuN9O5L0M++WG7cMPQID81WqANAhAQeADo0d8Ko6t6p+enT95Kp6/uTGAgAOZaz3wKvqTZn/9rMzkvxokrOT/FZ86AzozL73uo/k9ovecdGkxoGjNu4K/C1JLsn8N6altXZfkhdOaigA4NDG/RT63tbak1WVJKmqFUmW9jtYARaB1TTHinFX4J+pqncmObmqXp3ko0l8ewAADGTcgF+f5C+T3J3kl5L8YZJfmdRQAMChjXUKvbX2vST/cfQfADCwcT+Ffnee/Z73t5JsT/JvW2vfXOzBAICDG/dDbP8lyTNJfn+0/frR5WNJfidJX9/+DwCdGzfgl7TWLlmwfXdV/Vlr7ZKqumoSgwEABzfuh9hOraqf2LdRVa9Icupo8+lFnwoAOKRxV+DXJPlPVbUv2t9Ock1V/bUkvzaRyQCAgzpswKvqhCSvaq39WFW9IEm11h5dcJfbJzUcAHBghz2F3lp7JsmG0fVv7RdvAGAA455C/7OquiXJf07yV/t2tta+MJGpAIBDGjfgF48uf3XBvpbk7y3uOADAOMb9JrZ1kx4EABjfuCvwVNVlSV6WZOW+fa21Xz34IwCASRnr34FX1W8leV2StyapJFcmOXeCcwEAhzDuF7lc3Fr7+SR7Wmv/Jskrk5wzubEAgEMZN+CPjy6/W1UvSvJUkvMmMxIAcDjjvgf+8ao6PcnNSb6Q+U+gf2BSQwEAhzbup9BvHF39WFV9PMnK1tq3JjfW8WP9+vVH/Jht27ZNYBIAenIkn0K/OMmafY+pqrTWfndCcwEAhzBWwKvq95L8aJIvZf7vgifzp9EF/Dk62Gp638rcahuAAxl3BT6d5KWttTbJYQCA8Yz7KfR7kvz1SQ4CAIzvkCvwqprN/Kny5yf5SlX9eZK9+25vrb1msuMBAAdyuFPoW5NMJfnT/fb/3SS7JjLRBMzMzAw9wlHrbfbZ2dmhRwA4Lhwu4BuSvLO19uWFO6vqr5K8O8kHJzXYYrv3wW8MPcIROe+HT0nS19wvO+fMoUcAOG4cLuBr9o93krTWtlfVmsmMNDmvedtNQ48wtrs/PD9rLzNvff+7hh4B4LhyuA+xrTzEbScv5iAAwPgOF/DPV9Wb9t9ZVdck2TGZkQCAwzncKfR3JPmDqnpjvh/s6SQnJvnHE5wLADiEQwa8tbY7ycVVtS7JhaPdd7bW/njikx0n9r3XfSS3/9gbvd8McLwb94+ZfDrJpyc8CwAwprH/mAmTYTUNwNEY96tUAYBlRMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADo0GABr6oTquqLVfXx0fYZVfWpqrpvdLlqqNkAYLkbcgX+9iQ7F2xfn2Rba+2CJNtG2wDAAQwS8Ko6O8llST6wYPeGJLeNrt+W5LVLPBYAdGOoFfj7kvyrJN9bsG+qtfZwkowuXzjAXADQhWqtLe0LVv1skn/UWntzVV2a5F+21n62qh5trZ2+4H57WmvPeh+8qjYm2ZgkU1NTF23atOmwrzk3N5cnnnw6L5havUg/Bfv71u5dWXniipx//vmL+rxzc3NpTz2e86ZOX9TnZd79ux9NPe/kRT9uyej37uknsuosH2eZhD0P78nKFSsn8jv3zBNP5MWrHLdJeWDPnpywcvxjt27duh2tten99w8R8F9L8nNJnk6yMslpSbYkeXmSS1trD1fVWUnuaq295FDPNT093bZv337Y15yZmcm9D34jr3nbTc95fg5s6/vflZedc2ZmZ2cX9XlnZmayd9c9+dB1ly3q8zLvqpvvzEmrL1z045bMH7udu3fm8hsuX/TnJtly45asnVo7kd+5x3buzC1XXLGoz8v3Xbt5c05bO/6xq6oDBnzJT6G31n65tXZ2a21Nktcn+ePW2lVJtia5enS3q5PcsdSzAUAvltO/A39PkldX1X1JXj3aBgAOYMWQL95auyvJXaPr30yyfsh5AKAXy2kFDgCMScABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADokIADQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADokIADQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6JCAA0CHljzgVXVOVX26qnZW1b1V9fbR/jOq6lNVdd/octVSzwYAvRhiBf50kn/RWlub5CeTvKWqXprk+iTbWmsXJNk22gYADmDJA95ae7i19oXR9W8n2ZlkdZINSW4b3e22JK9d6tkAoBfVWhvuxavWJPmTJBcmeaC1dvqC2/a01p51Gr2qNibZmCRTU1MXbdq06bCvMzc3lyeefDovmFq9SJOzv2/t3pWVJ67I+eefv6jPOzc3l/bU4zlv6vRFfV7m3b/70dTzTl7045aMfu+efiKrzvJu2CTseXhPVq5YOZHfuWeeeCIvXuW4TcoDe/bkhJXjH7t169btaK1N779/sIBX1alJPpPkptbalqp6dJyALzQ9Pd22b99+2NeamZnJvQ9+I695203PdWwOYuv735WXnXNmZmdnF/V5Z2ZmsnfXPfnQdZct6vMy76qb78xJqy9c9OOWzB+7nbt35vIbLl/05ybZcuOWrJ1aO5Hfucd27swtV1yxqM/L9127eXNOWzv+sauqAwZ8kE+hV9XzknwsyYdba1tGu3dX1Vmj289K8sgQswFAD4b4FHol+WCSna219y64aWuSq0fXr05yx1LPBgC9WDHAa16S5OeS3F1VXxrte2eS9yS5vaquSfJAkisHmA0AurDkAW+t/dckdZCb1y/lLADQK9/EBgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADokIADQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANAhAQeADgk4AHRIwAGgQwIOAB0ScADokIADQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHRJwAOiQgANAhwQcADok4ADQIQEHgA4JOAB0SMABoEMCDgAdEnAA6JCAA0CHBBwAOiTgANChZRfwqvqZqvpqVc1V1fVDzwMAy9GyCnhVnZDk3yf5h0lemuQNVfXSYacCgOVnxdAD7OcVSeZaa3+RJFW1KcmGJF9ZjCff+v53LcbTMICrbr5z6BE4Sltu3DL0CByFazdvHnoEDmO5BXx1kgcXbD+U5CcW3qGqNibZONr8TlV9dYlmG8KZSb4x9BDjuv+/J1U19BjLQVfHLV/8X47b93V17L6Wrzl287o6bkmSrx3RsTv3QDuXW8AP9NO0H9ho7dYkty7NOMOqqu2ttemh5+DIOG79cuz6dLwet2X1HnjmV9znLNg+O8nXB5oFAJat5Rbwzye5oKrOq6oTk7w+ydaBZwKAZWdZnUJvrT1dVdcm+USSE5L8dmvt3oHHGtJx8VbBMchx65dj16fj8rhVa+3w9wIAlpXldgodABiDgANAhwR8Gamq7yy4/kdV9WhVfXzImRjPwmM32j6tqnZV1S1DzcTh7TtuVfXjVfXfqureqvpyVb1u6Nk4tAXH7tyq2lFVXxodv3829GxLZVl9iI0fcHOSU5L80tCDcFRuTPKZoYdgbN9N8vOttfuq6kVJdlTVJ1prjw48F4f3cJKLW2t7q+rUJPdU1dbW2jH/T5CtwJep1tq2JN8eeg6OXFVdlGQqySeHnoXxtNb+Z2vtvtH1ryd5JMmPDDsV42itPdla2zvaPCnHUdeOmx8UlkJV/VCS30hy3dCzcHSq6hVJTkzytaFnYTxVdU5VfTnzX8X9746H1Xci4LDY3pzkD1trDx72niw7VXVWkt9L8outte8NPQ/jaa092Fr7W0nOT3J1VU0NPdNS8B44LK5XJnlVVb05yalJTqyq77TW/G37Za6qTktyZ5Jfaa19duh5OHKtta9X1b1JXpXkmP9zagIOi6i19sZ916vqF5JMi/fyN/rq5j9I8ruttY8OPQ/jq6qzk3yztfZ4Va1KckmS9w481pIQ8GWqqv40yd9McmpVPZTkmtbaJwYeC45V/yTJTyX54dH/eCXJL7TWvjTYRIxrbZLfqKqW+b9o+euttbsHnmlJ+CpVAOiQD7EBQIcEHAA6JOAA0CEBB4AOCTgAdEjAAaBDAg4AHfp/BuurqhLm/HkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 504x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#mask = (df_charge['tRNA_annotation'] == 'Homo_sapiens_tRNA-Glu-CTC-1-1') & (df_charge['sample_name'] != 'U')\n",
    "mask = (df_charge['sample_name'] != 'U')\n",
    "\n",
    "fig, ax1 = plt.subplots(1, 1, figsize=(7, 5))\n",
    "\n",
    "g1 = sns.barplot(ax=ax1, x='barcode', y='charge', data=df_charge[mask], capsize=.1, errwidth=2, edgecolor='black', linewidth=2, alpha=0.7)\n",
    "g1.grid(True, axis='y')\n",
    "g1.set_ylabel('Charge');\n",
    "g1.set_xlabel('');\n",
    "#g1.set_xticklabels(g1.get_xticklabels(), rotation=90)\n",
    "\n",
    "\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfAAAAFgCAYAAABEyiulAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAYaElEQVR4nO3df5RfdX3n8ee7CRBYfiQUmRMDEjQcDaB1y4gK6pLG3drSIZYGfxyp0ENNexTR/cEWf3BUshzoQa2l2NNlWzUWa4oxlgy0KidLdGvrjwlQCUxdRrGBkAbRCcFCAsH3/vG9s4xh5js3M3Pnzid5Ps6Z8/3ez/d+733l3DO8uD/m3shMJElSWX6h7QCSJGn/WeCSJBXIApckqUAWuCRJBbLAJUkq0Ny2A0zFcccdl4sXL247hiRJjdm8efOjmfm8fceLLvDFixczMDDQdgxJkhoTEf8y1riH0CVJKpAFLklSgSxwSZIKZIFLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSpQYwUeEZ+KiEciYsuosWMj4vaIuL96XTDqs/dFxFBEfC8ifrWpXJIkHQia3AP/DPCGfcauADZm5inAxmqaiDgVeAtwWvWdP42IOQ1mkySpaI0VeGZ+HfjJPsMrgDXV+zXAG0eNr83MPZn5ADAEnNlUNkmSSjfTDzPpycztAJm5PSKOr8YXAd8cNd9D1dhzRMQqYBVAT08PmzZtai6tJEmz1Gx5GlmMMZZjzZiZNwI3AvT29uY555zTYCy1ra+vr+0I43riiSf2+ztHHHFEA0mmrr+/v+0IkvbTTBf4johYWO19LwQeqcYfAk4cNd8JwMMznE2z1J5tWyaeqQ0LXrjfX5mN/5bDFp3edgRJkzDTBb4BuAi4tnq9ZdT4X0XEx4HnA6cA357hbJrFbrr83LYjPMdb/3xwv78z2/4dF153W9sRJE1SYwUeEZ8HzgGOi4iHgA/RKe6bI+ISYCtwAUBm3hsRNwP3AXuBd2XmM01lk6bD5393adsRJB3EGivwzHzrOB8tH2f+q4Grm8ojSdKBxDuxSZJUIAtckqQCWeCSJBXIApckqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkAlngkiQVyAKXJKlAFrgkSQWywCVJKpAFLklSgSxwSZIKZIFLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSqQBS5JUoEscEmSCmSBS5JUIAtckqQCWeCSJBXIApckqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkAlngkiQVyAKXJKlAFrgkSQWywCVJKpAFLklSgSxwSZIKZIFLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSqQBS5JUoEscEmSCmSBS5JUIAtckqQCWeCSJBXIApckqUCtFHhE/OeIuDcitkTE5yNiXkQcGxG3R8T91euCNrJJklSCGS/wiFgEXAb0ZubpwBzgLcAVwMbMPAXYWE1LkqQxtHUIfS5weETMBY4AHgZWAGuqz9cAb2wnmiRJs9/cmV5hZm6LiI8CW4Enga9m5lcjoiczt1fzbI+I48f6fkSsAlYB9PT0sGnTphlKrjasWLGCfPo/8d3D5rcd5YB0wTteTBxyuL9HUoEiM2d2hZ1z218E3gzsBL4ArANuyMz5o+Ybzsyu58F7e3tzYGCgubBqXV9fH3u2beGmy89tO8oB6cLrbuOwRafT39/fdhRJ44iIzZnZu+94G4fQXw88kJk/ysyngfXAWcCOiFgIUL0+0kI2SZKK0EaBbwVeFRFHREQAy4FBYANwUTXPRcAtLWSTJKkIbZwD/1ZErAPuBPYCdwE3AkcCN0fEJXRK/oKZziZJUilmvMABMvNDwIf2Gd5DZ29ckiRNwDuxSZJUIAtckqQCWeCSJBXIApckqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkArVyK1VJkvbX8uX7f7ftjRs3NpBkdnAPXJKkArkHLkkqwlh70yN75QfynvZ43AOXJKlAFrgkSQXyELok6f/r6+trO8KklJi7v79/St+3wCVJP2fX4GDbEWqbe9JJQFmZAY5eunTKy7DAJUnPccPKlW1HqOW93/kOUE5egEvXrZuW5XgOXJKkAlngkiQVyEPokqQijBwu35/PPvGKVzQVp3XugUuSVCD3wCVJRTiQ96Ynwz1wSZIKZIFLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSqQBS5JUoEscEmSCmSBS5JUIAtckqQCWeCSJBVo3IeZRMSfADne55l5WSOJJEnShLrtgQ8Am4F5wC8D91c/LweeaTyZJEka17h74Jm5BiAiLgaWZebT1fSfAV+dkXSSJGlMdc6BPx84atT0kdWYJElqybh74KNcC9wVEXdU0/8B+HBjiSSpQcuXL9/v72zcuLGBJNLUTFjgmfnpiPg74JXV0BWZ+a/NxpIkSd1MWOAREcDrgRdm5lUR8YKIODMzv918PEmaXuPtTY/smbu3rVLUOQf+p8CrgbdW048Dn2wskSRJmlCdc+CvzMxfjoi7ADJzOCIObTiXJEnqos4e+NMRMYfqpi4R8TzgZ42mkiRJXdXZA78e+BJwfERcDawEPthoKknF6+vrazvCpJSWu7+/v+0Iakmdq9A/FxGbgeVAAG/MzMHGk0kq3uCOcv5TcdJRJwFlZV7as7TtCGpRnavQ/wL4k8z85KixD2fmh5sMJunAcP6V57cdoZbNn9gMlJN3/er1bUdQy+qcA/9V4DMR8fZRY+c1lEeSJNVQp8AfAV4HXBARn4yIuXQOpUuSpJbUKfDIzF2Z2Qf8CPgacEyzsSRJUjd1CnzDyJvqvPc1wA8byiNJkmqocxX6h/aZvhW4tbFEBxkfrCDNrJGL1fbn8zPee0ZTcaRJG3cPPCL+vnp9PCJ2jfp5PCJ2TWWlETE/ItZFxD9HxGBEvDoijo2I2yPi/up1wVTWIUnSgWzcPfDMfE31etR480zBHwNfzsyV1W1ZjwDeD2zMzGsj4grgCuAPGlj3rOKDFaSZ5d60DhTjFnhEHNvti5n5k8msMCKOpnNV+8XVcp4CnoqIFcA51WxrgE0cBAUuSdJkRGaO/UHEA3Tufz7Wn4xlZr5wUiuMeDlwI3Af8EvAZuA9wLbMnD9qvuHMfM5h9IhYBawC6OnpOWPt2rWTiTHrrV69GoArr7yy5STtGhoaIp9+kpN75rcd5YD0wI6dxCGHs2TJkmlf9tDQELv37mbBQs+GNWF4+zDz5s6b9m03NDTEM7t384IFbrembB0eZs68+ttu2bJlmzOzd9/xcQu8KRHRC3wTODszvxURfwzsAt5dp8BH6+3tzYGBgUbztsVD6B19fX3s2baFmy4/t+0oB6QLr7uNwxad3sj9tPv6+hjcMVjMnc1Ks371epb2LJ32bdfX18euwUFuWLlyWperZ126bh1HL62/7SJizAKv8zATqgvKTgHmjYxl5tdrZt3XQ8BDmfmtanodnfPdOyJiYWZuj4iFdG4gI0mSxlDnXui/S+cQ9wnA3cCrgH8EfmUyK8zMf42IByPixZn5PToPSbmv+rkIuLZ6vWUyyx9LaU8XGq207D4ZSZJmRp098PcArwC+mZnLIuIlwEemuN53A5+rrkD/AfA7dP6k7eaIuATYClwwxXX8nHsffHQ6F9e4k3/xCKCs3KedeFzbESTpoFGnwHdn5u6IICIOy8x/jogXT2WlmXk38Jzj+XT2xhtz3mVXN7n4aXXP5zpZS8m84foPtB1Bkg4qdQr8oYiYD/wNcHtEDAMPNxlKkiR1V+dWqr9Zvf1wRNxB50EmX240lSRJ6qrOw0yIiAUR8TLgcTpXkZ/eaCpJktRVnavQV9O5a9oPgJ9Vw8kkr0KXJElTV+cc+JuAF1W3PJUkSbNAnQLfAszHG6s0YuRq8/35/KVv84pvSTrY1Snwa4C7ImILsGdkMDPPayyVJEnqqk6BrwH+ELiHZ8+Ba5q4Ny1Jmow6Bf5oZl7feBJJklRbnQLfHBHXABv4+UPodzaWSpIkdVWnwP999fqqUWP+GZkkSS3qWuARMQfYkJl/NEN5JElSDV3vxJaZzwBebS5J0ixT5xD6P0TEDcBfA/82Mug5cEmS2lOnwM+qXq8aNeY5cEmSWlTnaWTLZiKIJEmqb8KnkUXEMRHx8YgYqH4+FhHHzEQ4SZI0tjqPE/0UnceIvqn62QV8uslQkiSpuzrnwF+Umb81avojEXF3Q3kkSVINdfbAn4yI14xMRMTZwJPNRZIkSROpswf++8Bnq/PeAfwEuLjJUJIkqbs6V6H/E/BLEXF0Nb2r8VSSJKmrCQs8Ig4DfgtYDMyNCAAy86ouX5MkSQ2qcwj9FuAxYDOjnkYmSZLaU6fAT8jMNzSeRJIk1VbnKvR/iIiXNp5EkiTVVmcP/DXAxRHxAJ1D6AFkZr6s0WSSJGlcdQr81xpPIUmS9kudPyP7l5kIIkmS6qtzDlySJM0yFrgkSQWqVeARcVJEvL56f3hEHNVsLEmS1E2d54G/A1gH/M9q6ATgbxrMJEmSJlBnD/xdwNl0ngNOZt4PHN9kKEmS1F2dAt+TmU+NTETEXCCbiyRJkiZSp8C/FhHvBw6PiP8IfAHobzaWJEnqpk6BXwH8CLgH+D3gb4EPNhlKkiR1V+dGLj8D/lf1I0mSZoE6zwO/h+ee834MGAD+R2b+uIlgkiRpfHXuhf53wDPAX1XTb6ledwGfAfqmP5YkSeqmToGfnZlnj5q+JyK+kZlnR8SFTQWTJEnjq3MR25ER8cqRiYg4EziymtzbSCpJktRVnT3wS4BPR8RIaT8OXBIR/w64prFkkiRpXF0LPCLmAK/NzJdGxDFAZObOUbPc3GQ4SZI0tq6H0DPzGWBF9f6xfcpbkiS1pM4h9G9ExA3AXwP/NjKYmXc2lkqSJHVVp8DPql6vGjWWwK9MfxxJklRHnTuxLZuJIJIkqb46e+BExLnAacC8kbHMvGr8b0iSpCZN+HfgEfFnwJuBdwMBXACc1HAuSZLURZ0buZyVmW8HhjPzI8CrgRObjSVJkrqpU+BPVq9PRMTzgaeBk5uLJEmSJlKnwG+NiPnAdcCdwA+BtVNdcUTMiYi7IuLWavrYiLg9Iu6vXhdMdR2SJB2oJizwzFydmTsz84t0zn2/JDOvnIZ1vwcYHDV9BbAxM08BNlbTkiRpDHWvQj8LWDwyf0SQmZ+d7Eoj4gTgXOBq4L9UwyuAc6r3a4BNwB9Mdh2SJB3IJizwiPhL4EXA3XSeCw6dG7lMusCBTwD/HThq1FhPZm4HyMztEXH8FJYvSdIBLTKz+wwRg8CpOdGMdVcY8RvAr2fmOyPiHOC/ZeZvRMTOzJw/ar7hzHzOefCIWAWsAujp6Tlj7dqJT8cPDQ2x+6m9HNOzaDr+CRrDYzu2Me/QuSxZsmRalzs0NEQ+/SQn98yf1uWq44EdO4lDDp/27QbV793e3SxY6OUsTRjePsy8ufMa+Z17ZvduXrDA7daUrcPDzJlXf9stW7Zsc2b27jtep8C/AFw2snc8VRFxDfDbdJ4lPg84GlgPvAI4p9r7XghsyswXd1tWb29vDgwMTLjOvr4+7n3wUc677Oop59fYNlz/AU478Tj6+/undbl9fX3s2baFmy4/d1qXq44Lr7uNwxadPu3bDTrbbnDHIOdfef60L1uwfvV6lvYsbeR3btfgIDesXDmty9WzLl23jqOX1t92ETFmgY97CD0i+ukcKj8KuC8ivg3sGfk8M8/b79Sd770PeF+1jnPo7IFfGBHXARcB11avt0xm+ZIkHQy6nQP/6Iyl6LgWuDkiLgG20rnjmyRJGkO3At9G58Kyb4wejIjXVZ9NWWZuonO1OZn5Y2D5dCxXkqQDXbe/A/8E8PgY409Un0mSpJZ0K/DFmfndfQczc4DO34RLkqSWdCvweV0+O3y6g0iSpPq6Ffh3IuId+w5WF5ltbi6SJEmaSLeL2N4LfCki3sazhd0LHAr8ZsO5JElSF+MWeGbuAM6KiGXA6dXwbZn5v2ckmSRJGteE90LPzDuAO2YgiyRJqqnO88AlSdIsY4FLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSqQBS5JUoEscEmSCmSBS5JUIAtckqQCWeCSJBXIApckqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkAlngkiQVyAKXJKlAFrgkSQWywCVJKpAFLklSgSxwSZIKZIFLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSqQBS5JUoEscEmSCmSBS5JUIAtckqQCWeCSJBXIApckqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkAlngkiQVyAKXJKlAFrgkSQWa8QKPiBMj4o6IGIyIeyPiPdX4sRFxe0TcX70umOlskiSVoo098L3Af83MpcCrgHdFxKnAFcDGzDwF2FhNS5KkMcx4gWfm9sy8s3r/ODAILAJWAGuq2dYAb5zpbJIklSIys72VRywGvg6cDmzNzPmjPhvOzOccRo+IVcAqgJ6enjPWrl074XqGhobY/dRejulZNE3Jta/Hdmxj3qFzWbJkybQud2hoiHz6SU7umT+ty1XHAzt2EoccPu3bDarfu727WbDQs2FNGN4+zLy58xr5nXtm925esMDt1pStw8PMmVd/2y1btmxzZvbuO95agUfEkcDXgKszc31E7KxT4KP19vbmwMDAhOvq6+vj3gcf5bzLrp5qbI1jw/Uf4LQTj6O/v39al9vX18eebVu46fJzp3W56rjwuts4bNHp077doLPtBncMcv6V50/7sgXrV69nac/SRn7ndg0OcsPKldO6XD3r0nXrOHpp/W0XEWMWeCtXoUfEIcAXgc9l5vpqeEdELKw+Xwg80kY2SZJK0MZV6AH8BTCYmR8f9dEG4KLq/UXALTOdTZKkUsxtYZ1nA78N3BMRd1dj7weuBW6OiEuArcAFLWSTJKkIM17gmfn3QIzz8fKZzCJJUqm8E5skSQWywCVJKpAFLklSgSxwSZIKZIFLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSqQBS5JUoEscEmSCmSBS5JUIAtckqQCWeCSJBXIApckqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkAlngkiQVyAKXJKlAFrgkSQWywCVJKpAFLklSgSxwSZIKZIFLklQgC1ySpAJZ4JIkFcgClySpQBa4JEkFssAlSSqQBS5JUoEscEmSCmSBS5JUIAtckqQCWeCSJBXIApckqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkAlngkiQVyAKXJKlAFrgkSQWywCVJKpAFLklSgSxwSZIKZIFLklSgWVfgEfGGiPheRAxFxBVt55EkaTaaVQUeEXOATwK/BpwKvDUiTm03lSRJs8/ctgPs40xgKDN/ABARa4EVwH3TsfAN139gOhajFlx43W1tR9AkrV+9vu0ImoRL161rO4ImMNsKfBHw4Kjph4BXjp4hIlYBq6rJn0bE92YoWxuOAx5tO0RdD/wTRETbMWaDorYbd/3Q7fasorbd9/m+266jqO0GwPf3a9udNNbgbCvwsf41+XMTmTcCN85MnHZFxEBm9radQ/vH7VYut12ZDtbtNqvOgdPZ4z5x1PQJwMMtZZEkadaabQX+HeCUiDg5Ig4F3gJsaDmTJEmzzqw6hJ6ZeyPiUuArwBzgU5l5b8ux2nRQnCo4ALndyuW2K9NBud0iMyeeS5IkzSqz7RC6JEmqwQKXJKlAFvgsEhE/HfX+yxGxMyJubTOT6hm97arpoyNiW0Tc0FYmTWxku0XEyyPiHyPi3oj4bkS8ue1s6m7UtjspIjZHxN3V9vv9trPNlFl1EZt+znXAEcDvtR1Ek7Ia+FrbIVTbE8DbM/P+iHg+sDkivpKZO1vOpYltB87KzD0RcSSwJSI2ZOYB/yfI7oHPUpm5EXi87RzafxFxBtADfLXtLKonM/9vZt5fvX8YeAR4XrupVEdmPpWZe6rJwziIeu2g+YdKMyEifgH4GHB521k0ORFxJnAo8P22s6ieiDgxIr5L51bcf3gw7H2DBS5Nt3cCf5uZD044p2adiFgI/CXwO5n5s7bzqJ7MfDAzXwYsAS6KiJ62M80Ez4FL0+vVwGsj4p3AkcChEfHTzPTZ9rNcRBwN3AZ8MDO/2XYe7b/MfDgi7gVeCxzwj1OzwKVplJlvG3kfERcDvZb37FfduvlLwGcz8wtt51F9EXEC8OPMfDIiFgBnAx9vOdaMsMBnqYj4P8BLgCMj4iHgksz8SsuxpAPVm4DXAb9Y/Y8XwMWZeXdriVTXUuBjEZF0nmj50cy8p+VMM8JbqUqSVCAvYpMkqUAWuCRJBbLAJUkqkAUuSVKBLHBJkgpkgUuSVCALXJKkAv0/RV7OsKafAgcAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 504x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#mask = (df_charge['tRNA_annotation'] == 'Homo_sapiens_tRNA-Glu-CTC-1-1') & (df_charge['sample_name'] != 'U')\n",
    "mask = (df_charge['sample_name'] != 'U')\n",
    "\n",
    "fig, ax1 = plt.subplots(1, 1, figsize=(7, 5))\n",
    "\n",
    "g1 = sns.barplot(ax=ax1, x='barcode', y='charge_norm', data=df_charge[mask], capsize=.1, errwidth=2, edgecolor='black', linewidth=2, alpha=0.7)\n",
    "g1.grid(True, axis='y')\n",
    "g1.set_ylabel('Charge normalized');\n",
    "g1.set_xlabel('');\n",
    "#g1.set_xticklabels(g1.get_xticklabels(), rotation=90)\n",
    "\n",
    "\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAfAAAAFgCAYAAABEyiulAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAfCElEQVR4nO3df5RdZX3v8feXZPLDSkgwZUgiEC6hbQTrD+KPgreXFLuqxSGWgpVKQUtLuypGvUaKtSpCc6WQtDZF60UQ4aqNmKaFSFvrTaWtrbUm+ANw6nUUG5iEwUAgtk3CJHzvH+fEDmFyZidz9jnzDO/XWlln9rP37P191snZn3n23mfvyEwkSVJZjuh2AZIk6dAZ4JIkFcgAlySpQAa4JEkFMsAlSSrQ1G4XMB5z587NhQsXdrsMSZJqs3nz5u2Z+aMHthcd4AsXLmTTpk3dLkOSpNpExL+N1u4hdEmSCmSAS5JUIANckqQCGeCSJBXIAJckqUAGuCRJBTLAJUkqkAEuSVKBDHBJkgpkgEuSVCADXJKkAtUW4BHxsYh4OCLuHdF2dER8PiK+3XydM2LeuyJiICK+FRE/V1ddkiRNBnWOwD8OvOqAtiuAjZl5MrCxOU1EPA94PXBK83c+HBFTaqxNkqSi1fY0ssz8+4hYeEDzMuDM5s+3AHcBv91sX5uZe4D7I2IAeCnwpbrqK1FfX1+3SzgsGzZs6HYJ0mEp9TMH1T53pfbPfUpDpx8n2puZ2wAyc1tEHNNsXwD884jlHmy2PU1EXApcCtDb28tdd91VX7UTzLJly8jhXd0u45BEz8xn1HukyWXZsmUM79nT7TIOWc/06ZU+d8uWLePJ4X31F9RGR/RMcZ/SNFGeBx6jtOVoC2bmDcANAEuWLMkzzzyzxrImltWrV7Nn8F4+8c6zu11KJRdedyfTF5zqX8sq1urVq9n6nQEuv+SibpdS2bU33cr8kxZV+tytXr2aXVt3cvPlazpQ2fi96drlzJw/y31KU6cDfCgi5jVH3/OAh5vtDwLHjVjuucDWDtcmSZokngmnBzod4HcAFwPXNF9vH9H+qYj4A2A+cDLwLx2uTSrCWWed1XL+xo0bO1SJNLENDg52u4RDsmDBqGeOD6q2AI+IP6VxwdrciHgQeB+N4L4tIi4BtgDnA2TmfRFxG/BNYC/w5sws68SMJGnCWbFiRbdLqGTVqlWH/Dt1XoV+wUFmjTp8yMyVwMq66pEmi5Ej7P2jcUfd0jOPd2KTJKlABrgkSQUywCVJKtBE+R64NKnV/ZWWutbv922licsAlzpkZ39/29c59YQTalv3rMWL275OSe1jgEsddP1557V1fW/7yldqWe9l69a1dX2S2s9z4JIkFcgAlySpQAa4JEkFMsAlSSqQAS5JUoG8Cl0qzP4rzw/W9sGXvKST5UjqEkfgkiQVyBG4VBhH2JLAANcktf8xm6Px0ZuSJgMPoUuSVCBH4C20GsWBI7mJbOR7s/999P2SNJk4ApckqUCOwFtwFCdJmqgcgUuSVCADXJKkAhngkiQVyACXJKlABrgkSQUywCVJKpABLklSgQxwSZIKZIBLklQgA1ySpAIZ4JIkFcgAlySpQAa4JEkFMsAlSSqQAS5JUoEMcEmSCmSAS5JUIANckqQCGeCSJBXIAJckqUAGuCRJBTLAJUkqkAEuSVKBDHBJkgpkgEuSVKCp3S5AAujr6yt2Gxs2bKhlvZLUigGuCaN/qL+W9Z5w5Am1rX9x7+K2r1OSqjDANaGc+55z277OzR/cXMu611+9vq3rk6RDMekCvO5DsR6GlSRNBJMuwAHue2B729d54nOeVdu6TzlubtvXKUma3CZlgAOcs3xlW9d3zydX1rLeO9a8u63rkyQ9M/g1MkmSCmSAS5JUoK4EeES8PSLui4h7I+JPI2JGRBwdEZ+PiG83X+d0ozZJkkrQ8QCPiAXAcmBJZp4KTAFeD1wBbMzMk4GNzWlJkjSKbh1CnwrMjIipwLOArcAy4Jbm/FuA13anNEmSJr7IzM5vNOKtwEpgF/A3mfmGiHgsM2ePWGZHZj7tMHpEXApcCtDb23va2rVrnzJ/YGCA3U/s5ajeBW2t+eMf/iAAb/ytt7V1vY8PDTJj2lQWLVo05rIDAwPk8C5O7J3d1hrqcv/QY0TPzMp92713N3Pmtf/MycfWfAyAX13+q21d745tO5gxdUbl/u3bvZvj55RxZmjLjh1MmVGtb5PZwMAAw3v2cOzc53S7lMoe2v4IPdOnV/5/+eTwPk489vgOVDZ+9z+0hSN6plTu2/DwML29vR2obPyGhobo6ekZtW9Lly7dnJlLDmzv+NfImue2lwEnAo8Bn4mIC6v+fmbeANwAsGTJkjzzzDOfMn/16tXc98D2tn/da7/Nj/9IW9d3xw23cMpxcyvdyGX16tXsGbyXT7zz7LbWUJdrP3on0xecWrlv/UP9tdyJbb+vH/H1tq5v/cfWs7h3ceX+7ezv5/rzzmtrDXW5Zt06Zi2u1rfJbPXq1Wz9zgCXX3JRt0up7MabbmX+SYsq/7/ctXUnN1++pgOVjd/v33gNM+fPqty3wcFBVqxY0YHKxu+jH/0oCxYsOKTPXDcOob8SuD8zv5+Zw8B64HRgKCLmATRfH+5CbZIkFaEbAb4FeHlEPCsiAjgL6AfuAC5uLnMxcHsXapMkqQgdP4SemV+OiHXA3cBe4Ks0Dok/G7gtIi6hEfLnd7o2SZJK0ZVbqWbm+4D3HdC8h8ZoXNIz1Flntd4FbNy4sUOVSBOfd2KTJKlAk/ZhJpLKM3KEvX807qhbGp0jcEmSCmSAS5JUIANckqQCGeCSJBXIAJckqUAGuCRJBTLAJUkqkAEuSVKBvJGLJqXNH9x80LbT3nZap8uRpLZzBC5JUoEcgWtScpQtabJzBC5JUoEMcEmSCuQhdEnj0tfXV+T6N2zYUMt6pU4xwCWN29CWx9q+ziPnTqtt3b3Hz277OqVOM8AltcV7l69u6/r+8FPvqmW9V615R1vXJ3WLAd7CPZ9c2bLt+W94dyfLkSTph7yITZKkAjkCb8ERtiRponIELklSgQxwSZIKZIBLklQgA1ySpAIZ4JIkFcgAlySpQAa4JEkFMsAlSSqQAS5JUoEMcEmSCmSAS5JUIANckqQCHfRhJhHxx0AebH5mLq+lIkmSNKZWTyPb1Hw9A3ge8Onm9PnA5jqLkvTM9IefelfLtrf/8gc6WY40oR00wDPzFoCIeCOwNDOHm9MfAf6mI9VJkqRRVXke+HzgSODR5vSzm22S1FaOsKXqqgT4NcBXI+ILzen/AVxZW0WSJGlMYwZ4Zt4cEX8FvKzZdEVmPlRvWZIkqZUxv0YWEQG8EnhBZt4OTIuIl9ZemSRJOqgq3wP/MPBTwAXN6R8AH6qtIkmSNKYq58BflpkvjoivAmTmjoiYVnNdkiSphSoj8OGImELzpi4R8aPAk7VWJUmSWqoS4GuAPweOiYiVwBeB/1VrVZIkqaUqV6F/MiI2A2cBAbw2M/trr0ySJB1UlavQbwJmZOaHMvP6zOyPiCvrL02SJB1MlUPoPwd8PCIuGtF2Tk31SJKkCqoE+MPATwPnR8SHImIqjUPpkiSpS6oEeGTmzszsA74P/B1wVL1lSZKkVqoE+B37f8jMK4EPAN+rqR5JklTBmAGeme87YPqzmfkz9ZUkSZLGctCvkUXEFzPzFRHxA5o3cdk/C8jMnFV7dZIkaVQHDfDMfEXz9ch2bzQiZgM3AqfS+OPgV4FvAZ8GFtI4RP+6zNzR7m1LkjQZHPQQekQc3erfOLf7R8BfZ+ZPAC8A+oErgI2ZeTKwsTktSZJG0epObJtpjI5H+8pYAv/tcDYYEbNofC3tjQCZ+QTwREQsA85sLnYLcBfw24ezDUmSJrvIzLGXaucGI14I3AB8k8boezPwVmAwM2ePWG5HZs4Z5fcvBS4F6O3tPW3t2rVPmT8wMMDuJ/ZyVO+CurrQVo8PDTJj2lQWLVo05rIDAwPk8C5O7J1df2FtcP/QY0TPzMp92713N3PmPe0tn7B2bNvBjKkzKvdv3+7dHD+njP5t2bGDKTOq923vE/uYd8xxHahs/LY9/ABTp02p3LfhPXs4du5zOlBZezy0/RF6pk+v3L8nh/dx4rHHd6Cy8bv/oS0c0XMI793wML29vR2obPyGhobo6ekZtW9Lly7dnJlLDmyvFOARMQc4GZixvy0z//5wioyIJcA/A2dk5pcj4o+AncBbqgT4SEuWLMlNmzY9pa2vr4/7HtjOOctXHk55HXfHmndzynFz2bBhw5jL9vX1sWfwXj7xzrM7UNn4XXjdnUxfcGrlvvUP9XPue87tQGXtsf7q9SzuXVy5fzv7+7n+vPM6UNn4XbZuHbMWV+/b0JbHeO/y1R2obPyuWvMOeo+fXblvW78zwOWXXDTmshPFtTfdyvyTFlXu366tO7n58jUdqGz83nTtcmbOn1W5b4ODg6xYsaIDlY3fqlWrWLBgwah9i4hRA7zKvdB/Dfh74HPA+5uvV46jzgeBBzPzy83pdcCLgaGImNfc5jwad4CTJEmjqHIjl7cCLwH+LTOXAi+icUe2w5KZDwEPRMSPN5vOonE4/Q7g4mbbxcDth7sNSZImuzEfJwrszszdEUFETM/Mfx0RvofrLcAnI2Ia8F3gTTT+mLgtIi4BtgDnj3MbkiRNWlUC/MHm97b/Avh8ROwAto5no5n5NeBpx/NpjMYlSdIYxgzwzPyF5o9XRsQXaDzI5K9rrUqSJLVU5Rw4ETEnIn4S+AGNi9BOrbUqSZLU0pgj8Ii4msZNV74LPNlsTsAHmkiS1CVVzoG/Djipecc0SZI0AVQ5hH4vMLvmOiRJ0iGoMgL/APDViLgX2LO/MTPPqa0qSZLUUpUAvwX4feAe/uscuCRJ6qIqAb49M8u4Ua4kSc8QVQJ8c0R8gMatTkceQr+7tqokSVJLVQL8Rc3Xl49o82tkkiR1UcsAj4gpwB2Z+YcdqkeSJFXQ8mtkmbkP8GpzSZImmCqH0P8pIq4HPg38x/5Gz4FLktQ9VQL89ObrVSPaPAcuSVIXVXka2dJOFCJJkqob81aqEXFURPxBRGxq/lsdEUd1ojhJkjS6KvdC/xiNx4i+rvlvJ3BznUVJkqTWqpwDPykzf3HE9Psj4ms11SNJkiqoMgLfFRGv2D8REWcAu+orSZIkjaXKCPw3gVub570DeBR4Y51FSZKk1qpchf514AURMas5vbP2qiRJUktjBnhETAd+EVgITI0IADLzqha/JkmSalTlEPrtwOPAZkY8jUySJHVPlQB/bma+qvZKJElSZVWuQv+niHh+7ZVIkqTKqozAXwG8MSLup3EIPYDMzJ+stTJJknRQVQL81bVXIUmSDkmVr5H9WycKkSRJ1VU5By5JkiYYA1ySpAJVCvCIOCEiXtn8eWZEHFlvWZIkqZUqzwP/dWAd8L+bTc8F/qLGmiRJ0hiqjMDfDJxB4zngZOa3gWPqLEqSJLVWJcD3ZOYT+yciYiqQ9ZUkSZLGUiXA/y4ifgeYGRE/C3wG2FBvWZIkqZUqAX4F8H3gHuA3gL8EfrfOoiRJUmtVbuTyJPDR5j9JkjQBVHke+D08/Zz348Am4Pcy85E6CpMkSQdX5V7ofwXsAz7VnH5983Un8HGgr/1lSZKkVqoE+BmZecaI6Xsi4h8z84yIuLCuwiRJ0sFVuYjt2RHxsv0TEfFS4NnNyb21VCVJklqqMgK/BLg5IvaH9g+ASyLiR4AP1FaZJEk6qJYBHhFTgP+emc+PiKOAyMzHRixyW53FSZKk0bU8hJ6Z+4BlzZ8fPyC8JUlSl1Q5hP6PEXE98GngP/Y3ZubdtVUlSZJaqhLgpzdfrxrRlsDPtL8cSZJURZU7sS3tRCGSJKm6KiNwIuJs4BRgxv62zLzq4L8hSZLqNOb3wCPiI8AvAW8BAjgfOKHmuiRJUgtVbuRyemZeBOzIzPcDPwUcV29ZkiSplSoBvqv5+p8RMR8YBk6sryRJkjSWKufAPxsRs4HrgLtpXIF+Y51FSZKk1qpchX5188c/i4jPAjMy8/Hxbrh5l7dNwGBmviYijqbxXfOFwPeA12XmjvFuR5KkyajKIXQi4vSI+GUaF7Mti4iL2rDttwL9I6avADZm5snAxua0JEkaRZWr0P8PsAp4BfCS5r8l49loRDwXOJunHopfBtzS/PkW4LXj2YYkSZNZZGbrBSL6geflWAseykYj1tF4ktmRwIrmIfTHMnP2iGV2ZOacUX73UuBSgN7e3tPWrl37lPkDAwPsfmIvR/UuaFe5tXp8aJAZ06ayaNGiMZcdGBggh3dxYu/s+gtrg/uHHiN6Zlbu2+69u5kz72lv+YS1Y9sOZkydUbl/+3bv5vg5ZfRvy44dTJlRvW97n9jHvGPK+HLKtocfYOq0KZX7NrxnD8fOfU4HKmuPh7Y/Qs/06ZX79+TwPk489vgOVDZ+9z+0hSN6DuG9Gx6mt7e3A5WN39DQED09PaP2benSpZsz82kD5yoB/hlgeWZua0eREfEa4Ocz87ci4kwOMcBHWrJkSW7atOkpbX19fdz3wHbOWb6yHeXW7o417+aU4+ayYcOGMZft6+tjz+C9fOKdZ3egsvG78Lo7mb7g1Mp96x/q59z3nNuBytpj/dXrWdy7uHL/dvb3c/1553WgsvG7bN06Zi2u3rehLY/x3uWrO1DZ+F215h30Hj+7ct+2fmeAyy9px1nDzrj2pluZf9Kiyv3btXUnN1++pgOVjd+brl3OzPmzKvdtcHCQFStWdKCy8Vu1ahULFiwYtW8RMWqAH/QitojYQOOK8yOBb0bEvwB79s/PzHMOs84zgHMi4udp3NltVkR8AhiKiHmZuS0i5gEPH+b6JUma9Fpdhb6qjg1m5ruAdwGMGIFfGBHXARcD1zRfb69j+5IkTQatAnwQ6M3MfxzZGBE/3ZzXbtcAt0XEJcAWGrdslSRJo2h1FfoHgR+M0v6fzXnjlpl3ZeZrmj8/kplnZebJzddH27ENSZImo1YBvjAzv3FgY2ZuonGzFUmS1CWtAnxGi3kz212IJEmqrlWAfyUifv3AxuY56s31lSRJksbS6iK2twF/HhFv4L8CewkwDfiFmuuSJEktHDTAM3MIOD0ilgKnNpvvzMy/7UhlkiTpoKo8jewLwBc6UIskSaqo0tPIJEnSxGKAS5JUIANckqQCGeCSJBXIAJckqUAGuCRJBTLAJUkqkAEuSVKBDHBJkgpkgEuSVCADXJKkAhngkiQVyACXJKlABrgkSQUywCVJKpABLklSgQxwSZIKZIBLklQgA1ySpAIZ4JIkFcgAlySpQAa4JEkFMsAlSSqQAS5JUoEMcEmSCmSAS5JUIANckqQCGeCSJBXIAJckqUAGuCRJBTLAJUkqkAEuSVKBDHBJkgpkgEuSVCADXJKkAhngkiQVyACXJKlABrgkSQUywCVJKpABLklSgQxwSZIKZIBLklQgA1ySpAIZ4JIkFcgAlySpQAa4JEkF6niAR8RxEfGFiOiPiPsi4q3N9qMj4vMR8e3m65xO1yZJUim6MQLfC7wjMxcDLwfeHBHPA64ANmbmycDG5rQkSRpFxwM8M7dl5t3Nn38A9AMLgGXALc3FbgFe2+naJEkqRVfPgUfEQuBFwJeB3szcBo2QB47pYmmSJE1okZnd2XDEs4G/A1Zm5vqIeCwzZ4+YvyMzn3YePCIuBS4F6O3tPW3t2rVPmT8wMMDuJ/ZyVO+CWutvl8eHBpkxbSqLFi0ac9mBgQFyeBcn9s6uv7A2uH/oMaJnZuW+7d67mznzyrn0Yce2HcyYOqNy//bt3s3xc8ro35YdO5gyo3rf9j6xj3nHHNeBysZv28MPMHXalMp9G96zh2PnPqcDlbXHQ9sfoWf69Mr9e3J4Hycee3wHKhu/+x/awhE9h/DeDQ/T29vbgcrGb2hoiJ6enlH7tnTp0s2ZueTA9q4EeET0AJ8FPpeZf9Bs+xZwZmZui4h5wF2Z+eOt1rNkyZLctGnTU9r6+vq474HtnLN8ZU3Vt9cda97NKcfNZcOGDWMu29fXx57Be/nEO8/uQGXjd+F1dzJ9wamV+9Y/1M+57zm3A5W1x/qr17O4d3Hl/u3s7+f6887rQGXjd9m6dcxaXL1vQ1se473LV3egsvG7as076D1+duW+bf3OAJdfclEHKmuPa2+6lfknLarcv11bd3Lz5Ws6UNn4vena5cycP6ty3wYHB1mxYkUHKhu/VatWsWDBglH7FhGjBng3rkIP4Cagf394N90BXNz8+WLg9k7XJklSKaZ2YZtnAL8C3BMRX2u2/Q5wDXBbRFwCbAHO70JtkiQVoeMBnplfBOIgs8/qZC2SJJXKO7FJklQgA1ySpAIZ4JIkFcgAlySpQAa4JEkFMsAlSSqQAS5JUoEMcEmSCmSAS5JUIANckqQCGeCSJBXIAJckqUAGuCRJBTLAJUkqkAEuSVKBDHBJkgpkgEuSVCADXJKkAhngkiQVyACXJKlABrgkSQUywCVJKpABLklSgQxwSZIKZIBLklQgA1ySpAIZ4JIkFcgAlySpQAa4JEkFMsAlSSqQAS5JUoEMcEmSCmSAS5JUIANckqQCGeCSJBXIAJckqUAGuCRJBTLAJUkqkAEuSVKBDHBJkgpkgEuSVCADXJKkAhngkiQVyACXJKlABrgkSQUywCVJKpABLklSgQxwSZIKZIBLklQgA1ySpAIZ4JIkFcgAlySpQAa4JEkFmnABHhGviohvRcRARFzR7XokSZqIJlSAR8QU4EPAq4HnARdExPO6W5UkSRPP1G4XcICXAgOZ+V2AiFgLLAO+eagrumPNu9tc2sRx4XV3druE2qy/en23S6jVZevWdbuE2ly15h3dLqE21950a7dLqNWbrl3e7RJqs2rVqm6XUJvIzG7X8EMRcR7wqsz8teb0rwAvy8zLRixzKXBpc/LHgW91sMS5wPYObq+TJnPfYHL3z76VaTL3DSZ3/zrdtxMy80cPbJxoI/AYpe0pf2Fk5g3ADZ0p56kiYlNmLunGtus2mfsGk7t/9q1Mk7lvMLn7N1H6NqHOgQMPAseNmH4usLVLtUiSNGFNtAD/CnByRJwYEdOA1wN3dLkmSZImnAl1CD0z90bEZcDngCnAxzLzvi6XNVJXDt13yGTuG0zu/tm3Mk3mvsHk7t+E6NuEuohNkiRVM9EOoUuSpAoMcEmSCmSAAxGxLyK+FhH3RsSGiJjdbJ8fES3vvBER34uIuR0ptHUdCyPi3gParoyIFaMsmxGxesT0ioi4ssI2Xh0RmyKiPyL+NSJWtdrORBARv9Ds7090u5Z2OpT3u0QR8e/drgE6s2+IiLsiYtOI6SURcVeF3/uxiPjL5m2n+yPitojojYgzI+KzY/duzPWP+R6UvC8Za98wUd+XkQzwhl2Z+cLMPBV4FHgzQGZuzczzultaLfYA5x7KHx4RcSpwPXBhZi4GTgW+W1N97XQB8EUa32iQDlWn9g3HRMSrqy4cETOAO4E/ycxFzc/knwBPu9lHzUrel1TZN0zo98UAf7ovAQvgqaOciJgSEasi4p6I+EZEvGXE77wlIu5uzithpLeXxlWUbz9wRkT0RcSXI+KrEfF/I6K3OetyYGVm/is0vjGQmR/uXMmHLiKeDZwBXIIBrvGrc99wHfC7BzY2t/MPzXXcHRGnN2f9MvClzNywf9nM/EJm3nvgOmpW5L7kEPYNE/p9McBHiMbDVM5i9O+eXwqcCLwoM38S+OSIedsz88U0/tIq5RDmh4A3RMRRB7R/EXh5Zr4IWEvjwwaNv5I3d7C+dngt8NeZ+f+ARyPixV2uR4XqwL7hS8CeiFh6QPvDwM821/FLwJpm+0T6PJa4L3kt1fYNE/p9McAbZkbE14BHgKOBz4+yzCuBj2TmXoDMfHTEvP1P4NgMLKyvzJYO9n3AUdszcydwK3DgUwyeC3wuIu4B3gmc0rYKO+8CGjsOmq8XdLGWdjuk91uHrZP7ht/j6aO9HuCjzc/jZ2g8pXFCKXRfcij7hgn7vhjgDbsy84XACcA0mue5DhAcfOe4p/m6j+7dHOcRYM4BbUcD25sX4XwtIn7zgPkfpHEI6UdGtP0xcH1mPh/4DWBGs/0+4LS2V12TiHgO8DPAjRHxPRo7kF+KiNHut1+ig77fXahlMmv7viEiPtf8PN44csHM/Fsan7eXj2h+OzAEvABY0qwBuvB5jIjjJsO+pMW+obj3xQAfITMfp/FX5IqI6Dlg9t8AvxkR+z+ER3e6vlYy89+BbRFxFvywvlcBX2xehPPCzPzIAb/zKHAbjQ/efkcBg82fLx7Rfh3wOxHxY831HxER/7Oe3rTFecCtmXlCZi7MzOOA+4FXdLmutmj1fne1sEmqnfuGzPy55ufx10aZvZL/OtQMjc/jtsx8EvgVGneoBPgUcHpEnL1/wYh4VUQ8/1D6dSgy84FJsi852L7h90p7XwzwA2TmV4Gv8/QLG24EtgDfiIiv07hYYaK5CPjd5iG/vwXen5nfGeN3VtN4NN5+VwKfiYh/YMRoLjO/AbwN+NOI6AfuBea1rfL2uwD48wPa/oyJ+b4drsN5v0vxrIh4cMS/rv+x2Il9Q2b+JfD9EU0fBi6OiH8Gfgz4j+Zyu4DX0LhI7tsR8U3gjTTOzbbLob4HpexLDnnfMMHelx/yVqqSJBXIEbgkSQUywCVJKpABLklSgQxwSZIKZIBLklQgA1ySpAIZ4JIkFej/A2EPW0cns23jAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 504x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "mask = (df_charge['tRNA_annotation'] == 'Homo_sapiens_mito_tRNA-Leu-TAA-1-1')\n",
    "\n",
    "fig, ax1 = plt.subplots(1, 1, figsize=(7, 5))\n",
    "\n",
    "g1 = sns.barplot(ax=ax1, x='sample_name', y='charge_norm', data=df_charge[mask], capsize=.1, errwidth=2, edgecolor='black', linewidth=2, alpha=0.7)\n",
    "g1.grid(True, axis='y')\n",
    "g1.set_ylabel('Charge normalized');\n",
    "g1.set_xlabel('');\n",
    "#g1.set_xticklabels(g1.get_xticklabels(), rotation=90)\n",
    "\n",
    "\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Escherichia_coli_str_K_12_substr_MG1655_tRNA-eColiLys-TTT-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Ala-TGC-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Arg-TCG-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Asn-GTT-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Asp-GTC-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Cys-GCA-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Gln-TTG-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Glu-TTC-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Gly-TCC-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-His-GTG-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Ile-GAT-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Leu-TAA-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Leu-TAG-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Lys-TTT-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Lys-TTT-1-1@Homo_sapiens_tRNA-Leu-CAG-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Met-CAT-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Phe-GAA-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Pro-TGG-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Ser-GCT-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Ser-TGA-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Thr-TGT-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Trp-TCA-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Tyr-GTA-1-1',\n",
       " 'Homo_sapiens_mito_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-chr21-2',\n",
       " 'Homo_sapiens_mito_tRNA-Val-TAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-1-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-1-1@Homo_sapiens_tRNA-Ala-AGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-10-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-10-1@Homo_sapiens_tRNA-Ala-AGC-12-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-10-1@Homo_sapiens_tRNA-Ala-AGC-12-1@Homo_sapiens_tRNA-Ala-AGC-13-1@Homo_sapiens_tRNA-Ala-AGC-16-1@Homo_sapiens_tRNA-Ala-AGC-17-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-11-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-11-1@Homo_sapiens_tRNA-Ala-AGC-15-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-11-1@Homo_sapiens_tRNA-Ala-AGC-8-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-11-1@Homo_sapiens_tRNA-Ala-AGC-9-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-11-1@Homo_sapiens_tRNA-Ala-AGC-chr6-93',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-12-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-12-1@Homo_sapiens_tRNA-Ala-AGC-13-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-12-1@Homo_sapiens_tRNA-Ala-AGC-13-1@Homo_sapiens_tRNA-Ala-AGC-16-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-12-1@Homo_sapiens_tRNA-Ala-AGC-13-1@Homo_sapiens_tRNA-Ala-AGC-16-1@Homo_sapiens_tRNA-Ala-AGC-17-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-13-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-14-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-15-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-15-1@Homo_sapiens_tRNA-Ala-AGC-8-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-19-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-2-1@Homo_sapiens_tRNA-Ala-AGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-2-1@Homo_sapiens_tRNA-Ala-AGC-5-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-2-1@Homo_sapiens_tRNA-Ala-AGC-7-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-3-1@Homo_sapiens_tRNA-Ala-TGC-1-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-4-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-4-1@Homo_sapiens_tRNA-Ala-CGC-1-1@Homo_sapiens_tRNA-Ala-CGC-2-1@Homo_sapiens_tRNA-Ala-TGC-2-1@Homo_sapiens_tRNA-Ala-TGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-4-1@Homo_sapiens_tRNA-Ala-CGC-1-1@Homo_sapiens_tRNA-Ala-TGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-5-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-7-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-8-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-8-1@Homo_sapiens_tRNA-Ala-AGC-9-1',\n",
       " 'Homo_sapiens_tRNA-Ala-AGC-9-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-1-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-1-1@Homo_sapiens_tRNA-Ala-CGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-1-1@Homo_sapiens_tRNA-Ala-CGC-2-1@Homo_sapiens_tRNA-Ala-TGC-2-1@Homo_sapiens_tRNA-Ala-TGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-1-1@Homo_sapiens_tRNA-Ala-CGC-3-1@Homo_sapiens_tRNA-Ala-TGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-1-1@Homo_sapiens_tRNA-Ala-TGC-1-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-1-1@Homo_sapiens_tRNA-Ala-TGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-2-1@Homo_sapiens_tRNA-Ala-TGC-1-1@Homo_sapiens_tRNA-Ala-TGC-3-1@Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-2-1@Homo_sapiens_tRNA-Ala-TGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-2-1@Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-3-1@Homo_sapiens_tRNA-Ala-TGC-4-1',\n",
       " 'Homo_sapiens_tRNA-Ala-CGC-4-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-1-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-1-1@Homo_sapiens_tRNA-Ala-TGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-1-1@Homo_sapiens_tRNA-Ala-TGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-1-1@Homo_sapiens_tRNA-Ala-TGC-3-1@Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-1-1@Homo_sapiens_tRNA-Ala-TGC-5-1@Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-1-1@Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-2-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-2-1@Homo_sapiens_tRNA-Ala-TGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-3-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-3-1@Homo_sapiens_tRNA-Ala-TGC-4-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-3-1@Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-4-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-5-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-5-1@Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-6-1',\n",
       " 'Homo_sapiens_tRNA-Ala-TGC-7-1',\n",
       " 'Homo_sapiens_tRNA-Arg-ACG-1-1',\n",
       " 'Homo_sapiens_tRNA-Arg-ACG-1-1@Homo_sapiens_tRNA-Arg-ACG-2-1',\n",
       " 'Homo_sapiens_tRNA-Arg-ACG-2-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCG-1-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCG-1-1@Homo_sapiens_tRNA-Arg-TCG-3-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCG-1-1@Homo_sapiens_tRNA-Arg-TCG-3-1@Homo_sapiens_tRNA-Arg-TCG-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCG-2-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCT-1-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCT-1-1@Homo_sapiens_tRNA-Arg-CCT-2-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCT-2-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCT-2-1@Homo_sapiens_tRNA-Arg-CCT-4-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCT-3-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCT-4-1',\n",
       " 'Homo_sapiens_tRNA-Arg-CCT-4-1@Homo_sapiens_tRNA-Arg-CCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-1-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-2-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-2-1@Homo_sapiens_tRNA-Arg-TCG-4-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-2-1@Homo_sapiens_tRNA-Arg-TCG-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-3-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-3-1@Homo_sapiens_tRNA-Arg-TCG-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-4-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-4-1@Homo_sapiens_tRNA-Arg-TCG-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCG-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-1-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-1-1@Homo_sapiens_tRNA-Arg-TCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-2-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-2-1@Homo_sapiens_tRNA-Arg-TCT-3-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-2-1@Homo_sapiens_tRNA-Arg-TCT-3-1@Homo_sapiens_tRNA-Arg-TCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-3-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-3-1@Homo_sapiens_tRNA-Arg-TCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Arg-TCT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-1-1@Homo_sapiens_tRNA-Asn-GTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-11-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-11-1@Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-11-1@Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-11-1@Homo_sapiens_tRNA-Asn-GTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-13-1@Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-15-1@Homo_sapiens_tRNA-Asn-GTT-19-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1@Homo_sapiens_tRNA-Asn-GTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1@Homo_sapiens_tRNA-Asn-GTT-7-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-7-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-2-1@Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-4-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-3-1@Homo_sapiens_tRNA-Asn-GTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-4-1@Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-4-1@Homo_sapiens_tRNA-Asn-GTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-5-1@Homo_sapiens_tRNA-Asn-GTT-9-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Asn-GTT-8-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-1-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-1-1@Homo_sapiens_tRNA-Asp-GTC-2-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-1-1@Homo_sapiens_tRNA-Asp-GTC-2-1@Homo_sapiens_tRNA-Asp-GTC-3-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-2-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-2-1@Homo_sapiens_tRNA-Asp-GTC-3-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-2-1@Homo_sapiens_tRNA-Asp-GTC-3-1@Homo_sapiens_tRNA-Asp-GTC-4-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-2-1@Homo_sapiens_tRNA-Asp-GTC-4-1',\n",
       " 'Homo_sapiens_tRNA-Asp-GTC-3-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-12-1@Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-19-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-3-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-19-1@Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-6-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-1-1@Homo_sapiens_tRNA-Cys-GCA-6-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-10-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-11-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-11-1@Homo_sapiens_tRNA-Cys-GCA-14-1@Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-11-1@Homo_sapiens_tRNA-Cys-GCA-2-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-11-1@Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-5-1@Homo_sapiens_tRNA-Cys-GCA-7-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-11-1@Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-11-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-12-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-13-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-13-1@Homo_sapiens_tRNA-Cys-GCA-20-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-13-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-14-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-14-1@Homo_sapiens_tRNA-Cys-GCA-2-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-14-1@Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-5-1@Homo_sapiens_tRNA-Cys-GCA-7-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-14-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-15-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-15-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-16-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-17-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-17-1@Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-5-1@Homo_sapiens_tRNA-Cys-GCA-7-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-17-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-18-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-19-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-3-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1@Homo_sapiens_tRNA-Ser-AGA-5-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-18-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-19-1@Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-8-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-19-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-2-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-20-1@Homo_sapiens_tRNA-Cys-GCA-5-1@Homo_sapiens_tRNA-Cys-GCA-7-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-4-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-5-1@Homo_sapiens_tRNA-Cys-GCA-7-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-7-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-2-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-21-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-3-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-3-1@Homo_sapiens_tRNA-Cys-GCA-4-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-3-1@Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1@Homo_sapiens_tRNA-Ser-AGA-5-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-4-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-4-1@Homo_sapiens_tRNA-Cys-GCA-8-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-5-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-5-1@Homo_sapiens_tRNA-Cys-GCA-7-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-5-1@Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-6-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-7-1@Homo_sapiens_tRNA-Cys-GCA-8-1@Homo_sapiens_tRNA-Cys-GCA-9-1@Homo_sapiens_tRNA-Ser-AGA-5-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-6-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-7-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-8-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-8-1@Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Cys-GCA-9-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-CTG-2-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-CTG-2-1@Homo_sapiens_tRNA-Gln-CTG-3-1@Homo_sapiens_tRNA-Gln-CTG-4-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-CTG-2-1@Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-CTG-5-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-CTG-6-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-CTG-6-1@Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-CTG-9-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-2-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-1-1@Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-2-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-2-1@Homo_sapiens_tRNA-Gln-CTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-2-1@Homo_sapiens_tRNA-Gln-CTG-3-1@Homo_sapiens_tRNA-Gln-CTG-4-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-2-1@Homo_sapiens_tRNA-Gln-CTG-4-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-3-1@Homo_sapiens_tRNA-Gln-CTG-4-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-4-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-5-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-6-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-6-1@Homo_sapiens_tRNA-Gln-TTG-1-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-6-1@Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-6-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-9-1',\n",
       " 'Homo_sapiens_tRNA-Gln-CTG-9-1@Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-1-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-2-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-2-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-1-1@Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-2-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-2-1@Homo_sapiens_tRNA-Gln-TTG-4-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-3-1',\n",
       " 'Homo_sapiens_tRNA-Gln-TTG-4-1',\n",
       " 'Homo_sapiens_tRNA-Glu-CTC-1-1',\n",
       " 'Homo_sapiens_tRNA-Glu-CTC-1-1@Homo_sapiens_tRNA-Glu-CTC-2-1',\n",
       " 'Homo_sapiens_tRNA-Glu-CTC-1-1@Homo_sapiens_tRNA-Glu-CTC-2-1@Homo_sapiens_tRNA-Glu-TTC-4-1',\n",
       " 'Homo_sapiens_tRNA-Glu-CTC-1-1@Homo_sapiens_tRNA-Glu-TTC-4-1',\n",
       " 'Homo_sapiens_tRNA-Glu-CTC-2-1',\n",
       " 'Homo_sapiens_tRNA-Glu-TTC-1-1',\n",
       " 'Homo_sapiens_tRNA-Glu-TTC-1-1@Homo_sapiens_tRNA-Glu-TTC-2-1',\n",
       " 'Homo_sapiens_tRNA-Glu-TTC-2-1',\n",
       " 'Homo_sapiens_tRNA-Glu-TTC-3-1',\n",
       " 'Homo_sapiens_tRNA-Glu-TTC-3-1@Homo_sapiens_tRNA-Glu-TTC-4-1',\n",
       " 'Homo_sapiens_tRNA-Glu-TTC-4-1',\n",
       " 'Homo_sapiens_tRNA-Gly-CCC-1-1',\n",
       " 'Homo_sapiens_tRNA-Gly-CCC-1-1@Homo_sapiens_tRNA-Gly-CCC-5-1',\n",
       " 'Homo_sapiens_tRNA-Gly-CCC-1-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Gly-CCC-2-1',\n",
       " 'Homo_sapiens_tRNA-Gly-CCC-4-1',\n",
       " 'Homo_sapiens_tRNA-Gly-CCC-5-1',\n",
       " 'Homo_sapiens_tRNA-Gly-GCC-1-1',\n",
       " 'Homo_sapiens_tRNA-Gly-GCC-1-1@Homo_sapiens_tRNA-Gly-GCC-2-1',\n",
       " 'Homo_sapiens_tRNA-Gly-GCC-2-1',\n",
       " 'Homo_sapiens_tRNA-Gly-GCC-2-1@Homo_sapiens_tRNA-Gly-GCC-3-1',\n",
       " 'Homo_sapiens_tRNA-Gly-TCC-1-1',\n",
       " 'Homo_sapiens_tRNA-Gly-TCC-1-1@Homo_sapiens_tRNA-Gly-TCC-2-1@Homo_sapiens_tRNA-Gly-TCC-3-1',\n",
       " 'Homo_sapiens_tRNA-Gly-TCC-2-1',\n",
       " 'Homo_sapiens_tRNA-Gly-TCC-2-1@Homo_sapiens_tRNA-Gly-TCC-3-1',\n",
       " 'Homo_sapiens_tRNA-Gly-TCC-3-1',\n",
       " 'Homo_sapiens_tRNA-His-GTG-1-1',\n",
       " 'Homo_sapiens_tRNA-His-GTG-1-1@Homo_sapiens_tRNA-His-GTG-2-1',\n",
       " 'Homo_sapiens_tRNA-His-GTG-2-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1@Homo_sapiens_tRNA-Ile-AAT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1@Homo_sapiens_tRNA-Ile-AAT-4-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1@Homo_sapiens_tRNA-Ile-AAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-6-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-1-1@Homo_sapiens_tRNA-Ile-AAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-3-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-6-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-6-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-2-1@Homo_sapiens_tRNA-Ile-AAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-3-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-3-1@Homo_sapiens_tRNA-Ile-AAT-6-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-4-1@Homo_sapiens_tRNA-Ile-AAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-4-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-4-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-4-1@Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-4-1@Homo_sapiens_tRNA-Ile-AAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-4-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-6-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-6-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-6-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-7-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-5-1@Homo_sapiens_tRNA-Leu-CAG-1-1@Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-6-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Ile-AAT-8-1',\n",
       " 'Homo_sapiens_tRNA-Ile-GAT-1-1',\n",
       " 'Homo_sapiens_tRNA-Ile-TAT-1-1',\n",
       " 'Homo_sapiens_tRNA-Ile-TAT-1-1@Homo_sapiens_tRNA-Ile-TAT-2-1',\n",
       " 'Homo_sapiens_tRNA-Ile-TAT-1-1@Homo_sapiens_tRNA-Ile-TAT-2-1@Homo_sapiens_tRNA-Ile-TAT-3-1',\n",
       " 'Homo_sapiens_tRNA-Ile-TAT-1-1@Homo_sapiens_tRNA-Ile-TAT-3-1',\n",
       " 'Homo_sapiens_tRNA-Ile-TAT-2-1',\n",
       " 'Homo_sapiens_tRNA-Ile-TAT-2-1@Homo_sapiens_tRNA-Ile-TAT-3-1',\n",
       " 'Homo_sapiens_tRNA-Ile-TAT-3-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-1-1@Homo_sapiens_tRNA-Leu-AAG-2-1@Homo_sapiens_tRNA-Leu-TAG-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-1-1@Homo_sapiens_tRNA-Leu-AAG-2-1@Homo_sapiens_tRNA-Leu-TAG-1-1@Homo_sapiens_tRNA-Leu-TAG-3-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-1-1@Homo_sapiens_tRNA-Leu-AAG-3-1@Homo_sapiens_tRNA-Leu-TAG-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-1-1@Homo_sapiens_tRNA-Leu-TAG-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-1-1@Homo_sapiens_tRNA-Leu-TAG-1-1@Homo_sapiens_tRNA-Leu-TAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-2-1@Homo_sapiens_tRNA-Leu-AAG-3-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-2-1@Homo_sapiens_tRNA-Leu-AAG-4-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-2-1@Homo_sapiens_tRNA-Leu-TAG-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-2-1@Homo_sapiens_tRNA-Leu-TAG-1-1@Homo_sapiens_tRNA-Leu-TAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-2-1@Homo_sapiens_tRNA-Leu-TAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-AAG-3-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-1-1@Homo_sapiens_tRNA-Leu-CAA-2-1@Homo_sapiens_tRNA-Leu-CAA-3-1@Homo_sapiens_tRNA-Leu-CAA-4-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-1-1@Homo_sapiens_tRNA-Leu-CAA-2-1@Homo_sapiens_tRNA-Leu-CAA-3-1@Homo_sapiens_tRNA-Leu-CAA-4-1@Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-1-1@Homo_sapiens_tRNA-Leu-CAA-3-1@Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-1-1@Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-2-1@Homo_sapiens_tRNA-Leu-CAA-3-1@Homo_sapiens_tRNA-Leu-CAA-4-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-2-1@Homo_sapiens_tRNA-Leu-CAA-3-1@Homo_sapiens_tRNA-Leu-CAA-4-1@Homo_sapiens_tRNA-Leu-CAG-1-1@Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-3-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-4-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAA-6-1@Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAG-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAG-1-1@Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAG-1-1@Homo_sapiens_tRNA-Leu-CAG-chr9-7',\n",
       " 'Homo_sapiens_tRNA-Leu-CAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-CAG-chr9-7',\n",
       " 'Homo_sapiens_tRNA-Leu-TAA-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-TAA-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-TAA-3-1',\n",
       " 'Homo_sapiens_tRNA-Leu-TAA-4-1',\n",
       " 'Homo_sapiens_tRNA-Leu-TAG-1-1',\n",
       " 'Homo_sapiens_tRNA-Leu-TAG-1-1@Homo_sapiens_tRNA-Leu-TAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-TAG-2-1',\n",
       " 'Homo_sapiens_tRNA-Leu-TAG-3-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-1-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-1-1@Homo_sapiens_tRNA-Lys-CTT-2-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-1-1@Homo_sapiens_tRNA-Lys-CTT-2-1@Homo_sapiens_tRNA-Lys-CTT-3-1@Homo_sapiens_tRNA-Lys-CTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-1-1@Homo_sapiens_tRNA-Lys-CTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-2-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-2-1@Homo_sapiens_tRNA-Lys-CTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-2-1@Homo_sapiens_tRNA-Lys-CTT-3-1@Homo_sapiens_tRNA-Lys-CTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-2-1@Homo_sapiens_tRNA-Lys-CTT-3-1@Homo_sapiens_tRNA-Lys-CTT-4-1@Homo_sapiens_tRNA-Lys-CTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-2-1@Homo_sapiens_tRNA-Lys-CTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-2-1@Homo_sapiens_tRNA-Lys-CTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Lys-CTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-1-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-1-1@Homo_sapiens_tRNA-Lys-TTT-2-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-1-1@Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-1-1@Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-4-1@Homo_sapiens_tRNA-Lys-TTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-1-1@Homo_sapiens_tRNA-Lys-TTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-1-1@Homo_sapiens_tRNA-Lys-TTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-11-1@Homo_sapiens_tRNA-Lys-TTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-3-1@Homo_sapiens_tRNA-Lys-TTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-4-1@Homo_sapiens_tRNA-Lys-TTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-4-1@Homo_sapiens_tRNA-Lys-TTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-4-1@Homo_sapiens_tRNA-Lys-TTT-7-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-4-1@Homo_sapiens_tRNA-Lys-TTT-9-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-2-1@Homo_sapiens_tRNA-Lys-TTT-6-1@Homo_sapiens_tRNA-Sup-TTA-1-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-3-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-3-1@Homo_sapiens_tRNA-Lys-TTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-3-1@Homo_sapiens_tRNA-Lys-TTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-3-1@Homo_sapiens_tRNA-Lys-TTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-4-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-5-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-6-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-6-1@Homo_sapiens_tRNA-Lys-TTT-9-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-6-1@Homo_sapiens_tRNA-Sup-TTA-1-1',\n",
       " 'Homo_sapiens_tRNA-Lys-TTT-9-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-1-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-1-1@Homo_sapiens_tRNA-Met-CAT-6-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-2-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-2-1@Homo_sapiens_tRNA-Met-CAT-3-1@Homo_sapiens_tRNA-Met-CAT-4-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-2-1@Homo_sapiens_tRNA-Met-CAT-4-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-2-1@Homo_sapiens_tRNA-Met-CAT-4-1@Homo_sapiens_tRNA-Met-CAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-2-1@Homo_sapiens_tRNA-Met-CAT-4-1@Homo_sapiens_tRNA-Met-CAT-7-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-3-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-4-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-4-1@Homo_sapiens_tRNA-Met-CAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-5-1',\n",
       " 'Homo_sapiens_tRNA-Met-CAT-6-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-1-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-1-1@Homo_sapiens_tRNA-Phe-GAA-2-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-1-1@Homo_sapiens_tRNA-Phe-GAA-2-1@Homo_sapiens_tRNA-Phe-GAA-3-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-1-1@Homo_sapiens_tRNA-Phe-GAA-2-1@Homo_sapiens_tRNA-Phe-GAA-3-1@Homo_sapiens_tRNA-Phe-GAA-4-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-2-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-2-1@Homo_sapiens_tRNA-Phe-GAA-3-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-2-1@Homo_sapiens_tRNA-Phe-GAA-3-1@Homo_sapiens_tRNA-Phe-GAA-4-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-3-1',\n",
       " 'Homo_sapiens_tRNA-Phe-GAA-4-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-AGG-2-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-CGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-2-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-CGG-1-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-CGG-2-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-TGG-1-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-TGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-2-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-CGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-1-1@Homo_sapiens_tRNA-Pro-TGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-AGG-2-1@Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-CGG-1-1',\n",
       " 'Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-CGG-2-1',\n",
       " 'Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-CGG-2-1@Homo_sapiens_tRNA-Pro-TGG-1-1@Homo_sapiens_tRNA-Pro-TGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-1-1',\n",
       " 'Homo_sapiens_tRNA-Pro-CGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-CGG-2-1',\n",
       " 'Homo_sapiens_tRNA-Pro-CGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-TGG-1-1',\n",
       " 'Homo_sapiens_tRNA-Pro-TGG-1-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-TGG-2-1',\n",
       " 'Homo_sapiens_tRNA-Pro-TGG-2-1@Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Pro-TGG-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-AGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-AGA-3-1@Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-AGA-4-1@Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-3-1@Homo_sapiens_tRNA-Ser-TGA-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-AGA-3-1@Homo_sapiens_tRNA-Ser-TGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-TGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-1-1@Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-AGA-3-1@Homo_sapiens_tRNA-Ser-AGA-4-1@Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-AGA-3-1@Homo_sapiens_tRNA-Ser-AGA-4-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-AGA-3-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-AGA-4-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-2-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-3-1@Homo_sapiens_tRNA-Ser-TGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-AGA-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-CGA-1-1',\n",
       " 'Homo_sapiens_tRNA-Ser-CGA-1-1@Homo_sapiens_tRNA-Ser-CGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-CGA-1-1@Homo_sapiens_tRNA-Ser-CGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-CGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-CGA-2-1@Homo_sapiens_tRNA-Ser-CGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-CGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-CGA-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-1-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-1-1@Homo_sapiens_tRNA-Ser-GCT-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-1-1@Homo_sapiens_tRNA-Ser-GCT-3-1@Homo_sapiens_tRNA-Ser-GCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-1-1@Homo_sapiens_tRNA-Ser-GCT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-1-1@Homo_sapiens_tRNA-Ser-GCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-2-1@Homo_sapiens_tRNA-Ser-GCT-3-1@Homo_sapiens_tRNA-Ser-GCT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-3-1@Homo_sapiens_tRNA-Ser-GCT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-3-1@Homo_sapiens_tRNA-Ser-GCT-4-1@Homo_sapiens_tRNA-Ser-GCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-3-1@Homo_sapiens_tRNA-Ser-GCT-4-1@Homo_sapiens_tRNA-Ser-GCT-6-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-3-1@Homo_sapiens_tRNA-Ser-GCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-4-1@Homo_sapiens_tRNA-Ser-GCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ser-GCT-5-1',\n",
       " 'Homo_sapiens_tRNA-Ser-TGA-1-1',\n",
       " 'Homo_sapiens_tRNA-Ser-TGA-2-1',\n",
       " 'Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-TGA-2-1@Homo_sapiens_tRNA-Ser-TGA-4-1',\n",
       " 'Homo_sapiens_tRNA-Ser-TGA-3-1',\n",
       " 'Homo_sapiens_tRNA-Ser-TGA-4-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-5-1@Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-5-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-AGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-4-1@Homo_sapiens_tRNA-Thr-AGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-5-1@Homo_sapiens_tRNA-Thr-AGT-6-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-5-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-AGT-5-1@Homo_sapiens_tRNA-Thr-TGT-1-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-1-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-AGT-5-1@Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-1-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-CGT-1-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-4-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-4-1@Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-4-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-6-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-1-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-2-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-CGT-1-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-3-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-4-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-4-1@Homo_sapiens_tRNA-Thr-AGT-6-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-5-1@Homo_sapiens_tRNA-Thr-AGT-6-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-6-1',\n",
       " 'Homo_sapiens_tRNA-Thr-AGT-6-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-CGT-1-1',\n",
       " 'Homo_sapiens_tRNA-Thr-CGT-1-1@Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-CGT-2-1',\n",
       " 'Homo_sapiens_tRNA-Thr-CGT-2-1@Homo_sapiens_tRNA-Thr-CGT-4-1',\n",
       " 'Homo_sapiens_tRNA-Thr-CGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-CGT-4-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-1-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-2-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-2-1@Homo_sapiens_tRNA-Thr-TGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-2-1@Homo_sapiens_tRNA-Thr-TGT-3-1@Homo_sapiens_tRNA-Thr-TGT-4-1@Homo_sapiens_tRNA-Thr-TGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-2-1@Homo_sapiens_tRNA-Thr-TGT-4-1@Homo_sapiens_tRNA-Thr-TGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-3-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-3-1@Homo_sapiens_tRNA-Thr-TGT-4-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-3-1@Homo_sapiens_tRNA-Thr-TGT-4-1@Homo_sapiens_tRNA-Thr-TGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-4-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-4-1@Homo_sapiens_tRNA-Thr-TGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-4-1@Homo_sapiens_tRNA-Thr-TGT-5-1@Homo_sapiens_tRNA-Thr-TGT-6-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-5-1',\n",
       " 'Homo_sapiens_tRNA-Thr-TGT-6-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-1-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-1-1@Homo_sapiens_tRNA-Trp-CCA-2-1@Homo_sapiens_tRNA-Trp-CCA-3-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-1-1@Homo_sapiens_tRNA-Trp-CCA-2-1@Homo_sapiens_tRNA-Trp-CCA-3-1@Homo_sapiens_tRNA-Trp-CCA-5-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-1-1@Homo_sapiens_tRNA-Trp-CCA-3-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-1-1@Homo_sapiens_tRNA-Trp-CCA-3-1@Homo_sapiens_tRNA-Trp-CCA-4-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-1-1@Homo_sapiens_tRNA-Trp-CCA-3-1@Homo_sapiens_tRNA-Trp-CCA-4-1@Homo_sapiens_tRNA-Trp-CCA-5-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-1-1@Homo_sapiens_tRNA-Trp-CCA-3-1@Homo_sapiens_tRNA-Trp-CCA-5-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-2-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-2-1@Homo_sapiens_tRNA-Trp-CCA-3-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-3-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-3-1@Homo_sapiens_tRNA-Trp-CCA-4-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-3-1@Homo_sapiens_tRNA-Trp-CCA-5-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-4-1',\n",
       " 'Homo_sapiens_tRNA-Trp-CCA-5-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-2-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-5-1@Homo_sapiens_tRNA-Tyr-GTA-7-1@Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-5-1@Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-3-1@Homo_sapiens_tRNA-Tyr-GTA-6-1@Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-4-1@Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-7-1@Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-1-1@Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-2-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-3-1@Homo_sapiens_tRNA-Tyr-GTA-5-1@Homo_sapiens_tRNA-Tyr-GTA-6-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-3-1@Homo_sapiens_tRNA-Tyr-GTA-5-1@Homo_sapiens_tRNA-Tyr-GTA-6-1@Homo_sapiens_tRNA-Tyr-GTA-7-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-3-1@Homo_sapiens_tRNA-Tyr-GTA-5-1@Homo_sapiens_tRNA-Tyr-GTA-6-1@Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-4-1@Homo_sapiens_tRNA-Tyr-GTA-5-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-5-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-2-1@Homo_sapiens_tRNA-Tyr-GTA-5-1@Homo_sapiens_tRNA-Tyr-GTA-7-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-3-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-3-1@Homo_sapiens_tRNA-Tyr-GTA-6-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-4-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-4-1@Homo_sapiens_tRNA-Tyr-GTA-5-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-5-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-6-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-7-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-8-1',\n",
       " 'Homo_sapiens_tRNA-Tyr-GTA-chr14-8@Homo_sapiens_tRNA-Tyr-GTA-chr21-2',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-AAC-4-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-5-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-AAC-5-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-AAC-5-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-AAC-5-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-5-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-AAC-chr6-84@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-AAC-chr6-84@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-5-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-5-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-5-1@Homo_sapiens_tRNA-Val-CAC-6-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-6-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-4-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-5-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-AAC-5-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-4-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-1-1@Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-6-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-AAC-4-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-2-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-4-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-4-1@Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-5-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-5-1@Homo_sapiens_tRNA-Val-CAC-4-1',\n",
       " 'Homo_sapiens_tRNA-Val-AAC-chr6-84',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-4-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-1-1@Homo_sapiens_tRNA-Val-CAC-6-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-4-1',\n",
       " 'Homo_sapiens_tRNA-Val-CAC-5-1',\n",
       " 'Homo_sapiens_tRNA-Val-TAC-1-1',\n",
       " 'Homo_sapiens_tRNA-Val-TAC-1-1@Homo_sapiens_tRNA-Val-TAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-TAC-1-1@Homo_sapiens_tRNA-Val-TAC-2-1@Homo_sapiens_tRNA-Val-TAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-TAC-1-1@Homo_sapiens_tRNA-Val-TAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-TAC-2-1',\n",
       " 'Homo_sapiens_tRNA-Val-TAC-3-1',\n",
       " 'Homo_sapiens_tRNA-Val-TAC-4-1',\n",
       " 'Homo_sapiens_tRNA-iMet-CAT-1-1',\n",
       " 'Homo_sapiens_tRNA-iMet-CAT-1-1@Homo_sapiens_tRNA-iMet-CAT-2-1',\n",
       " 'Homo_sapiens_tRNA-iMet-CAT-2-1'}"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_charge['tRNA_annotation'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['l3', 'l3', 'l3', ..., 'l3', 'l3', 'l3'], dtype=object)"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stat_df['barcode'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False, ..., False, False, False])"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stat_df['barcode'].values == 'l1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "stat_df = pd.read_csv('tmp_stat.csv', keep_default_na=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80944"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(stat_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "l3\n",
      "86.7669364603913\n"
     ]
    }
   ],
   "source": [
    "#for barcode in ['l1', 'l2', 'l3', 'l4']:\n",
    "for barcode in ['l3']:\n",
    "\n",
    "    mask = (stat_df['3p_cover']) & (stat_df['3p_non-temp'] == '') & (stat_df['barcode'] == barcode)\n",
    "    Aend = stat_df['align_3p_nt'] == 'A'\n",
    "    Cend = stat_df['align_3p_nt'] == 'C'\n",
    "\n",
    "    print(barcode)\n",
    "    print(sum(mask & Aend) / sum(mask &(Aend | Cend)) *100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9692503459181656"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(stat_df[mask]['3p_non-temp'].values == '') / len(stat_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "agg_cols = ['sample_name', 'replicate', 'barcode', 'tRNA_annotation', 'tRNA_annotation_len', 'unique_annotation', 'align_3p_nt', 'codon', 'anticodon', 'amino_acid']\n",
    "row_mask = (stat_df['3p_cover']) & (stat_df['3p_non-temp'] == '')\n",
    "\n",
    "tmp = stat_df.loc[row_mask, agg_cols]\n",
    "tmp['count'] = stat_df.loc[row_mask, ['align_3p_nt']]  # dummy for groupby count\n",
    "gdf = tmp.groupby(agg_cols, as_index=False).agg({\"count\": \"count\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Ala-TGC-1-1</td>\n",
       "      <td>72</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TGC</td>\n",
       "      <td>Ala</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Ala-TGC-1-1</td>\n",
       "      <td>72</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TGC</td>\n",
       "      <td>Ala</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Arg-TCG-1-1</td>\n",
       "      <td>68</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>CGA</td>\n",
       "      <td>TCG</td>\n",
       "      <td>Arg</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_tRNA-iMet-CAT-1-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>AUG</td>\n",
       "      <td>CAT</td>\n",
       "      <td>iMet</td>\n",
       "      <td>2037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_tRNA-iMet-CAT-1-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>AUG</td>\n",
       "      <td>CAT</td>\n",
       "      <td>iMet</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_tRNA-iMet-CAT-1-1@Homo_sapiens_tR...</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AUG</td>\n",
       "      <td>CAT</td>\n",
       "      <td>iMet</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_tRNA-iMet-CAT-1-1@Homo_sapiens_tR...</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>AUG</td>\n",
       "      <td>CAT</td>\n",
       "      <td>iMet</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>U</td>\n",
       "      <td>3</td>\n",
       "      <td>l3</td>\n",
       "      <td>Homo_sapiens_tRNA-iMet-CAT-2-1</td>\n",
       "      <td>75</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>AUG</td>\n",
       "      <td>CAT</td>\n",
       "      <td>iMet</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>710 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name  replicate barcode  \\\n",
       "0             U          3      l3   \n",
       "1             U          3      l3   \n",
       "2             U          3      l3   \n",
       "3             U          3      l3   \n",
       "4             U          3      l3   \n",
       "..          ...        ...     ...   \n",
       "705           U          3      l3   \n",
       "706           U          3      l3   \n",
       "707           U          3      l3   \n",
       "708           U          3      l3   \n",
       "709           U          3      l3   \n",
       "\n",
       "                                       tRNA_annotation  tRNA_annotation_len  \\\n",
       "0    Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1    Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2                   Homo_sapiens_mito_tRNA-Ala-TGC-1-1                   72   \n",
       "3                   Homo_sapiens_mito_tRNA-Ala-TGC-1-1                   72   \n",
       "4                   Homo_sapiens_mito_tRNA-Arg-TCG-1-1                   68   \n",
       "..                                                 ...                  ...   \n",
       "705                     Homo_sapiens_tRNA-iMet-CAT-1-1                   75   \n",
       "706                     Homo_sapiens_tRNA-iMet-CAT-1-1                   75   \n",
       "707  Homo_sapiens_tRNA-iMet-CAT-1-1@Homo_sapiens_tR...                   75   \n",
       "708  Homo_sapiens_tRNA-iMet-CAT-1-1@Homo_sapiens_tR...                   75   \n",
       "709                     Homo_sapiens_tRNA-iMet-CAT-2-1                   75   \n",
       "\n",
       "     unique_annotation align_3p_nt codon anticodon amino_acid  count  \n",
       "0                 True           A   AAA       TTT   eColiLys    178  \n",
       "1                 True           C   AAA       TTT   eColiLys    149  \n",
       "2                 True           A   GCA       TGC        Ala    108  \n",
       "3                 True           C   GCA       TGC        Ala     15  \n",
       "4                 True           A   CGA       TCG        Arg    174  \n",
       "..                 ...         ...   ...       ...        ...    ...  \n",
       "705               True           A   AUG       CAT       iMet   2037  \n",
       "706               True           C   AUG       CAT       iMet    203  \n",
       "707              False           A   AUG       CAT       iMet     10  \n",
       "708              False           C   AUG       CAT       iMet      2  \n",
       "709               True           A   AUG       CAT       iMet      4  \n",
       "\n",
       "[710 rows x 11 columns]"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../SWalign/U-1+2_UMI-trimmed_SWalign.json.gz'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jf1 = json_files[0]\n",
    "jf1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'U-1+2_UMI-trimmed'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'_'.join(jf1.split('/')[-1].split('_')[0:-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'U-1+2'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'G'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(UMIread.seq)[0:1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "86"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(str(UMIread.seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTCAGGATGGCCGAGCGGTCTAAGGCACTGCGTTCAGCTCGCAGTCTCCCGTGGAGGCGTGGGTTCGGATCCCACTCCTGACACCA'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(UMIread.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "k = list(SWres.keys())[400]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'score': 58,\n",
       " 'name': 'Homo_sapiens_tRNA-Trp-CCA-2-1',\n",
       " 'qpos': [[2, 75]],\n",
       " 'dpos': [[1, 75]],\n",
       " 'qseq': 'GACCTCGTCGCGCAATGGTAGCGCATCTGACTCCA-ATCAGAAGGTTGCGTGTTCATGTCACGTCGGGGTCACCA',\n",
       " 'aseq': '|||||||| ||||||||||||||| |||||||||| |||||||||||||||||||| ||||||||||||||||||',\n",
       " 'dseq': 'GACCTCGTGGCGCAATGGTAGCGCGTCTGACTCCAGATCAGAAGGTTGCGTGTTCAAGTCACGTCGGGGTCACCA',\n",
       " '5p_non-temp': 'C',\n",
       " '3p_non-temp': ''}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SWres[k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "75"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tRNA_id2len[SWres[k]['name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Homo_sapiens_tRNA-Trp-CCA-2-1'"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SWres[k]['name'].split('@')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SWres[k]['qseq'][-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
