# tRNA databse
This folder contains the tRNA transcript sequences that are used for alignment of the reads generated by tRNA sequencing.
The sequences have been formatted to work with subsequent processing so if new sequences, e.g. a new species, needs to be added they must be formatted correctly.


### Getting, merging and formatting tRNA sequences
tRNA transcripts have been annotated and deposited in the [gtRNAdb database](http://gtrnadb.ucsc.edu/).
However, a transcripts few appear to be misannotated and the mitochondrial tRNAs are missing.
The [mim-tRNAseq repo](https://github.com/nedialkova-lab/mim-tRNAseq) (Behrens et al. 2021) has compiled them and can be downloaded from my fork.
From these sequences, cyto/mito tRNA sequences are merged and formatted, introns are removed, post-transcriptional modifications added and Ecoli Lys and Thr tRNAs sequences are added because these sequences are used as a spike-in controls.
Note that the control sequences are hard-coded in the `tRNAdb_tool.py` script...



For human:
```
mkdir human
mkdir hg38
cd hg38
wget https://raw.githubusercontent.com/krdav/mim-tRNAseq/master/mimseq/data/hg38-eColitK/hg38-mitotRNAs.fa
wget https://raw.githubusercontent.com/krdav/mim-tRNAseq/master/mimseq/data/hg38-eColitK/hg38-tRNAs-all.fa
wget https://raw.githubusercontent.com/krdav/mim-tRNAseq/master/mimseq/data/hg38-eColitK/hg38-tRNAs-detailed.out
wget https://raw.githubusercontent.com/krdav/mim-tRNAseq/master/mimseq/data/hg38-eColitK/hg38-tRNAs_name_map.txt
cd ..
./tRNAdb_tool.py hg38/hg38-mitotRNAs.fa hg38/hg38-tRNAs-all.fa hg38/hg38-tRNAs-detailed.out hg38/hg38-tRNAs_name_map.txt > human/hg38-tRNAs.fa
```

For mouse:
```
mkdir mouse
mkdir mm10
cd mm10
wget https://raw.githubusercontent.com/krdav/mim-tRNAseq/master/mimseq/data/mm10-eColitK/mm10-mitotRNAs.fa
wget http://gtrnadb.ucsc.edu/genomes/eukaryota/Mmusc10/mm10-mature-tRNAs.fa
cd ..
./tRNAdb_tool.py mm10/mm10-mitotRNAs.fa mm10/mm10-mature-tRNAs.fa > mouse/mm10-tRNAs.fa
```

For Ecoli (gtrnadb puts CCA at the end of the mature Ecoli tRNAs but not the mature human tRNAs...):
```
mkdir ecoli
mkdir Esch_coli_K_12_MG1655
cd Esch_coli_K_12_MG1655
wget http://gtrnadb.ucsc.edu/genomes/bacteria/Esch_coli_K_12_MG1655/eschColi_K_12_MG1655-mature-tRNAs.fa
cd ..
./tRNAdb_tool_ecoli.py Esch_coli_K_12_MG1655/eschColi_K_12_MG1655-mature-tRNAs.fa > ecoli/ecoli-tRNAs.fa
```




### Make BLAST database for SWIPE
There are two ways of aligning tRNA sequencing reads to a reference: a) fast heuristics based alignment, or b) slow deterministic alignment.
This repo uses the latter (i.e. the Smith-Waterman algorithm), but the time spent on alignment is substantially decreased by usign a good implementation.
By informal tests, the fastest implementation is [SWIPE](https://github.com/torognes/swipe).
SWIPE requires a pre-made BLAST database of the reference, which needs to be made using `makeblastdb` with the `-blastdb_version 4` flag.

To make this database, step into the folder with the tRNA fasta file and execute the following command:  
`makeblastdb -dbtype prot -in <your_tRNA_transcripts.fa> -blastdb_version 4`

You may note the `-dbtype prot` flag, and wonder why you should specify the sequence as protein when it is DNA.
This is due to a quirk in SWIPE, which will only allow us to specify a substitution matrix for protein, not for DNA.
This behavior of SWIPE makes sense in most applications, except when you want to "mask" positions in the reference sequences.
However, specifying the sequence as protein and providing a custom substitution matrix will work fine.
Masking of the reference can then be done by replacing the masked nt. with "N" and using an appropiate substitution matrix.




