{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/Alicia-samples\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 39 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/Alicia-samples/data/AdapterRemoval\n"
     ]
    }
   ],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=4, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/Alicia-samples/data/BC_split\n",
      "Loaded results from previous run... Not running barcode split.\n"
     ]
    }
   ],
   "source": [
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=12, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/Alicia-samples/data/UMI_trimmed\n",
      "Loaded results from previous run... Not running UMI trimming.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A01</td>\n",
       "      <td>A01</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293T-cys @&amp;@ 293T-sergly @&amp;@ 293T-met @&amp;@ 293T...</td>\n",
       "      <td>Genotype-Time-Drug</td>\n",
       "      <td>WT-0</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>5598086</td>\n",
       "      <td>308346</td>\n",
       "      <td>5237383</td>\n",
       "      <td>5545729</td>\n",
       "      <td>99.064734</td>\n",
       "      <td>94.439937</td>\n",
       "      <td>5478089</td>\n",
       "      <td>509270</td>\n",
       "      <td>524272.802045</td>\n",
       "      <td>97.856464</td>\n",
       "      <td>97.138360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A02</td>\n",
       "      <td>A02</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293T-cys</td>\n",
       "      <td>Genotype-Time-Drug</td>\n",
       "      <td>WT-6</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>5362565</td>\n",
       "      <td>281781</td>\n",
       "      <td>5026667</td>\n",
       "      <td>5308448</td>\n",
       "      <td>98.990837</td>\n",
       "      <td>94.691838</td>\n",
       "      <td>5244422</td>\n",
       "      <td>506781</td>\n",
       "      <td>524264.267492</td>\n",
       "      <td>97.796894</td>\n",
       "      <td>96.665180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A03</td>\n",
       "      <td>A03</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293T-cys</td>\n",
       "      <td>Genotype-Time-Drug</td>\n",
       "      <td>WT-48</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>416891</td>\n",
       "      <td>97926</td>\n",
       "      <td>304409</td>\n",
       "      <td>402335</td>\n",
       "      <td>96.508440</td>\n",
       "      <td>75.660581</td>\n",
       "      <td>407214</td>\n",
       "      <td>179506</td>\n",
       "      <td>283156.761160</td>\n",
       "      <td>97.678770</td>\n",
       "      <td>63.394566</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                A01         A01          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "1                A02         A02          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "2                A03         A03          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2023-06-06/P03_R2.fastq.bz2     D501     D703    l1Sp   human   \n",
       "1  2023-06-06/P03_R2.fastq.bz2     D501     D703    l2Sp   human   \n",
       "2  2023-06-06/P03_R2.fastq.bz2     D501     D703    l3Sp   human   \n",
       "\n",
       "                                          plot_group            hue_name  \\\n",
       "0  293T-cys @&@ 293T-sergly @&@ 293T-met @&@ 293T...  Genotype-Time-Drug   \n",
       "1                                           293T-cys  Genotype-Time-Drug   \n",
       "2                                           293T-cys  Genotype-Time-Drug   \n",
       "\n",
       "  hue_value  hue_order P5_index_seq P7_index_seq         barcode_seq  N_total  \\\n",
       "0      WT-0          1     AGGCTATA     CGCTCATT    GGCTGCCATGCGACTA  5598086   \n",
       "1      WT-6          2     AGGCTATA     CGCTCATT  GGCTGCCATGCTGTCACG  5362565   \n",
       "2     WT-48          3     AGGCTATA     CGCTCATT    GGCTGCCATGCTGCGA   416891   \n",
       "\n",
       "     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  N_after_trim  \\\n",
       "0  308346  5237383   5545729             99.064734    94.439937       5478089   \n",
       "1  281781  5026667   5308448             98.990837    94.691838       5244422   \n",
       "2   97926   304409    402335             96.508440    75.660581        407214   \n",
       "\n",
       "   N_UMI_observed  N_UMI_expected  percent_seqs_after_UMI_trim  \\\n",
       "0          509270   524272.802045                    97.856464   \n",
       "1          506781   524264.267492                    97.796894   \n",
       "2          179506   283156.761160                    97.678770   \n",
       "\n",
       "   percent_UMI_obs-vs-exp  \n",
       "0               97.138360  \n",
       "1               96.665180  \n",
       "2               63.394566  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "# Note, the cDNA input amount is so large that it is very unlikely to sequence\n",
    "# the same PCR amplified DNA twice. Therefore, this processing step does not\n",
    "# attempt to merge possible UMI duplicates.\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=12, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/Alicia-samples/data/SWalign\n",
      "Loaded results from previous run... Not running alignment.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A01</td>\n",
       "      <td>A01</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293T-cys @&amp;@ 293T-sergly @&amp;@ 293T-met @&amp;@ 293T...</td>\n",
       "      <td>Genotype-Time-Drug</td>\n",
       "      <td>WT-0</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>5598086</td>\n",
       "      <td>308346</td>\n",
       "      <td>5237383</td>\n",
       "      <td>5545729</td>\n",
       "      <td>99.064734</td>\n",
       "      <td>94.439937</td>\n",
       "      <td>5478089</td>\n",
       "      <td>509270</td>\n",
       "      <td>524272.802045</td>\n",
       "      <td>97.856464</td>\n",
       "      <td>97.138360</td>\n",
       "      <td>5210288</td>\n",
       "      <td>65.751663</td>\n",
       "      <td>34.248337</td>\n",
       "      <td>4.458564</td>\n",
       "      <td>95.111416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A02</td>\n",
       "      <td>A02</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293T-cys</td>\n",
       "      <td>Genotype-Time-Drug</td>\n",
       "      <td>WT-6</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>5362565</td>\n",
       "      <td>281781</td>\n",
       "      <td>5026667</td>\n",
       "      <td>5308448</td>\n",
       "      <td>98.990837</td>\n",
       "      <td>94.691838</td>\n",
       "      <td>5244422</td>\n",
       "      <td>506781</td>\n",
       "      <td>524264.267492</td>\n",
       "      <td>97.796894</td>\n",
       "      <td>96.665180</td>\n",
       "      <td>4814273</td>\n",
       "      <td>65.548381</td>\n",
       "      <td>34.451619</td>\n",
       "      <td>4.396261</td>\n",
       "      <td>91.797971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A03</td>\n",
       "      <td>A03</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-06-06/P03_R1.fastq.bz2</td>\n",
       "      <td>2023-06-06/P03_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>293T-cys</td>\n",
       "      <td>Genotype-Time-Drug</td>\n",
       "      <td>WT-48</td>\n",
       "      <td>3</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGCGA</td>\n",
       "      <td>416891</td>\n",
       "      <td>97926</td>\n",
       "      <td>304409</td>\n",
       "      <td>402335</td>\n",
       "      <td>96.508440</td>\n",
       "      <td>75.660581</td>\n",
       "      <td>407214</td>\n",
       "      <td>179506</td>\n",
       "      <td>283156.761160</td>\n",
       "      <td>97.678770</td>\n",
       "      <td>63.394566</td>\n",
       "      <td>383222</td>\n",
       "      <td>76.933996</td>\n",
       "      <td>23.066004</td>\n",
       "      <td>11.045034</td>\n",
       "      <td>94.108258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate         fastq_mate1_filename  \\\n",
       "0                A01         A01          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "1                A02         A02          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "2                A03         A03          1  2023-06-06/P03_R1.fastq.bz2   \n",
       "\n",
       "          fastq_mate2_filename P5_index P7_index barcode species  \\\n",
       "0  2023-06-06/P03_R2.fastq.bz2     D501     D703    l1Sp   human   \n",
       "1  2023-06-06/P03_R2.fastq.bz2     D501     D703    l2Sp   human   \n",
       "2  2023-06-06/P03_R2.fastq.bz2     D501     D703    l3Sp   human   \n",
       "\n",
       "                                          plot_group            hue_name  \\\n",
       "0  293T-cys @&@ 293T-sergly @&@ 293T-met @&@ 293T...  Genotype-Time-Drug   \n",
       "1                                           293T-cys  Genotype-Time-Drug   \n",
       "2                                           293T-cys  Genotype-Time-Drug   \n",
       "\n",
       "  hue_value  hue_order P5_index_seq P7_index_seq         barcode_seq  N_total  \\\n",
       "0      WT-0          1     AGGCTATA     CGCTCATT    GGCTGCCATGCGACTA  5598086   \n",
       "1      WT-6          2     AGGCTATA     CGCTCATT  GGCTGCCATGCTGTCACG  5362565   \n",
       "2     WT-48          3     AGGCTATA     CGCTCATT    GGCTGCCATGCTGCGA   416891   \n",
       "\n",
       "     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  N_after_trim  \\\n",
       "0  308346  5237383   5545729             99.064734    94.439937       5478089   \n",
       "1  281781  5026667   5308448             98.990837    94.691838       5244422   \n",
       "2   97926   304409    402335             96.508440    75.660581        407214   \n",
       "\n",
       "   N_UMI_observed  N_UMI_expected  percent_seqs_after_UMI_trim  \\\n",
       "0          509270   524272.802045                    97.856464   \n",
       "1          506781   524264.267492                    97.796894   \n",
       "2          179506   283156.761160                    97.678770   \n",
       "\n",
       "   percent_UMI_obs-vs-exp  N_mapped  percent_single_annotation  \\\n",
       "0               97.138360   5210288                  65.751663   \n",
       "1               96.665180   4814273                  65.548381   \n",
       "2               63.394566    383222                  76.933996   \n",
       "\n",
       "   percent_multiple_annotation  percent_multiple_codons  Mapping_percent  \n",
       "0                    34.248337                 4.458564        95.111416  \n",
       "1                    34.451619                 4.396261        91.797971  \n",
       "2                    23.066004                11.045034        94.108258  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=2, min_score_align=15, \\\n",
    "                        common_seqs=common_seqs, overwrite_dir=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=4, overwrite=False, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/Alicia-samples/data/stats_collection\n",
      "Loaded results from previous run... Not running stats collection.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A01</td>\n",
       "      <td>A01</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A01</td>\n",
       "      <td>A01</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>4403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A01</td>\n",
       "      <td>A01</td>\n",
       "      <td>1</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode species  \\\n",
       "0                A01         A01          1    l1Sp   human   \n",
       "1                A01         A01          1    l1Sp   human   \n",
       "2                A01         A01          1    l1Sp   human   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  count  \n",
       "0               True     False           A   AAA       TTT   eColiLys    175  \n",
       "1               True     False           C   AAA       TTT   eColiLys   4403  \n",
       "2               True      True           A   AAA       TTT   eColiLys     66  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs, \\\n",
    "                             overwrite_dir=False)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=6, load_previous=True)\n",
    "# The dataframe returned is the \"ALL_stats_aggregate_filtered.csv\"\n",
    "# which is the aggregated data filtered to contain only the\n",
    "# most relevant columnns and requiring the 3' must be covered\n",
    "# and have no 3' non-template bases.\n",
    "# The CSV file output \"ALL_stats_aggregate.csv\" is the data\n",
    "# aggregated based on all values identical except\n",
    "# readID, 5p_UMI and 3p_BC. I.e. every information, except\n",
    "# the UMI seequence is maintained in the aggregated CSV.\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=False, pull_default=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  A01  A02  A03  A07  A09  A08  A10  A11  A12  A13  A06  A05  A16  A23  A24  A17  A22  A21  A25  A26  A20  A19  A15  A14  A18  A28  A33  A27  A29  A30  A31  A32  A34  A35  A36  A37  A39  A40  A38  A46  A42  A48  A45  A43  A44  A41  A47  A50  A53  A49  A52  A51  A58  A55  A54  A56  A57  A63  A59  A60  A62  A67  A64  A61  A66  A65  A72  A69  A68  A71  A70  A75  A76  A73  A74\n",
      "Now plotting sample:  A01  A02  A03  A05  A06  A07  A08  A09  A10  A11  A12  A13  A14  A15  A16  A17  A18  A19  A20  A21  A22  A23  A24  A25  A26  A27  A28  A29  A30  A31  A32  A33  A34  A35  A36  A37  A38  A39  A40  A41  A42  A43  A44  A45  A46  A47  A48  A49  A50  A51  A52  A53  A54  A55  A56  A57  A58  A59  A60  A61  A62  A63  A64  A65  A66  A67  A68  A69  A70  A71  A72  A73  A74  A75  A76\n",
      "Now collecting data for sample:  A01  A02  A03  A05  A06  A07  A08  A09  A10  A11  A12  A13  A16  A17  A24  A23  A22  A21  A25  A26  A20  A19  A15  A14  A18  A28  A33  A27  A29  A31  A30  A32  A34  A35  A36  A37  A40  A39  A38  A46  A42  A44  A48  A41  A43  A45  A47  A49  A53  A51  A55  A50  A52  A56  A54  A57  A59  A61  A58  A64  A60  A62  A66  A63  A65  A70  A67  A68  A69  A72  A73  A71  A76  A74  A75\n",
      "Now plotting sample:  A01  A02  A03  A05  A06  A07  A08  A09  A10  A11  A12  A13  A14  A15  A16  A17  A18  A19  A20  A21  A22  A23  A24  A25  A26  A27  A28  A29  A30  A31  A32  A33  A34  A35  A36  A37  A38  A39  A40  A41  A42  A43  A44  A45  A46  A47  A48  A49  A50  A51  A52  A53  A54  A55  A56  A57  A58  A59  A60  A61  A62  A63  A64  A65  A66  A67  A68  A69  A70  A71  A72  A73  A74  A75  A76\n",
      "Now plotting sample/group:  293T-cys  293T-his  293T-lys  293T-met  293T-phe  293T-sergly  293T-tyr  HCT-cys  HCT-his  HCT-lys  HCT-met  HCT-phe  HCT-sergly  HCT-tyr\n",
      "Now plotting sample/group:  293T-cys  293T-his  293T-lys  293T-met  293T-phe  293T-sergly  293T-tyr  HCT-cys  HCT-his  HCT-lys  HCT-met  HCT-phe  HCT-sergly  HCT-tyr\n",
      "Now plotting sample/group:  293T-cys  293T-his  293T-lys  293T-met  293T-phe  293T-sergly  293T-tyr  HCT-cys  HCT-his  HCT-lys  HCT-met  HCT-phe  HCT-sergly  HCT-tyr\n",
      "Now plotting sample/group:  293T-cys  293T-his  293T-lys  293T-met  293T-phe  293T-sergly  293T-tyr  HCT-cys  HCT-his  HCT-lys  HCT-met  HCT-phe  HCT-sergly  HCT-tyr"
     ]
    }
   ],
   "source": [
    "# Coverage plots:\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_norm', n_jobs=12)\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_norm', n_jobs=12)\n",
    "\n",
    "# Charge plots:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_charge', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)\n",
    "\n",
    "# RPM plots:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_rpm', \\\n",
    "                        group=True, min_obs=500, charge_plot=False)\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_rpm', \\\n",
    "                        group=True, min_obs=500, charge_plot=False)\n",
    "\n",
    "# Ecoli control plots:\n",
    "plot_obj.plot_Ecoli_ctr(plot_name='ecoli-ctr_charge', charge_plot=True)\n",
    "plot_obj.plot_Ecoli_ctr(plot_name='ecoli-ctr_rpm', charge_plot=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  A01  A02  A03  A05  A06  A07  A08  A09  A12  A13  A17  A18  A15  A14  A16  A11  A10  A19  A22  A23  A21  A20  A24  A25  A26  A27  A29  A28  A30  A32  A34  A31  A33  A35  A40  A36  A37  A38  A39  A45  A41  A42  A43  A44  A49  A48  A47  A46  A52  A54  A51  A50  A53  A55  A56  A57  A58  A63  A59  A62  A61  A60  A66  A68  A64  A65  A72  A67  A73  A70  A69  A76  A75  A71  A74\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  A01  A07  A02  A03  A05  A06  A08  A09  A12  A13  A18  A17  A15  A11  A14  A16  A10  A19  A21  A23  A22  A20  A25  A24  A26  A27  A28  A33  A32  A34  A31  A29  A30  A35  A39  A37  A36  A38  A46  A41  A43  A45  A40  A42  A44  A47  A48  A51  A54  A50  A49  A53  A52  A55  A56  A60  A57  A61  A58  A62  A63  A59  A64  A68  A71  A66  A65  A67  A70  A72  A69  A74  A73  A75  A76\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  A02  A01  A03  A05  A06  A07  A08  A09  A12  A16  A13  A17  A18  A15  A14  A11  A10  A19  A20  A22  A25  A21  A26  A23  A24  A27  A32  A34  A29  A28  A33  A31  A30  A37  A36  A35  A40  A38  A44  A39  A43  A41  A48  A42  A47  A45  A49  A46  A50  A52  A56  A53  A54  A51  A55  A57  A58  A59  A61  A64  A60  A63  A62  A66  A65  A67  A73  A68  A71  A72  A69  A75  A70  A74  A76\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# Logo plots of 5/3 prime non-template sequence:\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=8)\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=8)\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write charge/rpm data to csv file:\n",
    "fnam = 'data/plotting/alicia_res_AA.csv'\n",
    "plot_obj.charge_filt['aa'].reset_index(drop=True).to_csv(fnam, index=False)\n",
    "fnam = 'data/plotting/alicia_res_codon.csv'\n",
    "plot_obj.charge_filt['codon'].reset_index(drop=True).to_csv(fnam, index=False)\n",
    "fnam = 'data/plotting/alicia_res_tr.csv'\n",
    "plot_obj.charge_filt['tr'].reset_index(drop=True).to_csv(fnam, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     common_seqs=common_seqs, ignore_common_count=False, \\\n",
    "                     overwrite_dir=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting stats from:  A01  A02  A03  A05  A06  A07  A10  A11  A09  A14  A12  A13  A08  A15  A16  A17  A18  A19  A20  A21  A22  A23  A24  A25  A26  A28  A27  A29  A30  A31  A32  A33  A35  A36  A37  A34  A38  A39  A41  A43  A40  A42  A46  A47  A44  A45  A48  A49  A50  A52  A53  A51  A55  A54  A57  A56  A58  A60  A59  A61  A64  A62  A63  A65  A67  A66  A70  A68  A71  A69  A73  A72  A74  A75  A76"
     ]
    }
   ],
   "source": [
    "TM_obj.find_muts(n_jobs=6, unique_anno=False, fix_end=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap difference matrix for methionine depleted samples:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='Met_tr-mut_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   sample_pairs=[['A01', 'A01', 'A20', 'A20', \\\n",
    "                                                  'A39', 'A39', 'A58', 'A58'], \\\n",
    "                                                 ['A10', 'A11', 'A29', 'A30', \\\n",
    "                                                  'A48', 'A49', 'A67', 'A68']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   topN=10, topN_select='max_diff')\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='Met_tr-gap_matrix_comp_top10-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=True, \\\n",
    "                                   min_count_show=500, \\\n",
    "                                   sample_pairs=[['A01', 'A01', 'A20', 'A20', \\\n",
    "                                                  'A39', 'A39', 'A58', 'A58'], \\\n",
    "                                                 ['A10', 'A11', 'A29', 'A30', \\\n",
    "                                                  'A48', 'A49', 'A67', 'A68']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   topN=10, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write transcript mutations/gaps to csv file:\n",
    "TM_obj.write_transcript_mut(csv_name='mut-matrix_right-aligned', \\\n",
    "                            sample_list=list(sample_df['sample_name_unique'].values))\n",
    "TM_obj.write_transcript_mut(csv_name='gap-matrix_right-aligned', gap_only=True, \\\n",
    "                            sample_list=list(sample_df['sample_name_unique'].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
