{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 25 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked['mouse'] = '{}/tRNA_database_masked/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 5\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=4, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=12, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/UMI_trimmed\n",
      "Loaded results from previous run... Not running UMI trimming.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>TAAGGCAT</td>\n",
       "      <td>3732885</td>\n",
       "      <td>1727046</td>\n",
       "      <td>1639602</td>\n",
       "      <td>3366648</td>\n",
       "      <td>90.188902</td>\n",
       "      <td>48.701320</td>\n",
       "      <td>3683016</td>\n",
       "      <td>494716</td>\n",
       "      <td>523821.623056</td>\n",
       "      <td>98.664063</td>\n",
       "      <td>94.443600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-2</td>\n",
       "      <td>L</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>CGTACTCG</td>\n",
       "      <td>4737214</td>\n",
       "      <td>1714150</td>\n",
       "      <td>2726027</td>\n",
       "      <td>4440177</td>\n",
       "      <td>93.729711</td>\n",
       "      <td>61.394557</td>\n",
       "      <td>4674289</td>\n",
       "      <td>504438</td>\n",
       "      <td>524217.593246</td>\n",
       "      <td>98.671688</td>\n",
       "      <td>96.226835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-1+2</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>ATGCAGTA</td>\n",
       "      <td>4910826</td>\n",
       "      <td>1292817</td>\n",
       "      <td>3318699</td>\n",
       "      <td>4611516</td>\n",
       "      <td>93.905099</td>\n",
       "      <td>71.965466</td>\n",
       "      <td>4844079</td>\n",
       "      <td>508574</td>\n",
       "      <td>524237.070610</td>\n",
       "      <td>98.640819</td>\n",
       "      <td>97.012216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate       fastq_mate1_filename  \\\n",
       "0                L-1           L          1  2021-11-22/L_R1.fastq.bz2   \n",
       "1                L-2           L          2  2021-11-22/L_R1.fastq.bz2   \n",
       "2              L-1+2           L          3  2021-11-22/L_R1.fastq.bz2   \n",
       "\n",
       "        fastq_mate2_filename P5_index P7_index barcode species plot_group  \\\n",
       "0  2021-11-22/L_R2.fastq.bz2     D501     D701      l1   human  Pilot-exp   \n",
       "1  2021-11-22/L_R2.fastq.bz2     D501     D701      l2   human  Pilot-exp   \n",
       "2  2021-11-22/L_R2.fastq.bz2     D501     D701      l3   human  Pilot-exp   \n",
       "\n",
       "    hue_name hue_value  hue_order P5_index_seq P7_index_seq barcode_seq  \\\n",
       "0  Treatment    No Leu          2     AGGCTATA     ATTACTCG    TAAGGCAT   \n",
       "1  Treatment    No Leu          2     AGGCTATA     ATTACTCG    CGTACTCG   \n",
       "2  Treatment    No Leu          2     AGGCTATA     ATTACTCG    ATGCAGTA   \n",
       "\n",
       "   N_total     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  \\\n",
       "0  3732885  1727046  1639602   3366648             90.188902    48.701320   \n",
       "1  4737214  1714150  2726027   4440177             93.729711    61.394557   \n",
       "2  4910826  1292817  3318699   4611516             93.905099    71.965466   \n",
       "\n",
       "   N_after_trim  N_UMI_observed  N_UMI_expected  percent_seqs_after_UMI_trim  \\\n",
       "0       3683016          494716   523821.623056                    98.664063   \n",
       "1       4674289          504438   524217.593246                    98.671688   \n",
       "2       4844079          508574   524237.070610                    98.640819   \n",
       "\n",
       "   percent_UMI_obs-vs-exp  \n",
       "0               94.443600  \n",
       "1               96.226835  \n",
       "2               97.012216  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "# Note, the cDNA input amount is so large that it is very unlikely to sequence\n",
    "# the same PCR amplified DNA twice. Therefore, this processing step does not\n",
    "# attempt to merge possible UMI duplicates.\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False, check_input=False)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=12, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/SWalign\n",
      "Loaded results from previous run... Not running alignment.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>TAAGGCAT</td>\n",
       "      <td>3732885</td>\n",
       "      <td>1727046</td>\n",
       "      <td>1639602</td>\n",
       "      <td>3366648</td>\n",
       "      <td>90.188902</td>\n",
       "      <td>48.701320</td>\n",
       "      <td>3683016</td>\n",
       "      <td>494716</td>\n",
       "      <td>523821.623056</td>\n",
       "      <td>98.664063</td>\n",
       "      <td>94.443600</td>\n",
       "      <td>3419143</td>\n",
       "      <td>87.400439</td>\n",
       "      <td>12.599561</td>\n",
       "      <td>0.723953</td>\n",
       "      <td>92.835410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-2</td>\n",
       "      <td>L</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>CGTACTCG</td>\n",
       "      <td>4737214</td>\n",
       "      <td>1714150</td>\n",
       "      <td>2726027</td>\n",
       "      <td>4440177</td>\n",
       "      <td>93.729711</td>\n",
       "      <td>61.394557</td>\n",
       "      <td>4674289</td>\n",
       "      <td>504438</td>\n",
       "      <td>524217.593246</td>\n",
       "      <td>98.671688</td>\n",
       "      <td>96.226835</td>\n",
       "      <td>4475974</td>\n",
       "      <td>88.112755</td>\n",
       "      <td>11.887245</td>\n",
       "      <td>0.785617</td>\n",
       "      <td>95.757323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-1+2</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>ATGCAGTA</td>\n",
       "      <td>4910826</td>\n",
       "      <td>1292817</td>\n",
       "      <td>3318699</td>\n",
       "      <td>4611516</td>\n",
       "      <td>93.905099</td>\n",
       "      <td>71.965466</td>\n",
       "      <td>4844079</td>\n",
       "      <td>508574</td>\n",
       "      <td>524237.070610</td>\n",
       "      <td>98.640819</td>\n",
       "      <td>97.012216</td>\n",
       "      <td>4644107</td>\n",
       "      <td>87.785488</td>\n",
       "      <td>12.214512</td>\n",
       "      <td>0.802716</td>\n",
       "      <td>95.871826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate       fastq_mate1_filename  \\\n",
       "0                L-1           L          1  2021-11-22/L_R1.fastq.bz2   \n",
       "1                L-2           L          2  2021-11-22/L_R1.fastq.bz2   \n",
       "2              L-1+2           L          3  2021-11-22/L_R1.fastq.bz2   \n",
       "\n",
       "        fastq_mate2_filename P5_index P7_index barcode species plot_group  \\\n",
       "0  2021-11-22/L_R2.fastq.bz2     D501     D701      l1   human  Pilot-exp   \n",
       "1  2021-11-22/L_R2.fastq.bz2     D501     D701      l2   human  Pilot-exp   \n",
       "2  2021-11-22/L_R2.fastq.bz2     D501     D701      l3   human  Pilot-exp   \n",
       "\n",
       "    hue_name hue_value  hue_order P5_index_seq P7_index_seq barcode_seq  \\\n",
       "0  Treatment    No Leu          2     AGGCTATA     ATTACTCG    TAAGGCAT   \n",
       "1  Treatment    No Leu          2     AGGCTATA     ATTACTCG    CGTACTCG   \n",
       "2  Treatment    No Leu          2     AGGCTATA     ATTACTCG    ATGCAGTA   \n",
       "\n",
       "   N_total     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  \\\n",
       "0  3732885  1727046  1639602   3366648             90.188902    48.701320   \n",
       "1  4737214  1714150  2726027   4440177             93.729711    61.394557   \n",
       "2  4910826  1292817  3318699   4611516             93.905099    71.965466   \n",
       "\n",
       "   N_after_trim  N_UMI_observed  N_UMI_expected  percent_seqs_after_UMI_trim  \\\n",
       "0       3683016          494716   523821.623056                    98.664063   \n",
       "1       4674289          504438   524217.593246                    98.671688   \n",
       "2       4844079          508574   524237.070610                    98.640819   \n",
       "\n",
       "   percent_UMI_obs-vs-exp  N_mapped  percent_single_annotation  \\\n",
       "0               94.443600   3419143                  87.400439   \n",
       "1               96.226835   4475974                  88.112755   \n",
       "2               97.012216   4644107                  87.785488   \n",
       "\n",
       "   percent_multiple_annotation  percent_multiple_codons  Mapping_percent  \n",
       "0                    12.599561                 0.723953        92.835410  \n",
       "1                    11.887245                 0.785617        95.757323  \n",
       "2                    12.214512                 0.802716        95.871826  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        overwrite_dir=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=6, overwrite=False, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/stats_collection\n",
      "Collecting stats from:  L-1  L-2  L-1+2  A-2  A-1+2  A-NaCl  Rich-1  Rich-2  Rich-1+2  Rich-NaCl  L-NaCl  A-1  2U1  U-1  0U1  Fix1  U-2  U-NaCl  U-1+2  0U2  2U2  Fix3  8U1  4U1  FT  Fix2  HVS2  HVV1  HAV2  4U2  HAV1  HVP2  8U2  HAS2  HVS1  HCV2  HVP1  HR30V1  HVV2  HCV1  HR30P1  HAS1  BVV1  HCS2  HR30V2  HCS1  HR30P2  HR30S2  BVV2  HR20V1  BVP2  BVS1  HR20P1  BVS2  HR30S1  HR40P1  HR40V2  HR20S2  HR20S1  BVP1  HR20V2  HR40S2  HR80V1  HR80P1  Tu3  HR40P2  HR80V2  HR80P2  HR40S1  HR40V1  HR20P2  Tu1  HR80S1  Li2  Tu2  Mu2  BAV1  Li1  Tu4  Mu1  Mu4  CyA  HR80S2  BAR2  BAR1  BVR1  BAV2  CyP  Li4  Mu3  BVR2  Li3"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>7291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>6288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>3787</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode  \\\n",
       "0                L-1           L          1      l1   \n",
       "1                L-1           L          1      l1   \n",
       "2                L-1           L          1      l1   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  count  \n",
       "0               True     False           A   AAA       TTT   eColiLys   7291  \n",
       "1               True     False           C   AAA       TTT   eColiLys   6288  \n",
       "2               True      True           A   AAA       TTT   eColiLys   3787  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, \\\n",
    "                             overwrite_dir=False)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=12, load_previous=False)\n",
    "# The dataframe returned is the \"ALL_stats_aggregate_filtered.csv\"\n",
    "# which is the aggregated data filtered to contain only the\n",
    "# most relevant columnns and requiring the 3' must be covered\n",
    "# and have no 3' non-template bases.\n",
    "# The CSV file output \"ALL_stats_aggregate.csv\" is the data\n",
    "# aggregated based on all values identical except\n",
    "# readID, 5p_UMI and 3p_BC. I.e. every information, except\n",
    "# the UMI seequence is maintained in the aggregated CSV.\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=True, pull_default=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  L-1  L-2  L-1+2  L-NaCl  A-1  A-2  A-1+2  A-NaCl  U-1  Rich-1  U-1+2  Rich-1+2  Rich-2  Rich-NaCl  U-NaCl  U-2  Fix2  Fix1  FT  0U1  2U2  0U2  Fix3  4U1  8U2  HVV2  HVS2  4U2  HVS1  2U1  HVV1  8U1  HAS2  HVP2  HAS1  HAV1  HCV1  HCV2  HVP1  HAV2  HR30V1  HR30V2  HR30P1  HCS2  HCS1  HR30S1  HR30S2  BVV1  BVV2  BVP1  HR30P2  BVP2  HR20V2  BVS1  HR20S2  BVS2  HR20P1  HR40P1  HR40V2  HR20V1  HR20P2  HR40V1  HR20S1  HR40S2  HR80V2  HR40P2  HR80S2  HR80P1  HR80V1  Li1  Tu2  HR80P2  Tu3  HR40S1  Li2  Tu4  Mu2  Li3  HR80S1  Mu1  Mu4  BVR2  Tu1  BVR1  Mu3  CyP  CyA  BAV2  Li4  BAV1  BAR2  BAR1\n",
      "Now plotting sample:  L-1  L-2  L-1+2  L-NaCl  A-1  A-2  A-1+2  A-NaCl  Rich-1  Rich-2  Rich-1+2  Rich-NaCl  U-1  U-2  U-1+2  U-NaCl  Fix1  Fix2  Fix3  FT  0U1  0U2  2U1  2U2  4U1  4U2  8U1  8U2  HVV1  HVV2  HVS1  HVS2  HVP1  HVP2  HAV1  HAV2  HAS1  HAS2  HCV1  HCV2  HCS1  HCS2  HR30V1  HR30V2  HR30S1  HR30S2  HR30P1  HR30P2  BVV1  BVV2  BVS1  BVS2  BVP1  BVP2  HR20V1  HR20V2  HR20S1  HR20S2  HR20P1  HR20P2  HR40V1  HR40V2  HR40S1  HR40S2  HR40P1  HR40P2  HR80V1  HR80V2  HR80S1  HR80S2  HR80P1  HR80P2  Tu1  Tu2  Tu3  Tu4  Li1  Li2  Li3  Li4  Mu1  Mu2  Mu3  Mu4  BAV1  BAV2  BVR1  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now collecting data for sample:  L-1  L-1+2  L-2  L-NaCl  A-1  A-2  A-1+2  A-NaCl  U-1  Rich-1  Rich-NaCl  Rich-2  Rich-1+2  U-1+2  U-2  U-NaCl  Fix2  0U1  Fix1  4U2  Fix3  FT  2U1  4U1  0U2  HVS1  HVS2  HVP1  HVV1  2U2  HVP2  8U2  8U1  HAV2  HVV2  HAS1  HAV1  HCS1  HAS2  HR30V2  HCV1  HR30S1  HR30V1  HCV2  HR30S2  HCS2  BVV2  HR30P2  BVV1  HR30P1  BVS2  BVS1  HR20V2  HR20S1  BVP2  HR20P2  HR40V2  HR20P1  HR40P2  HR20V1  BVP1  HR40V1  HR20S2  HR80V2  HR80S1  HR80V1  HR40S1  Tu2  Tu1  HR80S2  HR40P1  Li1  Tu4  Tu3  Li2  HR40S2  Li4  Li3  Mu1  HR80P1  BVR1  Mu2  Mu3  Mu4  HR80P2  BAR1  BAV1  BAV2  BVR2  BAR2  CyP  CyA\n",
      "Now plotting sample:  L-1  L-2  L-1+2  L-NaCl  A-1  A-2  A-1+2  A-NaCl  Rich-1  Rich-2  Rich-1+2  Rich-NaCl  U-1  U-2  U-1+2  U-NaCl  Fix1  Fix2  Fix3  FT  0U1  0U2  2U1  2U2  4U1  4U2  8U1  8U2  HVV1  HVV2  HVS1  HVS2  HVP1  HVP2  HAV1  HAV2  HAS1  HAS2  HCV1  HCV2  HCS1  HCS2  HR30V1  HR30V2  HR30S1  HR30S2  HR30P1  HR30P2  BVV1  BVV2  BVS1  BVS2  BVP1  BVP2  HR20V1  HR20V2  HR20S1  HR20S2  HR20P1  HR20P2  HR40V1  HR40V2  HR40S1  HR40S2  HR40P1  HR40P2  HR80V1  HR80V2  HR80S1  HR80S2  HR80P1  HR80P2  Tu1  Tu2  Tu3  Tu4  Li1  Li2  Li3  Li4  Mu1  Mu2  Mu3  Mu4  BAV1  BAV2  BVR1  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now collecting data for sample:  L-1  L-NaCl  L-2  L-1+2  A-1  A-2  A-NaCl  A-1+2  Rich-1  Rich-NaCl  U-1  Rich-2  Rich-1+2  Fix1  U-2  U-1+2  U-NaCl  Fix3  Fix2  FT  0U2  0U1  2U1  4U1  4U2  8U1  8U2  2U2  HVV2  HVV1  HVP2  HVP1  HVS2  HVS1  HAV2  HAS2  HAV1  HAS1  HCS1  HCS2  HCV1  HCV2  HR30V1  HR30S1  HR30V2  HR30S2  HR30P1  HR30P2  BVS2  BVV1  BVP2  BVP1  BVS1  BVV2  HR20V2  HR20S1  HR20V1  HR20P2  HR40V1  HR20P1  HR20S2  HR40S2  HR40S1  HR40P2  HR80V1  HR80S1  HR80S2  HR80V2  HR40V2  HR40P1  HR80P1  Tu2  Tu1  Tu3  Tu4  Li1  HR80P2  Li4  Li2  Li3  Mu3  Mu4  Mu2  Mu1  BVR2  BAV1  BAV2  BVR1  BAR1  BAR2  CyA  CyP\n",
      "Now plotting logo plot.\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  L-1+2  L-2  L-1  L-NaCl  A-1  A-2  A-1+2  A-NaCl  Rich-2  U-1  Rich-1  U-1+2  Rich-NaCl  Rich-1+2  U-2  U-NaCl  Fix3  Fix1  0U2  2U1  Fix2  0U1  FT  8U1  8U2  HVV2  4U1  HVV1  HVS2  HAV1  4U2  HVP2  2U2  HAV2  HAS2  HVS1  HAS1  HCV1  HCV2  HVP1  HR30V2  HR30S1  HCS1  HR30P2  HCS2  HR30P1  BVS2  HR30V1  BVV2  BVV1  BVS1  HR30S2  HR20S2  HR20V1  HR20V2  HR20P1  HR40V2  HR40S2  BVP2  BVP1  HR20S1  HR80V2  HR40S1  HR20P2  HR80S1  HR80V1  Tu1  HR40P2  HR40V1  HR40P1  Tu2  HR80P1  Tu4  Tu3  Li1  Li2  Li3"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Worker-4 died unexpectedly",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_3360/2213123483.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      8\u001b[0m                        seq_len_percentile=99, n_jobs=4)\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m \u001b[0mplot_obj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot_UMI_logo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m8\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
      "\u001b[0;32m~/tRNA-charge-seq/src/plotting.py\u001b[0m in \u001b[0;36mplot_UMI_logo\u001b[0;34m(self, plot_name, n_jobs, verbose)\u001b[0m\n\u001b[1;32m    674\u001b[0m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    675\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mWorkerPool\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 676\u001b[0;31m             \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_collect_UMI_mat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    677\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_run_logomaker\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplot_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    678\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36mmap\u001b[0;34m(self, func, iterable_of_args, iterable_len, max_tasks_active, chunk_size, n_splits, worker_lifespan, progress_bar, progress_bar_position, concatenate_numpy_output, worker_init, worker_exit, task_timeout, worker_init_timeout, worker_exit_timeout, progress_bar_options)\u001b[0m\n\u001b[1;32m    368\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0miterable_len\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mhasattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterable_of_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__len__'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    369\u001b[0m             \u001b[0miterable_len\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterable_of_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 370\u001b[0;31m         results = self.map_unordered(func, ((args_idx, args) for args_idx, args in enumerate(iterable_of_args)),\n\u001b[0m\u001b[1;32m    371\u001b[0m                                      \u001b[0miterable_len\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_tasks_active\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchunk_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_splits\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_lifespan\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    372\u001b[0m                                      \u001b[0mprogress_bar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprogress_bar_position\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_init\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_exit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtask_timeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36mmap_unordered\u001b[0;34m(self, func, iterable_of_args, iterable_len, max_tasks_active, chunk_size, n_splits, worker_lifespan, progress_bar, progress_bar_position, worker_init, worker_exit, task_timeout, worker_init_timeout, worker_exit_timeout, progress_bar_options)\u001b[0m\n\u001b[1;32m    437\u001b[0m         \"\"\"\n\u001b[1;32m    438\u001b[0m         \u001b[0;31m# Simply call imap and cast it to a list. This make sure all elements are there before returning\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m         return list(self.imap_unordered(func, iterable_of_args, iterable_len, max_tasks_active, chunk_size,\n\u001b[0m\u001b[1;32m    440\u001b[0m                                         \u001b[0mn_splits\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_lifespan\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprogress_bar\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprogress_bar_position\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_init\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m                                         \u001b[0mworker_exit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtask_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_init_timeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mworker_exit_timeout\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36mimap_unordered\u001b[0;34m(self, func, iterable_of_args, iterable_len, max_tasks_active, chunk_size, n_splits, worker_lifespan, progress_bar, progress_bar_position, worker_init, worker_exit, task_timeout, worker_init_timeout, worker_exit_timeout, progress_bar_options)\u001b[0m\n\u001b[1;32m    662\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    663\u001b[0m                         \u001b[0;31m# Check worker status (e.g., restarts). This can yield intermediate results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 664\u001b[0;31m                         \u001b[0;32mfor\u001b[0m \u001b[0mresults\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_check_worker_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    665\u001b[0m                             \u001b[0;32myield\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mresults\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    666\u001b[0m                             \u001b[0mn_active\u001b[0m \u001b[0;34m-=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/mpire/pool.py\u001b[0m in \u001b[0;36m_check_worker_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    224\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_worker_comms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mRuntimeError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr_msg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    225\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mterminate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 226\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr_msg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    227\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m         \u001b[0;31m# Check for worker_init/task/worker_exit timeouts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Worker-4 died unexpectedly"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_behrens_norm', n_jobs=8)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_behrens_norm', n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_UMI_logo(n_jobs=8)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  Barcode test  293t-NoLeu  293t-NoArg  293t-NoGln  HCT116-Depletion\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  Barcode test  293t-NoLeu  293t-NoArg  293t-NoGln  HCT116-Depletion\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  Barcode test  293t-NoLeu  293t-NoArg  293t-NoGln  HCT116-Depletion\n",
      "Now plotting sample/group:  Charge-titration  Acylation-half-life  Barcode test  293t-NoLeu  293t-NoArg  293t-NoGln  HCT116-Depletion\n",
      "Now plotting sample pairs:  (100p - 85p)  (100p - 0p)  (85p - 40p)  (85p - 0p)  (40p - 0p)  (40h - 40h_NoOx)\n",
      "Now plotting sample pairs:  (100p - 85p)  (100p - 0p)  (85p - 40p)  (85p - 0p)  (40p - 0p)  (40h - 40h_NoOx)\n",
      "Now plotting sample pairs:  (L1 - R1)  (L1 - L6)  (R1 - R6)\n",
      "Now plotting sample pairs:  (L1 - R1)  (L1 - L6)  (R1 - R6)\n",
      "Now plotting sample pairs:  (40h_1 - 40h_2)  (40h_NoOx_1 - 40h_NoOx_2)"
     ]
    }
   ],
   "source": [
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_abundance_grp', \\\n",
    "                        group=True, min_obs=500)\n",
    "\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_abundance_grp', \\\n",
    "                        group=True, min_obs=500)\n",
    "\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)\n",
    "\n",
    "plot_obj.plot_abundance(plot_type='aa', plot_name='aa_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['100p', '100p', '85p', '85p', '40p', '40h'], \\\n",
    "                                           ['85p', '0p', '40p', '0p', '0p', '40h_NoOx']], \\\n",
    "                             plot_type='codon', plot_name='codon_abundance_corr', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_pairs=[['100p', '100p', '85p', '85p', '40p', '40h'], \\\n",
    "                                           ['85p', '0p', '40p', '0p', '0p', '40h_NoOx']], \\\n",
    "                             plot_type='transcript', plot_name='tr_abundance_corr', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['L1', 'L1', 'R1'], ['R1', 'L6', 'R6']], \\\n",
    "                             plot_type='codon', plot_name='codon_abundance_corr_L-R', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['L1', 'L1', 'R1'], ['R1', 'L6', 'R6']], \\\n",
    "                             plot_type='transcript', plot_name='tr_abundance_corr_L-R', \\\n",
    "                             min_obs=500, charge_plot=False)\n",
    "\n",
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['100p1', '85p1', '55p1', '40p1', '0p1', '40h_1', '40h_NoOx_1'], \\\n",
    "                                           ['100p3', '85p4', '55p3', '40p2', '0p2', '40h_2', '40h_NoOx_2']], \\\n",
    "                             plot_type='codon', plot_name='codon_charge_corr', \\\n",
    "                             min_obs=500, charge_plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the TM_analysis to find the most mutated positions in the tRNA transcripts.\n",
    "# Then generate a new version of the tRNA transcripts database with these positions masked\n",
    "# Then, using the masked sequences, re-run the alignment, stats collected and plotting\n",
    "\n",
    "# Re-run using gap_open = -3 and gap_extension = -2 to reflect how gaps are more tolerated\n",
    "\n",
    "### The masked tRNA database will only be used for alignment\n",
    "# All other steps will use the old unmasked database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     overwrite_dir=False)\n",
    "\n",
    "\n",
    "#TM_obj.find_muts(n_jobs=12, unique_anno=True)\n",
    "#TM_obj.find_muts(n_jobs=4)\n",
    "\n",
    "#TM_obj.plot_transcript_logo()\n",
    "\n",
    "\n",
    "\n",
    "# tr_mut_out = TM_obj.plot_transcript_mut(topN=30, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat', min_count_show=10)\n",
    "#tr_mut_out = TM_obj.plot_transcript_mut(topN=30, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat2')\n",
    "\n",
    "\n",
    "#tr_cov_out = TM_obj.plot_transcript_cov(topN=40, no_plot_return=False, mito=False, plot_name='test_heat_cov', sort_rows=True)\n",
    "#tr_mut_out = TM_obj.plot_transcript_mut(topN=40, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat2', sort_rows=tr_cov_out[2])\n",
    "\n",
    "\n",
    "#TM_obj.mask_tRNA_database(min_mut_freq=0.5, min_pos_count=50, min_tr_count=100, frac_max_score=0.90)\n",
    "\n",
    "#tRNA_database_masked = TM_obj.write_masked_tRNA_database(out_dir='tRNA_database_masked')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TM_obj.find_muts(n_jobs=10, unique_anno=True, fix_end=True, \\\n",
    "#                 sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "#                              '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3',\n",
    "#                              'L1', 'L3', 'L6', 'L8', 'L12', \\\n",
    "#                              'R1', 'R3', 'R6', 'R8', 'R12'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting stats from:  L-1  L-2  A-2  A-1  A-1+2  L-NaCl  A-NaCl  L-1+2  Rich-1  Rich-2  U-1+2  Rich-1+2  Fix3  U-2  Rich-NaCl  U-1  Fix1  FT  Fix2  U-NaCl  0U2  2U2  2U1  0U1  HVV1  HVS1  HVP1  4U1  4U2  8U1  HVP2  HAV1  HAV2  HVS2  8U2  HAS2  HVV2  HAS1  HCV2  HCV1  HCS2  HR30V2  HR30P2  HCS1  HR30P1  HR30V1  HR30S2  BVV1  HR30S1  BVP1  BVV2  BVS1  BVP2  BVS2  HR20S2  HR20P1  HR20V2  HR20P2  HR40P2  HR20V1  HR40V2  HR20S1  HR40S2  HR80S2  HR40V1  HR80S1  HR80V1  Li1  HR80P2  HR40S1  Tu4  Tu2  Tu3  HR40P1  Tu1  HR80V2  Li2  HR80P1  Li4  Mu1  BAV1  Mu2  Mu4  BAR2  Mu3  Li3  BVR2  BAR1  BVR1  CyP  BAV2  CyA"
     ]
    }
   ],
   "source": [
    "#alicia_samples.remove('R4')\n",
    "TM_obj.find_muts(n_jobs=10, unique_anno=False, fix_end=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Arg', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Arg')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Leu', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=1000, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Leu')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Asn', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Asn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='NoLeu-NoArg_tr-mut_matrix_comp_AA-Ala', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False, gap_only=False, \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_unique_pairs=[['L1', 'L8',  'L1', 'L6',  'R1', 'R8',  'R1', 'R6'], \\\n",
    "                                                        ['L6', 'L12', 'L8', 'L12', 'R6', 'R12', 'R8', 'R12']], \\\n",
    "                                   freq_avg_weighted=True, \\\n",
    "                                   anno_substring_compare='Ala')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.plot_transcript_logo(sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "                                         '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3', \\\n",
    "                                         'L1', 'L3', 'L6', 'R1', 'R3', 'R6'], \\\n",
    "                            plot_name='tr-muts_logo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr_cov_out = TM_obj.plot_transcript_cov(topN=40, no_plot_return=True, mito=False, \\\n",
    "                                        plot_name='tr-cov_matrix', sort_rows=True, \\\n",
    "                                        sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "                                                     '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3', \\\n",
    "                                                     'L1', 'L3', 'L6', 'R1', 'R3', 'R6'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr_mut_out = TM_obj.plot_transcript_mut(topN=40, no_plot_return=True, mito=False, \\\n",
    "                                        gap_only=False, min_count_show=1000, \\\n",
    "                                        plot_name='tr-mut_matrix', \\\n",
    "                                        sample_list=['100p1', '100p3', '0p1', '0p2', '0p4', \\\n",
    "                                                     '0m_1', '40h_NoOx_1', '40h_NoOx_2', '0m_3', \\\n",
    "                                                     'L1', 'L3', 'L6', 'R1', 'R3', 'R6'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.write_transcript_mut(list(TM_obj.tr_muts.keys()), \\\n",
    "                            csv_name='first-gen_tRNAseq_mut-matrix')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "TM_obj.write_transcript_mut(list(TM_obj.tr_muts.keys()), \\\n",
    "                            csv_name='first-gen_tRNAseq_mut-matrix_left-aligned')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
