{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float64'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:518: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  setattr(self, word, getattr(machar, word).flat[0])\n",
      "/home/sulab/anaconda3/lib/python3.9/site-packages/numpy/core/getlimits.py:89: UserWarning: The value of the smallest subnormal for <class 'numpy.float32'> type is zero.\n",
      "  return self._float_to_str(self.smallest_subnormal)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2, copy\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen\n",
      "Repo is in: /home/sulab/tRNA-charge-seq\n",
      "Using minimum read length: 25 (after merge)\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked = dict()\n",
    "tRNA_database_masked['human'] = '{}/tRNA_database_masked/human/human-tRNAs.fa'.format(homedir)\n",
    "tRNA_database_masked['mouse'] = '{}/tRNA_database_masked/mouse/mouse-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 10\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 5\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Input settings ###\n",
    "sample_list_fnam = 'sample_list.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=1e4)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/AdapterRemoval\n"
     ]
    }
   ],
   "source": [
    "### Run AdapterRemoval ###\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN, overwrite_dir=False, \\\n",
    "                  check_input=False)\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=2, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/BC_split\n",
      "Loaded results from previous run... Not running barcode split.\n"
     ]
    }
   ],
   "source": [
    "### Split files based on barcodes ###\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df, overwrite_dir=False)\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=6, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/UMI_trimmed\n",
      "Loaded results from previous run... Not running UMI trimming.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>TAAGGCAT</td>\n",
       "      <td>3732885</td>\n",
       "      <td>1727039</td>\n",
       "      <td>1639607</td>\n",
       "      <td>3366646</td>\n",
       "      <td>90.188849</td>\n",
       "      <td>48.701497</td>\n",
       "      <td>98.663848</td>\n",
       "      <td>94.441311</td>\n",
       "      <td>3683008</td>\n",
       "      <td>523821.615940</td>\n",
       "      <td>2500000</td>\n",
       "      <td>494704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-2</td>\n",
       "      <td>L</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>CGTACTCG</td>\n",
       "      <td>4737211</td>\n",
       "      <td>1714169</td>\n",
       "      <td>2726081</td>\n",
       "      <td>4440250</td>\n",
       "      <td>93.731312</td>\n",
       "      <td>61.394764</td>\n",
       "      <td>98.671497</td>\n",
       "      <td>96.224165</td>\n",
       "      <td>4674277</td>\n",
       "      <td>524217.591635</td>\n",
       "      <td>2500000</td>\n",
       "      <td>504424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-1+2</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>ATGCAGTA</td>\n",
       "      <td>4910788</td>\n",
       "      <td>1292825</td>\n",
       "      <td>3318665</td>\n",
       "      <td>4611490</td>\n",
       "      <td>93.905296</td>\n",
       "      <td>71.965135</td>\n",
       "      <td>98.640524</td>\n",
       "      <td>97.015269</td>\n",
       "      <td>4844027</td>\n",
       "      <td>524237.065559</td>\n",
       "      <td>2500000</td>\n",
       "      <td>508590</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate       fastq_mate1_filename  \\\n",
       "0                L-1           L          1  2021-11-22/L_R1.fastq.bz2   \n",
       "1                L-2           L          2  2021-11-22/L_R1.fastq.bz2   \n",
       "2              L-1+2           L          3  2021-11-22/L_R1.fastq.bz2   \n",
       "\n",
       "        fastq_mate2_filename P5_index P7_index barcode species plot_group  \\\n",
       "0  2021-11-22/L_R2.fastq.bz2     D501     D701      l1   human  Pilot-exp   \n",
       "1  2021-11-22/L_R2.fastq.bz2     D501     D701      l2   human  Pilot-exp   \n",
       "2  2021-11-22/L_R2.fastq.bz2     D501     D701      l3   human  Pilot-exp   \n",
       "\n",
       "    hue_name hue_value  hue_order P5_index_seq P7_index_seq barcode_seq  \\\n",
       "0  Treatment    No Leu          2     AGGCTATA     ATTACTCG    TAAGGCAT   \n",
       "1  Treatment    No Leu          2     AGGCTATA     ATTACTCG    CGTACTCG   \n",
       "2  Treatment    No Leu          2     AGGCTATA     ATTACTCG    ATGCAGTA   \n",
       "\n",
       "   N_total     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  \\\n",
       "0  3732885  1727039  1639607   3366646             90.188849    48.701497   \n",
       "1  4737211  1714169  2726081   4440250             93.731312    61.394764   \n",
       "2  4910788  1292825  3318665   4611490             93.905296    71.965135   \n",
       "\n",
       "   percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  N_after_trim  \\\n",
       "0                    98.663848               94.441311       3683008   \n",
       "1                    98.671497               96.224165       4674277   \n",
       "2                    98.640524               97.015269       4844027   \n",
       "\n",
       "   N_UMI_expected  N_after_downsample  N_UMI_observed  \n",
       "0   523821.615940             2500000          494704  \n",
       "1   524217.591635             2500000          504424  \n",
       "2   524237.065559             2500000          508590  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Split files based on barcodes ###\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df, overwrite_dir=False, \\\n",
    "                       downsample_absolute=2.5e6)\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=4, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/SWalign\n",
      "Loaded results from previous run... Not running alignment.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>N_after_downsample</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>percent_multiple_codons</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>TAAGGCAT</td>\n",
       "      <td>3732885</td>\n",
       "      <td>1727039</td>\n",
       "      <td>1639607</td>\n",
       "      <td>3366646</td>\n",
       "      <td>90.188849</td>\n",
       "      <td>48.701497</td>\n",
       "      <td>98.663848</td>\n",
       "      <td>94.441311</td>\n",
       "      <td>3683008</td>\n",
       "      <td>523821.615940</td>\n",
       "      <td>2500000</td>\n",
       "      <td>494704</td>\n",
       "      <td>2320697</td>\n",
       "      <td>88.412964</td>\n",
       "      <td>11.587036</td>\n",
       "      <td>0.718965</td>\n",
       "      <td>63.010914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-2</td>\n",
       "      <td>L</td>\n",
       "      <td>2</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l2</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>CGTACTCG</td>\n",
       "      <td>4737211</td>\n",
       "      <td>1714169</td>\n",
       "      <td>2726081</td>\n",
       "      <td>4440250</td>\n",
       "      <td>93.731312</td>\n",
       "      <td>61.394764</td>\n",
       "      <td>98.671497</td>\n",
       "      <td>96.224165</td>\n",
       "      <td>4674277</td>\n",
       "      <td>524217.591635</td>\n",
       "      <td>2500000</td>\n",
       "      <td>504424</td>\n",
       "      <td>2394116</td>\n",
       "      <td>88.263016</td>\n",
       "      <td>11.736984</td>\n",
       "      <td>0.782168</td>\n",
       "      <td>51.218959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-1+2</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>2021-11-22/L_R1.fastq.bz2</td>\n",
       "      <td>2021-11-22/L_R2.fastq.bz2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l3</td>\n",
       "      <td>human</td>\n",
       "      <td>Pilot-exp</td>\n",
       "      <td>Treatment</td>\n",
       "      <td>No Leu</td>\n",
       "      <td>2</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>ATGCAGTA</td>\n",
       "      <td>4910788</td>\n",
       "      <td>1292825</td>\n",
       "      <td>3318665</td>\n",
       "      <td>4611490</td>\n",
       "      <td>93.905296</td>\n",
       "      <td>71.965135</td>\n",
       "      <td>98.640524</td>\n",
       "      <td>97.015269</td>\n",
       "      <td>4844027</td>\n",
       "      <td>524237.065559</td>\n",
       "      <td>2500000</td>\n",
       "      <td>508590</td>\n",
       "      <td>2396555</td>\n",
       "      <td>87.979871</td>\n",
       "      <td>12.020129</td>\n",
       "      <td>0.798938</td>\n",
       "      <td>49.474435</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate       fastq_mate1_filename  \\\n",
       "0                L-1           L          1  2021-11-22/L_R1.fastq.bz2   \n",
       "1                L-2           L          2  2021-11-22/L_R1.fastq.bz2   \n",
       "2              L-1+2           L          3  2021-11-22/L_R1.fastq.bz2   \n",
       "\n",
       "        fastq_mate2_filename P5_index P7_index barcode species plot_group  \\\n",
       "0  2021-11-22/L_R2.fastq.bz2     D501     D701      l1   human  Pilot-exp   \n",
       "1  2021-11-22/L_R2.fastq.bz2     D501     D701      l2   human  Pilot-exp   \n",
       "2  2021-11-22/L_R2.fastq.bz2     D501     D701      l3   human  Pilot-exp   \n",
       "\n",
       "    hue_name hue_value  hue_order P5_index_seq P7_index_seq barcode_seq  \\\n",
       "0  Treatment    No Leu          2     AGGCTATA     ATTACTCG    TAAGGCAT   \n",
       "1  Treatment    No Leu          2     AGGCTATA     ATTACTCG    CGTACTCG   \n",
       "2  Treatment    No Leu          2     AGGCTATA     ATTACTCG    ATGCAGTA   \n",
       "\n",
       "   N_total     N_CC    N_CCA  N_CCA+CC  CCA+CC_percent_total  percent_CCA  \\\n",
       "0  3732885  1727039  1639607   3366646             90.188849    48.701497   \n",
       "1  4737211  1714169  2726081   4440250             93.731312    61.394764   \n",
       "2  4910788  1292825  3318665   4611490             93.905296    71.965135   \n",
       "\n",
       "   percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  N_after_trim  \\\n",
       "0                    98.663848               94.441311       3683008   \n",
       "1                    98.671497               96.224165       4674277   \n",
       "2                    98.640524               97.015269       4844027   \n",
       "\n",
       "   N_UMI_expected  N_after_downsample  N_UMI_observed  N_mapped  \\\n",
       "0   523821.615940             2500000          494704   2320697   \n",
       "1   524217.591635             2500000          504424   2394116   \n",
       "2   524237.065559             2500000          508590   2396555   \n",
       "\n",
       "   percent_single_annotation  percent_multiple_annotation  \\\n",
       "0                  88.412964                    11.587036   \n",
       "1                  88.263016                    11.736984   \n",
       "2                  87.979871                    12.020129   \n",
       "\n",
       "   percent_multiple_codons  Mapping_percent  \n",
       "0                 0.718965        63.010914  \n",
       "1                 0.782168        51.218959  \n",
       "2                 0.798938        49.474435  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, \\\n",
    "                        gap_penalty=6, extension_penalty=3, min_score_align=15, \\\n",
    "                        overwrite_dir=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=6, overwrite=False, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /home/sulab/tRNA-charge-seq/projects/tRNAseq_first-gen/data/stats_collection\n",
      "Loaded results from previous run... Not running stats collection.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>5p_cover</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>align_gap</th>\n",
       "      <th>fmax_score&gt;0.9</th>\n",
       "      <th>count</th>\n",
       "      <th>UMIcount</th>\n",
       "      <th>UMI_percent_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>337</td>\n",
       "      <td>337</td>\n",
       "      <td>100.032047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>4270</td>\n",
       "      <td>4270</td>\n",
       "      <td>100.407676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L-1</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>l1</td>\n",
       "      <td>human</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>A</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>317</td>\n",
       "      <td>317</td>\n",
       "      <td>100.030139</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode species  \\\n",
       "0                L-1           L          1      l1   human   \n",
       "1                L-1           L          1      l1   human   \n",
       "2                L-1           L          1      l1   human   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "2  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "\n",
       "   unique_annotation  5p_cover align_3p_nt codon anticodon amino_acid  \\\n",
       "0               True     False           A   AAA       TTT   eColiLys   \n",
       "1               True     False           A   AAA       TTT   eColiLys   \n",
       "2               True     False           A   AAA       TTT   eColiLys   \n",
       "\n",
       "   align_gap  fmax_score>0.9  count  UMIcount  UMI_percent_exp  \n",
       "0      False           False    337       337       100.032047  \n",
       "1      False            True   4270      4270       100.407676  \n",
       "2       True           False    317       317       100.030139  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, \\\n",
    "                             overwrite_dir=False)\n",
    "stats_df = stats_obj.run_parallel(n_jobs=6, load_previous=True)\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, sample_df, overwrite_dir=False, pull_default=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ecoli control:\n",
    "plot_obj.plot_Ecoli_ctr(min_obs=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample/group:  143B-ETCinhib  Atpenin-rescue  CytB-test  FCCP-test  Fix-test  H1299-ETCinhib  H1299-rotenone  Liver-tissue  Muscle-tissue  NaCl-Control  Pilot-exp  Tumor-tissue  UCPH-titration\n",
      "Now plotting sample/group:  143B-ETCinhib  Atpenin-rescue  CytB-test  FCCP-test  Fix-test  H1299-ETCinhib  H1299-rotenone  Liver-tissue  Muscle-tissue  NaCl-Control  Pilot-exp  Tumor-tissue  UCPH-titration"
     ]
    }
   ],
   "source": [
    "# Codon abundance barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_abundance_grp', \\\n",
    "                        group=True, min_obs=500)\n",
    "\n",
    "# Codon charge barchart:\n",
    "plot_obj.plot_abundance(plot_type='codon', plot_name='codon_charge_grp', \\\n",
    "                        group=True, min_obs=500, charge_plot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  L-1  L-2  L-1+2  A-2  A-1+2  Rich-1  Rich-2  Rich-1+2  A-NaCl  L-NaCl  Rich-NaCl  A-1  U-1  Fix2  0U2  U-NaCl  2U2  FT  U-2  Fix3  Fix1  U-1+2  2U1  0U1  HAV2  HAV1  8U1  HVV1  4U1  HVV2  4U2  8U2  HVS1  HVP2  HVS2  HVP1  HAS1  HCV1  HAS2  HCV2  HCS1  HR30S1  HR30P1  HR30V2  HR30P2  HCS2  HR30V1  HR30S2  BVV1  BVV2  BVS1  BVS2  BVP1  HR20V1  HR40V1  HR20S1  BVP2  HR20S2  HR20V2  HR20P1  HR40P1  HR20P2  HR40S2  HR40S1  HR80P2  HR40V2  HR80S1  HR40P2  HR80V2  HR80V1  HR80P1  HR80S2  Tu2  Tu1  Mu2  Tu3  Li3  Tu4  Li1  Li2  Li4  Mu1  Mu4  Mu3  BAV1  BAV2  BVR1  BVR2  BAR2  BAR1  CyP  CyA\n",
      "Now plotting sample:  L-1  L-2  L-1+2  L-NaCl  A-1  A-2  A-1+2  A-NaCl  Rich-1  Rich-2  Rich-1+2  Rich-NaCl  U-1  U-2  U-1+2  U-NaCl  Fix1  Fix2  Fix3  FT  0U1  0U2  2U1  2U2  4U1  4U2  8U1  8U2  HVV1  HVV2  HVS1  HVS2  HVP1  HVP2  HAV1  HAV2  HAS1  HAS2  HCV1  HCV2  HCS1  HCS2  HR30V1  HR30V2  HR30S1  HR30S2  HR30P1  HR30P2  BVV1  BVV2  BVS1  BVS2  BVP1  BVP2  HR20V1  HR20V2  HR20S1  HR20S2  HR20P1  HR20P2  HR40V1  HR40V2  HR40S1  HR40S2  HR40P1  HR40P2  HR80V1  HR80V2  HR80S1  HR80S2  HR80P1  HR80P2  Tu1  Tu2  Tu3  Tu4  Li1  Li2  Li3  Li4  Mu1  Mu2  Mu3  Mu4  BAV1  BAV2  BVR1  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now collecting data for sample:  L-1  L-1+2  L-2  L-NaCl  A-1  A-2  A-NaCl  A-1+2  Rich-1  Rich-2  Rich-1+2  Rich-NaCl  FT  Fix1  0U1  Fix2  U-1  U-2  2U2  2U1  Fix3  U-NaCl  0U2  U-1+2  8U1  HVS2  HAV1  HVS1  4U1  8U2  4U2  HVV1  HVV2  HAS1  HVP1  HVP2  HAV2  HAS2  HCV2  HCV1  HCS1  HCS2  HR30P1  HR30S1  HR30V2  HR30S2  HR30V1  HR30P2  BVV1  BVV2  BVS1  BVP1  BVS2  HR40V1  HR20V1  BVP2  HR40S2  HR20V2  HR20P2  HR20S1  HR20P1  HR20S2  HR40S1  HR40P1  Tu1  HR40V2  HR80V2  HR80V1  HR40P2  Li4  HR80P2  HR80S2  HR80P1  HR80S1  Tu3  Tu2  Tu4  Li1  Li2  Li3  BAV2  Mu1  Mu3  Mu2  Mu4  BAV1  BVR1  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now plotting sample:  L-1  L-2  L-1+2  L-NaCl  A-1  A-2  A-1+2  A-NaCl  Rich-1  Rich-2  Rich-1+2  Rich-NaCl  U-1  U-2  U-1+2  U-NaCl  Fix1  Fix2  Fix3  FT  0U1  0U2  2U1  2U2  4U1  4U2  8U1  8U2  HVV1  HVV2  HVS1  HVS2  HVP1  HVP2  HAV1  HAV2  HAS1  HAS2  HCV1  HCV2  HCS1  HCS2  HR30V1  HR30V2  HR30S1  HR30S2  HR30P1  HR30P2  BVV1  BVV2  BVS1  BVS2  BVP1  BVP2  HR20V1  HR20V2  HR20S1  HR20S2  HR20P1  HR20P2  HR40V1  HR40V2  HR40S1  HR40S2  HR40P1  HR40P2  HR80V1  HR80V2  HR80S1  HR80S2  HR80P1  HR80P2  Tu1  Tu2  Tu3  Tu4  Li1  Li2  Li3  Li4  Mu1  Mu2  Mu3  Mu4  BAV1  BAV2  BVR1  BVR2  BAR1  BAR2  CyP  CyA"
     ]
    }
   ],
   "source": [
    "# Coverage plots for cyto/mito transcripts:\n",
    "plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_cyto_behrens_norm', n_jobs=12)\n",
    "\n",
    "plot_obj.plot_coverage(compartment='mito', plot_type='behrens', y_norm=True, \\\n",
    "                       plot_name='cov_plot_mito_behrens_norm', n_jobs=12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  L-1  L-2  L-1+2  L-NaCl  A-NaCl  A-1  A-2  A-1+2  Rich-NaCl  Rich-1  Rich-2  Rich-1+2  U-1  U-1+2  U-2  U-NaCl  Fix1  Fix2  Fix3  FT  0U1  0U2  2U2  2U1  4U1  8U2  4U2  8U1  HVV1  HVS1  HVS2  HVV2  HVP1  HVP2  HAV1  HAS1  HAV2  HAS2  HCV1  HCV2  HCS1  HCS2  HR30V1  HR30S1  HR30V2  HR30S2  HR30P2  HR30P1  BVV1  BVP1  BVS2  BVV2  BVS1  BVP2  HR20V1  HR20S1  HR20S2  HR20V2  HR20P1  HR40S1  HR20P2  HR40V1  HR40V2  HR40S2  HR40P1  HR40P2  HR80V1  HR80P1  HR80V2  HR80S1  HR80S2  Tu2  HR80P2  Tu1  Tu3  Tu4  Li1  Li2  Li3  Li4  Mu1  Mu2  BAV1  Mu3  Mu4  BVR1  BAV2  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# UMI logo:\n",
    "plot_obj.plot_UMI_logo(n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now collecting data for sample:  L-1  L-2  L-1+2  L-NaCl  A-1  A-2  A-NaCl  A-1+2  Rich-1  Rich-NaCl  Rich-2  Rich-1+2  U-1  U-2  U-1+2  U-NaCl  Fix1  Fix2  FT  Fix3  0U2  0U1  2U1  2U2  4U2  8U1  4U1  8U2  HVV1  HVV2  HVS1  HVS2  HVP1  HAV1  HVP2  HAV2  HAS1  HAS2  HCV1  HCV2  HCS1  HR30V1  HCS2  HR30V2  HR30S1  HR30S2  HR30P1  BVV1  HR30P2  BVS1  BVS2  BVV2  BVP1  BVP2  HR20V1  HR20S1  HR20V2  HR20S2  HR20P1  HR40S1  HR20P2  HR40V1  HR40V2  HR40S2  HR40P1  HR80V1  HR40P2  HR80S2  HR80S1  HR80V2  HR80P1  Tu1  HR80P2  Tu2  Tu3  Li1  Tu4  Li2  Li3  Mu1  Li4  Mu2  Mu4  Mu3  BAV1  BAV2  BVR1  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  L-1  L-2  L-NaCl  L-1+2  A-1  A-2  A-NaCl  A-1+2  Rich-1  Rich-NaCl  Rich-2  Rich-1+2  U-1  U-1+2  U-2  U-NaCl  Fix1  Fix2  FT  Fix3  0U1  0U2  2U1  2U2  4U1  8U1  4U2  8U2  HVV1  HVV2  HVS1  HVS2  HVP1  HAV1  HVP2  HAV2  HAS1  HAS2  HCV1  HCV2  HCS1  HCS2  HR30V1  HR30V2  HR30S1  HR30S2  HR30P1  HR30P2  BVV1  BVP1  BVS1  BVS2  BVV2  BVP2  HR20V2  HR20V1  HR20S1  HR20S2  HR20P2  HR20P1  HR40V1  HR40S2  HR40V2  HR40S1  HR40P1  HR40P2  HR80V1  HR80P1  HR80V2  HR80S1  HR80S2  HR80P2  Tu2  Tu1  Tu3  Tu4  Li1  Li2  Li3  Li4  Mu1  Mu2  BAV1  Mu3  Mu4  BAV2  BVR1  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now plotting logo plot.\n",
      "Now collecting data for sample:  L-1  L-2  L-1+2  L-NaCl  A-1  A-2  A-NaCl  A-1+2  Rich-2  Rich-1+2  Rich-1  Rich-NaCl  U-1  U-1+2  U-2  U-NaCl  Fix1  Fix2  FT  Fix3  2U1  0U2  0U1  2U2  8U1  4U1  4U2  8U2  HVV1  HVV2  HVS1  HVS2  HVP1  HVP2  HAV2  HAV1  HAS1  HCV2  HAS2  HCV1  HCS1  HCS2  HR30V1  HR30V2  HR30S1  HR30S2  HR30P1  HR30P2  BVV1  BVS1  BVV2  BVP1  BVS2  BVP2  HR20V2  HR20S1  HR20V1  HR20S2  HR20P2  HR20P1  HR40V1  HR40S2  HR40V2  HR40S1  HR40P1  HR40P2  HR80V1  HR80P1  HR80V2  HR80S1  HR80S2  Tu2  HR80P2  Tu1  Tu3  Li1  Tu4  Li2  Li3  Li4  Mu1  Mu2  BAV1  Mu3  Mu4  BAV2  BVR1  BVR2  BAR1  BAR2  CyP  CyA\n",
      "Now plotting logo plot."
     ]
    }
   ],
   "source": [
    "# Non-template nucleotides:\n",
    "plot_obj.plot_non_temp(end='5p', plot_name='_5p-non-template_logo', \\\n",
    "                       seq_len_percentile=99, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_logo', \\\n",
    "                       seq_len_percentile=99.9, n_jobs=4)\n",
    "\n",
    "plot_obj.plot_non_temp(end='3p', plot_name='_3p-non-template_3p-cover_logo', \\\n",
    "                       seq_len_percentile=99.9, _3p_cover=True, n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Folder exists and overwrite set to false... Doing nothing.\n"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=False, \\\n",
    "                     overwrite_dir=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find mutations and save them:\n",
    "if False:\n",
    "    TM_obj.find_muts(n_jobs=8, unique_anno=True)\n",
    "    TM_obj.pickle_muts_write(pickle_name='saved_muts_unique-anno.pickle')\n",
    "else:\n",
    "    TM_obj.pickle_muts_read(pickle_name='saved_muts_unique-anno.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap/RT stops for Atpenin rescue:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='143B-Atp-Rot-rescue_mut_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='mut', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['BVP', 'BVP', 'BVP', 'BAV', 'BAV', 'BVR'], \\\n",
    "                                                 ['BAV', 'BVR', 'BAR', 'BVR', 'BAR', 'BAR']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='143B-Atp-Rot-rescue_gap_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='gap', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['BVP', 'BVP', 'BVP', 'BAV', 'BAV', 'BVR'], \\\n",
    "                                                 ['BAV', 'BVR', 'BAR', 'BVR', 'BAR', 'BAR']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='143B-Atp-Rot-rescue_RTstops_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='RTstops', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['BVP', 'BVP', 'BVP', 'BAV', 'BAV', 'BVR'], \\\n",
    "                                                 ['BAV', 'BVR', 'BAR', 'BVR', 'BAR', 'BAR']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap/RT stops for tRNA pilot experiment:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='pilot-exp_mut_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='mut', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['Rich', 'Rich', 'Rich'], \\\n",
    "                                                 ['L',    'A',    'U']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='pilot-exp_gap_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='gap', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['Rich', 'Rich', 'Rich'], \\\n",
    "                                                 ['L',    'A',    'U']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='pilot-exp_RTstops_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='RTstops', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['Rich', 'Rich', 'Rich'], \\\n",
    "                                                 ['L',    'A',    'U']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap/RT stops for UCPH titration:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='UCPH_mut_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='mut', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['0U', '0U', '0U'], \\\n",
    "                                                 ['2U', '4U', '8U']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='UCPH_gap_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='gap', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['0U', '0U', '0U'], \\\n",
    "                                                 ['2U', '4U', '8U']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='UCPH_RTstops_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='RTstops', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['0U', '0U', '0U'], \\\n",
    "                                                 ['2U', '4U', '8U']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap/RT stops for H1299 rotenone titration:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='H1299-rot-tit_mut_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='mut', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['HVV',   'HVV',   'HVV',   'HVV',   'HVV'], \\\n",
    "                                                 ['HR20V', 'HR30V', 'HR40V', 'HR80V', 'HR20P']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='H1299-rot-tit_gap_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='gap', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['HVV',   'HVV',   'HVV',   'HVV',   'HVV'], \\\n",
    "                                                 ['HR20V', 'HR30V', 'HR40V', 'HR80V', 'HR20P']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='H1299-rot-tit_RTstops_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='RTstops', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['HVV',   'HVV',   'HVV',   'HVV',   'HVV'], \\\n",
    "                                                 ['HR20V', 'HR30V', 'HR40V', 'HR80V', 'HR20P']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mutation/gap/RT stops for H1299 rotenone, vehicle vs. pyruvate:\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='H1299-rot-pyr_mut_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='mut', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['HVV', 'HR20V', 'HR30V', 'HR40V', 'HR80V'], \\\n",
    "                                                 ['HVP', 'HR20P', 'HR30P', 'HR40P', 'HR80P']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='H1299-rot-pyr_gap_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='gap', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['HVV', 'HR20V', 'HR30V', 'HR40V', 'HR80V'], \\\n",
    "                                                 ['HVP', 'HR20P', 'HR30P', 'HR40P', 'HR80P']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')\n",
    "\n",
    "TM_obj.plot_transcript_mut_compare(species='human', \\\n",
    "                                   plot_name='H1299-rot-pyr_RTstops_matrix_top20-max-diff', \\\n",
    "                                   no_plot_return=True, \\\n",
    "                                   mito=False,\n",
    "                                   data_type='RTstops', \\\n",
    "                                   min_count_show=400, \\\n",
    "                                   sample_pairs=[['HVV', 'HR20V', 'HR30V', 'HR40V', 'HR80V'], \\\n",
    "                                                 ['HVP', 'HR20P', 'HR30P', 'HR40P', 'HR80P']], \\\n",
    "                                   sample_pairs_col='sample_name', \\\n",
    "                                   freq_avg_weighted=False, \\\n",
    "                                   topN=20, topN_select='max_diff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
