{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, shutil, bz2, random, resource, warnings, subprocess, copy, re, glob, json\n",
    "from pathlib import Path\n",
    "import xml.etree.ElementTree as ET\n",
    "from subprocess import Popen, PIPE, STDOUT\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "from Bio import SeqIO, bgzf\n",
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "from Bio import Seq, SeqIO, SearchIO, SeqRecord\n",
    "from mpire import WorkerPool\n",
    "import jellyfish\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen\n",
      "Repo is in: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq\n",
      "Using minimum read length: 49 (after merge)\n",
      "Using minimum alignemnt score: 20\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from utils.read_processing import index_to_sample_df, downsample_raw_input, AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from utils.alignment import SWIPE_align\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_folder = 'data'\n",
    "seq_folder = 'raw_fastq'\n",
    "seq_folder_noDS = seq_folder # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/utils/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/utils/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 20\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# The minimum alignment score.\n",
    "# The alignment score is calculated as 1 per match,\n",
    "# -2 per mismatch, -6 per gap opening and -1 per gap extension.\n",
    "MIN_SCORE_ALIGN = 20\n",
    "print('Using minimum alignemnt score: {}'.format(MIN_SCORE_ALIGN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list_fnam = 'sample_list_P1-2-3.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if True:\n",
    "    sample_df, inp_file_df, seq_folder = downsample_raw_input(sample_df, inp_file_df, NBdir, data_folder, seq_folder_noDS, downsample_absolute=1e4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(inp_file_df, NBdir, data_folder, seq_folder, AdapterRemoval_dir, MIN_READ_LEN)\n",
    "AdapterRemoval_dir_abs = AR_obj.make_dir(overwrite=True)\n",
    "# inp_file_df = AR_obj.run_serial()\n",
    "inp_file_df = AR_obj.run_parallel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(sample_df, inp_file_df, NBdir, data_folder, AdapterRemoval_dir_abs, BC_dir)\n",
    "BC_dir_abs = BCsplit_obj.make_dir(overwrite=True)\n",
    "# sample_df, inp_file_df = BCsplit_obj.run_serial()\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Perform Kmer analysis on unmapped reads ###\n",
    "if False: # barcode analysis is typically suficient \n",
    "    kmer_obj = Kmer_analysis(inp_file_df, index_df, BC_dir_abs)\n",
    "    # Add a filter to avoid Kmers from the end of tRNA sequences:\n",
    "    kmer_obj.filter_3p_fasta(tRNA_database['human'])\n",
    "    # Add to this filter the constant region of the adapters:\n",
    "    kmer_obj.filter_window_BC(filter_window=(0, 11))\n",
    "    # Search for Kmers:\n",
    "    all_kmer = kmer_obj.search_unmapped(search_size=13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>l5Sp</td>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>l12Sp</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>l4Sp</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>l6Sp</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>l10Sp</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>l9Sp</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>l3Sp</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>l8Sp</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>l1Sp</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>l2Sp</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>l7Sp</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>l11Sp</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name  Count\n",
       "0    l5Sp    163\n",
       "1   l12Sp    108\n",
       "2    l4Sp     65\n",
       "3    l6Sp     49\n",
       "4   l10Sp     48\n",
       "5    l9Sp     48\n",
       "6    l3Sp     42\n",
       "7    l8Sp     39\n",
       "8    l1Sp     37\n",
       "9    l2Sp     31\n",
       "10   l7Sp     22\n",
       "11  l11Sp      3"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Perform barcode analysis on unmapped reads ###\n",
    "# Recall that adapters look like this:\n",
    "# GGCTGCCATGC    GACTA\n",
    "# GGCTGCCATGCA   AGTGC\n",
    "# GGCTGCCATGCTG  TCACG\n",
    "# GGCTGCCATGCAAC CTGAT\n",
    "# With the barcode as the rightmost 5 nt.\n",
    "# Search these 5 nt. barcodes by specifying BC_size_3p=5\n",
    "bc_analysis_obj = BC_analysis(inp_file_df, index_df, BC_dir_abs, BC_size_3p=5)\n",
    "# Search for barcodes in the unmapped reads.\n",
    "# For the summary output, filter by a max distance\n",
    "# to any barcode of 1 (group_dist=1), then group by barcode name:\n",
    "bc_analysis_df = bc_analysis_obj.search_unmapped(group_dist=1)\n",
    "bc_analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>P1_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P1_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>942</td>\n",
       "      <td>38</td>\n",
       "      <td>901</td>\n",
       "      <td>939</td>\n",
       "      <td>99.681529</td>\n",
       "      <td>95.953142</td>\n",
       "      <td>929</td>\n",
       "      <td>865</td>\n",
       "      <td>928.178310</td>\n",
       "      <td>98.619958</td>\n",
       "      <td>93.193300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>P2_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P2_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>673</td>\n",
       "      <td>14</td>\n",
       "      <td>648</td>\n",
       "      <td>662</td>\n",
       "      <td>98.365527</td>\n",
       "      <td>97.885196</td>\n",
       "      <td>663</td>\n",
       "      <td>621</td>\n",
       "      <td>662.581602</td>\n",
       "      <td>98.514116</td>\n",
       "      <td>93.724305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>P3_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P3_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>709</td>\n",
       "      <td>35</td>\n",
       "      <td>666</td>\n",
       "      <td>701</td>\n",
       "      <td>98.871650</td>\n",
       "      <td>95.007133</td>\n",
       "      <td>692</td>\n",
       "      <td>631</td>\n",
       "      <td>691.544180</td>\n",
       "      <td>97.602257</td>\n",
       "      <td>91.245074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate P5_index P7_index barcode species  \\\n",
       "0              100p1        100p          1     D501     D701    l1Sp   human   \n",
       "1              100p2        100p          2     D501     D702    l2Sp   human   \n",
       "2              100p3        100p          3     D501     D703    l3Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq     fastq_mate1_filename  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  P1_R1_DSA-10k.fastq.bz2   \n",
       "1     TCCGGAGA   GGCTGCCATGCAAGTGC  P2_R1_DSA-10k.fastq.bz2   \n",
       "2     CGCTCATT  GGCTGCCATGCTGTCACG  P3_R1_DSA-10k.fastq.bz2   \n",
       "\n",
       "      fastq_mate2_filename  N_total  N_CC  N_CCA  N_CCA+CC  \\\n",
       "0  P1_R2_DSA-10k.fastq.bz2      942    38    901       939   \n",
       "1  P2_R2_DSA-10k.fastq.bz2      673    14    648       662   \n",
       "2  P3_R2_DSA-10k.fastq.bz2      709    35    666       701   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.681529    95.953142           929             865   \n",
       "1             98.365527    97.885196           663             621   \n",
       "2             98.871650    95.007133           692             631   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \n",
       "0      928.178310                    98.619958               93.193300  \n",
       "1      662.581602                    98.514116               93.724305  \n",
       "2      691.544180                    97.602257               91.245074  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "# Note, the cDNA input amount is so large that it is very unlikely to sequence\n",
    "# the same PCR amplified DNA twice. Therefore, this processing step does not\n",
    "# attempt to merge possible UMI duplicates.\n",
    "UMItrim_obj = UMI_trim(sample_df, NBdir, data_folder, BC_dir_abs, UMI_dir)\n",
    "UMI_dir_abs = UMItrim_obj.make_dir(overwrite=True)\n",
    "# sample_df = UMItrim_obj.run_serial()\n",
    "sample_df = UMItrim_obj.run_parallel()\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Swipe on:  100p2  100p1  100p3  100p4  85p4  85p3  85p2  70p2  85p1  70p1  70p3  55p1  55p2  55p4  70p4  40p1  40p2  55p3  40p4  25p1  25p3  25p2  10p1  10p3  40p3  0p1  25p4  0p4  10p2  10p4  0p3  0p2  0p_90m3  0p_90m1  0p_90m2  0p_90m4\n",
      "Collecting alignment statistics, from sample:  100p1  100p2  100p3  100p4  85p1  85p2  85p3  85p4  70p1  70p2  70p3  70p4  55p1  55p2  55p3  55p4  40p1  40p2  40p3  40p4  25p1  25p2  25p3  25p4  10p1  10p2  10p3  10p4  0p1  0p2  0p3  0p4  0p_90m1  0p_90m2  0p_90m3  0p_90m4"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>P1_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P1_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>942</td>\n",
       "      <td>38</td>\n",
       "      <td>901</td>\n",
       "      <td>939</td>\n",
       "      <td>99.681529</td>\n",
       "      <td>95.953142</td>\n",
       "      <td>929</td>\n",
       "      <td>865</td>\n",
       "      <td>928.178310</td>\n",
       "      <td>98.619958</td>\n",
       "      <td>93.193300</td>\n",
       "      <td>918</td>\n",
       "      <td>74.509804</td>\n",
       "      <td>25.490196</td>\n",
       "      <td>98.815931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>P2_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P2_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>673</td>\n",
       "      <td>14</td>\n",
       "      <td>648</td>\n",
       "      <td>662</td>\n",
       "      <td>98.365527</td>\n",
       "      <td>97.885196</td>\n",
       "      <td>663</td>\n",
       "      <td>621</td>\n",
       "      <td>662.581602</td>\n",
       "      <td>98.514116</td>\n",
       "      <td>93.724305</td>\n",
       "      <td>652</td>\n",
       "      <td>74.079755</td>\n",
       "      <td>25.920245</td>\n",
       "      <td>98.340875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>P3_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P3_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>709</td>\n",
       "      <td>35</td>\n",
       "      <td>666</td>\n",
       "      <td>701</td>\n",
       "      <td>98.871650</td>\n",
       "      <td>95.007133</td>\n",
       "      <td>692</td>\n",
       "      <td>631</td>\n",
       "      <td>691.544180</td>\n",
       "      <td>97.602257</td>\n",
       "      <td>91.245074</td>\n",
       "      <td>673</td>\n",
       "      <td>72.956909</td>\n",
       "      <td>27.043091</td>\n",
       "      <td>97.254335</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate P5_index P7_index barcode species  \\\n",
       "0              100p1        100p          1     D501     D701    l1Sp   human   \n",
       "1              100p2        100p          2     D501     D702    l2Sp   human   \n",
       "2              100p3        100p          3     D501     D703    l3Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq     fastq_mate1_filename  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  P1_R1_DSA-10k.fastq.bz2   \n",
       "1     TCCGGAGA   GGCTGCCATGCAAGTGC  P2_R1_DSA-10k.fastq.bz2   \n",
       "2     CGCTCATT  GGCTGCCATGCTGTCACG  P3_R1_DSA-10k.fastq.bz2   \n",
       "\n",
       "      fastq_mate2_filename  N_total  N_CC  N_CCA  N_CCA+CC  \\\n",
       "0  P1_R2_DSA-10k.fastq.bz2      942    38    901       939   \n",
       "1  P2_R2_DSA-10k.fastq.bz2      673    14    648       662   \n",
       "2  P3_R2_DSA-10k.fastq.bz2      709    35    666       701   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.681529    95.953142           929             865   \n",
       "1             98.365527    97.885196           663             621   \n",
       "2             98.871650    95.007133           692             631   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0      928.178310                    98.619958               93.193300   \n",
       "1      662.581602                    98.514116               93.724305   \n",
       "2      691.544180                    97.602257               91.245074   \n",
       "\n",
       "   N_mapped  percent_single_annotation  percent_multiple_annotation  \\\n",
       "0       918                  74.509804                    25.490196   \n",
       "1       652                  74.079755                    25.920245   \n",
       "2       673                  72.956909                    27.043091   \n",
       "\n",
       "   Mapping_percent  \n",
       "0        98.815931  \n",
       "1        98.340875  \n",
       "2        97.254335  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(tRNA_database, sample_df, NBdir, data_folder, UMI_dir_abs, align_dir, SWIPE_score_mat, gap_penalty=6, extension_penalty=1, min_score_align=20)\n",
    "align_dir_abs = align_obj.make_dir(overwrite=True)\n",
    "# sample_df = align_obj.run_serial(dry_run=False, verbose=False, overwrite=False)\n",
    "sample_df = align_obj.run_parallel()\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
