{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys, shutil, bz2\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 50)\n",
    "import numpy as np\n",
    "\n",
    "### Plotting imports ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib as mpl\n",
    "from matplotlib.patches import StepPatch\n",
    "import matplotlib.ticker as ticker\n",
    "import matplotlib.gridspec as gridspec\n",
    "import logomaker as lm\n",
    "palette = list(mcolors.TABLEAU_COLORS.keys())\n",
    "sns.set_theme(style=\"ticks\", palette=\"muted\")\n",
    "sns.set_context(\"talk\")\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notebook is in: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen\n",
      "Repo is in: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq\n",
      "Using minimum read length: 49 (after merge)\n",
      "Using minimum alignemnt score: 20\n"
     ]
    }
   ],
   "source": [
    "# Navigate back to NBdir in case of re-running a code block:\n",
    "if not 'NBdir' in globals():\n",
    "    NBdir = os.getcwd()\n",
    "print('Notebook is in: {}'.format(NBdir))\n",
    "os.chdir(NBdir)  # If you changed the current working dir, this will take you back to the notebook dir.\n",
    "\n",
    "# Define the path to the repo folder.\n",
    "# Change if necessary.\n",
    "homedir = '/'.join(NBdir.split('/')[0:-2])\n",
    "print('Repo is in: {}'.format(homedir))\n",
    "sys.path.insert(1, homedir)\n",
    "from src.misc import index_to_sample_df, downsample_raw_input, read_tRNAdb_info, sample_df_to_dict\n",
    "from src.read_processing import AR_merge, BC_split, Kmer_analysis, BC_analysis, UMI_trim\n",
    "from src.alignment import SWIPE_align\n",
    "from src.stats_collection import STATS_collection\n",
    "from src.plotting import TRNA_plot\n",
    "from src.transcript_mutations import TM_analysis\n",
    "\n",
    "# These are default folder names for data and raw fastq files\n",
    "# relative to the folder in which this notebook is in:\n",
    "data_dir = 'data'\n",
    "seq_dir = 'raw_fastq'\n",
    "seq_dir_noDS = seq_dir # Not downsampled\n",
    "\n",
    "# These folder names are used in subsequent processing steps\n",
    "# to dump data. Best to not change:\n",
    "AdapterRemoval_dir = 'AdapterRemoval'\n",
    "BC_dir = 'BC_split'\n",
    "UMI_dir = 'UMI_trimmed'\n",
    "align_dir = 'SWalign'\n",
    "stats_dir = 'stats_collection'\n",
    "TM_dir = 'transcript_mutations'\n",
    "plotting_dir = 'plotting'\n",
    "tRNA_database = dict()\n",
    "tRNA_database['human'] = '{}/tRNA_database/human/hg38-tRNAs.fa'.format(homedir)\n",
    "tRNA_database['mouse'] = '{}/tRNA_database/mouse/mm10-tRNAs.fa'.format(homedir)\n",
    "# Read information (length, codon etc) of tRNAs into dictionary:\n",
    "tRNA_data = read_tRNAdb_info(tRNA_database)\n",
    "SWIPE_score_mat = '{}/utils/nuc_score-matrix.txt'.format(homedir)\n",
    "SWIPE_score_mat2 = '{}/utils/nuc_score-matrix_2.txt'.format(homedir) # For masked reference sequences\n",
    "# tRNA sequencing yields many duplicated reads.\n",
    "# Adding these commonly seen sequences to a list prevents duplicated alignment:\n",
    "common_seqs = '{}/utils/common-seqs.fasta.bz2'.format(homedir)\n",
    "# ^^ That one is too big, take the top 1k instead:\n",
    "common_seqs = '{}/utils/common-seqs_1k.fasta.bz2'.format(homedir)\n",
    "\n",
    "# Define minimum read length based on minimum insert size:\n",
    "MIN_INSERT_LEN = 20\n",
    "UMI_LEN = 10\n",
    "BC_MAX_LEN = 19\n",
    "MIN_READ_LEN = MIN_INSERT_LEN + UMI_LEN + BC_MAX_LEN\n",
    "print('Using minimum read length: {} (after merge)'.format(MIN_READ_LEN))\n",
    "\n",
    "# The minimum alignment score.\n",
    "# The alignment score is calculated as 1 per match,\n",
    "# -2 per mismatch, -6 per gap opening and -1 per gap extension.\n",
    "# Better to set relatively low, since additional filtering can\n",
    "# be applied later.\n",
    "MIN_SCORE_ALIGN = 20\n",
    "print('Using minimum alignemnt score: {}'.format(MIN_SCORE_ALIGN))\n",
    "\n",
    "# Read index information:\n",
    "index_list_fnam = 'index_list.xlsx'\n",
    "index_df = pd.read_excel('{}/utils/{}'.format(homedir, index_list_fnam))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list_fnam = 'sample_list_P1-2-3.xlsx'\n",
    "sample_df = pd.read_excel('{}/{}'.format(NBdir, sample_list_fnam))\n",
    "# Add barcode sequences:\n",
    "sample_df = index_to_sample_df(sample_df, index_df)\n",
    "# Read elementary info (replicate, barcode, species)\n",
    "# for each unique sample name into a dictionary:\n",
    "sample_dict = sample_df_to_dict(sample_df)\n",
    "# Get filenames from the sample information:\n",
    "inp_file_df = sample_df[['fastq_mate1_filename', 'fastq_mate2_filename', 'P5_index', 'P7_index', 'P5_index_seq', 'P7_index_seq']].copy().drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "# Downsample:\n",
    "if False:\n",
    "    sample_df, inp_file_df, seq_dir = downsample_raw_input(sample_df, inp_file_df, NBdir, data_dir, seq_dir_noDS, downsample_absolute=5e5)\n",
    "\n",
    "# Make a dictionary with paths used for data processing:\n",
    "dir_dict = dict(NBdir = NBdir,\n",
    "                data_dir = data_dir,\n",
    "                seq_dir = seq_dir,\n",
    "                AdapterRemoval_dir = AdapterRemoval_dir,\n",
    "                BC_dir = BC_dir,\n",
    "                UMI_dir = UMI_dir,\n",
    "                align_dir = align_dir,\n",
    "                stats_dir = stats_dir,\n",
    "                TM_dir = TM_dir,\n",
    "                plotting_dir = plotting_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/AdapterRemoval\n"
     ]
    }
   ],
   "source": [
    "# Run AdapterRemoval:\n",
    "AR_obj = AR_merge(dir_dict, inp_file_df, MIN_READ_LEN)\n",
    "AdapterRemoval_dir_abs = AR_obj.make_dir(overwrite=False)\n",
    "# inp_file_df = AR_obj.run_serial()\n",
    "inp_file_df = AR_obj.run_parallel(n_jobs=4, overwrite=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/BC_split\n",
      "Loaded results from previous run... Not running barcode split.\n"
     ]
    }
   ],
   "source": [
    "# Split files based on barcodes:\n",
    "BCsplit_obj = BC_split(dir_dict, sample_df, inp_file_df)\n",
    "BC_dir_abs = BCsplit_obj.make_dir(overwrite=False)\n",
    "# sample_df, inp_file_df = BCsplit_obj.run_serial()\n",
    "sample_df, inp_file_df = BCsplit_obj.run_parallel(n_jobs=4, load_previous=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Perform Kmer analysis on unmapped reads ###\n",
    "if False: # barcode analysis is typically suficient \n",
    "    kmer_obj = Kmer_analysis(dir_dict, inp_file_df, index_df)\n",
    "    # Add a filter to avoid Kmers from the end of tRNA sequences:\n",
    "    kmer_obj.filter_3p_fasta(tRNA_database['human'])\n",
    "    # Add to this filter the constant region of the adapters:\n",
    "    kmer_obj.filter_window_BC(filter_window=(0, 11))\n",
    "    # Search for Kmers:\n",
    "    all_kmer = kmer_obj.search_unmapped(search_size=13)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/BC_split/BC_analysis\n",
      "Loaded results from previous run... Not running barcode analysis.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>l5Sp</td>\n",
       "      <td>26700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>l4Sp</td>\n",
       "      <td>10041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>l10Sp</td>\n",
       "      <td>6808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>l6Sp</td>\n",
       "      <td>6016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>l1Sp</td>\n",
       "      <td>5905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>l3Sp</td>\n",
       "      <td>5826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>l8Sp</td>\n",
       "      <td>5806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>l12Sp</td>\n",
       "      <td>5783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>l9Sp</td>\n",
       "      <td>5434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>l2Sp</td>\n",
       "      <td>4797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>l7Sp</td>\n",
       "      <td>2362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>l11Sp</td>\n",
       "      <td>1009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name  Count\n",
       "0    l5Sp  26700\n",
       "1    l4Sp  10041\n",
       "2   l10Sp   6808\n",
       "3    l6Sp   6016\n",
       "4    l1Sp   5905\n",
       "5    l3Sp   5826\n",
       "6    l8Sp   5806\n",
       "7   l12Sp   5783\n",
       "8    l9Sp   5434\n",
       "9    l2Sp   4797\n",
       "10   l7Sp   2362\n",
       "11  l11Sp   1009"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Perform barcode analysis on unmapped reads ###\n",
    "# Recall that adapters look like this:\n",
    "# GGCTGCCATGC    GACTA\n",
    "# GGCTGCCATGCA   AGTGC\n",
    "# GGCTGCCATGCTG  TCACG\n",
    "# GGCTGCCATGCAAC CTGAT\n",
    "# With the barcode as the rightmost 5 nt.\n",
    "# Search these 5 nt. barcodes by specifying BC_size_3p=5\n",
    "bc_analysis_obj = BC_analysis(dir_dict, inp_file_df, index_df, BC_size_3p=5, overwrite=False)\n",
    "# Search for barcodes in the unmapped reads.\n",
    "# For the summary output, filter by a max distance\n",
    "# to any barcode of 1 (group_dist=1), then group by barcode name:\n",
    "bc_analysis_df = bc_analysis_obj.search_unmapped(group_dist=1, load_previous=True)\n",
    "bc_analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using existing folder because overwrite set to false: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/UMI_trimmed\n",
      "Loaded results from previous run... Not running UMI trimming.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>P1_R1_DSA-500k.fastq.bz2</td>\n",
       "      <td>P1_R2_DSA-500k.fastq.bz2</td>\n",
       "      <td>46652</td>\n",
       "      <td>2106</td>\n",
       "      <td>44255</td>\n",
       "      <td>46361</td>\n",
       "      <td>99.376233</td>\n",
       "      <td>95.457389</td>\n",
       "      <td>45900</td>\n",
       "      <td>37221</td>\n",
       "      <td>43948.202015</td>\n",
       "      <td>98.388065</td>\n",
       "      <td>84.692885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>P2_R1_DSA-500k.fastq.bz2</td>\n",
       "      <td>P2_R2_DSA-500k.fastq.bz2</td>\n",
       "      <td>35150</td>\n",
       "      <td>1294</td>\n",
       "      <td>33661</td>\n",
       "      <td>34955</td>\n",
       "      <td>99.445235</td>\n",
       "      <td>96.298098</td>\n",
       "      <td>34621</td>\n",
       "      <td>28011</td>\n",
       "      <td>33502.694907</td>\n",
       "      <td>98.495021</td>\n",
       "      <td>83.608200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>P3_R1_DSA-500k.fastq.bz2</td>\n",
       "      <td>P3_R2_DSA-500k.fastq.bz2</td>\n",
       "      <td>36506</td>\n",
       "      <td>1718</td>\n",
       "      <td>34501</td>\n",
       "      <td>36219</td>\n",
       "      <td>99.213828</td>\n",
       "      <td>95.256633</td>\n",
       "      <td>35923</td>\n",
       "      <td>29454</td>\n",
       "      <td>34719.984436</td>\n",
       "      <td>98.403002</td>\n",
       "      <td>84.832987</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate P5_index P7_index barcode species  \\\n",
       "0              100p1        100p          1     D501     D701    l1Sp   human   \n",
       "1              100p2        100p          2     D501     D702    l2Sp   human   \n",
       "2              100p3        100p          3     D501     D703    l3Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq      fastq_mate1_filename  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  P1_R1_DSA-500k.fastq.bz2   \n",
       "1     TCCGGAGA   GGCTGCCATGCAAGTGC  P2_R1_DSA-500k.fastq.bz2   \n",
       "2     CGCTCATT  GGCTGCCATGCTGTCACG  P3_R1_DSA-500k.fastq.bz2   \n",
       "\n",
       "       fastq_mate2_filename  N_total  N_CC  N_CCA  N_CCA+CC  \\\n",
       "0  P1_R2_DSA-500k.fastq.bz2    46652  2106  44255     46361   \n",
       "1  P2_R2_DSA-500k.fastq.bz2    35150  1294  33661     34955   \n",
       "2  P3_R2_DSA-500k.fastq.bz2    36506  1718  34501     36219   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.376233    95.457389         45900           37221   \n",
       "1             99.445235    96.298098         34621           28011   \n",
       "2             99.213828    95.256633         35923           29454   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \n",
       "0    43948.202015                    98.388065               84.692885  \n",
       "1    33502.694907                    98.495021               83.608200  \n",
       "2    34719.984436                    98.403002               84.832987  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Generate UMI stats and write final trimmed tRNA sequences ###\n",
    "# Note, the cDNA input amount is so large that it is very unlikely to sequence\n",
    "# the same PCR amplified DNA twice. Therefore, this processing step does not\n",
    "# attempt to merge possible UMI duplicates.\n",
    "UMItrim_obj = UMI_trim(dir_dict, sample_df)\n",
    "UMI_dir_abs = UMItrim_obj.make_dir(overwrite=False)\n",
    "# sample_df = UMItrim_obj.run_serial()\n",
    "sample_df = UMItrim_obj.run_parallel(n_jobs=4, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Using existing folder because overwrite set to false: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/SWalign\n",
      "Loaded results from previous run... Not running alignment.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>P1_R1_DSA-500k.fastq.bz2</td>\n",
       "      <td>P1_R2_DSA-500k.fastq.bz2</td>\n",
       "      <td>46652</td>\n",
       "      <td>2106</td>\n",
       "      <td>44255</td>\n",
       "      <td>46361</td>\n",
       "      <td>99.376233</td>\n",
       "      <td>95.457389</td>\n",
       "      <td>45900</td>\n",
       "      <td>37221</td>\n",
       "      <td>43948.202015</td>\n",
       "      <td>98.388065</td>\n",
       "      <td>84.692885</td>\n",
       "      <td>45479</td>\n",
       "      <td>73.724136</td>\n",
       "      <td>26.275864</td>\n",
       "      <td>99.082789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>P2_R1_DSA-500k.fastq.bz2</td>\n",
       "      <td>P2_R2_DSA-500k.fastq.bz2</td>\n",
       "      <td>35150</td>\n",
       "      <td>1294</td>\n",
       "      <td>33661</td>\n",
       "      <td>34955</td>\n",
       "      <td>99.445235</td>\n",
       "      <td>96.298098</td>\n",
       "      <td>34621</td>\n",
       "      <td>28011</td>\n",
       "      <td>33502.694907</td>\n",
       "      <td>98.495021</td>\n",
       "      <td>83.608200</td>\n",
       "      <td>34199</td>\n",
       "      <td>75.075295</td>\n",
       "      <td>24.924705</td>\n",
       "      <td>98.781087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>P3_R1_DSA-500k.fastq.bz2</td>\n",
       "      <td>P3_R2_DSA-500k.fastq.bz2</td>\n",
       "      <td>36506</td>\n",
       "      <td>1718</td>\n",
       "      <td>34501</td>\n",
       "      <td>36219</td>\n",
       "      <td>99.213828</td>\n",
       "      <td>95.256633</td>\n",
       "      <td>35923</td>\n",
       "      <td>29454</td>\n",
       "      <td>34719.984436</td>\n",
       "      <td>98.403002</td>\n",
       "      <td>84.832987</td>\n",
       "      <td>34788</td>\n",
       "      <td>72.424399</td>\n",
       "      <td>27.575601</td>\n",
       "      <td>96.840464</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate P5_index P7_index barcode species  \\\n",
       "0              100p1        100p          1     D501     D701    l1Sp   human   \n",
       "1              100p2        100p          2     D501     D702    l2Sp   human   \n",
       "2              100p3        100p          3     D501     D703    l3Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq      fastq_mate1_filename  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  P1_R1_DSA-500k.fastq.bz2   \n",
       "1     TCCGGAGA   GGCTGCCATGCAAGTGC  P2_R1_DSA-500k.fastq.bz2   \n",
       "2     CGCTCATT  GGCTGCCATGCTGTCACG  P3_R1_DSA-500k.fastq.bz2   \n",
       "\n",
       "       fastq_mate2_filename  N_total  N_CC  N_CCA  N_CCA+CC  \\\n",
       "0  P1_R2_DSA-500k.fastq.bz2    46652  2106  44255     46361   \n",
       "1  P2_R2_DSA-500k.fastq.bz2    35150  1294  33661     34955   \n",
       "2  P3_R2_DSA-500k.fastq.bz2    36506  1718  34501     36219   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.376233    95.457389         45900           37221   \n",
       "1             99.445235    96.298098         34621           28011   \n",
       "2             99.213828    95.256633         35923           29454   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0    43948.202015                    98.388065               84.692885   \n",
       "1    33502.694907                    98.495021               83.608200   \n",
       "2    34719.984436                    98.403002               84.832987   \n",
       "\n",
       "   N_mapped  percent_single_annotation  percent_multiple_annotation  \\\n",
       "0     45479                  73.724136                    26.275864   \n",
       "1     34199                  75.075295                    24.924705   \n",
       "2     34788                  72.424399                    27.575601   \n",
       "\n",
       "   Mapping_percent  \n",
       "0        99.082789  \n",
       "1        98.781087  \n",
       "2        96.840464  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database, sample_df, SWIPE_score_mat, gap_penalty=6, extension_penalty=1, min_score_align=20, common_seqs=common_seqs)\n",
    "align_dir_abs = align_obj.make_dir(overwrite=False)\n",
    "# sample_df = align_obj.run_serial(dry_run=False, verbose=False, overwrite=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=4, load_previous=True)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Using existing folder because overwrite set to false: /Users/krdav/Google Drive/MCB/Sullivan_lab/tRNA-charge-seq/projects/tRNAseq_third-gen/data/stats_collection\n",
      "Loaded results from previous run... Not running stats collection.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>barcode</th>\n",
       "      <th>tRNA_annotation</th>\n",
       "      <th>tRNA_annotation_len</th>\n",
       "      <th>unique_annotation</th>\n",
       "      <th>align_3p_nt</th>\n",
       "      <th>codon</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>amino_acid</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0p1</td>\n",
       "      <td>0p</td>\n",
       "      <td>1</td>\n",
       "      <td>l5Sp</td>\n",
       "      <td>Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...</td>\n",
       "      <td>76</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>AAA</td>\n",
       "      <td>TTT</td>\n",
       "      <td>eColiLys</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0p1</td>\n",
       "      <td>0p</td>\n",
       "      <td>1</td>\n",
       "      <td>l5Sp</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Ala-TGC</td>\n",
       "      <td>72</td>\n",
       "      <td>True</td>\n",
       "      <td>C</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TGC</td>\n",
       "      <td>Ala</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0p1</td>\n",
       "      <td>0p</td>\n",
       "      <td>1</td>\n",
       "      <td>l5Sp</td>\n",
       "      <td>Homo_sapiens_mito_tRNA-Arg-TCG</td>\n",
       "      <td>68</td>\n",
       "      <td>True</td>\n",
       "      <td>A</td>\n",
       "      <td>CGA</td>\n",
       "      <td>TCG</td>\n",
       "      <td>Arg</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate barcode  \\\n",
       "0                0p1          0p          1    l5Sp   \n",
       "1                0p1          0p          1    l5Sp   \n",
       "2                0p1          0p          1    l5Sp   \n",
       "\n",
       "                                     tRNA_annotation  tRNA_annotation_len  \\\n",
       "0  Escherichia_coli_str_K_12_substr_MG1655_tRNA-e...                   76   \n",
       "1                     Homo_sapiens_mito_tRNA-Ala-TGC                   72   \n",
       "2                     Homo_sapiens_mito_tRNA-Arg-TCG                   68   \n",
       "\n",
       "   unique_annotation align_3p_nt codon anticodon amino_acid  count  \n",
       "0               True           C   AAA       TTT   eColiLys     12  \n",
       "1               True           C   GCA       TGC        Ala     37  \n",
       "2               True           A   CGA       TCG        Arg      1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Collect alignment statistics ###\n",
    "stats_obj = STATS_collection(dir_dict, tRNA_data, sample_df, common_seqs=common_seqs)\n",
    "stats_dir_abs = stats_obj.make_dir(overwrite=False)\n",
    "# stats_df = stats_obj.run_serial()\n",
    "stats_df = stats_obj.run_parallel(n_jobs=4, load_previous=True)\n",
    "# The dataframe returned is the \"ALL_stats_aggregate_filtered.csv\"\n",
    "# which is the aggregated data filtered to contain only the\n",
    "# most relevant columnns and requiring the 3' must be covered\n",
    "# and have no 3' non-template bases.\n",
    "# The CSV file output \"ALL_stats_aggregate.csv\" is the data\n",
    "# aggregated based on all values identical except\n",
    "# readID, 5p_UMI and 3p_BC. I.e. every information, except\n",
    "# the UMI seequence is maintained in the aggregated CSV.\n",
    "stats_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Folder exists and overwrite set to false... Doing nothing.\n"
     ]
    }
   ],
   "source": [
    "### Generate standard tRNAseq data plots ###\n",
    "plot_obj = TRNA_plot(dir_dict, pull_default=True)\n",
    "plotting_dir_abs = plot_obj.make_dir(overwrite=False)\n",
    "plot_obj.get_charge_df()\n",
    "\n",
    "# Plot UMI nucleotide content logo per sample:\n",
    "#plot_obj.plot_UMI_logo()\n",
    "\n",
    "# Plot the 5p non-template nucleotide logo per sample,\n",
    "# using only those at, or below, the legth determined as a percentile:\n",
    "#plot_obj.plot_non_temp('5p', '_5p-non-template_logo', seq_len_percentile=99)\n",
    "\n",
    "# 3p non-template nucleotides are more rare, therefore 99.9 percentile:\n",
    "#plot_obj.plot_non_temp('3p', '_3p-non-template_logo', seq_len_percentile=99.9)\n",
    "\n",
    "# 3p non-template nucleotides, but require 3p coverage.\n",
    "# This is more likely to reflect either:\n",
    "# 1) Erronous CCA addition, or 2) additional bases in the adapter.\n",
    "#plot_obj.plot_non_temp('3p', '_3p-non-template_3p-cover_logo', seq_len_percentile=99.9, _3p_cover=True)\n",
    "\n",
    "# Plot the 5p to 3p coverage for each amino acid:\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='needle', aa_norm=False, plot_name='cov_plot_cyto_needle')\n",
    "#plot_obj.plot_coverage(compartment='mito', plot_type='needle', aa_norm=False, plot_name='cov_plot_mito_needle')\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', aa_norm=False, plot_name='cov_plot_cyto_behrens')\n",
    "#plot_obj.plot_coverage(compartment='mito', plot_type='behrens', aa_norm=False, plot_name='cov_plot_mito_behrens')\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='needle', aa_norm=False, #plot_name='cov_plot_cyto_needle_100p', sample_list=['100p1', '100p2', '100p3', '100p4'])\n",
    "#plot_obj.plot_coverage(compartment='cyto', plot_type='behrens', y_norm=True, plot_name='cov_plot_cyto_behrens_norm')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample/group:  Charge-titration"
     ]
    }
   ],
   "source": [
    "plot_obj.plot_abundance(bc_list_exl=['l4Sp', 'l7Sp'], sample_list_exl=['70p2', '70p3', '70p4'], charge_plot=False, group=True, plot_type='aa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample pairs:  (100p - 0p)  (85p - 40p)"
     ]
    }
   ],
   "source": [
    "plot_obj.plot_abundance_corr(sample_pairs=[['100p', '85p'], ['0p', '40p']], log=False, charge_plot=False, plot_type='transcript', min_obs=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Now plotting sample pairs:  (100p1 - 100p2)  (100p2 - 100p3)  (100p3 - 100p4)"
     ]
    }
   ],
   "source": [
    "plot_obj.plot_abundance_corr(sample_unique_pairs=[['100p1', '100p2', '100p3'], ['100p2', '100p3', '100p4']], log=False, charge_plot=True, plot_type='transcript', min_obs=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the TM_analysis to find the most mutated positions in the tRNA transcripts.\n",
    "# Then generate a new version of the tRNA transcripts database with these positions masked\n",
    "# Then, using the masked sequences, re-run the alignment, stats collected and plotting\n",
    "\n",
    "# Re-run using gap_open = -3 and gap_extension = -2 to reflect how gaps are more tolerated\n",
    "\n",
    "### The masked tRNA database will only be used for alignment\n",
    "# All other steps will use the old unmasked database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 623,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences...\n",
      "Collecting stats from:  100p1  100p2  100p4  100p3  85p4  85p2  85p3  85p1  70p3  70p1  70p2  70p4  55p2  55p1  55p3  55p4  40p1  40p2  40p3  25p1  40p4  25p2  10p2  25p3  10p4  25p4  0p2  10p3  10p1  0p1  0p3  0p4  0p_90m1  0p_90m3  0p_90m2  0p_90m4"
     ]
    }
   ],
   "source": [
    "### Perform transcript mutation analysis ###\n",
    "TM_obj = TM_analysis(dir_dict, sample_df, tRNA_database, pull_default=True, common_seqs=common_seqs, ignore_common_count=True)\n",
    "TM_dir_abs = TM_obj.make_dir(overwrite=True)\n",
    "\n",
    "#TM_obj.find_muts(n_jobs=4, sample_list=['70p1', '70p2', '70p3', '70p4'])\n",
    "TM_obj.find_muts(n_jobs=4, unique_anno=True)\n",
    "#TM_obj.find_muts(n_jobs=4)\n",
    "TM_obj.fix_end()\n",
    "\n",
    "#TM_obj.plot_transcript_logo()\n",
    "\n",
    "\n",
    "\n",
    "# tr_mut_out = TM_obj.plot_transcript_mut(topN=30, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat', min_count_show=10)\n",
    "#tr_mut_out = TM_obj.plot_transcript_mut(topN=30, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat2')\n",
    "\n",
    "\n",
    "#tr_cov_out = TM_obj.plot_transcript_cov(topN=40, no_plot_return=False, mito=False, plot_name='test_heat_cov', sort_rows=True)\n",
    "#tr_mut_out = TM_obj.plot_transcript_mut(topN=40, no_plot_return=False, mito=False, gap_only=False, plot_name='test_heat2', sort_rows=tr_cov_out[2])\n",
    "\n",
    "\n",
    "TM_obj.mask_tRNA_database(min_mut_freq=0.5, min_pos_count=50, min_tr_count=100, frac_max_score=0.90)\n",
    "\n",
    "tRNA_database_masked = TM_obj.write_masked_tRNA_database(out_dir='tRNA_database_masked')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 615,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEDCAYAAADayhiNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUI0lEQVR4nO3dbVBU58GH8T9sRCi7pYLtJjbyIq10ICEhATpWqtiIaNPYqR1TG0IbQYVUQxpDQkwt03bsEI0JBtqMob6k0UzHEhsbW1sNrdNW/AQ6fZIKzAQYYDLYJvQlLCKycp4PGe7JBkWQs7sC128mM/E+Z8/e5wvXnpfdE2JZliUAACSFBnsCAIAbB1EAABhEAQBgEAUAgEEUAADGTcGewPVKTk7W0NCQnE5nsKcCAJOGx+NRaGiozp07d8Xlk/ZIYWhoSNxNCwDjY1mWhoaGrrp80h4pDB8hNDQ0BHkmADB5pKenj7p80h4pAADsRxQAAAZRAAAYRAEAYIw7Ck1NTUpJSdH58+d9xnNycpSUlDTiv3//+99mnbfeekv5+flKS0tTVlaWnn/+eQ0ODk58LwAAthjX3Uetra0qKiqS1+v1Ge/r61NXV5cef/xxZWZm+iz75Cc/KUnq6OjQQw89pLS0NO3atUutra2qrKyUx+NReXn5BHcDAGCHMUXB6/Xq0KFDeu655zRjxowRy1taWmRZlu655x4lJiZecRs1NTVyuVx68cUXFRYWpsWLFys8PFzbtm1TUVGR3G73xPYEADBhYzp91NjYqJ07d6qgoEClpaUjljc1NSk8PFzx8fFX3UZ9fb2WLFmisLAwM7Z8+XJdvnxZp06dGv/MAQC2G9ORQmJiourq6hQTE6Pf/OY3I5a3tLQoKipKmzdvVn19vS5fvqzs7Gw9/fTT+vSnP63+/n51d3crISHB53XR0dFyOp1qb2+3Z2/GwdN/WW3d/bZtb94tEXJGOGzbHgAEw5iiMHv27FGXNzc36/3339fnP/955efnq62tTVVVVfrOd76j119/Xb29vZJ0xd8pioyMlMfjGTF+rW/d9fb2yuVyjWX6V9TW3a+yX7Re9+s/bvv6RKXO43eYAExutvzMxdatW2VZlu644w5JH/5BT0xM1AMPPKA33nhDixcvHvX1oaHcGQsANwJbopCamjpi7O6775bL5VJzc7PuvfdeSR/epfRxHo/nip/4r/WbRtc6kgAAjN+EP6JfuHBBhw8fVnNzs8/40NCQBgcHNWvWLEVGRsrtdqujo8NnnZ6eHvX19Y241gAACI4JR2HmzJl65pln9LOf/cxn/M9//rMuXrxovrewcOFCnTx5UpcuXTLrHD9+XA6HY8R3GwAAwTHhKDgcDn3ve9/Tm2++qW3btun06dN6+eWXVVZWpnvuuUdf/OIXJUnr1q3Te++9pw0bNujkyZPav3+/KioqdP/992vOnDkT3hEAwMTZck1h7dq1cjqdeuWVV1RbW6uoqCitWbNGjzzyiFknMTFR+/bt044dO1RSUqJZs2Zp7dq1PusAAIIrxJqkjy8bvtB8vQ/Z+b82D7ekAph2rvW3k3tBAQAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYIw7Ck1NTUpJSdH58+d9xk+dOqVvfvObuuOOO/SVr3xF+/btG/Hat956S/n5+UpLS1NWVpaef/55DQ4OXv/sAQC2GlcUWltbVVRUJK/X6zN+5swZFRcXa968eaqurtZ9992nHTt2aO/evWadjo4OPfTQQ5o5c6Z27dqlgoIC7d+/XxUVFfbsCQBgwm4ay0per1eHDh3Sc889pxkzZoxYXlVVpeTkZD377LOSpEWLFsnr9Wr37t3Kz89XWFiYampq5HK59OKLLyosLEyLFy9WeHi4tm3bpqKiIrndbnv3DAAwbmM6UmhsbNTOnTtVUFCg0tJSn2UDAwNqaGjQsmXLfMZzc3P1wQcf6MyZM5Kk+vp6LVmyRGFhYWad5cuX6/Llyzp16tRE9wMAYIMxRSExMVF1dXXatGmTHA6Hz7Kuri4NDg4qISHBZzwuLk6S1N7erv7+fnV3d49YJzo6Wk6nU+3t7RPZBwCATcZ0+mj27NlXXdbb2ytJcjqdPuORkZGSJI/Hc9V1htfzeDwjxtPT00edU29vr1wu1+gTBwCMy4RvSbUsa/Q3CA0d0zoAgOAb05HCaIY/rff19fmMD3/6d7lc5gjh4+sMr3elT/wNDQ2jvu+1jiQAAOM34Y/osbGxcjgc6uzs9Bkf/ndCQoIiIyPldrvV0dHhs05PT4/6+vpGXGsAAATHhKMwc+ZMpaen68SJEz6niY4fPy6Xy6XbbrtNkrRw4UKdPHlSly5d8lnH4XAoMzNzotMAANjAlpP5Dz/8sM6cOaPHHntMf/nLX7Rr1y7t3btXRUVFioiIkCStW7dO7733njZs2KCTJ0+aL67df//9mjNnjh3TAABMkC1RWLBggaqrq9Xa2qqNGzfq6NGjevLJJ7V+/XqzTmJiovbt26cLFy6opKRE+/fv19q1a/WDH/zAjikAAGww7gvNq1at0qpVq0aM5+TkKCcnZ9TXpqen69e//vV43xIAECDcCwoAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAAjJvs2pDX69Vdd92lgYEBn/FPfOITOnv2rCTp1KlTqqys1DvvvKOYmBg9+OCDKigosGsKAIAJsi0K7e3tGhgY0Pbt2xUfH2/GQ0M/PBg5c+aMiouLtWLFCj366KNqbGzUjh07ZFmWCgsL7ZoGAGACbItCc3OzQkNDlZubq4iIiBHLq6qqlJycrGeffVaStGjRInm9Xu3evVv5+fkKCwuzayoAgOtk2zWFpqYmxcbGXjEIAwMDamho0LJly3zGc3Nz9cEHH+jMmTN2TQMAMAG2RaGlpUVhYWEqLCxUWlqaMjIyVF5eLo/Ho66uLg0ODiohIcHnNXFxcZI+PPUEAAg+W08feTwerV69WsXFxXr77bdVXV2t9vZ2bd68WZLkdDp9XhMZGSlJ8ng8I7aXnp4+6vv19vbK5XLZNHsAgGRjFCorKxUVFaWkpCRJUkZGhmJiYvTEE0+ovr5+1NcOX4wGAASXbVHIzMwcMZadne3z776+Pp9/Dx8hXOkTf0NDw6jvd60jCQDA+NnyEb2np0e1tbXq6uryGb948aIkKSYmRg6HQ52dnT7Lh//98WsNAIDgsCUKISEhKi8v18GDB33Gjx07JofDoS996UtKT0/XiRMnZFmWWX78+HG5XC7ddtttdkwDADBBtpw+io6OVl5eng4cOCCn06n09HQ1NjZq9+7dysvLU1xcnB5++GGtXbtWjz32mL7xjW/o7Nmz2rt3rx5//PEr3sYKAAg8264plJWVye126/Dhw6qpqZHb7VZJSYnWrVsnSVqwYIGqq6tVVVWljRs3yu1268knn+RnLgDgBmJbFGbMmKH169dr/fr1V10nJydHOTk5dr0lAMBm3AsKADCIAgDAsO30EYAbg6f/stq6+23Z1rxbIuSMcNiyLUwORAGYYtq6+1X2i1ZbtrV9faJS5zmvvSKmDKKAKY1PzcD4EAVMaXxqBsaHC80AAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDbzRPI/zkA4BrIQrTCD/5AOBaOH0EADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwbgr2BABgqvP0X1Zbd79t25t3S4ScEQ7btvdRRAEA/Kytu19lv2i1bXvb1ycqdZ7Ttu19FKePAAAGUQAAGJw+AjDp2XnO3p/n6ycDogBg0rPznL0/z9dPBgE/ffS73/1O9957r1JTU7VixQodOXIk0FMAAFxFQKNw7NgxlZaWKisrSz//+c+VmZmpsrIy/fGPfwzkNAAAVxHQ00eVlZVasWKFtmzZIkn68pe/rP/973964YUXtHz58kBOBQBwBQE7Uujq6lJnZ6eWLVvmM56bm6u2tjZ1dXUFaioAgKsI2JFCW1ubJCkhIcFnPC4uTpLU3t6uuXPnmvH09PRRt9fb2zum9a7GsqQhy7qu117J2hMhCgmxbXN+Yec+T4b9ldjniWKf7XEj/b3p7e1VyCgvDlgUhv+IO52+V/UjIyMlSR6PZ9zbHG3HxjIXl8t1Xa+fjNjn6YF9vjGFhEgOG0szkX0OCQlRaOjVTxIFLArWNSr58Uk2NDT4bS7DRxf+fI8bDfs8PbDP04M/9zlg1xSGi9bX1+czPnyEcCNXHgCmi4BFYfhaQmdnp894R0eHz3IAQPAELApxcXG69dZbR3wn4cSJE4qPj9ecOXMCNRUAwFUE9HsKGzdu1JYtWxQVFaXs7Gz96U9/0h/+8AdVVlYGchoAgKsIaBRWrVqlS5cuad++faqtrdXcuXO1fft2ffWrXw3kNAAAVxHwH8Rbs2aN1qxZE+i3BQCMAc9TAAAYIda1vkAAAJg2OFIAABhEAQBgEAUAgDHtojCdn/zW1NSklJQUnT9/PthT8auhoSH96le/0n333ae0tDQtXbpUFRUV1/Wji5OFZVl6+eWXlZubq9TUVK1cuVJHjx4N9rQCZtOmTcrJyQn2NPzO6/UqNTVVSUlJPv+lpaXZ9h7T6hnNw09+++53v6usrCzV1dWprKxM4eHhU/4hP62trSoqKpLX6w32VPxuz5492rVrlwoLC7VgwQK1t7erqqpK77zzjvbu3Rvs6fnFSy+9pKqqKj3yyCO688479de//lWlpaVyOBxT/ntAv/3tb/Xmm28qNjY22FPxu/b2dg0MDGj79u2Kj48346P96um4WdPI0qVLre9///s+Y48++qi1fPnyIM3I/wYHB62DBw9aaWlpVmZmpjV//nyru7s72NPym6GhISsjI8P60Y9+5DP++9//3po/f7517ty5IM3Mfy5dumRlZGRYP/nJT3zGH3zwQevb3/52kGYVGOfPn7cyMjKsRYsWWUuXLg32dPzujTfesL7whS9YFy5c8Nt7TJvTR9P1yW+NjY3auXOnCgoKVFpaGuzp+F1fX59Wrlypr33taz7j8+bNkzTyBxmnAofDoQMHDmjDhg0+4zNmzNDAwECQZhUYW7du1cKFC7VgwYJgTyUgmpqaFBsbq4iICL+9x7SJwlie/DYVJSYmqq6uTps2bZLD4Qj2dPzO6XRq69atuvvuu33G6+rqJEmf+9zngjEtvwoNDVVSUpLcbrcsy9L777+vmpoanT59Wt/61reCPT2/qa2t1T/+8Q/98Ic/DPZUAqalpUVhYWEqLCxUWlqaMjIyVF5ebuv1smlzTcEfT36bDGbPnh3sKQTd3//+d9XU1Gjp0qVKTEwM9nT86sSJEyopKZEkZWdna+XKlUGekX+8++67qqioUEVFhaKjo4M9nYBpbm6Wx+PR6tWrVVxcrLffflvV1dVqb2/XK6+8ct1Po/yoaRMFa5xPfsPU0NjYqOLiYt16663atm1bsKfjd8nJyTp48KBaWlr0wgsvaMOGDfrlL39pyx+LG4VlWXr66ae1ePFi5ebmBns6AVVZWamoqCglJSVJkjIyMhQTE6MnnnhCp0+f1sKFCyf8HtMmCjz5bfo5duyYnnrqKcXHx2vPnj2aNWtWsKfkd3PnztXcuXOVkZEhp9OpsrIynT17VnfddVewp2abV199VS0tLTp69Ki5m274Q5/X65XD4ZhSEfyozMzMEWPZ2dmSPjyKsCMK0+bjMU9+m17279+vzZs3684779Srr76qz3zmM8Gekt/897//1ZEjR/TPf/7TZzw5OVmS9K9//SsY0/Kb48eP6z//+Y+ysrKUkpKilJQUHTlyRJ2dnUpJSdHrr78e7Cn6RU9Pj2pra0fcFHPx4kVJsu1Dz7SJAk9+mz5qa2v1zDPPaMWKFdqzZ8+UPwocGhrSU089pUOHDvmM19fXS5Lmz58fjGn5zY9//GO99tprPv8tWbJEN998s/n/qSgkJETl5eU6ePCgz/ixY8fkcDhG3FxxvabN6SOJJ79NBz09PfrpT3+qz372s8rLy9O5c+d8lsfGxk65C5PR0dF64IEHVFNTo/DwcN1+++1qbGzUSy+9pNWrV5vbcaeKK+3Ppz71KYWFhen2228PwowCIzo6Wnl5eTpw4ICcTqfS09PV2Nio3bt3Ky8vz9xJOVHTKgo8+W3q+9vf/qb+/n69++67ysvLG7F8x44d+vrXvx6EmfnXli1bdMstt+i1115TdXW1br75ZpWUlKiwsDDYU4ONysrK5Ha7dfjwYdXU1MjtdqukpETr1q2z7T14ngIAwJg21xQAANdGFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCA8f/yF4DDAC61bAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(TM_obj.mask_stats['human']['mask_count'], bins=15);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mask_count': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "        1, 0, 0, 5, 0, 0, 0, 5, 0, 3, 0, 3, 2, 3, 0, 5, 0, 3, 3, 2, 0, 3,\n",
       "        3, 3, 3, 2, 0, 2, 5, 5, 4, 2, 2, 2, 2, 2, 2, 0, 2, 4, 2, 4, 0, 1,\n",
       "        0, 0, 0, 0, 2, 0, 0, 0, 2, 2, 0, 0, 0, 2, 2, 2, 2, 0, 0, 2, 0, 1,\n",
       "        1, 1, 2, 0, 2, 0, 2, 2, 0, 0, 2, 0, 0, 2, 0, 2, 0, 0, 0, 2, 2, 2,\n",
       "        2, 2, 2, 2, 2, 2, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 2, 0, 0, 1,\n",
       "        1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 3, 0, 3, 3, 3, 3, 3, 3, 3, 0,\n",
       "        0, 1, 1, 1, 3, 3, 3, 3, 3, 3, 3, 3, 0, 0, 4, 4, 1, 0, 0, 0, 0, 0,\n",
       "        0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 3, 3,\n",
       "        3, 3, 0, 0, 3, 3, 3, 3, 0, 0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 0, 0, 0,\n",
       "        0, 0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 0, 0, 5, 5, 0, 0, 5, 0, 0, 0, 0,\n",
       "        0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "        0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 2, 2]),\n",
       " 'mask_sum': 307,\n",
       " 'mask_mean': 1.0848056537102473,\n",
       " 'mask_frac': 0.4664310954063604}"
      ]
     },
     "execution_count": 616,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TM_obj.mask_stats['human']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 617,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEDCAYAAADayhiNAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUI0lEQVR4nO3dbVBU58GH8T9sRCi7pYLtJjbyIq10ICEhATpWqtiIaNPYqR1TG0IbQYVUQxpDQkwt03bsEI0JBtqMob6k0UzHEhsbW1sNrdNW/AQ6fZIKzAQYYDLYJvQlLCKycp4PGe7JBkWQs7sC128mM/E+Z8/e5wvXnpfdE2JZliUAACSFBnsCAIAbB1EAABhEAQBgEAUAgEEUAADGTcGewPVKTk7W0NCQnE5nsKcCAJOGx+NRaGiozp07d8Xlk/ZIYWhoSNxNCwDjY1mWhoaGrrp80h4pDB8hNDQ0BHkmADB5pKenj7p80h4pAADsRxQAAAZRAAAYRAEAYIw7Ck1NTUpJSdH58+d9xnNycpSUlDTiv3//+99mnbfeekv5+flKS0tTVlaWnn/+eQ0ODk58LwAAthjX3Uetra0qKiqS1+v1Ge/r61NXV5cef/xxZWZm+iz75Cc/KUnq6OjQQw89pLS0NO3atUutra2qrKyUx+NReXn5BHcDAGCHMUXB6/Xq0KFDeu655zRjxowRy1taWmRZlu655x4lJiZecRs1NTVyuVx68cUXFRYWpsWLFys8PFzbtm1TUVGR3G73xPYEADBhYzp91NjYqJ07d6qgoEClpaUjljc1NSk8PFzx8fFX3UZ9fb2WLFmisLAwM7Z8+XJdvnxZp06dGv/MAQC2G9ORQmJiourq6hQTE6Pf/OY3I5a3tLQoKipKmzdvVn19vS5fvqzs7Gw9/fTT+vSnP63+/n51d3crISHB53XR0dFyOp1qb2+3Z2/GwdN/WW3d/bZtb94tEXJGOGzbHgAEw5iiMHv27FGXNzc36/3339fnP/955efnq62tTVVVVfrOd76j119/Xb29vZJ0xd8pioyMlMfjGTF+rW/d9fb2yuVyjWX6V9TW3a+yX7Re9+s/bvv6RKXO43eYAExutvzMxdatW2VZlu644w5JH/5BT0xM1AMPPKA33nhDixcvHvX1oaHcGQsANwJbopCamjpi7O6775bL5VJzc7PuvfdeSR/epfRxHo/nip/4r/WbRtc6kgAAjN+EP6JfuHBBhw8fVnNzs8/40NCQBgcHNWvWLEVGRsrtdqujo8NnnZ6eHvX19Y241gAACI4JR2HmzJl65pln9LOf/cxn/M9//rMuXrxovrewcOFCnTx5UpcuXTLrHD9+XA6HY8R3GwAAwTHhKDgcDn3ve9/Tm2++qW3btun06dN6+eWXVVZWpnvuuUdf/OIXJUnr1q3Te++9pw0bNujkyZPav3+/KioqdP/992vOnDkT3hEAwMTZck1h7dq1cjqdeuWVV1RbW6uoqCitWbNGjzzyiFknMTFR+/bt044dO1RSUqJZs2Zp7dq1PusAAIIrxJqkjy8bvtB8vQ/Z+b82D7ekAph2rvW3k3tBAQAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYBAFAIBBFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCAQRQAAAZRAAAYRAEAYIw7Ck1NTUpJSdH58+d9xk+dOqVvfvObuuOOO/SVr3xF+/btG/Hat956S/n5+UpLS1NWVpaef/55DQ4OXv/sAQC2GlcUWltbVVRUJK/X6zN+5swZFRcXa968eaqurtZ9992nHTt2aO/evWadjo4OPfTQQ5o5c6Z27dqlgoIC7d+/XxUVFfbsCQBgwm4ay0per1eHDh3Sc889pxkzZoxYXlVVpeTkZD377LOSpEWLFsnr9Wr37t3Kz89XWFiYampq5HK59OKLLyosLEyLFy9WeHi4tm3bpqKiIrndbnv3DAAwbmM6UmhsbNTOnTtVUFCg0tJSn2UDAwNqaGjQsmXLfMZzc3P1wQcf6MyZM5Kk+vp6LVmyRGFhYWad5cuX6/Llyzp16tRE9wMAYIMxRSExMVF1dXXatGmTHA6Hz7Kuri4NDg4qISHBZzwuLk6S1N7erv7+fnV3d49YJzo6Wk6nU+3t7RPZBwCATcZ0+mj27NlXXdbb2ytJcjqdPuORkZGSJI/Hc9V1htfzeDwjxtPT00edU29vr1wu1+gTBwCMy4RvSbUsa/Q3CA0d0zoAgOAb05HCaIY/rff19fmMD3/6d7lc5gjh4+sMr3elT/wNDQ2jvu+1jiQAAOM34Y/osbGxcjgc6uzs9Bkf/ndCQoIiIyPldrvV0dHhs05PT4/6+vpGXGsAAATHhKMwc+ZMpaen68SJEz6niY4fPy6Xy6XbbrtNkrRw4UKdPHlSly5d8lnH4XAoMzNzotMAANjAlpP5Dz/8sM6cOaPHHntMf/nLX7Rr1y7t3btXRUVFioiIkCStW7dO7733njZs2KCTJ0+aL67df//9mjNnjh3TAABMkC1RWLBggaqrq9Xa2qqNGzfq6NGjevLJJ7V+/XqzTmJiovbt26cLFy6opKRE+/fv19q1a/WDH/zAjikAAGww7gvNq1at0qpVq0aM5+TkKCcnZ9TXpqen69e//vV43xIAECDcCwoAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAAjJvs2pDX69Vdd92lgYEBn/FPfOITOnv2rCTp1KlTqqys1DvvvKOYmBg9+OCDKigosGsKAIAJsi0K7e3tGhgY0Pbt2xUfH2/GQ0M/PBg5c+aMiouLtWLFCj366KNqbGzUjh07ZFmWCgsL7ZoGAGACbItCc3OzQkNDlZubq4iIiBHLq6qqlJycrGeffVaStGjRInm9Xu3evVv5+fkKCwuzayoAgOtk2zWFpqYmxcbGXjEIAwMDamho0LJly3zGc3Nz9cEHH+jMmTN2TQMAMAG2RaGlpUVhYWEqLCxUWlqaMjIyVF5eLo/Ho66uLg0ODiohIcHnNXFxcZI+PPUEAAg+W08feTwerV69WsXFxXr77bdVXV2t9vZ2bd68WZLkdDp9XhMZGSlJ8ng8I7aXnp4+6vv19vbK5XLZNHsAgGRjFCorKxUVFaWkpCRJUkZGhmJiYvTEE0+ovr5+1NcOX4wGAASXbVHIzMwcMZadne3z776+Pp9/Dx8hXOkTf0NDw6jvd60jCQDA+NnyEb2np0e1tbXq6uryGb948aIkKSYmRg6HQ52dnT7Lh//98WsNAIDgsCUKISEhKi8v18GDB33Gjx07JofDoS996UtKT0/XiRMnZFmWWX78+HG5XC7ddtttdkwDADBBtpw+io6OVl5eng4cOCCn06n09HQ1NjZq9+7dysvLU1xcnB5++GGtXbtWjz32mL7xjW/o7Nmz2rt3rx5//PEr3sYKAAg8264plJWVye126/Dhw6qpqZHb7VZJSYnWrVsnSVqwYIGqq6tVVVWljRs3yu1268knn+RnLgDgBmJbFGbMmKH169dr/fr1V10nJydHOTk5dr0lAMBm3AsKADCIAgDAsO30EYAbg6f/stq6+23Z1rxbIuSMcNiyLUwORAGYYtq6+1X2i1ZbtrV9faJS5zmvvSKmDKKAKY1PzcD4EAVMaXxqBsaHC80AAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDbzRPI/zkA4BrIQrTCD/5AOBaOH0EADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwiAIAwCAKAACDKAAADKIAADCIAgDAIAoAAIMoAAAMogAAMIgCAMAgCgAAgygAAAyiAAAwbgr2BABgqvP0X1Zbd79t25t3S4ScEQ7btvdRRAEA/Kytu19lv2i1bXvb1ycqdZ7Ttu19FKePAAAGUQAAGJw+AjDp2XnO3p/n6ycDogBg0rPznL0/z9dPBgE/ffS73/1O9957r1JTU7VixQodOXIk0FMAAFxFQKNw7NgxlZaWKisrSz//+c+VmZmpsrIy/fGPfwzkNAAAVxHQ00eVlZVasWKFtmzZIkn68pe/rP/973964YUXtHz58kBOBQBwBQE7Uujq6lJnZ6eWLVvmM56bm6u2tjZ1dXUFaioAgKsI2JFCW1ubJCkhIcFnPC4uTpLU3t6uuXPnmvH09PRRt9fb2zum9a7GsqQhy7qu117J2hMhCgmxbXN+Yec+T4b9ldjniWKf7XEj/b3p7e1VyCgvDlgUhv+IO52+V/UjIyMlSR6PZ9zbHG3HxjIXl8t1Xa+fjNjn6YF9vjGFhEgOG0szkX0OCQlRaOjVTxIFLArWNSr58Uk2NDT4bS7DRxf+fI8bDfs8PbDP04M/9zlg1xSGi9bX1+czPnyEcCNXHgCmi4BFYfhaQmdnp894R0eHz3IAQPAELApxcXG69dZbR3wn4cSJE4qPj9ecOXMCNRUAwFUE9HsKGzdu1JYtWxQVFaXs7Gz96U9/0h/+8AdVVlYGchoAgKsIaBRWrVqlS5cuad++faqtrdXcuXO1fft2ffWrXw3kNAAAVxHwH8Rbs2aN1qxZE+i3BQCMAc9TAAAYIda1vkAAAJg2OFIAABhEAQBgEAUAgDHtojCdn/zW1NSklJQUnT9/PthT8auhoSH96le/0n333ae0tDQtXbpUFRUV1/Wji5OFZVl6+eWXlZubq9TUVK1cuVJHjx4N9rQCZtOmTcrJyQn2NPzO6/UqNTVVSUlJPv+lpaXZ9h7T6hnNw09+++53v6usrCzV1dWprKxM4eHhU/4hP62trSoqKpLX6w32VPxuz5492rVrlwoLC7VgwQK1t7erqqpK77zzjvbu3Rvs6fnFSy+9pKqqKj3yyCO688479de//lWlpaVyOBxT/ntAv/3tb/Xmm28qNjY22FPxu/b2dg0MDGj79u2Kj48346P96um4WdPI0qVLre9///s+Y48++qi1fPnyIM3I/wYHB62DBw9aaWlpVmZmpjV//nyru7s72NPym6GhISsjI8P60Y9+5DP++9//3po/f7517ty5IM3Mfy5dumRlZGRYP/nJT3zGH3zwQevb3/52kGYVGOfPn7cyMjKsRYsWWUuXLg32dPzujTfesL7whS9YFy5c8Nt7TJvTR9P1yW+NjY3auXOnCgoKVFpaGuzp+F1fX59Wrlypr33taz7j8+bNkzTyBxmnAofDoQMHDmjDhg0+4zNmzNDAwECQZhUYW7du1cKFC7VgwYJgTyUgmpqaFBsbq4iICL+9x7SJwlie/DYVJSYmqq6uTps2bZLD4Qj2dPzO6XRq69atuvvuu33G6+rqJEmf+9zngjEtvwoNDVVSUpLcbrcsy9L777+vmpoanT59Wt/61reCPT2/qa2t1T/+8Q/98Ic/DPZUAqalpUVhYWEqLCxUWlqaMjIyVF5ebuv1smlzTcEfT36bDGbPnh3sKQTd3//+d9XU1Gjp0qVKTEwM9nT86sSJEyopKZEkZWdna+XKlUGekX+8++67qqioUEVFhaKjo4M9nYBpbm6Wx+PR6tWrVVxcrLffflvV1dVqb2/XK6+8ct1Po/yoaRMFa5xPfsPU0NjYqOLiYt16663atm1bsKfjd8nJyTp48KBaWlr0wgsvaMOGDfrlL39pyx+LG4VlWXr66ae1ePFi5ebmBns6AVVZWamoqCglJSVJkjIyMhQTE6MnnnhCp0+f1sKFCyf8HtMmCjz5bfo5duyYnnrqKcXHx2vPnj2aNWtWsKfkd3PnztXcuXOVkZEhp9OpsrIynT17VnfddVewp2abV199VS0tLTp69Ki5m274Q5/X65XD4ZhSEfyozMzMEWPZ2dmSPjyKsCMK0+bjMU9+m17279+vzZs3684779Srr76qz3zmM8Gekt/897//1ZEjR/TPf/7TZzw5OVmS9K9//SsY0/Kb48eP6z//+Y+ysrKUkpKilJQUHTlyRJ2dnUpJSdHrr78e7Cn6RU9Pj2pra0fcFHPx4kVJsu1Dz7SJAk9+mz5qa2v1zDPPaMWKFdqzZ8+UPwocGhrSU089pUOHDvmM19fXS5Lmz58fjGn5zY9//GO99tprPv8tWbJEN998s/n/qSgkJETl5eU6ePCgz/ixY8fkcDhG3FxxvabN6SOJJ79NBz09PfrpT3+qz372s8rLy9O5c+d8lsfGxk65C5PR0dF64IEHVFNTo/DwcN1+++1qbGzUSy+9pNWrV5vbcaeKK+3Ppz71KYWFhen2228PwowCIzo6Wnl5eTpw4ICcTqfS09PV2Nio3bt3Ky8vz9xJOVHTKgo8+W3q+9vf/qb+/n69++67ysvLG7F8x44d+vrXvx6EmfnXli1bdMstt+i1115TdXW1br75ZpWUlKiwsDDYU4ONysrK5Ha7dfjwYdXU1MjtdqukpETr1q2z7T14ngIAwJg21xQAANdGFAAABlEAABhEAQBgEAUAgEEUAAAGUQAAGEQBAGAQBQCA8f/yF4DDAC61bAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(TM_obj.mask_stats['human']['mask_count'], bins=15);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 576,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mask_count': array([ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,\n",
       "         0,  0,  0,  0,  0,  1,  0,  0,  5,  0,  0,  0,  5,  0,  4,  0,  4,\n",
       "         0,  4,  0,  5,  0,  4,  4,  0,  0,  5,  5,  5,  5,  3,  0,  3,  5,\n",
       "         5,  4,  2,  0,  0,  0,  0,  0,  0,  0,  4,  0,  4,  0,  0,  0,  0,\n",
       "         0,  0,  5,  0,  0,  0,  5,  5,  0,  0,  0,  5,  5,  5,  5,  4,  0,\n",
       "         5,  4,  2,  2,  2,  2,  3,  3,  2,  3,  3,  3,  2,  3,  3,  2,  3,\n",
       "         3,  3,  0,  0,  3,  3,  3,  2,  3,  3,  3,  2,  2,  2,  2,  2,  0,\n",
       "         0,  3,  3,  3,  0,  1,  1,  3,  5,  3,  5,  5,  1,  1,  0,  1,  1,\n",
       "         1,  0,  0,  1,  1,  1,  1,  2,  4,  0,  4,  4,  4,  4,  4,  4,  3,\n",
       "         0,  0,  3,  3,  3,  4,  4,  4,  4,  3,  3,  3,  3,  0,  0,  5,  5,\n",
       "         2,  0,  0,  0,  4,  4,  0,  1,  0,  0,  1,  1,  1,  1,  0,  0,  0,\n",
       "         1,  0,  4,  4,  4,  4,  1,  1,  0,  3,  0,  3,  3,  0,  0, 13, 13,\n",
       "        13, 13,  0,  4,  4,  4,  4,  4,  4,  4,  0,  0,  0,  0,  0,  0,  0,\n",
       "         0,  0,  5,  5,  5,  5,  5,  3,  0,  5,  5,  5,  8,  8,  6,  6,  8,\n",
       "         6,  0,  0,  3,  0,  0,  0,  0,  3,  3,  3,  3,  2,  2,  2,  2,  2,\n",
       "         0,  3,  3,  3,  3,  3,  3,  3,  3,  0,  3,  3,  0,  3,  3,  0,  2,\n",
       "         0,  0,  0,  2,  2,  1,  1,  1,  0,  4,  4]),\n",
       " 'mask_sum': 615,\n",
       " 'mask_frac': 0.6183745583038869}"
      ]
     },
     "execution_count": 576,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TM_obj.mask_stats['human']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  100p2  100p1  100p3  100p4  85p4  85p2  85p3  85p1  70p3  70p2  70p1  55p1  55p3  55p4  70p4  40p1  40p2  55p2  25p1  40p4  25p2  10p1  25p4  10p3  10p2  25p3  0p2  40p3  0p4  0p1  10p4  0p3  0p_90m4  0p_90m1  0p_90m2  0p_90m3  common-seqs\n",
      "Collecting alignment statistics, from sample:  100p1  100p2  100p3  100p4  85p2  85p1  85p3  85p4  70p3  70p1  70p2  70p4  55p2  55p1  55p3  55p4  40p1  40p2  40p4  40p3  25p1  25p4  25p2  25p3  10p2  10p1  10p3  10p4  0p1  0p2  0p3  0p4  0p_90m3  0p_90m1  0p_90m2  0p_90m4  common-seqs"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>P1_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P1_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>944</td>\n",
       "      <td>38</td>\n",
       "      <td>902</td>\n",
       "      <td>940</td>\n",
       "      <td>99.576271</td>\n",
       "      <td>95.957447</td>\n",
       "      <td>931</td>\n",
       "      <td>866</td>\n",
       "      <td>930.174768</td>\n",
       "      <td>98.622881</td>\n",
       "      <td>93.100784</td>\n",
       "      <td>922.0</td>\n",
       "      <td>74.511931</td>\n",
       "      <td>25.488069</td>\n",
       "      <td>99.033298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>P2_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P2_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>673</td>\n",
       "      <td>14</td>\n",
       "      <td>648</td>\n",
       "      <td>662</td>\n",
       "      <td>98.365527</td>\n",
       "      <td>97.885196</td>\n",
       "      <td>663</td>\n",
       "      <td>622</td>\n",
       "      <td>662.581602</td>\n",
       "      <td>98.514116</td>\n",
       "      <td>93.875230</td>\n",
       "      <td>653.0</td>\n",
       "      <td>73.813170</td>\n",
       "      <td>26.186830</td>\n",
       "      <td>98.491704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>P3_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P3_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>709</td>\n",
       "      <td>35</td>\n",
       "      <td>666</td>\n",
       "      <td>701</td>\n",
       "      <td>98.871650</td>\n",
       "      <td>95.007133</td>\n",
       "      <td>692</td>\n",
       "      <td>633</td>\n",
       "      <td>691.544180</td>\n",
       "      <td>97.602257</td>\n",
       "      <td>91.534282</td>\n",
       "      <td>674.0</td>\n",
       "      <td>72.848665</td>\n",
       "      <td>27.151335</td>\n",
       "      <td>97.398844</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate P5_index P7_index barcode species  \\\n",
       "0              100p1        100p          1     D501     D701    l1Sp   human   \n",
       "1              100p2        100p          2     D501     D702    l2Sp   human   \n",
       "2              100p3        100p          3     D501     D703    l3Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq     fastq_mate1_filename  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  P1_R1_DSA-10k.fastq.bz2   \n",
       "1     TCCGGAGA   GGCTGCCATGCAAGTGC  P2_R1_DSA-10k.fastq.bz2   \n",
       "2     CGCTCATT  GGCTGCCATGCTGTCACG  P3_R1_DSA-10k.fastq.bz2   \n",
       "\n",
       "      fastq_mate2_filename  N_total  N_CC  N_CCA  N_CCA+CC  \\\n",
       "0  P1_R2_DSA-10k.fastq.bz2      944    38    902       940   \n",
       "1  P2_R2_DSA-10k.fastq.bz2      673    14    648       662   \n",
       "2  P3_R2_DSA-10k.fastq.bz2      709    35    666       701   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.576271    95.957447           931             866   \n",
       "1             98.365527    97.885196           663             622   \n",
       "2             98.871650    95.007133           692             633   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0      930.174768                    98.622881               93.100784   \n",
       "1      662.581602                    98.514116               93.875230   \n",
       "2      691.544180                    97.602257               91.534282   \n",
       "\n",
       "   N_mapped  percent_single_annotation  percent_multiple_annotation  \\\n",
       "0     922.0                  74.511931                    25.488069   \n",
       "1     653.0                  73.813170                    26.186830   \n",
       "2     674.0                  72.848665                    27.151335   \n",
       "\n",
       "   Mapping_percent  \n",
       "0        99.033298  \n",
       "1        98.491704  \n",
       "2        97.398844  "
      ]
     },
     "execution_count": 599,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database, sample_df, SWIPE_score_mat2, gap_penalty=3, extension_penalty=2, min_score_align=20, common_seqs=common_seqs)\n",
    "align_dir_abs = align_obj.make_dir(overwrite=True)\n",
    "# sample_df = align_obj.run_serial(dry_run=False, verbose=False, overwrite=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=4)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 618,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using common sequences to prevent duplicated alignment.\n",
      "Running Swipe on:  100p1  100p3  100p2  100p4  85p4  85p2  85p3  70p3  85p1  70p1  70p2  55p1  55p2  55p4  40p1  70p4  40p2  55p3  40p3  25p3  25p4  10p1  25p1  25p2  10p4  10p2  10p3  40p4  0p2  0p1  0p3  0p4  0p_90m1  0p_90m3  0p_90m2  0p_90m4  common-seqs\n",
      "Collecting alignment statistics, from sample:  100p2  100p3  100p1  100p4  85p2  85p3  85p4  85p1  70p2  70p1  70p3  70p4  55p2  55p1  55p4  55p3  40p1  40p2  40p3  40p4  25p1  25p2  25p3  25p4  10p2  10p3  10p1  10p4  0p1  0p2  0p3  0p4  0p_90m1  0p_90m3  0p_90m2  0p_90m4  common-seqs"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name_unique</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>replicate</th>\n",
       "      <th>P5_index</th>\n",
       "      <th>P7_index</th>\n",
       "      <th>barcode</th>\n",
       "      <th>species</th>\n",
       "      <th>plot_group</th>\n",
       "      <th>hue_name</th>\n",
       "      <th>hue_value</th>\n",
       "      <th>hue_order</th>\n",
       "      <th>P5_index_seq</th>\n",
       "      <th>P7_index_seq</th>\n",
       "      <th>barcode_seq</th>\n",
       "      <th>fastq_mate1_filename</th>\n",
       "      <th>fastq_mate2_filename</th>\n",
       "      <th>N_total</th>\n",
       "      <th>N_CC</th>\n",
       "      <th>N_CCA</th>\n",
       "      <th>N_CCA+CC</th>\n",
       "      <th>CCA+CC_percent_total</th>\n",
       "      <th>percent_CCA</th>\n",
       "      <th>N_after_trim</th>\n",
       "      <th>N_UMI_observed</th>\n",
       "      <th>N_UMI_expected</th>\n",
       "      <th>percent_seqs_after_UMI_trim</th>\n",
       "      <th>percent_UMI_obs-vs-exp</th>\n",
       "      <th>N_mapped</th>\n",
       "      <th>percent_single_annotation</th>\n",
       "      <th>percent_multiple_annotation</th>\n",
       "      <th>Mapping_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100p1</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>D501</td>\n",
       "      <td>D701</td>\n",
       "      <td>l1Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>ATTACTCG</td>\n",
       "      <td>GGCTGCCATGCGACTA</td>\n",
       "      <td>P1_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P1_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>944</td>\n",
       "      <td>38</td>\n",
       "      <td>902</td>\n",
       "      <td>940</td>\n",
       "      <td>99.576271</td>\n",
       "      <td>95.957447</td>\n",
       "      <td>931</td>\n",
       "      <td>866</td>\n",
       "      <td>930.174768</td>\n",
       "      <td>98.622881</td>\n",
       "      <td>93.100784</td>\n",
       "      <td>927.0</td>\n",
       "      <td>75.512406</td>\n",
       "      <td>24.487594</td>\n",
       "      <td>99.570354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100p2</td>\n",
       "      <td>100p</td>\n",
       "      <td>2</td>\n",
       "      <td>D501</td>\n",
       "      <td>D702</td>\n",
       "      <td>l2Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>TCCGGAGA</td>\n",
       "      <td>GGCTGCCATGCAAGTGC</td>\n",
       "      <td>P2_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P2_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>673</td>\n",
       "      <td>14</td>\n",
       "      <td>648</td>\n",
       "      <td>662</td>\n",
       "      <td>98.365527</td>\n",
       "      <td>97.885196</td>\n",
       "      <td>663</td>\n",
       "      <td>622</td>\n",
       "      <td>662.581602</td>\n",
       "      <td>98.514116</td>\n",
       "      <td>93.875230</td>\n",
       "      <td>656.0</td>\n",
       "      <td>72.713415</td>\n",
       "      <td>27.286585</td>\n",
       "      <td>98.944193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100p3</td>\n",
       "      <td>100p</td>\n",
       "      <td>3</td>\n",
       "      <td>D501</td>\n",
       "      <td>D703</td>\n",
       "      <td>l3Sp</td>\n",
       "      <td>human</td>\n",
       "      <td>Charge-titration</td>\n",
       "      <td>Percent charge</td>\n",
       "      <td>100p</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGCTATA</td>\n",
       "      <td>CGCTCATT</td>\n",
       "      <td>GGCTGCCATGCTGTCACG</td>\n",
       "      <td>P3_R1_DSA-10k.fastq.bz2</td>\n",
       "      <td>P3_R2_DSA-10k.fastq.bz2</td>\n",
       "      <td>709</td>\n",
       "      <td>35</td>\n",
       "      <td>666</td>\n",
       "      <td>701</td>\n",
       "      <td>98.871650</td>\n",
       "      <td>95.007133</td>\n",
       "      <td>692</td>\n",
       "      <td>633</td>\n",
       "      <td>691.544180</td>\n",
       "      <td>97.602257</td>\n",
       "      <td>91.534282</td>\n",
       "      <td>677.0</td>\n",
       "      <td>73.264402</td>\n",
       "      <td>26.735598</td>\n",
       "      <td>97.832370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name_unique sample_name  replicate P5_index P7_index barcode species  \\\n",
       "0              100p1        100p          1     D501     D701    l1Sp   human   \n",
       "1              100p2        100p          2     D501     D702    l2Sp   human   \n",
       "2              100p3        100p          3     D501     D703    l3Sp   human   \n",
       "\n",
       "         plot_group        hue_name hue_value  hue_order P5_index_seq  \\\n",
       "0  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "1  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "2  Charge-titration  Percent charge      100p          1     AGGCTATA   \n",
       "\n",
       "  P7_index_seq         barcode_seq     fastq_mate1_filename  \\\n",
       "0     ATTACTCG    GGCTGCCATGCGACTA  P1_R1_DSA-10k.fastq.bz2   \n",
       "1     TCCGGAGA   GGCTGCCATGCAAGTGC  P2_R1_DSA-10k.fastq.bz2   \n",
       "2     CGCTCATT  GGCTGCCATGCTGTCACG  P3_R1_DSA-10k.fastq.bz2   \n",
       "\n",
       "      fastq_mate2_filename  N_total  N_CC  N_CCA  N_CCA+CC  \\\n",
       "0  P1_R2_DSA-10k.fastq.bz2      944    38    902       940   \n",
       "1  P2_R2_DSA-10k.fastq.bz2      673    14    648       662   \n",
       "2  P3_R2_DSA-10k.fastq.bz2      709    35    666       701   \n",
       "\n",
       "   CCA+CC_percent_total  percent_CCA  N_after_trim  N_UMI_observed  \\\n",
       "0             99.576271    95.957447           931             866   \n",
       "1             98.365527    97.885196           663             622   \n",
       "2             98.871650    95.007133           692             633   \n",
       "\n",
       "   N_UMI_expected  percent_seqs_after_UMI_trim  percent_UMI_obs-vs-exp  \\\n",
       "0      930.174768                    98.622881               93.100784   \n",
       "1      662.581602                    98.514116               93.875230   \n",
       "2      691.544180                    97.602257               91.534282   \n",
       "\n",
       "   N_mapped  percent_single_annotation  percent_multiple_annotation  \\\n",
       "0     927.0                  75.512406                    24.487594   \n",
       "1     656.0                  72.713415                    27.286585   \n",
       "2     677.0                  73.264402                    26.735598   \n",
       "\n",
       "   Mapping_percent  \n",
       "0        99.570354  \n",
       "1        98.944193  \n",
       "2        97.832370  "
      ]
     },
     "execution_count": 618,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Align reads to database of reference tRNAs ###\n",
    "align_obj = SWIPE_align(dir_dict, tRNA_database_masked, sample_df, SWIPE_score_mat2, gap_penalty=3, extension_penalty=2, min_score_align=20, common_seqs=common_seqs)\n",
    "align_dir_abs = align_obj.make_dir(overwrite=True)\n",
    "# sample_df = align_obj.run_serial(dry_run=False, verbose=False, overwrite=False)\n",
    "sample_df = align_obj.run_parallel(n_jobs=4)\n",
    "sample_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
